// @krn/live-bridge/src/liveStreamViewer/index.ts
import {
    EGiftTab,
    IJoinFansGroupParams,
    EJoinFansGroupSource,
    ISendGiftToAnchorParams,
    ISendGiftToAnchorResult,
} from './interface';
import { liveRnBridgeExecute } from '../utils/liveRnBridgeExecute';
import type {
    IMockSelfSendGiftEffect,
    IOpenLiveGiftPanel,
    IQueryFollowStatusParams,
    IQueryFollowStatusResult,
    IRoleInfos,
    ISendGiftToGuestParams,
    ISyncWealthGradeStatus,
    ISendGiftResponseExtraInfo,
    ILiveSendNoPanelGiftResponse,
} from './interface';

/**
 * 通知客户端mock 本地礼物特效 </br>
 * 支持版本: 9.5.20+ </br>
 * @param params
 */
export function mockSelfSendGiftEffect(
    params: IMockSelfSendGiftEffect,
): Promise<object> {
    return liveRnBridgeExecute('mockSelfSendGiftEffect', { param: params });
}

/**
 * 直播间拉起小黄车 </br>
 * 支持版本: 9.0.20+ </br>
 */
export function openCart(): Promise<object> {
    return liveRnBridgeExecute('openCart', {});
}

/**
 * 同步等级保护、等级隐身、特殊身份的状态 </br>
 * 支持版本: 9.0.20+ </br>
 * @param params
 */
export function syncWealthGradeStatus(
    params: ISyncWealthGradeStatus,
): Promise<object> {
    return liveRnBridgeExecute('syncWealthGradeStatus', params);
}

/**
 * 查询观众的关注状态 </br>
 * 支持版本: 8.3.40+ </br>
 * @param uid
 */
export function queryFollowStatus(
    params: IQueryFollowStatusParams,
): Promise<IQueryFollowStatusResult> {
    return liveRnBridgeExecute('queryFollowStatus', { param: params });
}

/**
 * 打开礼物面板 </br>
 * 支持版本: 8.1.30+ </br>
 * @param params
 */
export function openLiveGiftPanel(params: IOpenLiveGiftPanel): Promise<object> {
    return liveRnBridgeExecute('openLiveGiftPanel', params);
}

/**
 * 给嘉宾送礼礼物面板 </br>
 * 支持版本: 8.0.20+ </br>
 * @param params
 */
export function sendGiftToGuest(
    params: ISendGiftToGuestParams,
): Promise<object> {
    return liveRnBridgeExecute('sendGiftToGuest', params);
}

/**
 * 给主播送礼物 </br>
 * @param params
 */
export function sendGiftToAnchor(
    params: ISendGiftToAnchorParams,
): Promise<ISendGiftToAnchorResult> {
    return liveRnBridgeExecute('sendGiftToAnchor', params);
}

/**
 * 直接加入粉丝团 </br>
 * 支持版本：9.10.20+ </br>
 * @param params
 */
export function joinFansGroup(params: IJoinFansGroupParams): Promise<object> {
    return liveRnBridgeExecute('joinFansGroup', params);
}

export { EGiftTab, EJoinFansGroupSource };

export type {
    IMockSelfSendGiftEffect,
    IRoleInfos,
    ISyncWealthGradeStatus,
    IQueryFollowStatusResult,
    IOpenLiveGiftPanel,
    ISendGiftToGuestParams,
    IQueryFollowStatusParams,
    IJoinFansGroupParams,
    ISendGiftToAnchorParams,
    ISendGiftToAnchorResult,
    ISendGiftResponseExtraInfo,
    ILiveSendNoPanelGiftResponse,
};
