// @krn/live-bridge/src/liveStreamViewer/interface.ts
/* eslint-disable max-lines */
import { IBridgeResult } from '../common';
import { ICdnUrl } from '../liveStream/interface';

export interface IMockSelfSendGiftEffect {
    liveStreamId: string;
    comboKey: string;
    comboCount: number;
    giftSentInfo: string;
}

export interface IRoleInfos {
    /**
     * 0 代表普通用户，1代码神秘人
     */
    roleType: 1 | 0;
    /**
     *  特殊身份的名称
     */
    roleName: string;
    /**
     * 头像, cdnUrl数组
     */
    roleHeadUrls: [{ url: string }];
}

export interface ISyncWealthGradeStatus {
    /**
     *  0表示隐身 1表示可见
     */
    visible: 1 | 0;
    roleInfos: IRoleInfos[];
}

export interface IQueryFollowStatusResult extends IBridgeResult {
    /**
     * 1 已关注，0 未关注
     */
    isFollowed: 1 | 0;
}

export enum EGiftTab {
    /**
     * 普通tab
     */
    NORMAL_GIFT = 'NormalGift',
    /**
     * 背包
     */
    PACKET_GIFT = 'PacketGift',
    /**
     * 特权
     */
    PRIVILEGE_GIFT = 'PrivilegeGift',
    /**
     * 粉丝团
     */
    FANS_GROUP_GIFT = 'FansGroupGift',
    /**
     * 道具
     */
    PROPS_PANEL = 'PropsPanel',
}

export interface IOpenLiveGiftPanel {
    /**
     * 埋点用，范围:0～32
     */
    source: number;
    /**
     * iOS从8.0.10开始支持指定giftTab
     *  iOS、Android 8.1.30（含）后开始支持道具tab：PropPanel
     */
    giftTab?: EGiftTab;
}

export interface ISendGiftToGuestParams {
    /**
     * 埋点用，范围:0～32
     */
    source: number;
    guest: {
        // eslint-disable-next-line camelcase
        user_id: number;
        // eslint-disable-next-line camelcase
        user_name: string;
        /**
         * android 8.0.20修复，之前版本无头像，兜底方案使用headurls
         */
        headurl: string;
        headurls: ICdnUrl[];
    };
    /**
     * 服务端会下发，用来请求嘉宾列表
     */
    bizMap: { chatLiveStreamId: string };
    /**
     * 送礼场景，服务端下发，用来识别支付分成策略
     */
    scene: string;
}

export interface ISendGiftToAnchorParams {
    /**
     * 礼物id
     */
    giftId?: string;
    /**
     * 礼物数量
     */
    count?: number;
    /**
     * 礼物token
     */
    giftToken?: string;
    /**
     * 送礼埋点需要的参数，会通过送礼接口传到服务端,不需要可以传空, 结构 { key1 : value1, key2: value2 }
     */
    logExtraInfo?: { [key: string]: unknown };
    /**
     * 连送时的key,不是连送传0
     */
    comboKey?: number;
    /**
     * 连送的个数
     */
    comboCount?: number;
    /**
     * 服务端是否异步送礼,异步送礼时不支持连送
     */
    isAsync?: boolean;
    /**
     * 传空表示不需要使用 native 统一送礼埋点，仅用于客户端埋点
     */
    loggerParams?: {
        /**
         * 默认 unknown, 也就是0
         */
        giftEntryType?: number;
        /**
         * 默认 unknown, 也就是0
         */
        giftBoxSourceType?: number;
    };
}

export interface ISendGiftToAnchorResult {
    /**
     * 送礼失败的错误信息
     */
    error?: string;

    /**
     * 送礼成功的返回结果 Android: 10.1.10+ iOS: 10.1.10+
     */
    response?: ILiveSendNoPanelGiftResponse;
}

export interface ILiveSendNoPanelGiftResponse {
    /**
     * 快币送礼后的加密信息，用于直播等级业务计算积分
     */
    simpleMsg?: string;
    /**
     * 用于埋点的信息,json格式
     */
    csLogCorrelateInfo?: object;
    /**
     * 用于接受服务端的提示文案
     */
    msg?: string;
    /**
     * ktraceId
     */
    ktraceId?: string;
    /**
     * 黄钻余额
     */
    xZuan?: number;
    /**
     * 黄钻对应的rmb价值，单位 fen
     */
    withdrawAmount?: number;
    /**
     * 快币余额
     */
    ksCoin?: number;
    /**
     * 快贝余额，因为只有部分接口会返回这个字段，所以默认值为-1
     */
    kshell?: number;
    /**
     * 用于校验
     */
    version?: number;
    /**
     * 快贝和普通礼物不走一个所以没有version，需要用server的时间戳判定
     */
    serverTimeStamp?: number;
    /**
     * ShowAccountProtectAlert
     */
    showAccountProtectAlert?: boolean;
    /**
     * showBindPhoneAlert
     */
    showBindPhoneAlert?: boolean;
    /**
     * starLevel
     */
    starLevel?: number;
    /**
     * subStarLevel
     */
    subStarLevel?: number;
    /**
     * styleTypeValue
     */
    styleTypeValue?: number;
    /**
     * 大额充值或消费后提示绑定手机号
     */
    needBindMobile?: boolean;
    /**
     * 是否参加充值活动
     */
    showFirstDepositWidget?: boolean;
    /**
     * 充值活动的提示文案，如 充值有礼
     */
    walletPageDepositEntranceHint?: string;
    /**
     * 参加充值活动的类型
     */
    depositEventType?: string;
    /**
     * 其他信息
     */
    extraInfo?: ISendGiftResponseExtraInfo;
}

export interface ISendGiftResponseExtraInfo {
    /**
     * 亲密度增加的toast的内容
     */
    intimacyIncreaseToast?: string;
}

export interface IQueryFollowStatusParams {
    uid: string;
}

export const enum EJoinFansGroupSource {
    UNKNOWN = 'UNKNOWN',
    /**
     * 粉丝团面板
     */
    FANS_GROUP_PANEL = 'FANS_GROUP_PANEL',
    /**
     * 直播间评论feed区域
     */
    LIVE_COMMENT_FEED_AREA = 'LIVE_COMMENT_FEED_AREA',
    /**
     * 通过粉丝团红包加入粉丝团
     */
    RED_PACK_FANS_GROUP_PANEL = 'RED_PACK_FANS_GROUP_PANEL',
    /**
     * 通过电商信任卡拉起粉丝团弹框
     */
    MERCHANT_TRUST_TASK = 'MERCHANT_TRUST_TASK',
    /**
     * 礼物面板
     */
    GIFT_PANEL = 'GIFT_PANEL',
    /**
     * 人气榜面板吸底按钮
     */
    LIVE_POPULARITY_RANK_LIST = 'LIVE_POPULARITY_RANK_LIST',
    /**
     * 粉丝团快捷礼物
     */
    CHEAP_BUTTON = 'CHEAP_BUTTON',
    /**
     * 个人profile页加入粉丝团按钮
     */
    PERSONEL_CARD_FANS_GROUP_BUTTON = 'PERSONEL_CARD_FANS_GROUP_BUTTON',
}

export interface IJoinFansGroupParams {
    // 必选，标记来源
    source: EJoinFansGroupSource;
    // 可选，标记是否展示确认弹框
    needShowConfirmDialog?: false;
    // 可选，标记是否展示加入弹框
    needShowFansGroupDialog?: false;
}
