// @locallife/biz-component/lib/module/components/poiInfo/index.js
import React from 'react';
import { View, Text, Image, TouchableOpacity } from 'react-native';
import { KidIcon } from '@kid-ui/krn';
import styles from './styles';
import { screenUnit } from '@lux/ts-lbs-krn-utils';
import { getDistance } from '../../utils/decimal';
import { formatHotPeople } from '@lux/ts-lbs-krn-utils';
const {
  screenSize
} = screenUnit;
export const PoiInfo = _ref => {
  let {
    productData,
    poiData,
    isShowPhoneIcon,
    isShowMinDistance,
    toPOIList,
    toPOIDetail,
    toGuide,
    isShowPopupVisible
  } = _ref;
  // 点击电话弹窗
  const handlePhoneIcon = () => {
    isShowPopupVisible && isShowPopupVisible();
  };
  return /*#__PURE__*/React.createElement(View, {
    style: styles.container
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.header
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.title
  }, "\u9002\u7528\u95E8\u5E97", (productData === null || productData === void 0 ? void 0 : productData.poiSize) > 1 && /*#__PURE__*/React.createElement(Text, null, "\uFF08", productData === null || productData === void 0 ? void 0 : productData.poiSize, " \u5BB6\uFF09")), (productData === null || productData === void 0 ? void 0 : productData.poiSize) > 1 && /*#__PURE__*/React.createElement(TouchableOpacity, {
    activeOpacity: 1,
    onPress: toPOIList
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.rowFlexAlignCenter
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.moreStoreLink
  }, "\u67E5\u770B\u5168\u90E8"), /*#__PURE__*/React.createElement(KidIcon, {
    kid: 'common_base_next_24',
    kidColor: 'color_base_black_6',
    size: screenSize(14)
  })))), /*#__PURE__*/React.createElement(TouchableOpacity, {
    activeOpacity: 1,
    onPress: toPOIDetail
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.content, styles.rowFlexAlignCenter]
  }, /*#__PURE__*/React.createElement(Image, {
    style: styles.poiImage,
    source: {
      uri: poiData.image
    }
  }), /*#__PURE__*/React.createElement(View, {
    style: styles.poiDescWrapper
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.poiNameWrap
  }, /*#__PURE__*/React.createElement(Text, {
    numberOfLines: 1,
    style: [styles.poiName, isShowMinDistance && styles.poiNameWithMinDistance]
  }, poiData.poiName), isShowMinDistance && /*#__PURE__*/React.createElement(View, {
    style: styles.minDistance
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.minDistanceText
  }, "\u79BB\u6211\u6700\u8FD1"))), /*#__PURE__*/React.createElement(View, {
    style: styles.poiCategoryContainer
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.subCategory
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.tag
  }, poiData.subCate)), !!poiData.avgPrice && /*#__PURE__*/React.createElement(View, {
    style: [styles.subCategory, styles.subCategoryDistance]
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.tag
  }, "\uFFE5", Math.floor(+poiData.avgPrice / 100), "/\u4EBA")), !!formatHotPeople(poiData.visitCount) && /*#__PURE__*/React.createElement(View, {
    style: [styles.subCategory, styles.subCategoryDistance]
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.tag
  }, formatHotPeople(poiData.visitCount), "\u4EBA\u60F3\u53BB")))), isShowPhoneIcon && /*#__PURE__*/React.createElement(TouchableOpacity, {
    activeOpacity: 1,
    onPress: handlePhoneIcon
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.poiPhoneWrapper
  }, /*#__PURE__*/React.createElement(KidIcon, {
    kid: 'common_baseflat_telephone_24',
    kidColor: 'color_base_black_8',
    size: screenSize(24)
  }), /*#__PURE__*/React.createElement(Text, {
    style: styles.poiPhoneText
  }, "\u7535\u8BDD"))))), /*#__PURE__*/React.createElement(View, {
    style: styles.dividingLine
  }), /*#__PURE__*/React.createElement(TouchableOpacity, {
    activeOpacity: 1,
    onPress: toGuide
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.positionWrapper, styles.rowFlexAlignCenter]
  }, !!poiData.distance && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(KidIcon, {
    kid: 'common_list_location_grey',
    kidColor: 'color_base_black_11',
    size: screenSize(14)
  }), /*#__PURE__*/React.createElement(Text, {
    style: styles.distance
  }, getDistance(poiData.distance)), /*#__PURE__*/React.createElement(Text, {
    style: styles.divide
  }, "|")), /*#__PURE__*/React.createElement(Text, {
    style: styles.address,
    ellipsizeMode: "tail",
    numberOfLines: 1
  }, poiData.address), /*#__PURE__*/React.createElement(KidIcon, {
    kid: 'common_base_next_24',
    kidColor: 'color_base_black_6',
    size: screenSize(14)
  }))));
};
