// @locallife/biz-component/lib/module/components/poiInfo/styles.js
import { PRODUCT_DETAIL_COMPONENT_MARGIN_GAP, PRODUCT_DETAIL_PAGE_HORIZONTAL_PADDING, PRODUCT_DETAIL_COMPONENT_HORIZONTAL_PADDING } from '../../utils/config';
import { screenUnit } from '@lux/ts-lbs-krn-utils';
import { StyleSheet } from 'react-native';
const {
  screenSize
} = screenUnit;
export default StyleSheet.create({
  container: {
    display: 'flex',
    paddingHorizontal: screenSize(PRODUCT_DETAIL_PAGE_HORIZONTAL_PADDING),
    position: 'relative',
    paddingVertical: screenSize(PRODUCT_DETAIL_COMPONENT_HORIZONTAL_PADDING),
    backgroundColor: '#ffffff',
    marginTop: screenSize(PRODUCT_DETAIL_COMPONENT_MARGIN_GAP),
    borderRadius: screenSize(8)
  },
  header: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: screenSize(16),
    alignItems: 'center'
  },
  title: {
    fontFamily: 'PingFangSC-Medium',
    fontSize: screenSize(18),
    color: '#222222',
    fontWeight: '500'
  },
  moreStoreLink: {
    color: '#9c9c9c',
    fontSize: screenSize(14)
  },
  rowFlexAlignCenter: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center'
  },
  content: {
    marginBottom: screenSize(12)
  },
  poiImage: {
    width: screenSize(44),
    height: screenSize(44),
    borderRadius: screenSize(4)
  },
  poiDescWrapper: {
    display: 'flex',
    flexGrow: 1,
    flexShrink: 1,
    paddingLeft: screenSize(12)
  },
  poiName: {
    fontSize: screenSize(16),
    fontFamily: 'PingFangSC-Medium',
    fontWeight: '500',
    maxWidth: screenSize(254)
  },
  poiNameWithMinDistance: {
    maxWidth: screenSize(206)
  },
  poiCategoryContainer: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    color: '#666666'
  },
  subCategory: {
    fontSize: screenSize(14),
    lineHeight: screenSize(16),
    marginTop: screenSize(6),
    color: '#9C9C9C',
    flexDirection: 'row',
    alignItems: 'center'
  },
  subCategoryDistance: {
    marginHorizontal: screenSize(6)
  },
  positionWrapper: {
    marginTop: screenSize(12)
  },
  address: {
    fontSize: screenSize(14),
    lineHeight: screenSize(17),
    color: '#666666',
    flexWrap: 'wrap',
    flexShrink: 1,
    flexGrow: 1
  },
  divide: {
    fontSize: screenSize(12),
    lineHeight: screenSize(16),
    color: '#B5B5B6',
    flexGrow: 0,
    flexShrink: 0,
    paddingHorizontal: screenSize(6)
  },
  dividingLine: {
    borderBottomColor: '#E6E6E6',
    borderBottomWidth: screenSize(0.5)
  },
  distance: {
    fontSize: screenSize(14),
    lineHeight: screenSize(17),
    color: '#666666',
    flexGrow: 0,
    flexShrink: 0
  },
  positionIcon: {
    width: screenSize(12),
    height: screenSize(12),
    marginRight: screenSize(2)
  },
  minDistance: {
    borderRadius: screenSize(3),
    borderColor: 'rgb(223, 223, 223)',
    borderStyle: 'solid',
    borderWidth: screenSize(1),
    paddingHorizontal: screenSize(4),
    height: screenSize(16),
    marginLeft: screenSize(4),
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center'
  },
  minDistanceText: {
    color: '#666666',
    fontSize: screenSize(11)
  },
  poiNameWrap: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  tag: {
    color: '#666666',
    fontSize: screenSize(14)
  },
  poiPhoneWrapper: {
    width: screenSize(24)
  },
  poiPhoneText: {
    marginTop: screenSize(2),
    fontSize: screenSize(11),
    color: '#666666'
  },
  close: {
    position: 'absolute',
    right: 0,
    padding: screenSize(10)
  }
});
