// @kid-ui/krn/src/components/alert/dialog-component.tsx
import React from 'react';
import { DialogComponent as Dialog, DialogProps } from '../dialog';
import useThemeValue from '../../theme/useThemeValue';

const titleColors = { light: '#222222', dark: '#e6e6e6' };
const descTextColors = { light: '#666666', dark: '#b5b5b6' };
const borderColors = { light: '#eaeaea', dark: 'rgba(255,255,255,0.04)' };
const backgroundColor = { light: '#fff', dark: '#2b2b2f' };
const leftButtonColors = { light: '#666666', dark: '#b5b5b6' };
const rightButtonColors = { light: '#5181fb', dark: '#4f7ef8' };

export const DialogComponent: React.FC<DialogProps> = (
    props,
) => {
    const {
        popupSet = {},
        borderColor,
        leftButtonColor,
        rightButtonColor,
        titleColor,
        descTextColor,
        wrapBackGroundColor,
        ...rest
    } = props;

    const _leftButtonColor = useThemeValue(leftButtonColor || leftButtonColors);
    const _rightButtonColor = useThemeValue(rightButtonColor || rightButtonColors);
    const _borderColor = useThemeValue(borderColor || borderColors);
    const _titleColor = useThemeValue(titleColor || titleColors);
    const _descTextColor = useThemeValue(descTextColor || descTextColors);
    const _wrapBackGroundColor = useThemeValue(wrapBackGroundColor || backgroundColor);
    
    const defaultOpts: DialogProps = {
        leftButtonColor: _leftButtonColor,
        rightButtonColor: _rightButtonColor,
        borderColor: _borderColor,
        titleColor: _titleColor,
        descTextColor: _descTextColor,
        wrapBackGroundColor: _wrapBackGroundColor,
    };

    return <Dialog {...defaultOpts} {...rest}></Dialog>;
};
