// @locallife/biz-hooks/src/useSafeMemo/demo/index.tsx
/* eslint-disable react/jsx-no-bind */
import React, { useState } from 'react';
import { Text, StyleSheet, SafeAreaView, Button, View } from 'react-native';
import useSafeMemo from '../index';

const styles = StyleSheet.create({
    container: {
        flex: 1,
        width: '100%',
        display: 'flex',
        alignItems: 'center',
    },
    text: {
        backgroundColor: '#aaa',
        color: '#ffffff',
        textAlign: 'center',
        width: '100%',
        height: 50,
        lineHeight: 50,
        marginBottom: 10,
    },
    bottom: {
        display: 'flex',
        flexDirection: 'row',
    },
});

export default () => {
    const [counter, setCounter] = useState(0);

    const display = useSafeMemo(() => {
        return counter * 10;
    }, [counter]);

    return (
        <SafeAreaView style={styles.container}>
            <Text style={styles.text}>值{display}</Text>
            <View style={styles.bottom}>
                <Button
                    title="加10"
                    onPress={() => {
                        setCounter(counter + 1);
                    }}
                />
                <Button
                    title="减10"
                    onPress={() => {
                        setCounter(counter - 1);
                    }}
                />
            </View>
        </SafeAreaView>
    );
};
