// @krn/live-bridge/dist/liveStreamViewer/index.js
import { EGiftTab, } from './interface';
import { liveRnBridgeExecute } from '../utils/liveRnBridgeExecute';
/**
 * 通知客户端mock 本地礼物特效 </br>
 * 支持版本: 9.5.20+ </br>
 * @param params
 */
export function mockSelfSendGiftEffect(params) {
    return liveRnBridgeExecute('mockSelfSendGiftEffect', { param: params });
}
/**
 * 直播间拉起小黄车 </br>
 * 支持版本: 9.0.20+ </br>
 */
export function openCart() {
    return liveRnBridgeExecute('openCart', {});
}
/**
 * 同步等级保护、等级隐身、特殊身份的状态 </br>
 * 支持版本: 9.0.20+ </br>
 * @param params
 */
export function syncWealthGradeStatus(params) {
    return liveRnBridgeExecute('syncWealthGradeStatus', params);
}
/**
 * 查询观众的关注状态 </br>
 * 支持版本: 8.3.40+ </br>
 * @param uid
 */
export function queryFollowStatus(params) {
    return liveRnBridgeExecute('queryFollowStatus', { param: params });
}
/**
 * 打开礼物面板 </br>
 * 支持版本: 8.1.30+ </br>
 * @param params
 */
export function openLiveGiftPanel(params) {
    return liveRnBridgeExecute('openLiveGiftPanel', params);
}
/**
 * 给嘉宾送礼礼物面板 </br>
 * 支持版本: 8.0.20+ </br>
 * @param params
 */
export function sendGiftToGuest(params) {
    return liveRnBridgeExecute('sendGiftToGuest', params);
}
/**
 * 给主播送礼物 </br>
 * @param params
 */
export function sendGiftToAnchor(params) {
    return liveRnBridgeExecute('sendGiftToAnchor', params);
}
/**
 * 直接加入粉丝团 </br>
 * 支持版本：9.10.20+ </br>
 * @param params
 */
export function joinFansGroup(params) {
    return liveRnBridgeExecute('joinFansGroup', params);
}
export { EGiftTab };
