// @locallife/base-component/lib/commonjs/boundary/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _krn = require("@kid-ui/krn");

var _reactNative = require("react-native");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  }
});

const noop = () => {};
/**
 * 这个组件不建议放到页面的最外层使用，这会导致页面加载时无法创建子节点，影响秒开
 * @param props
 * @constructor
 */


function Boundary(props) {
  const {
    isLoading,
    isError,
    isEmpty,
    refetch = noop,
    children,
    containerStyle,
    EmptyComponent,
    LoadingComponent,
    ErrorComponent
  } = props;
  const bgColor = (0, _krn.useThemeValue)('cs_common_background_secondary');
  const _style = [styles.container, {
    backgroundColor: bgColor
  }, containerStyle];

  if (isLoading) {
    // 正在请求返回loading
    if (LoadingComponent) {
      return LoadingComponent;
    }

    return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      style: _style
    }, /*#__PURE__*/_react.default.createElement(_krn.KidLoading, null));
  }

  if (isError) {
    if (ErrorComponent) {
      return ErrorComponent;
    } // 请求失败返回错误展位图，点击refetch重试


    return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      style: _style
    }, /*#__PURE__*/_react.default.createElement(_krn.KidNetError, {
      buttonBgColor: 'cs_common_background_primary',
      onPress: refetch
    }));
  }

  if (isEmpty) {
    if (EmptyComponent) {
      return EmptyComponent;
    }

    return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      style: _style
    }, /*#__PURE__*/_react.default.createElement(_krn.KidIcon, {
      kid: 'local_server_emptystate_nocontent_96',
      size: 84
    }));
  } // 返回子组件


  return children;
}

Boundary.defaultProps = {
  containerStyle: undefined,
  isEmpty: false
};

var _default = /*#__PURE__*/_react.default.memo(Boundary);

exports.default = _default;
