// @lux/x-tools/core/deadline.js
'use strict';

// 时间
var ratios = [1000, 60, 60, 24, 30, 12];
var timeCycle = ['s', 'm', 'h', 'D', 'M', 'Y'];
var patternRegExp = /(YYYY|MM|M|DD|D|hh|h|mm|m|ss|s)/g;
var defaultFormatStrings = {
    s: 'ss秒',
    m: 'mm分钟ss秒',
    h: 'hh小时mm分钟ss秒',
    D: 'DD天 hh小时mm分钟ss秒',
    M: 'MM个月DD天 hh小时mm分钟ss秒',
    Y: 'YYYY年MM个月DD天 hh小时mm分钟ss秒',
};
function padStartZero(value, zero) {
    if (zero === void 0) { zero = true; }
    var stringVal = value.toString();
    return zero ? stringVal.padStart(2, '0') : stringVal;
}
/**
 * 倒计时格式化
 * @param timestamp 剩余毫秒时间戳
 * @returns
 */
function deadline(timestamp, options) {
    if (options === void 0) { options = {}; }
    // 最大周期
    var maxCycle = options.cycle || 'D';
    // 计算时的取整方式
    var defaultFormatString = defaultFormatStrings[maxCycle];
    // 最大周期对应的索引
    var maxIndex = timeCycle.indexOf(maxCycle);
    // 毫秒数
    var millseconds = timestamp % 1000;
    var arr = [0, 0, 0, 0, 0, 0];
    var index = 0;
    var rest = Math.floor(timestamp / ratios[index]);
    for (index = 1; index <= maxIndex && rest > 0; index++) {
        var ratio = ratios[index];
        arr[index - 1] = rest % ratio;
        rest = Math.floor(rest / ratio);
    }
    arr[index - 1] = rest;
    var years = arr[5];
    var months = arr[4];
    var days = arr[3];
    var hours = arr[2];
    var minutes = arr[1];
    var seconds = arr[0];
    function format(pattern) {
        if (pattern === void 0) { pattern = defaultFormatString; }
        return pattern.replace(patternRegExp, function (p) {
            switch (p) {
                case 'YYYY':
                    return "".concat(years);
                case 'MM':
                    return padStartZero(months);
                case 'DD':
                    return padStartZero(days);
                case 'hh':
                    return padStartZero(hours);
                case 'mm':
                    return padStartZero(minutes);
                case 'ss':
                    return padStartZero(seconds);
                case 'M':
                    return padStartZero(months, false);
                case 'D':
                    return padStartZero(days, false);
                case 'h':
                    return padStartZero(hours, false);
                case 'm':
                    return padStartZero(minutes, false);
                case 's':
                    return padStartZero(seconds, false);
                default:
                    return '';
            }
        });
    }
    function parse() {
        return {
            years: years,
            months: months,
            days: days,
            hours: hours,
            minutes: minutes,
            seconds: seconds,
            millseconds: millseconds,
            monthsPad: padStartZero(months),
            daysPad: padStartZero(days),
            hoursPad: padStartZero(hours),
            minutesPad: padStartZero(minutes),
            secondsPad: padStartZero(seconds),
        };
    }
    return {
        /**
         * 获取解析后的年月日时分秒数据
         * @returns
         */
        parse: parse,
        /**
         * 格式化
         * @param pattern 模板
         * @returns
         */
        format: format,
    };
}

module.exports = deadline;
