// @krn/live-bridge/src/common/index.ts
import {
    getTheme,
    getLanguage,
    addIOSThemeChange,
    getLocalizedString,
    ELanguageType,
    ETheme,
} from './env';
import { request } from './network';
import {
    hasInstalledApp,
    APP_PKG_NAME_MAP,
    open,
    openWithInfo,
    mobileQuickLoginInfo,
} from './utils';
import { reportPageRenderTime, reportAppDidMount } from './performance';
import { setStorage, getStorage } from './storage';
import { selectCity, showToast } from './ui';
import type { IRequestParams } from './network';
import type { IBridgeResult, IInitAPPProps } from './interface';
import type {
    TIOSThemeChangeListener,
    TIOSThemeChangeListenerRemoveFn,
} from './env';
import type { IGetStorageParam, ISetStorageParam } from './storage';
import type {
    ISelectCityResult,
    ISelectCityParams,
    IShowToastParams,
} from './ui';
import type { IOpenWithInfoParams, IMobileQuickLoginInfoResult } from './utils';

export {
    request,
    getTheme,
    getLanguage,
    addIOSThemeChange,
    getLocalizedString,
    hasInstalledApp,
    reportAppDidMount,
    reportPageRenderTime,
    setStorage,
    getStorage,
    selectCity,
    showToast,
    open,
    openWithInfo,
    mobileQuickLoginInfo,
    APP_PKG_NAME_MAP,
    ETheme,
    ELanguageType,
};

export type {
    IRequestParams,
    TIOSThemeChangeListener,
    TIOSThemeChangeListenerRemoveFn,
    IBridgeResult,
    IInitAPPProps,
    IGetStorageParam,
    ISetStorageParam,
    ISelectCityResult,
    ISelectCityParams,
    IShowToastParams,
    IOpenWithInfoParams,
    IMobileQuickLoginInfoResult,
};
