// @krn/live-bridge/src/common/ui.ts
import { IBridgeResult } from './interface';
import { bridgeCenterExecute } from '../utils/bridgeCenterExecute';

export interface ISelectCityResult extends IBridgeResult {
    data?: {
        /**
         * 城市编码
         */
        cityCode?: string;
        /**
         * 城市名称
         */
        cityName?: string;
        /**
         * 城市编码
         */
        code?: string;
        /**
         * 省份名称
         */
        provinceName?: string;
    };
}

export interface ISelectCityParams {
    /**
     * 默认城市
     */
    defaultCityCode?: string;
}

export interface IShowToastParams {
    type?: 'success' | 'normal' | 'error';
    isAddToWindow?: boolean; // 用于控制9.9.30之后toast在安卓上的渲染层级
    text: string;
}

/**
 * 城市选择
 * @param params
 */
export function selectCity(
    params?: ISelectCityParams,
): Promise<ISelectCityResult> {
    const paramStr = params ? JSON.stringify(params) : '';
    return bridgeCenterExecute('ui', 'selectCity', paramStr);
}

/**
 * 弹 Toast
 * @param params
 */
export function showToast(params: IShowToastParams): void {
    const paramStr = params ? JSON.stringify(params) : '';
    // toast 实测不会有 resolve
    bridgeCenterExecute('ui', 'showToast', paramStr)
        .then((res) => {})
        .catch((error) => {});
}
