// @krn/live-bridge/src/common/utils.ts
import { NativeModules, Platform } from 'react-native';
import { Result } from 'kdsbridge/types/KRNBasic';
import { bridgeCenterExecute } from '../utils/bridgeCenterExecute';

export interface IOpenWithInfoParams {
    /**
     * 跳转的 url
     */
    url: string;
    /**
     * 0 不带动画，1 带动画 默认 1
     */
    animate?: 0 | 1;
    /**
     * 辅助精确定位当前 View
     */
    reactTag?: number;
}
export interface IMobileQuickLoginInfoResult {
    /**
     * 结果
     */
    result: number;
    /**
     * 本机号码（带遮掩码）
     */
    phoneNumber: string;
    /**
     * 运营商类型：0：未知；1：中国移动；2：中国联通；3：中国电信；9：其他运营商
     */
    operatorType: 0 | 1 | 2 | 3 | 9;
    /**
     * 运营商token
     */
    token: string;
    /**
     * 运营商accessCode
     */
    accessCode?: string;
}
/**
 * 市场上常用 App 的包名
 */
export const APP_PKG_NAME_MAP = {
    WECHAT: Platform.select({
        android: 'com.tencent.mm',
        ios: 'wechat',
    }) as string,
    QQ: Platform.select({
        android: 'com.tencent.mobileqq',
        ios: 'mqq',
    }) as string,
    ALIPAY: Platform.select({
        android: 'com.eg.android.AlipayGphone',
        ios: 'alipay',
    }) as string,
};

/**
 * 判断是否安装某个 APP
 * @param pkgName {@link APP_PKG_NAME_MAP}
 */
export function hasInstalledApp(
    pkgName: string | keyof typeof APP_PKG_NAME_MAP,
): Promise<boolean> {
    return new Promise((resolve, reject) => {
        if (typeof NativeModules?.KRNBasic?.hasInstalledApp === 'function') {
            // @ts-ignore
            NativeModules.KRNBasic.hasInstalledApp(pkgName, (result) => {
                if (result.result === 1) {
                    resolve(true);
                } else {
                    resolve(false);
                }
            });
        } else {
            reject(
                new Error(
                    ' NativeModules.KRNBasic.hasInstalledApp not function',
                ),
            );
        }
    });
}

/**
 * 打开页面
 * @param url
 * @param callback
 */
export function open(url: string, callback?: (r: Result) => void) {
    return new Promise((resolve, reject) => {
        if (typeof NativeModules?.KRNBasic?.open === 'function') {
            NativeModules.KRNBasic.open(url, callback || ((r: Result) => {}));
        } else {
            reject(new Error(' NativeModules.KRNBasic.open not function'));
        }
    });
}

/**
 * 页面跳转
 * @param params
 * @param callback
 */
export function openWithInfo(params: IOpenWithInfoParams) {
    return new Promise((resolve, reject) => {
        if (typeof NativeModules?.KRNBasic?.openWithInfo === 'function') {
            NativeModules.KRNBasic.openWithInfo(params);
        } else {
            reject(new Error(' NativeModules.KRNBasic.open not function'));
        }
    });
}

/**
 * 获取一键登录信息
 * https://docs.corp.kuaishou.com/k/home/VIeHQqc-L2PU/fcADhAQXiOv6yTYsms8_5cJGN
 */
export function mobileQuickLoginInfo(): Promise<IMobileQuickLoginInfoResult> {
    return bridgeCenterExecute('social', 'mobileQuickLoginInfo');
}
