// @locallife/biz-hooks/src/useOnTextLayout/demo/index.tsx
import React from 'react';
import { Text } from 'react-native';
import useOnTextLayout from '..';
import { StyleSheet } from 'react-native';

const textStyle = StyleSheet.create({
    text: {
        fontSize: 14,
        color: 'black',
        marginBottom: 20,
    },
    originText: {
        fontSize: 20,
        color: 'red',
    },
    hasNextLine: {
        fontSize: 20,
        color: 'orange',
    },
    singleLine: {
        fontSize: 20,
        color: 'blue',
    },
    nextLine: {
        fontSize: 20,
        color: 'green',
    },
});

export default () => {
    const text: string =
        '鲁迅（1881-1936），原名周樟寿，后改名周树人，字豫才，浙江绍兴人。20世纪中国重要作家，新文化运动的领导人、左翼文化运动的支持者。毛泽东评价他是伟大的无产阶级的文学家、思想家、革命家，是中国文化革命的主将，也被人民称为“民族魂”。';

    const [textInfo, onTextLayout] = useOnTextLayout(text);

    return (
        <>
            <Text style={textStyle.originText}>展示原文</Text>
            <Text
                style={textStyle.text}
                // @ts-ignore
                onTextLayout={onTextLayout}
            >
                {textInfo.text}
            </Text>

            <Text style={textStyle.hasNextLine}>是否有下一行</Text>
            <Text style={textStyle.text}>
                {textInfo.hasNextLine ? '有下一行' : '没下一行'}
            </Text>

            <Text style={textStyle.singleLine}>展示一行</Text>
            <Text style={textStyle.text}>{textInfo.firstLineText}</Text>

            <Text style={textStyle.nextLine}>展示剩余部分</Text>
            <Text style={textStyle.text}>{textInfo.nextLineText}</Text>
        </>
    );
};
