// @kds/refresh-list/dist/constant.js
import React, { Component } from 'react';
import { Text } from 'react-native';
export var PullState;
(function (PullState) {
    PullState[PullState["INIT"] = 0] = "INIT";
    PullState[PullState["FINISH"] = 1] = "FINISH";
    PullState[PullState["PULLING"] = 2] = "PULLING";
    PullState[PullState["LOADING"] = 3] = "LOADING";
    PullState[PullState["HOLDING"] = 4] = "HOLDING";
})(PullState || (PullState = {}));
export var PushingState;
(function (PushingState) {
    PushingState[PushingState["INIT"] = 1] = "INIT";
    PushingState[PushingState["PREPARE"] = 2] = "PREPARE";
    PushingState[PushingState["LOADING"] = 3] = "LOADING";
    PushingState[PushingState["COMPLETE"] = 4] = "COMPLETE";
})(PushingState || (PushingState = {}));
export class DefaultRefreshControl extends Component {
    constructor(props) {
        super(props);
        this.state = {
            text: '下拉刷新',
        };
    }
    setRefreshControlParams(refreshControlParams) {
        if (refreshControlParams.pullState === PullState.INIT) {
            this.setState({
                text: '下拉刷新',
            });
        }
        else if (refreshControlParams.pullState === PullState.PULLING) {
            this.setState({
                text: refreshControlParams.movePercent >= 100 ? '松开加载' : '下拉刷新',
            });
        }
        else if (refreshControlParams.pullState === PullState.LOADING) {
            this.setState({
                text: '拼命加载中...',
            });
        }
        else if (refreshControlParams.pullState === PullState.HOLDING) {
            this.setState({
                text: '加载成功',
            });
        }
    }
    render() {
        return <Text>{this.state.text}</Text>;
    }
}
