// @locallife/biz-component/lib/commonjs/components/packageDetail/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _tsLbsKrnUtils = require("@lux/ts-lbs-krn-utils");
var _reactNative = require("react-native");
var _config = require("../../utils/config");
const {
  screenSize
} = _tsLbsKrnUtils.screenUnit;
var _default = _reactNative.StyleSheet.create({
  container: {
    display: 'flex',
    paddingHorizontal: screenSize(_config.PRODUCT_DETAIL_PAGE_HORIZONTAL_PADDING),
    position: 'relative',
    paddingVertical: screenSize(_config.PRODUCT_DETAIL_COMPONENT_HORIZONTAL_PADDING),
    backgroundColor: '#ffffff',
    marginTop: screenSize(_config.PRODUCT_DETAIL_COMPONENT_MARGIN_GAP),
    borderRadius: screenSize(16)
  },
  title: {
    fontSize: screenSize(18),
    fontFamily: 'PingFangSC-Medium',
    fontWeight: '500',
    color: '#222222',
    lineHeight: screenSize(25)
  },
  subtitle: {
    fontSize: screenSize(16),
    fontFamily: 'PingFang SC',
    fontWeight: '500',
    color: '#222222',
    lineHeight: screenSize(20),
    paddingLeft: screenSize(15)
  },
  subContent: {
    fontSize: screenSize(14),
    lineHeight: screenSize(18),
    color: '#666666',
    marginTop: screenSize(12)
  },
  subContentDistance: {
    marginLeft: screenSize(12)
  },
  subContentLine: {
    paddingBottom: screenSize(13),
    borderBottomWidth: screenSize(0.5),
    borderBottomColor: '#E6E6E6'
  },
  subContentListDistance: {
    paddingLeft: screenSize(12)
  },
  subGroupsLine: {
    marginTop: screenSize(13),
    marginLeft: screenSize(12),
    height: screenSize(0.1),
    backgroundColor: '#ECECEC'
  },
  packageDetailItemWrapper: {
    marginTop: screenSize(15)
  },
  packageDetailItemDistance: {
    marginTop: screenSize(6),
    marginLeft: screenSize(15)
  },
  packageDetailItemRow: {
    width: '100%',
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  packageDetailRowTitleWrapper: {
    flexShrink: 1,
    marginRight: screenSize(24)
  },
  packageDetailTitle: {
    width: screenSize(237)
  },
  packageDetailRowTitle: {
    fontSize: screenSize(16),
    fontWeight: '400',
    lineHeight: screenSize(22),
    flexShrink: 1,
    color: '#666'
  },
  packageDetailRowNum: {
    fontSize: screenSize(16),
    color: '#9c9c9c'
  },
  packageDetailRowPriceAll: {
    textAlign: 'right',
    justifyContent: 'flex-end',
    width: screenSize(53)
  },
  packageDetailRowPrice: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    fontSize: screenSize(16),
    fontWeight: '400',
    color: '#222'
  },
  packageDetailRowPriceUnit: {
    fontSize: screenSize(12),
    fontWeight: '500',
    textAlign: 'right',
    marginRight: screenSize(2),
    color: '#222'
  },
  rowFlexAlignCenter: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center'
  },
  packageDetailPaddingWrapper: {
    paddingLeft: screenSize(18)
  },
  priceItemContianer: {
    marginTop: screenSize(20)
  },
  lookMore: {
    fontSize: screenSize(16),
    color: '#666666'
  },
  priceItemArrow: {
    marginTop: screenSize(12)
  },
  moreIcon: {
    width: screenSize(14),
    height: screenSize(14)
  },
  graphicDetailDescContainer: {
    marginTop: screenSize(12)
  },
  graphicDetailImg: {
    borderRadius: screenSize(6),
    overflow: 'hidden',
    marginTop: screenSize(16)
  },
  graphicDetailText: {
    fontSize: screenSize(14),
    lineHeight: screenSize(17),
    fontWeight: '400',
    color: '#222222',
    textAlign: 'justify',
    marginVertical: screenSize(2)
  },
  priceSubtitle: {
    marginBottom: 0
  },
  customizationUlPoint: {
    backgroundColor: '#C6C6C6'
  }
});
exports.default = _default;
