// @locallife/biz-component/src/utils/config.ts
export const PRODUCT_DETAIL_COMPONENT_MARGIN_GAP = 12;
export const PRODUCT_DETAIL_PAGE_HORIZONTAL_MARGIN = 11;
export const PRODUCT_DETAIL_PAGE_HORIZONTAL_PADDING = 12;
export const PRODUCT_DETAIL_COMPONENT_HORIZONTAL_PADDING = 14;

// 退款规则
export enum REFUNDED_RULE {
    NO_WRITE_OFF_REFUND_ANY_TIME = '未使用随时退、过期自动退',
    CAN_NOT_REFUND = '不可退',
    WITH_CONDITION_REFUND = '有条件退，具体规则请咨询客服',
    EFFECTIVE_NOT_REFUND_EXPIRE_AUTO_REFUND = '有效期内不支持发起退款，过期未使用，平台将自动退款',
}

// 限购规则文案映射关系
export const BUY_LIMIT_RULE = {
    UNLIMITED_PURCHASE: [
        {
            key: 'onLimit',
            text: '不限购',
        },
    ],
    SINGLE_ORDER_AMOUNT: [
        {
            key: 'singleOrderLimitNum',
            text: '每笔订单限购{num}份',
        },
    ],
    ORDER: [
        {
            key: 'orderLeastNum',
            text: '每笔订单最少购买{num}份，',
        },
        {
            key: 'orderMostNum',
            text: '最多购买{num}份',
        },
    ],
    MOBILE_PHONE: [
        {
            key: 'phoneInDay',
            text: '同一手机号在{num}天内，',
        },
        {
            key: 'phoneMostNum',
            text: '最多购买{num}份',
        },
    ],
    CARD_ID: [
        {
            key: 'identityCardInDay',
            text: '同一身份证号在{num}天内，',
        },
        {
            key: 'identityCardMostNum',
            text: '最多购买{num}份',
        },
    ],
    ACCOUNT: [
        {
            key: 'accountInDay',
            text: '同一账号{num}天内，',
        },
        {
            key: 'accountMostNum',
            text: '最多购买{num}份',
        },
    ],
    MOBILE_PHONE_NO_TIME: [
        {
            key: 'phoneNoTimeLimitNum',
            text: '每个手机号最多购买{num}份',
        },
    ],
    MOBILE_PHONE_PLAY_TIME: [
        {
            key: 'phonePlayTimeLimitNum',
            text: '每个手机号游玩日期最多购买{num}份',
        },
    ],
    CARD_ID_PLAY_TIME: [
        {
            key: 'idCardPlayTimeLimitNum',
            text: '每个身份证游玩日期最多购买{num}份',
        },
    ],
    ACCOUNT_NO_TIME: [
        {
            key: 'accountNoTimeLimitNum',
            text: '每个账号最多购买{num}份',
        },
    ],
};
