// @kid-ui/krn/src/components/toast/index.tsx
import React from 'react';
import { Toast as toast, ToastProps as BaseProps  } from './Toast';
import { KidLoading } from '../loading';
import { KidToastComponent, ToastProps as baseProps } from './toast-component';

export interface ToastProps extends BaseProps, baseProps {}

export const Toast = {
    ...toast,
    loading(opts = {}) {
        const defaultOpts = {
            position: toast.position.CENTER,
            data: '加载中...',
            icon: <KidLoading type={'#ffffff'} style={{marginBottom: 13}}/>,
            backgroundColor: 'rgba(0, 0, 0, 0.7)',
            contentStyle: { width: 110, height: 110, borderRadius: 8, padding: 0},
            isShowShadow: false,
            duration: 0,
            showMask: true,
        };
        return toast.show(Object.assign({}, defaultOpts, opts));
    },
    kidShow(opts: ToastProps = {}) {
        const {
            toastStyle = {},
            text = '',
            textStyle = {},
            kid = 'style_toast_text',
            percent = 0,
            getPercentText,
            ...rest
        } = opts;
        const defaultProps = {
            toastStyle,
            text,
            textStyle,
            kid,
            percent,
            getPercentText
        };
        const defaultOpts: BaseProps = {
            position: toast.position.CENTER,
            backgroundColor: 'transparent',
            isShowShadow: false,
            duration: 0,
            icon: <KidToastComponent {...defaultProps}/>,
            showMask: true,
        };
        return toast.show(Object.assign({}, defaultOpts, rest));
    },
};
