// @locallife/biz-component/lib/module/components/contactPopup/index.js
import React from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import { KidIcon } from '@kid-ui/krn';
import styles from './styles';
import { screenUnit, bridge } from '@lux/ts-lbs-krn-utils';
const {
  screenSize
} = screenUnit;
const {
  makePhoneCall
} = bridge;
export const ContactPopup = _ref => {
  let {
    phoneList
  } = _ref;
  /**
   * 拨打商家电话
   */
  const callSeller = phone => {
    makePhoneCall([phone]);
  };
  /**
   * @param item 当前这条信息
   * @returns 是否只有一条电话数据
   */
  const isOnlyPhone = item => {
    return typeof item === 'string' || typeof item === 'number';
  };
  /**
   * 处理点击
   * @param item 当前这条信息
   */
  const handlePress = item => {
    var _item$custom;
    if (item !== null && item !== void 0 && (_item$custom = item.custom) !== null && _item$custom !== void 0 && _item$custom.customPress) {
      var _item$custom2;
      item === null || item === void 0 ? void 0 : (_item$custom2 = item.custom) === null || _item$custom2 === void 0 ? void 0 : _item$custom2.customPress();
    } else {
      var _item$custom3, _item$custom3$phonePr;
      callSeller(isOnlyPhone(item) ? item : item.phone);
      item === null || item === void 0 ? void 0 : (_item$custom3 = item.custom) === null || _item$custom3 === void 0 ? void 0 : (_item$custom3$phonePr = _item$custom3.phonePress) === null || _item$custom3$phonePr === void 0 ? void 0 : _item$custom3$phonePr.call(_item$custom3);
    }
  };
  return /*#__PURE__*/React.createElement(View, {
    style: styles.popupContent
  }, phoneList.map((item, index) => {
    var _item$custom4;
    return /*#__PURE__*/React.createElement(TouchableOpacity, {
      key: index,
      activeOpacity: 1,
      onPress: () => handlePress(item)
    }, index !== 0 && /*#__PURE__*/React.createElement(View, {
      style: styles.dividingLine
    }), /*#__PURE__*/React.createElement(View, {
      style: styles.popupItem
    }, /*#__PURE__*/React.createElement(View, {
      style: styles.popupItemContent
    }, !isOnlyPhone(item) && !!(item !== null && item !== void 0 && item.title) && /*#__PURE__*/React.createElement(Text, {
      style: styles.popupItemDesc
    }, item === null || item === void 0 ? void 0 : item.title, ' '), /*#__PURE__*/React.createElement(Text, {
      style: styles.popupItemDesc
    }, isOnlyPhone(item) ? item : item === null || item === void 0 ? void 0 : item.phone)), /*#__PURE__*/React.createElement(View, {
      style: styles.popupItemIcon
    }, /*#__PURE__*/React.createElement(KidIcon, {
      kid: (item === null || item === void 0 ? void 0 : (_item$custom4 = item.custom) === null || _item$custom4 === void 0 ? void 0 : _item$custom4.icon) || 'common_baseflat_telephone_24',
      kidColor: 'color_base_black_8',
      size: screenSize(24)
    }))));
  }));
};
