// @locallife/biz-hooks/lib/commonjs/useOnDarkModeChange/demo/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = _reactNative.StyleSheet.create({
  container: {
    alignItems: 'center'
  },
  row: {
    flexDirection: 'row'
  },
  smallText: {
    fontSize: 13
  },
  bigText: {
    fontSize: 18
  },
  red: {
    color: 'red'
  },
  margin: {
    marginTop: 20
  }
});

var _default = () => {
  const {
    changing,
    isDarkMode
  } = (0, _.default)(isDark => {
    console.log(`isDark = ${isDark}`);
  });
  console.log(`changing = ${changing}`);
  let result;

  switch (isDarkMode) {
    case true:
      result = '是';
      break;

    case false:
      result = '否';
      break;

    default:
      result = '未知';
      break;
  }

  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.row
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.bigText
  }, "\u5F53\u524D\u662F\u5426\u662F\u6DF1\u8272\u6A21\u5F0F\uFF1A"), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.bigText, styles.red]
  }, result)), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.smallText, styles.margin]
  }, "\u8BF7\u5230\u8BBE\u7F6E\u4E2D\u5207\u6362\u6DF1\u8272\u6A21\u5F0F"));
};

exports.default = _default;
