// @locallife/biz-hooks/src/useOnDarkModeChange/demo/index.tsx
import React from 'react';
import { Text, View, StyleSheet } from 'react-native';
import useOnDarkModeChange from '..';

const styles = StyleSheet.create({
    container: {
        alignItems: 'center',
    },
    row: {
        flexDirection: 'row',
    },
    smallText: {
        fontSize: 13,
    },
    bigText: {
        fontSize: 18,
    },
    red: {
        color: 'red',
    },
    margin: {
        marginTop: 20,
    },
});

export default () => {
    const { changing, isDarkMode } = useOnDarkModeChange((isDark: boolean) => {
        console.log(`isDark = ${isDark}`);
    });
    console.log(`changing = ${changing}`);
    let result;
    switch (isDarkMode) {
        case true:
            result = '是';
            break;
        case false:
            result = '否';
            break;
        default:
            result = '未知';
            break;
    }
    return (
        <View style={styles.container}>
            <View style={styles.row}>
                <Text style={styles.bigText}>当前是否是深色模式：</Text>
                <Text style={[styles.bigText, styles.red]}>{result}</Text>
            </View>
            <Text style={[styles.smallText, styles.margin]}>
                请到设置中切换深色模式
            </Text>
        </View>
    );
};
