// @krn/cli/lib/commands/update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./../utils");
const cli_shared_1 = require("@krn/cli-shared");
const Listr = require("listr");
const command_1 = require("./command");
class Update extends command_1.default {
    async run() {
        this.runAnalytics({
            id: this.id,
            CLIVersion: this.config.version,
        });
        if (cli_shared_1.isInstalledGlobally) {
            await new Listr([
                {
                    title: '更新CLI版本',
                    task: (_, task) => {
                        return (0, utils_1.execAsync)('yarn global add @krn/cli --registry=https://npm.corp.kuaishou.com').then(() => {
                            task.title = '升级成功';
                        });
                    },
                },
            ]).run();
        }
        else {
            utils_1.Logger.error('本地命令不能使用该命令，请使用 yarn add @krn/cli -D');
        }
    }
}
exports.default = Update;
Update.hidden = false;
Update.description = '升级CLI';
Update.examples = [`krn update`];
