// @kid-ui/krn/lib/components/pull-refresh/refresh-control.js
// 先放这，之后再迁移
import React, { Component } from 'react';
import { PullState } from '@kds/refresh-list';
import { KidLoading, LoadingType } from '../loading';
export default function getKsRefreshControl(type = LoadingType.highlight) {
    return class KsRefreshControl extends Component {
        constructor(props) {
            super(props);
            this.state = {
                animated: false,
                percent: 0,
            };
        }
        setRefreshControlParams(refreshControlParams) {
            if (refreshControlParams.pullState === PullState.PULLING) {
                this.setState({
                    animated: false,
                    percent: (Math.max(refreshControlParams.movePercent - 50, 0) / (100 - 50)) * 100,
                });
            }
            else if (refreshControlParams.pullState === PullState.LOADING) {
                this.setState({
                    animated: true,
                });
            }
            else if (refreshControlParams.pullState === PullState.HOLDING) {
                this.setState({
                    animated: true,
                });
            }
        }
        render() {
            const { animated, percent } = this.state;
            return <KidLoading type={type} animated={animated} percent={percent}/>;
        }
    };
}
