// @lux/ts-lbs-krn-utils/base/src/screenUnit.ts
import { Dimensions, PixelRatio } from 'react-native';

export const DeviceWidth = Dimensions.get('window').width;

export const DeviceHeight = Dimensions.get('window').height;

export const FontScale = PixelRatio.getFontScale();

export const pixelRatio = PixelRatio.get();

export function screenSize(size: number, ceil = false, designW = 414) {
    const isMinus = size < 0;
    if (size === 0) {
        return 0;
    }

    // 换算到屏幕物理像素点
    const physicalPixels =
        (size * PixelRatio.getPixelSizeForLayoutSize(DeviceWidth)) / designW;

    // 物理像素点取整，最小返回 1 像素
    const roundPhysicalPixels = Math.max(
        1,
        (ceil ? Math.ceil : Math.floor)(
            isMinus ? -physicalPixels : physicalPixels,
        ),
    );

    // 换算到逻辑像素点
    const pixels = (roundPhysicalPixels / pixelRatio).toFixed(1);
    return isMinus ? -pixels : +pixels;
}

export function screenSize375(size: number, ceil = false) {
    return screenSize(size, ceil, 375);
}

export function screenSize414(size: number, ceil = false) {
    return screenSize(size, ceil, 414);
}
