// @locallife/biz-component/src/components/purchaseInfo/PricePopup.tsx
import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { screenUnit } from '@lux/ts-lbs-krn-utils';
import CommonText from '../commonText';

const { screenSize } = screenUnit;

const styles = StyleSheet.create({
    contentContainer: {
        paddingTop: screenSize(20),
        marginBottom: screenSize(24),
        paddingHorizontal: screenSize(19),
        borderTopColor: '#eaeaea',
        borderTopWidth: screenSize(0.5),
    },
    descContainer: {
        marginTop: screenSize(18),
        marginLeft: screenSize(10),
    },
    title: {
        fontSize: screenSize(14),
        lineHeight: screenSize(20),
        fontWeight: '500',
        color: '#222222',
        marginLeft: screenSize(10),
    },
    desc: {
        fontSize: screenSize(14),
        lineHeight: screenSize(20),
        fontWeight: '400',
        color: '#666666',
        textAlign: 'justify',
        marginTop: screenSize(6),
        marginLeft: screenSize(10),
    },
});

interface PopupParams {
    visible: boolean;
    onClose: () => void;
}

export const PricePopup = () => {
    return (
        <View style={styles.contentContainer}>
            <View style={[styles.descContainer, { marginTop: 0 }]}>
                <CommonText>
                    <Text style={styles.title}>团购价及其他未划线展示价格</Text>
                </CommonText>
                <Text style={styles.desc}>
                    商家在本平台展示的商品/服务的售价，如存疑请先咨询提供价格的商家。实际成交价格可能会因为使用优惠券而发生变化。
                </Text>
            </View>
            <View style={styles.descContainer}>
                <CommonText>
                    <Text style={styles.title}>划线价/门市价/门店价</Text>
                </CommonText>
                <Text style={styles.desc}>
                    商家/服务在特定时间、地域、实体门店或网络平台销售的吊牌价、专柜标价、厂商建议零售价，由于市场行情波动，该价格仅供参考。
                </Text>
            </View>
        </View>
    );
};
