// @locallife/biz-hooks/lib/commonjs/useABTest/demo/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _index = _interopRequireDefault(require("../index"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    width: '100%',
    display: 'flex',
    alignItems: 'center'
  },
  text: {
    backgroundColor: '#aaa',
    color: '#ffffff',
    textAlign: 'center',
    width: '100%',
    height: 50,
    lineHeight: 50,
    marginBottom: 10
  },
  bottom: {
    display: 'flex',
    flexDirection: 'row'
  }
});

const ABKey = 'highlight_equity_card_word_client';

var _default = () => {
  const display = (0, _index.default)(ABKey);
  return /*#__PURE__*/_react.default.createElement(_reactNative.SafeAreaView, {
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.text
  }, "key\uFF1A", ABKey), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.text
  }, "value\uFF1A", String(display)));
};

exports.default = _default;
