// @locallife/biz-hooks/lib/module/useCountdown/demo/index.js
/* eslint-disable no-alert */
import React from 'react';
import useCountdown from '..';
import { Text, StyleSheet, View } from 'react-native';
const textStyle = StyleSheet.create({
  title: {
    fontSize: 18,
    color: 'blue'
  },
  margin: {
    marginTop: 20
  }
});
export default (() => {
  const {
    formattedRes: formattedResult1
  } = useCountdown({
    endTime: Date.now() + 86400000
  });
  const {
    formattedRes: formattedResult2
  } = useCountdown({
    // 指定更新的时间
    endTime: 1704038399999,
    useDay: true
  });
  const {
    formattedRes: formattedResult3
  } = useCountdown({
    endTime: Date.now() + 10000,
    onEnd: () => {
      alert('倒计时结束');
    }
  });
  const {
    formattedRes: formattedResult4
  } = useCountdown({
    endTime: 1704038399999,
    useDay: true,
    useServerTime: true
  });
  return /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(Text, {
    style: textStyle.title
  }, "\u5012\u8BA1\u65F6\u4E00\u5929"), /*#__PURE__*/React.createElement(Text, null, formattedResult1.hour, " \u65F6 ", formattedResult1.minute, " \u5206", ' ', formattedResult1.second, " \u79D2"), /*#__PURE__*/React.createElement(Text, {
    style: [textStyle.title, textStyle.margin]
  }, "\u5012\u8BA1\u65F6\u52302023\u5E74\u5E74\u5E95"), /*#__PURE__*/React.createElement(Text, null, formattedResult2.day, " \u5929 ", formattedResult2.hour, " \u65F6", ' ', formattedResult2.minute, " \u5206 ", formattedResult2.second, " \u79D2"), /*#__PURE__*/React.createElement(Text, {
    style: [textStyle.title, textStyle.margin]
  }, "\u5012\u8BA1\u65F6\u7ED3\u675F\u76D1\u542C"), /*#__PURE__*/React.createElement(Text, null, formattedResult3.second, " \u79D2 "), /*#__PURE__*/React.createElement(Text, {
    style: [textStyle.title, textStyle.margin]
  }, "\u6D4B\u8BD5\u670D\u52A1\u5668\u65F6\u95F4\uFF0C\u5728\u8BBE\u7F6E\u4E2D\u8C03\u6574\u624B\u673A\u65F6\u95F4\u4E0D\u751F\u6548"), /*#__PURE__*/React.createElement(Text, null, formattedResult4.day, " \u5929 ", formattedResult4.hour, " \u65F6", ' ', formattedResult4.minute, " \u5206 ", formattedResult4.second, " \u79D2"));
});
