// @locallife/biz-hooks/lib/commonjs/useSafeMemo/demo/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _index = _interopRequireDefault(require("../index"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* eslint-disable react/jsx-no-bind */
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    width: '100%',
    display: 'flex',
    alignItems: 'center'
  },
  text: {
    backgroundColor: '#aaa',
    color: '#ffffff',
    textAlign: 'center',
    width: '100%',
    height: 50,
    lineHeight: 50,
    marginBottom: 10
  },
  bottom: {
    display: 'flex',
    flexDirection: 'row'
  }
});

var _default = () => {
  const [counter, setCounter] = (0, _react.useState)(0);
  const display = (0, _index.default)(() => {
    return counter * 10;
  }, [counter]);
  return /*#__PURE__*/_react.default.createElement(_reactNative.SafeAreaView, {
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.text
  }, "\u503C", display), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.bottom
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Button, {
    title: "\u52A010",
    onPress: () => {
      setCounter(counter + 1);
    }
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Button, {
    title: "\u51CF10",
    onPress: () => {
      setCounter(counter - 1);
    }
  })));
};

exports.default = _default;
