// @locallife/biz-component/lib/commonjs/components/newUserZoneEntrance/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactNative = require("react-native");
var _default = _reactNative.StyleSheet.create({
  rootContainer: {
    width: 392,
    height: 84,
    overflow: 'hidden',
    borderRadius: 16
  },
  bgStyle: {
    width: '100%',
    height: '100%'
  },
  disCountArea: {
    flexDirection: 'row',
    marginTop: 8,
    marginLeft: 25,
    justifyContent: 'flex-start',
    alignItems: 'flex-end'
  },
  disCountContentType: {
    fontFamily: 'Avenir Next',
    fontWeight: '700',
    fontSize: 20,
    lineHeight: 24,
    color: '#000000',
    ..._reactNative.Platform.select({
      android: {
        marginBottom: 4
      },
      ios: {
        marginBottom: 3
      }
    })
  },
  disCountContent: {
    fontFamily: 'Avenir Next',
    fontWeight: '700',
    fontSize: 36,
    lineHeight: 44,
    color: '#000000',
    ..._reactNative.Platform.select({
      android: {
        marginBottom: 0
      },
      ios: {
        marginBottom: -4
      }
    })
  },
  disCountTitleImg: {
    marginLeft: 3,
    height: 21,
    width: 134,
    marginBottom: 9
  }
});
exports.default = _default;
