// @krn/live-bridge/src/liveCommon/interface.ts
import { IBridgeResult } from '../common';

export interface IDispatchLiveRouterResult extends IBridgeResult {
    /**
     * 失败原因
     */
    failReason: string;
}

export interface IDispatchLiveRouterParams {
    /**
     * 直播内部路由 scheme，具体参加直播路由表
     */
    path: string;
    /**
     * 跳转后是否展示本容器，false 会关闭
     */
    keepDisplayWebView: boolean;
}

export interface ILiveRouterParams {
    /**
     * 路由 url，详情参考下方描述中的路由表链接
     */
    url: string;
    /**
     * 是否需要关闭当前页, 默认为不关闭
     */
    dismiss?: boolean;
}

/**
 * 魔表资源下载优先级
 */
export enum MagicResPriority {
    /**
     * Background优先级
     */
    Background = 1,
    /**
     * Utility优先级
     */
    Utility = 2,
    /**
     * User_Init优先级
     */
    UserInit = 3,
    /**
     * 立即下载
     */
    Immediate = 4,
}

/**
 * 魔表资源参数
 */
export interface IGetMagicResParams {
    /**
     * 资源ID列表
     */
    resourceIdList: string[];
    /**
     * 资源中台分组
     */
    subBiz: string;
    /**
     * 资源下载优先级
     */
    priority: MagicResPriority;
}

/**
 * 魔表资源返回参数
 */
export interface IMagicResPathResult {
    /**
     * 是否全部下载完成
     */
    isAllSuccess: boolean;
    /**
     * 资源路径Map
     */
    resMap: object;
}

export interface ILiveBridgeResult<T> {
    /**
     1 : 成功
     -1 : 通用失败错误码
     -2 : API 不存在
     JS 侧需判断 result === 1，此时 data 字段可用
     */
    result: number;
    /**
     * 业务错误信息，可选
     */
    // eslint-disable-next-line camelcase
    error_msg: string;
    /**
     * 可选，默认为 null。如果有值，其结构定义见下方
     */
    data?: T;
}
