// @locallife/biz-hooks/lib/commonjs/useFeedMount/demo/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _2 = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// 本Demo展示了useFeedMount的使用，通过该Hook可以对列表每个item展现的上报打点进行控制，保证每个item只会上报一次展现点
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1
  },
  text: {
    width: '100%',
    height: 300,
    textAlign: 'center',
    lineHeight: 300,
    color: '#ffffff',
    fontSize: 22
  },
  blue: {
    backgroundColor: 'blue'
  },
  orange: {
    backgroundColor: 'orange'
  },
  purple: {
    backgroundColor: 'purple'
  },
  green: {
    backgroundColor: 'green'
  },
  bottom: {
    backgroundColor: 'black',
    color: '#ffffff',
    textAlign: 'center',
    lineHeight: 50
  }
});

class ItemModel {
  constructor() {
    _defineProperty(this, "data", void 0);

    _defineProperty(this, "index", void 0);
  }

}

var _default = () => {
  const onFeedMount = (0, _2.default)(item => item.index, item => {
    console.log(`第${item.index + 1}个文本 已展现, 上报该展现点`);
  });
  let data = Array(100).fill({
    data: '0',
    index: '0'
  });
  data = data.map((_, index) => {
    return {
      data: `第${index + 1}个文本`,
      index: index
    };
  });
  const backgroundArray = [styles.blue, styles.orange, styles.purple, styles.green];

  const renderItem = _ref => {
    let {
      index
    } = _ref;
    return /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
      style: [styles.text, backgroundArray[index % backgroundArray.length]]
    }, data[index].data);
  };

  const onViewableItemChange = (0, _react.useRef)(_ref2 => {
    let {
      changed
    } = _ref2;
    changed.forEach(viewToken => {
      if (viewToken.isViewable) {
        onFeedMount && onFeedMount(viewToken.item, viewToken.index);
        console.log(`第${viewToken.index + 1}个文本 is visible`);
      } else {
        console.log(`第${viewToken.index + 1}个文本 is not visible`);
      }
    });
  });
  return /*#__PURE__*/_react.default.createElement(_reactNative.SafeAreaView, {
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_reactNative.FlatList, {
    data: data,
    renderItem: renderItem,
    keyExtractor: (_, index) => `${index}`,
    onViewableItemsChanged: onViewableItemChange.current
  }));
};

exports.default = _default;
