// @locallife/biz-hooks/lib/commonjs/useLoading/demo/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _index = _interopRequireDefault(require("../index"));

var _reactNative = require("react-native");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const LoadingDemo = () => {
  const styles = _reactNative.StyleSheet.create({
    container: {
      width: '100%',
      padding: 20
    },
    itemWrapper: {
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'space-between',
      width: '100%',
      padding: 30,
      marginBottom: 20,
      backgroundColor: '#eee'
    },
    label: {
      color: '#999',
      position: 'absolute',
      left: 3,
      top: 3
    },
    button: {
      marginLeft: 30
    }
  });

  const {
    status: status1,
    withLoading: withLoading1,
    isLoading: isLoading1
  } = (0, _index.default)({
    onlyOnce: true
  });
  const {
    status: status2,
    withLoading: withLoading2,
    isLoading: isLoading2
  } = (0, _index.default)();
  const {
    status: status3,
    withLoading: withLoading3,
    isLoading: isLoading3
  } = (0, _index.default)({
    onlyOnce: true
  });

  const mainRequest = () => new Promise(res => {
    setTimeout(() => {
      res('');
    }, 3000);
  });

  const handleRefresh2 = (0, _react.useCallback)(() => {
    withLoading2(mainRequest);
  }, [withLoading2]);
  const handleRefresh3 = (0, _react.useCallback)(() => {
    withLoading3(mainRequest);
  }, [withLoading3]);
  const getStatus = (0, _react.useCallback)(status => {
    if (status === 0) {
      return 'initial';
    }

    if (status === 1) {
      return 'loading';
    }

    if (status === 2) {
      return 'success';
    }

    if (status === 3) {
      return 'fail';
    }
  }, []);
  (0, _react.useEffect)(() => {
    withLoading1(mainRequest); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.itemWrapper
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.label
  }, getStatus(status1)), /*#__PURE__*/_react.default.createElement(_reactNative.Text, null, "\u4E09\u79D2\u63A5\u53E3\uFF08\u81EA\u52A8\u8BF7\u6C42\uFF09\uFF1A", isLoading1 ? 'loading...' : 'success!')), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.itemWrapper
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.label
  }, getStatus(status2)), /*#__PURE__*/_react.default.createElement(_reactNative.Text, null, "\u4E09\u79D2\u63A5\u53E3\uFF1A", isLoading2 ? 'loading...' : 'success!'), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.button,
    onPress: handleRefresh2
  }, "\u5237\u65B0")), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.itemWrapper
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.label
  }, getStatus(status3)), /*#__PURE__*/_react.default.createElement(_reactNative.Text, null, "\u4E09\u79D2\u63A5\u53E3\uFF08\u4EC5load\u4E00\u6B21\uFF09\uFF1A", isLoading3 ? 'loading...' : 'success!'), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.button,
    onPress: handleRefresh3
  }, "\u5237\u65B0")));
};

var _default = LoadingDemo;
exports.default = _default;
