// @locallife/page/lib/commonjs/customQueryClient.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CustomQueryClient = void 0;

var _reactQuery = require("react-query");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const defaultQueryClient = new _reactQuery.QueryClient({
  defaultOptions: {
    queries: {
      cacheTime: 500,
      // 防止setTimeout频繁创建
      onError: error => {
        console.log(`QueryClient queries err=${JSON.stringify(error)}`);
      },
      retry: 0
    },
    mutations: {
      onError: error => {
        console.log(`QueryClient mutations err=${JSON.stringify(error)}`);
      },
      retry: 0
    }
  }
});

class CustomQueryClient {
  static set(client) {
    this.queryClient = client;
  }

  static get() {
    return this.queryClient;
  }

}

exports.CustomQueryClient = CustomQueryClient;

_defineProperty(CustomQueryClient, "queryClient", defaultQueryClient);
