// @locallife/page/lib/commonjs/errorScene.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _krn = require("@kid-ui/krn");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = _reactNative.StyleSheet.create({
  errorContainer: {
    position: 'absolute',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1
  }
}); // 错误页面


const ErrorScene = props => {
  const bgColor = (0, _krn.useKidColor)('cs_common_background_secondary');
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.errorContainer, {
      backgroundColor: bgColor
    }]
  }, /*#__PURE__*/_react.default.createElement(_krn.KidNetError, {
    onPress: props.onRetry,
    kid: "style_emptystate_default",
    emptyStyle: ""
  }));
};

var _default = ErrorScene;
exports.default = _default;
