// @locallife/page/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Alpha", {
  enumerable: true,
  get: function () {
    return _alpha.Alpha;
  }
});
Object.defineProperty(exports, "AlphaProps", {
  enumerable: true,
  get: function () {
    return _alpha.AlphaProps;
  }
});
Object.defineProperty(exports, "CustomQueryClient", {
  enumerable: true,
  get: function () {
    return _customQueryClient.CustomQueryClient;
  }
});
Object.defineProperty(exports, "ErrorScene", {
  enumerable: true,
  get: function () {
    return _errorScene.default;
  }
});
Object.defineProperty(exports, "ILCPImageContextProps", {
  enumerable: true,
  get: function () {
    return _pageContext.ILCPImageContextProps;
  }
});
Object.defineProperty(exports, "Lazy", {
  enumerable: true,
  get: function () {
    return _lazy.Lazy;
  }
});
Object.defineProperty(exports, "LoadingScene", {
  enumerable: true,
  get: function () {
    return _loadingScene.default;
  }
});
Object.defineProperty(exports, "OP_IMMEDIATE", {
  enumerable: true,
  get: function () {
    return _alpha.OP_IMMEDIATE;
  }
});
Object.defineProperty(exports, "OP_INLINE", {
  enumerable: true,
  get: function () {
    return _alpha.OP_INLINE;
  }
});
Object.defineProperty(exports, "OP_NEAREST_IDLE_INTERACTIVE", {
  enumerable: true,
  get: function () {
    return _alpha.OP_NEAREST_IDLE_INTERACTIVE;
  }
});
Object.defineProperty(exports, "OP_NEAREST_IDLE_JS", {
  enumerable: true,
  get: function () {
    return _alpha.OP_NEAREST_IDLE_JS;
  }
});
Object.defineProperty(exports, "OP_NEXT_DRAWFRAME", {
  enumerable: true,
  get: function () {
    return _alpha.OP_NEXT_DRAWFRAME;
  }
});
Object.defineProperty(exports, "Page", {
  enumerable: true,
  get: function () {
    return _page.Page;
  }
});
Object.defineProperty(exports, "PageContext", {
  enumerable: true,
  get: function () {
    return _pageContext.PageContext;
  }
});
Object.defineProperty(exports, "PageNoOpt", {
  enumerable: true,
  get: function () {
    return _pageNoOpt.PageNoOpt;
  }
});
Object.defineProperty(exports, "TAppProps", {
  enumerable: true,
  get: function () {
    return _page.TAppProps;
  }
});
Object.defineProperty(exports, "TPageProps", {
  enumerable: true,
  get: function () {
    return _page.TPageProps;
  }
});
Object.defineProperty(exports, "usePagePreFetch", {
  enumerable: true,
  get: function () {
    return _hooks.usePagePreFetch;
  }
});
Object.defineProperty(exports, "usePageQuery", {
  enumerable: true,
  get: function () {
    return _hooks.usePageQuery;
  }
});

var _page = require("./page");

var _pageNoOpt = require("./pageNoOpt");

var _alpha = require("./alpha");

var _customQueryClient = require("./customQueryClient");

var _pageContext = require("./pageContext");

var _hooks = require("./hooks");

var _lazy = require("./lazy");

var _loadingScene = _interopRequireDefault(require("./loadingScene"));

var _errorScene = _interopRequireDefault(require("./errorScene"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
