// @locallife/page/lib/commonjs/page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Page = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _alpha = require("./alpha");

var _loadingScene = _interopRequireDefault(require("./loadingScene"));

var _errorScene = _interopRequireDefault(require("./errorScene"));

var _customQueryClient = require("./customQueryClient");

var _reactQuery = require("react-query");

var _pageContext = require("./pageContext");

var _log = require("./log");

var _utils = require("./utils");

var _AppPropsProvider2 = require("./_AppPropsProvider");

var _useLcpCollect = require("./hooks/useLcpCollect");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1
  }
});

var EAppState; // 工程脚手架

(function (EAppState) {
  EAppState[EAppState["LOADING"] = 0] = "LOADING";
  EAppState[EAppState["ERROR"] = 1] = "ERROR";
  EAppState[EAppState["COMPLETED"] = 2] = "COMPLETED";
})(EAppState || (EAppState = {}));

const Page = props => {
  _AppPropsProvider2._AppPropsProvider.init(props.appProps);

  (0, _log.useReportContainer)(props);
  const t2TimeStampRef = (0, _react.useRef)(Date.now());
  const [loadedData, setLoadedData] = (0, _react.useState)(false);
  const [appState, setAppState] = (0, _react.useState)(props.needLoading || undefined === props.needLoading ? EAppState.LOADING : EAppState.COMPLETED);
  const handleResult = (0, _react.useCallback)(() => {
    const queryState = _customQueryClient.CustomQueryClient.get().getQueryState(props.queryKey);

    if (queryState.status === 'success') {
      var _props$onSuccess;

      t2TimeStampRef.current = queryState.dataUpdatedAt;
      (0, _log.publishRubas)({
        event: (0, _log.getPageRubasKey)(props, queryState.errorUpdateCount ? _log.PageRubasRule.apiRetrySuccess : _log.PageRubasRule.apiSuccess)
      });
      setLoadedData(true);
      setAppState(EAppState.COMPLETED);
      (_props$onSuccess = props.onSuccess) === null || _props$onSuccess === void 0 ? void 0 : _props$onSuccess.call(props);
    } else if (queryState.status === 'error') {
      var _props$onError;

      if (queryState.errorUpdateCount === 1) {
        (0, _log.publishRubas)({
          event: (0, _log.getPageRubasKey)(props, _log.PageRubasRule.apiFailed),
          payload: {
            local_life_network_error_code: (0, _utils.getErrorCode)(queryState === null || queryState === void 0 ? void 0 : queryState.error),
            local_life_network_error_message: (0, _utils.getErrorMessage)(queryState === null || queryState === void 0 ? void 0 : queryState.error)
          }
        });
      }

      setAppState(EAppState.ERROR);
      (_props$onError = props.onError) === null || _props$onError === void 0 ? void 0 : _props$onError.call(props);
    }
  }, []);
  const handleException = (0, _react.useCallback)(() => {
    var _props$onError2;

    setAppState(EAppState.ERROR);
    (_props$onError2 = props.onError) === null || _props$onError2 === void 0 ? void 0 : _props$onError2.call(props);
    (0, _log.publishRubas)({
      event: (0, _log.getPageRubasKey)(props, _log.PageRubasRule.apiFailed)
    });
  }, []); // 请求数据

  const requestData = (0, _react.useCallback)(() => {
    setAppState(EAppState.LOADING); // t1上报

    if (!props.noReport) {
      (0, _log.pageLog)('分阶段耗时', `t1=${Date.now() - props.appProps.onCreateTimestamp}`);

      _reactNative.NativeModules.RootViewPageBridge.rootViewDidMountWith(props.appProps.rootTag, {});

      (0, _log.publishRubas)({
        event: 'local_life_page_on_load_view_t1',
        payload: (0, _utils.getRubasTime)(Date.now())
      });
    }

    if (props.queryKey && props.onRequest) {
      props.onRequest().then(handleResult).catch(handleException);
    } else if (!props.noReport) {
      // 无网络请求或者数据处理时，t2、t3上报
      const time2 = Date.now() - props.appProps.onCreateTimestamp;

      _reactNative.NativeModules.RootViewPageBridge.reportPageRenderTime(props.appProps.rootTag, time2, time2, {});
    }
  }, []);
  (0, _react.useMemo)(requestData, []);
  const {
    lcpImages
  } = (0, _useLcpCollect.useLcpCollect)();
  (0, _react.useEffect)(() => {
    if (loadedData) {
      // 有网络请求或者数据处理时，t2、t3上报
      if (!props.noReport) {
        const t3 = Date.now() - props.appProps.onCreateTimestamp;
        const t2 = t2TimeStampRef.current - props.appProps.onCreateTimestamp;

        _reactNative.NativeModules.RootViewPageBridge.reportPageRenderTime(props.appProps.rootTag, t2, t3, {});

        (0, _log.pageLog)('分阶段耗时', `t2=${t2},t3=${t3}`);
        (0, _log.publishRubas)({
          event: 'local_life_page_request_t2',
          payload: (0, _utils.getRubasTime)(t2TimeStampRef.current)
        });
        (0, _log.publishRubas)({
          event: 'local_life_page_render_t3',
          payload: (0, _utils.getRubasTime)(Date.now())
        });
      }

      (0, _log.publishRubas)({
        event: (0, _log.getPageRubasKey)(props, _log.PageRubasRule.pageRender)
      });
    }
  }, [loadedData]); // 渲染组件

  function renderContent() {
    const needLoading = !!(props.needLoading || undefined === props.needLoading || props.onRequest && props.queryKey);
    const needError = !!(props.needError || undefined === props.needError);
    return /*#__PURE__*/_react.default.createElement(_reactQuery.QueryClientProvider, {
      client: _customQueryClient.CustomQueryClient.get()
    }, /*#__PURE__*/_react.default.createElement(_pageContext.PageContext.Provider, {
      value: {
        loadedData: loadedData,
        lcpImages: lcpImages
      }
    }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      style: styles.container
    }, /*#__PURE__*/_react.default.createElement(_alpha.Alpha, {
      target: props.contentPager,
      schedule: needLoading ? _alpha.OP_NEXT_DRAWFRAME : _alpha.OP_INLINE,
      data: props.appProps
    }), needLoading && EAppState.LOADING === appState && /*#__PURE__*/_react.default.createElement(_alpha.Alpha, {
      target: props.LoadingComponent || _loadingScene.default,
      schedule: _alpha.OP_INLINE
    }), needError && EAppState.ERROR === appState && /*#__PURE__*/_react.default.createElement(_alpha.Alpha, {
      target: props.ErrorComponent || _errorScene.default,
      schedule: _alpha.OP_INLINE,
      data: {
        onRetry: () => {
          _customQueryClient.CustomQueryClient.get().refetchQueries(props.queryKey).then(handleResult).catch(handleException);
        }
      }
    }))));
  }

  return renderContent();
};

exports.Page = Page;
