// @kds/react-native-svg/src/elements/NativeComponents.tsx
import { requireNativeComponent as rnc, HostComponent } from 'react-native';

const SVGNameMap: Map<string, HostComponent<unknown>> = new Map<string, HostComponent<unknown>>();
const getRNC = (name: string): HostComponent<unknown> => {
    if(SVGNameMap.has(name)) {
        return SVGNameMap.get(name)!;
    } else {
        SVGNameMap.set(name, rnc(name));
        return SVGNameMap.get(name)!;
    }
}

export const RNSVGSvg = getRNC('RNSVGSvgView');
export const RNSVGCircle = getRNC('RNSVGCircle');
export const RNSVGClipPath = getRNC('RNSVGClipPath');
export const RNSVGDefs = getRNC('RNSVGDefs');
export const RNSVGEllipse = getRNC('RNSVGEllipse');
export const RNSVGForeignObject = getRNC('RNSVGForeignObject');
export const RNSVGGroup = getRNC('RNSVGGroup');
export const RNSVGImage = getRNC('RNSVGImage');
export const RNSVGLine = getRNC('RNSVGLine');
export const RNSVGLinearGradient = getRNC('RNSVGLinearGradient');
export const RNSVGMarker = getRNC('RNSVGMarker');
export const RNSVGMask = getRNC('RNSVGMask');
export const RNSVGPath = getRNC('RNSVGPath');
export const RNSVGPattern = getRNC('RNSVGPattern');
export const RNSVGRadialGradient = getRNC('RNSVGRadialGradient');
export const RNSVGRect = getRNC('RNSVGRect');
export const RNSVGSymbol = getRNC('RNSVGSymbol');
export const RNSVGText = getRNC('RNSVGText');
export const RNSVGTextPath = getRNC('RNSVGTextPath');
export const RNSVGTSpan = getRNC('RNSVGTSpan');
export const RNSVGUse = getRNC('RNSVGUse');
