// @locallife/biz-component/src/components/contactPopup/index.tsx
import React from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import { KidIcon } from '@kid-ui/krn';
import styles from './styles';
import { screenUnit, bridge } from '@lux/ts-lbs-krn-utils';

const { screenSize } = screenUnit;
export interface phoneListItem {
    // 手机号
    phone?: string | number;
    // 标题
    title?: string;
    // 自定义配置
    custom?: {
        // 自定义点击事件
        customPress?: () => void;
        // 自定义右侧icon
        icon?: string;
        // 点击电话后自定义触发事件
        phonePress?: () => void;
    };
}
export interface PopupParams {
    phoneList: phoneListItem[] | Array<string | number>;
}
const { makePhoneCall } = bridge;

export const ContactPopup = ({ phoneList }: PopupParams) => {
    /**
     * 拨打商家电话
     */
    const callSeller = (phone) => {
        makePhoneCall([phone]);
    };
    /**
     * @param item 当前这条信息
     * @returns 是否只有一条电话数据
     */
    const isOnlyPhone = (item) => {
        return typeof item === 'string' || typeof item === 'number';
    };
    /**
     * 处理点击
     * @param item 当前这条信息
     */
    const handlePress = (item) => {
        if (item?.custom?.customPress) {
            item?.custom?.customPress();
        } else {
            callSeller(isOnlyPhone(item) ? item : item.phone);
            item?.custom?.phonePress?.();
        }
    };

    return (
        <View style={styles.popupContent}>
            {phoneList.map((item, index) => (
                <TouchableOpacity
                    key={index}
                    activeOpacity={1}
                    onPress={() => handlePress(item)}
                >
                    {index !== 0 && <View style={styles.dividingLine} />}
                    <View style={styles.popupItem}>
                        <View style={styles.popupItemContent}>
                            {!isOnlyPhone(item) && !!item?.title && (
                                <Text style={styles.popupItemDesc}>
                                    {item?.title}{' '}
                                </Text>
                            )}
                            <Text style={styles.popupItemDesc}>
                                {isOnlyPhone(item) ? item : item?.phone}
                            </Text>
                        </View>

                        <View style={styles.popupItemIcon}>
                            <KidIcon
                                kid={
                                    item?.custom?.icon ||
                                    'common_baseflat_telephone_24'
                                }
                                kidColor={'color_base_black_8'}
                                size={screenSize(24)}
                            />
                        </View>
                    </View>
                </TouchableOpacity>
            ))}
        </View>
    );
};
