// @locallife/biz-component/src/components/contactPopup/styles.ts
import { screenUnit, iphone } from '@lux/ts-lbs-krn-utils';
import { StyleSheet } from 'react-native';
const { screenSize } = screenUnit;
export default StyleSheet.create({
    rowFlexAlignCenter: {
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center',
    },
    popupItem: {
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
        height: screenSize(58),
    },
    popupItemContent: {
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center',
        width: screenSize(352),
    },
    popupItemDesc: {
        fontWeight: '500',
        fontSize: screenSize(16),
        color: '#222222',
    },
    popupItemIcon: {
        flexGrow: 0,
        width: screenSize(24),
    },
    phoneText: {
        marginTop: screenSize(2),
        fontSize: screenSize(11),
        color: '#666666',
    },
    dividingLine: {
        borderBottomColor: '#f0f0f0',
        borderBottomWidth: screenSize(1),
    },
    popupContent: {
        paddingHorizontal: screenSize(19),
        paddingBottom: iphone.safeAreaBottom,
    },
});
