// @locallife/biz-component/lib/module/components/newUserZoneEntrance/index.js
import React from 'react';
import { View, Text, TouchableWithoutFeedback } from 'react-native';
import { isUserLogin, needLogin, showToast } from '@locallife/utils';
import KwaiImage from '@kds/image';
import styles from './styles';
import { isEligibility } from './api';
import { ShowWithData } from '@locallife/base-component';
/**
 * 新人专区入口
 *      货架列表页Banner位置
 *      POI页
 */
export const NewUserZoneEntrance = /*#__PURE__*/React.memo(entranceData => {
  var _entranceData$data, _entranceData$data2, _entranceData$data2$p, _entranceData$data3, _entranceData$data3$p, _entranceData$data3$p2;
  function doCheckEligibility() {
    // 核验资格
    isEligibility({
      source: 1 // banner检查是否新用户取source=1的固定值
    }).then(result => {
      if (result !== null && result !== void 0 && result.data) {
        if (result.data.newUser) {
          // 有资格
          entranceData.handleClick(true);
        } else {
          var _entranceData$refresh;
          // 无资格
          entranceData.handleClick(false);
          // 提示
          showToast('您当前不满足新用户身份哦，看看其他商品吧。');
          // 刷新
          (_entranceData$refresh = entranceData.refreshData) === null || _entranceData$refresh === void 0 ? void 0 : _entranceData$refresh.call(entranceData);
        }
      } else {
        // 数据返回异常
        showToast('数据异常，请重试。');
      }
    }).catch(() => {
      // 异常提示
      showToast('网络异常，请重试。');
    });
  }
  function entranceHandleClick() {
    if (isUserLogin()) {
      // 已经登录
      doCheckEligibility();
    } else {
      // 未登录，不跳转
      entranceData.handleClick(false);
      // 唤起登录页面
      needLogin(() => {
        // 成功登录
        doCheckEligibility();
      }, () => {
        // 登录失败，不做处理
      });
    }
  }
  return (_entranceData$data = entranceData.data) !== null && _entranceData$data !== void 0 && _entranceData$data.icon ?
  /*#__PURE__*/
  // eslint-disable-next-line react/jsx-no-bind
  React.createElement(TouchableWithoutFeedback, {
    onPress: entranceHandleClick
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.rootContainer, entranceData.contentContainerStyle]
  }, /*#__PURE__*/React.createElement(KwaiImage, {
    style: [styles.bgStyle, entranceData.bgStyle],
    source: {
      uri: entranceData.data.icon
    },
    resizeMode: 'cover'
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.disCountArea, entranceData.disCountAreaStyle]
  }, /*#__PURE__*/React.createElement(Text, {
    style: [styles.disCountContentType, entranceData.disCountContentTypeStyle]
  }, '¥'), /*#__PURE__*/React.createElement(ShowWithData, {
    data: (_entranceData$data2 = entranceData.data) === null || _entranceData$data2 === void 0 ? void 0 : (_entranceData$data2$p = _entranceData$data2.params) === null || _entranceData$data2$p === void 0 ? void 0 : _entranceData$data2$p.price
  }, /*#__PURE__*/React.createElement(Text, {
    style: [styles.disCountContent, entranceData.disCountContentStyle]
  }, entranceData.data.params.price)), /*#__PURE__*/React.createElement(ShowWithData, {
    data: ((_entranceData$data3 = entranceData.data) === null || _entranceData$data3 === void 0 ? void 0 : (_entranceData$data3$p = _entranceData$data3.params) === null || _entranceData$data3$p === void 0 ? void 0 : (_entranceData$data3$p2 = _entranceData$data3$p.subIcons) === null || _entranceData$data3$p2 === void 0 ? void 0 : _entranceData$data3$p2.length) > 0
  }, /*#__PURE__*/React.createElement(KwaiImage, {
    style: [styles.disCountTitleImg, entranceData.disCountTitleImgStyle],
    source: {
      uri: entranceData.data.params.subIcons[0]
    },
    resizeMode: 'cover'
  })))))) : /*#__PURE__*/React.createElement(View, null);
});
