// @locallife/biz-component/lib/module/components/newUserZoneEntrance/styles.js
import { Platform, StyleSheet } from 'react-native';
export default StyleSheet.create({
  rootContainer: {
    width: 392,
    height: 84,
    overflow: 'hidden',
    borderRadius: 16
  },
  bgStyle: {
    width: '100%',
    height: '100%'
  },
  disCountArea: {
    flexDirection: 'row',
    marginTop: 8,
    marginLeft: 25,
    justifyContent: 'flex-start',
    alignItems: 'flex-end'
  },
  disCountContentType: {
    fontFamily: 'Avenir Next',
    fontWeight: '700',
    fontSize: 20,
    lineHeight: 24,
    color: '#000000',
    ...Platform.select({
      android: {
        marginBottom: 4
      },
      ios: {
        marginBottom: 3
      }
    })
  },
  disCountContent: {
    fontFamily: 'Avenir Next',
    fontWeight: '700',
    fontSize: 36,
    lineHeight: 44,
    color: '#000000',
    ...Platform.select({
      android: {
        marginBottom: 0
      },
      ios: {
        marginBottom: -4
      }
    })
  },
  disCountTitleImg: {
    marginLeft: 3,
    height: 21,
    width: 134,
    marginBottom: 9
  }
});
