// @locallife/biz-request/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BaseRequestImpl", {
  enumerable: true,
  get: function () {
    return _baseRequest.BaseRequestImpl;
  }
});
Object.defineProperty(exports, "BizRequestConfig", {
  enumerable: true,
  get: function () {
    return _BizRequestConfig.BizRequestConfig;
  }
});
Object.defineProperty(exports, "BizRequestImpl", {
  enumerable: true,
  get: function () {
    return _BizRequestImpl.BizRequestImpl;
  }
});
Object.defineProperty(exports, "CommonHeaders", {
  enumerable: true,
  get: function () {
    return _baseRequest.CommonHeaders;
  }
});
Object.defineProperty(exports, "CommonParams", {
  enumerable: true,
  get: function () {
    return _baseRequest.CommonParams;
  }
});
Object.defineProperty(exports, "CommonRequestParams", {
  enumerable: true,
  get: function () {
    return _baseRequest.CommonRequestParams;
  }
});
Object.defineProperty(exports, "DefaultBizRequestConfig", {
  enumerable: true,
  get: function () {
    return _BizRequestConfig.DefaultBizRequestConfig;
  }
});
Object.defineProperty(exports, "DefaultRequestConfig", {
  enumerable: true,
  get: function () {
    return _baseRequest.DefaultRequestConfig;
  }
});
Object.defineProperty(exports, "IRequestParam", {
  enumerable: true,
  get: function () {
    return _baseRequest.IRequestParam;
  }
});
Object.defineProperty(exports, "Interceptor", {
  enumerable: true,
  get: function () {
    return _baseRequest.Interceptor;
  }
});
Object.defineProperty(exports, "InterceptorModel", {
  enumerable: true,
  get: function () {
    return _baseRequest.InterceptorModel;
  }
});
Object.defineProperty(exports, "KSBaseResponse", {
  enumerable: true,
  get: function () {
    return _baseRequest.KSBaseResponse;
  }
});
Object.defineProperty(exports, "RequestConfig", {
  enumerable: true,
  get: function () {
    return _baseRequest.RequestConfig;
  }
});
Object.defineProperty(exports, "RequestMethod", {
  enumerable: true,
  get: function () {
    return _baseRequest.RequestMethod;
  }
});
Object.defineProperty(exports, "TParams", {
  enumerable: true,
  get: function () {
    return _baseRequest.TParams;
  }
});
exports.bizRequest = void 0;
Object.defineProperty(exports, "request", {
  enumerable: true,
  get: function () {
    return _baseRequest.request;
  }
});

var _BizRequestImpl = require("./request/BizRequestImpl");

var _BizRequestConfig = require("./model/BizRequestConfig");

var _init = require("./init");

var _baseRequest = require("@locallife/base-request");

// @ts-ignore
const bizRequest = _baseRequest.request;
exports.bizRequest = bizRequest;

bizRequest.bizInit = function (commonRequestParams, interceptorParams) {
  this.init(commonRequestParams);

  _init.bizRequestImpl.init(interceptorParams);
};
