// @locallife/biz-component/lib/commonjs/components/contactPopup/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _tsLbsKrnUtils = require("@lux/ts-lbs-krn-utils");
var _reactNative = require("react-native");
const {
  screenSize
} = _tsLbsKrnUtils.screenUnit;
var _default = _reactNative.StyleSheet.create({
  rowFlexAlignCenter: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center'
  },
  popupItem: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    height: screenSize(58)
  },
  popupItemContent: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    width: screenSize(352)
  },
  popupItemDesc: {
    fontWeight: '500',
    fontSize: screenSize(16),
    color: '#222222'
  },
  popupItemIcon: {
    flexGrow: 0,
    width: screenSize(24)
  },
  phoneText: {
    marginTop: screenSize(2),
    fontSize: screenSize(11),
    color: '#666666'
  },
  dividingLine: {
    borderBottomColor: '#f0f0f0',
    borderBottomWidth: screenSize(1)
  },
  popupContent: {
    paddingHorizontal: screenSize(19),
    paddingBottom: _tsLbsKrnUtils.iphone.safeAreaBottom
  }
});
exports.default = _default;
