// @locallife/log/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "LocalLifeCustomLogger", {
  enumerable: true,
  get: function () {
    return _LocalLifeCustomLogger.default;
  }
});
Object.defineProperty(exports, "LocalLifeDebugLogger", {
  enumerable: true,
  get: function () {
    return _LocalLifeDebugLogger.default;
  }
});
Object.defineProperty(exports, "LocalLifeInitLogger", {
  enumerable: true,
  get: function () {
    return _LocalLifeInitLogger.default;
  }
});
Object.defineProperty(exports, "LocalLifeJinJingLogger", {
  enumerable: true,
  get: function () {
    return _LocalLifeJinJingLogger.default;
  }
});
Object.defineProperty(exports, "LocalLifeKrnLogger", {
  enumerable: true,
  get: function () {
    return _localLifeKrnLogger.default;
  }
});
Object.defineProperty(exports, "LocalLifeRequestLogger", {
  enumerable: true,
  get: function () {
    return _LocalLifeRequestLogger.default;
  }
});
Object.defineProperty(exports, "LocalLifeUrlLogger", {
  enumerable: true,
  get: function () {
    return _LocalLifeUrlLogger.default;
  }
});
Object.defineProperty(exports, "ReportType", {
  enumerable: true,
  get: function () {
    return _ReportType.ReportType;
  }
});
Object.defineProperty(exports, "localLifeBizLogger", {
  enumerable: true,
  get: function () {
    return _LocalLifeBizLogger.localLifeBizLogger;
  }
});

var _LocalLifeCustomLogger = _interopRequireDefault(require("./LocalLifeCustomLogger"));

var _LocalLifeDebugLogger = _interopRequireDefault(require("./LocalLifeDebugLogger"));

var _LocalLifeJinJingLogger = _interopRequireDefault(require("./LocalLifeJinJingLogger"));

var _LocalLifeUrlLogger = _interopRequireDefault(require("./LocalLifeUrlLogger"));

var _LocalLifeInitLogger = _interopRequireDefault(require("./LocalLifeInitLogger"));

var _LocalLifeRequestLogger = _interopRequireDefault(require("./LocalLifeRequestLogger"));

var _LocalLifeBizLogger = require("./LocalLifeBizLogger");

var _ReportType = require("./ReportType");

var _localLifeKrnLogger = _interopRequireDefault(require("./localLifeKrnLogger"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
