// @locallife/log/lib/commonjs/LocalLifeBizLogger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.localLifeBizLogger = void 0;

var _LocalLifeInitLogger = _interopRequireDefault(require("./LocalLifeInitLogger"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class LocalLifeBizLogger {
  innerLog(tag, message) {
    console.log(`[@locallife/log][${tag}]:${typeof message === 'string' ? message : JSON.stringify(message)}`);
  }

  click(action, params, contentPackage, other) {
    const weblog = _LocalLifeInitLogger.default.getWeblog();

    weblog.sendImmediately('CLICK', {
      action,
      params: params,
      contentPackage: contentPackage,
      ...other
    });
    this.innerLog('click log', JSON.stringify({
      action,
      params,
      contentPackage
    }));
  }

  show(action, params, contentPackage, other) {
    const weblog = _LocalLifeInitLogger.default.getWeblog();

    weblog.sendImmediately('SHOW', {
      action,
      params: params,
      contentPackage: contentPackage,
      ...other
    });
  }

  event(action, params, contentPackage) {
    const weblog = _LocalLifeInitLogger.default.getWeblog();

    this.innerLog('event log', JSON.stringify({
      action,
      params,
      contentPackage
    }));
    weblog.sendImmediately('CLICK', {
      action,
      params: params,
      type: 'BACKGROUND_TASK_EVENT'
    });
  }

  pageShow(pvParams) {
    const weblog = _LocalLifeInitLogger.default.getWeblog();

    const pvEventOption = _LocalLifeInitLogger.default.getPvEventOption();

    weblog.sendImmediately('PV', {
      type: 'enter',
      page: pvEventOption.page,
      params: { ...pvEventOption.params,
        ...pvParams
      },
      coPage: !!pvEventOption.coPage
    });
  }

}

const localLifeBizLogger = new LocalLifeBizLogger();
exports.localLifeBizLogger = localLifeBizLogger;
