// @locallife/log/lib/commonjs/LocalLifeFmpLogger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reactNative = require("react-native");

var _LocalLifeInitLogger = _interopRequireDefault(require("./LocalLifeInitLogger"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class LocalLifeFmpLogger {
  // 上报 t1
  // 白屏时间（T1）
  static handleT1() {
    var _NativeModules$RootVi;

    if (this.rootTag) {
      throw new Error('You send fmp log already.');
    }

    let {
      rootTag,
      onCreateTimestamp,
      extra
    } = _LocalLifeInitLogger.default.getAppModel();

    this.rootTag = rootTag;
    this.t1 = onCreateTimestamp;
    this.extra = extra;
    _reactNative.NativeModules === null || _reactNative.NativeModules === void 0 ? void 0 : (_NativeModules$RootVi = _reactNative.NativeModules.RootViewPageBridge) === null || _NativeModules$RootVi === void 0 ? void 0 : _NativeModules$RootVi.rootViewDidMountWith(this.rootTag, {});
  } // 第一次接口上报时间(T2)


  static handleT2() {
    if (this.t2) {
      return;
    }

    if (this.t1) {
      this.t2 = Date.now() - this.t1;
    }
  } // 首屏时间（T3）


  static handleT3() {
    if (this.t3) {
      return;
    }

    if (this.t1) {
      this.t3 = Date.now() - this.t1;
    }

    if (this.t2 && this.t3 && this.t3 > this.t2) {
      var _NativeModules$RootVi2;

      _reactNative.NativeModules === null || _reactNative.NativeModules === void 0 ? void 0 : (_NativeModules$RootVi2 = _reactNative.NativeModules.RootViewPageBridge) === null || _NativeModules$RootVi2 === void 0 ? void 0 : _NativeModules$RootVi2.reportPageRenderTime(this.rootTag, this.t2, this.t3, this.extra);
    }
  }

}

exports.default = LocalLifeFmpLogger;

_defineProperty(LocalLifeFmpLogger, "rootTag", void 0);

_defineProperty(LocalLifeFmpLogger, "t1", void 0);

_defineProperty(LocalLifeFmpLogger, "t2", void 0);

_defineProperty(LocalLifeFmpLogger, "t3", void 0);

_defineProperty(LocalLifeFmpLogger, "extra", void 0);
