// @kid-ui/krn/lib/components/city-selector/utils.js
import { NativeModules, StatusBar, Platform, Dimensions } from 'react-native';
import { rem } from '../../utils';
export const screenWidth = Dimensions.get('screen').width;
export const screenHeight = Dimensions.get('screen').height;
export const windowWidth = Dimensions.get('window').width;
export const windowHeight = Dimensions.get('window').height;
export const isAndroid = Platform.OS === 'android';
export const isIOS = Platform.OS === 'ios';
export const TITLE_BAR_HEIGHT = 44;
export const fontWeight = {
    Thin: Platform.OS === 'ios' ? '100' : 'normal',
    UltraLight: Platform.OS === 'ios' ? '200' : 'normal',
    Light: Platform.OS === 'ios' ? '300' : 'normal',
    Regular: Platform.OS === 'ios' ? '400' : 'normal',
    Medium: Platform.OS === 'ios' ? '500' : 'bold',
    Semibold: Platform.OS === 'ios' ? '600' : 'bold',
    Bold: Platform.OS === 'ios' ? '700' : 'bold',
    Heavy: Platform.OS === 'ios' ? '800' : 'bold',
    Black: Platform.OS === 'ios' ? '900' : 'bold',
};
export class UIUtils {
    static isDark(theme) {
        return theme === 'dark';
    }
    static isIphoneX() {
        return isIOS && windowHeight >= 812;
    }
    static getStatusBarHeight() {
        var _a, _b;
        let height;
        if (isIOS) {
            height = this.isIphoneX() ? 44 : 20;
        }
        else if ((_b = (_a = NativeModules.KSAppearance) === null || _a === void 0 ? void 0 : _a.canImmersive) === null || _b === void 0 ? void 0 : _b.call(_a)) {
            height = StatusBar.currentHeight || 0;
        }
        else {
            height = 0;
        }
        return height;
    }
    static getBottomBarHeight() {
        let height;
        if (Platform.OS === 'ios') {
            height = UIUtils.isIphoneX() ? rem(34) : 0;
        }
        else {
            height = 0;
        }
        return height;
    }
}
export function ShowWithData(props) {
    if (!props.data) {
        return null;
    }
    return props.children;
}
