// @lux/pkg-report/lib/utils/const.js
"use strict";
/**
 * package.json path
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSH_URL_RE = exports.HTTPS_URL_RE = exports.reportHost = exports.reportStagingHost = exports.luxLibPrefix = exports.luxPrefix = exports.appRootPackageJsonPath = exports.appRootPath = void 0;
const app_root_path_1 = __importDefault(require("app-root-path"));
/**
 * 当前运行脚本的应用项目路径
 * @TODO 在本地使用 npm link 测试的时候路径是真实的存储路径：
 * /Users/linkfly/Documents/workspace/lux-cli/node_modules/.pnpm/app-root-path@3.0.0/node_modules/app-root-path
 */
exports.appRootPath = app_root_path_1.default.path;
/**
 * 用户项目目录下的 package.json 路径
 */
exports.appRootPackageJsonPath = `${exports.appRootPath}/package.json`;
/**
 * 用于识别是否是 lux 组件
 */
exports.luxPrefix = '@lux';
/**
 * 用于识别是否是 lux 组件库
 * @TODO 这样的识别还是有问题，因为 sharp-ui-plus 没有覆盖到
 */
exports.luxLibPrefix = '@lux-lib';
/**
 * 上报地址：STAGING
 */
exports.reportStagingHost = 'http://data-center-base.web-ops.staging.kuaishou.com/rest/collect/report';
/**
 * 上报地址：ONLINE
 */
exports.reportHost = 'https://lux.corp.kuaishou.com/rest/collect/report';
/**
 * 匹配仓库名正则
 * 通过https 形式clone的仓库
 * @example https://git.corp.kuaishou.com/mfe/platform/lux-cli.git
 */
exports.HTTPS_URL_RE = /^http.*(\.com|\.cn)\/(.*).git$/;
/**
 * 匹配仓库名正则
 * 通过 ssh 形式 clone 的仓库
 * @example git@git.corp.kuaishou.com:mfe/platform/lux-cli.git
 */
exports.SSH_URL_RE = /^git.*.com:(.+).git$/;
