// @locallife/biz-hooks/src/useCountdown/demo/index.tsx
/* eslint-disable no-alert */
import React from 'react';
import useCountdown from '..';
import { Text, StyleSheet, View } from 'react-native';

const textStyle = StyleSheet.create({
    title: {
        fontSize: 18,
        color: 'blue',
    },
    margin: {
        marginTop: 20,
    },
});

export default () => {
    const { formattedRes: formattedResult1 } = useCountdown({
        endTime: Date.now() + 86400000,
    });
    const { formattedRes: formattedResult2 } = useCountdown({
        // 指定更新的时间
        endTime: 1704038399999,
        useDay: true,
    });
    const { formattedRes: formattedResult3 } = useCountdown({
        endTime: Date.now() + 10000,
        onEnd: () => {
            alert('倒计时结束');
        },
    });
    const { formattedRes: formattedResult4 } = useCountdown({
        endTime: 1704038399999,
        useDay: true,
        useServerTime: true,
    });

    return (
        <View>
            <Text style={textStyle.title}>倒计时一天</Text>
            <Text>
                {formattedResult1.hour} 时 {formattedResult1.minute} 分{' '}
                {formattedResult1.second} 秒
            </Text>

            <Text style={[textStyle.title, textStyle.margin]}>
                倒计时到2023年年底
            </Text>
            <Text>
                {formattedResult2.day} 天 {formattedResult2.hour} 时{' '}
                {formattedResult2.minute} 分 {formattedResult2.second} 秒
            </Text>

            <Text style={[textStyle.title, textStyle.margin]}>
                倒计时结束监听
            </Text>
            <Text>{formattedResult3.second} 秒 </Text>

            <Text style={[textStyle.title, textStyle.margin]}>
                测试服务器时间，在设置中调整手机时间不生效
            </Text>
            <Text>
                {formattedResult4.day} 天 {formattedResult4.hour} 时{' '}
                {formattedResult4.minute} 分 {formattedResult4.second} 秒
            </Text>
        </View>
    );
};
