// @locallife/biz-component/src/components/poiInfo/index.tsx
import React from 'react';
import { View, Text, Image, TouchableOpacity } from 'react-native';
import { KidIcon } from '@kid-ui/krn';
import styles from './styles';
import { IPoiData } from './types';
import { screenUnit } from '@lux/ts-lbs-krn-utils';
import { getDistance } from '../../utils/decimal';
import { formatHotPeople } from '@lux/ts-lbs-krn-utils';

const { screenSize } = screenUnit;

interface IpoiData {
    productData: any;
    poiData: IPoiData;
    isShowPhoneIcon: boolean;
    isShowMinDistance: boolean;
    toPOIList: () => void;
    toPOIDetail: () => void;
    toGuide: () => void;
    isShowPopupVisible: () => void;
}

export const PoiInfo = ({
    productData,
    poiData,
    isShowPhoneIcon,
    isShowMinDistance,
    toPOIList,
    toPOIDetail,
    toGuide,
    isShowPopupVisible,
}: IpoiData) => {
    // 点击电话弹窗
    const handlePhoneIcon = () => {
        isShowPopupVisible && isShowPopupVisible();
    };
    return (
        <View style={styles.container}>
            {/* 只有多于一个才展示门店列表跳转和(N)家 */}
            <View style={styles.header}>
                <Text style={styles.title}>
                    适用门店
                    {productData?.poiSize > 1 && (
                        <Text>（{productData?.poiSize} 家）</Text>
                    )}
                </Text>
                {productData?.poiSize > 1 && (
                    <TouchableOpacity activeOpacity={1} onPress={toPOIList}>
                        <View style={styles.rowFlexAlignCenter}>
                            <Text style={styles.moreStoreLink}>查看全部</Text>
                            <KidIcon
                                kid={'common_base_next_24'}
                                kidColor={'color_base_black_6'}
                                size={screenSize(14)}
                            />
                        </View>
                    </TouchableOpacity>
                )}
            </View>
            <TouchableOpacity activeOpacity={1} onPress={toPOIDetail}>
                <View style={[styles.content, styles.rowFlexAlignCenter]}>
                    <Image
                        style={styles.poiImage}
                        source={{ uri: poiData.image }}
                    />
                    <View style={styles.poiDescWrapper}>
                        <View style={styles.poiNameWrap}>
                            <Text
                                numberOfLines={1}
                                style={[
                                    styles.poiName,
                                    isShowMinDistance &&
                                        styles.poiNameWithMinDistance,
                                ]}
                            >
                                {poiData.poiName}
                            </Text>
                            {isShowMinDistance && (
                                <View style={styles.minDistance}>
                                    <Text style={styles.minDistanceText}>
                                        离我最近
                                    </Text>
                                </View>
                            )}
                        </View>
                        <View style={styles.poiCategoryContainer}>
                            <View style={styles.subCategory}>
                                <Text style={styles.tag}>
                                    {poiData.subCate}
                                </Text>
                            </View>
                            {!!poiData.avgPrice && (
                                <View
                                    style={[
                                        styles.subCategory,
                                        styles.subCategoryDistance,
                                    ]}
                                >
                                    <Text style={styles.tag}>
                                        ￥{Math.floor(+poiData.avgPrice / 100)}
                                        /人
                                    </Text>
                                </View>
                            )}
                            {!!formatHotPeople(poiData.visitCount) && (
                                <View
                                    style={[
                                        styles.subCategory,
                                        styles.subCategoryDistance,
                                    ]}
                                >
                                    <Text style={styles.tag}>
                                        {formatHotPeople(poiData.visitCount)}
                                        人想去
                                    </Text>
                                </View>
                            )}
                        </View>
                    </View>
                    {isShowPhoneIcon && (
                        <TouchableOpacity
                            activeOpacity={1}
                            onPress={handlePhoneIcon}
                        >
                            <View style={styles.poiPhoneWrapper}>
                                <KidIcon
                                    kid={'common_baseflat_telephone_24'}
                                    kidColor={'color_base_black_8'}
                                    size={screenSize(24)}
                                />
                                <Text style={styles.poiPhoneText}>电话</Text>
                            </View>
                        </TouchableOpacity>
                    )}
                </View>
            </TouchableOpacity>
            <View style={styles.dividingLine} />
            <TouchableOpacity activeOpacity={1} onPress={toGuide}>
                <View
                    style={[styles.positionWrapper, styles.rowFlexAlignCenter]}
                >
                    {!!poiData.distance && (
                        <>
                            <KidIcon
                                kid={'common_list_location_grey'}
                                kidColor={'color_base_black_11'}
                                size={screenSize(14)}
                            />
                            <Text style={styles.distance}>
                                {getDistance(poiData.distance)}
                            </Text>
                            <Text style={styles.divide}>|</Text>
                        </>
                    )}
                    <Text
                        style={styles.address}
                        ellipsizeMode="tail"
                        numberOfLines={1}
                    >
                        {poiData.address}
                    </Text>
                    <KidIcon
                        kid={'common_base_next_24'}
                        kidColor={'color_base_black_6'}
                        size={screenSize(14)}
                    />
                </View>
            </TouchableOpacity>
        </View>
    );
};
