// @locallife/biz-hooks/lib/module/useSafeMemo/demo/index.js
/* eslint-disable react/jsx-no-bind */
import React, { useState } from 'react';
import { Text, StyleSheet, SafeAreaView, Button, View } from 'react-native';
import useSafeMemo from '../index';
const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: '100%',
    display: 'flex',
    alignItems: 'center'
  },
  text: {
    backgroundColor: '#aaa',
    color: '#ffffff',
    textAlign: 'center',
    width: '100%',
    height: 50,
    lineHeight: 50,
    marginBottom: 10
  },
  bottom: {
    display: 'flex',
    flexDirection: 'row'
  }
});
export default (() => {
  const [counter, setCounter] = useState(0);
  const display = useSafeMemo(() => {
    return counter * 10;
  }, [counter]);
  return /*#__PURE__*/React.createElement(SafeAreaView, {
    style: styles.container
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.text
  }, "\u503C", display), /*#__PURE__*/React.createElement(View, {
    style: styles.bottom
  }, /*#__PURE__*/React.createElement(Button, {
    title: "\u52A010",
    onPress: () => {
      setCounter(counter + 1);
    }
  }), /*#__PURE__*/React.createElement(Button, {
    title: "\u51CF10",
    onPress: () => {
      setCounter(counter - 1);
    }
  })));
});
