// @kid-ui/krn/lib/components/range-slider/active-line.js
import React from 'react';
import { StyleSheet, View } from 'react-native';
export default function ActiveLine(props) {
    const { height = 10, dotSize = 35, screenWidth, width, left = 0, containerBgColor = '#ccc', barBgColor = 'gray', containerStyle, barStyle, render, } = props;
    const style = [
        styles.container,
        {
            height: height,
            top: (dotSize - height) / 2,
            borderRadius: height / 2,
            backgroundColor: containerBgColor
        },
        containerStyle
    ];
    if (typeof render === 'function') {
        style.push({ backgroundColor: 'transparent' });
    }
    return (<>
      <View style={style}>
        {typeof render === 'function' ? (render({ screenWidth })) : (<>
            <View style={[
                styles.bar,
                {
                    width,
                    left,
                    height,
                    backgroundColor: barBgColor
                },
                barStyle
            ]}/>
          </>)}
      </View>
    </>);
}
const styles = StyleSheet.create({
    container: {
        position: 'absolute',
        width: '100%',
        backgroundColor: '#ccc',
        overflow: 'hidden',
    },
    bar: {
        backgroundColor: 'gray',
    },
});
