// @kid-ui/krn/lib/components/range-slider/dot.js
import React from 'react';
import { StyleSheet, Animated } from 'react-native';
export default function Dot({ size = 35, panResponder, translateX, style, render }) {
    return (<>
      {typeof render === 'function' ? (<Animated.View {...panResponder.panHandlers} style={[
                styles.dot,
                {
                    transform: [{ translateX }],
                    height: size,
                    width: size,
                    borderRadius: size / 2,
                },
                style,
            ]}>
          {render(styles.dot)}
        </Animated.View>) : (<Animated.View {...panResponder.panHandlers} style={[
                styles.dot,
                {
                    transform: [{ translateX }],
                    height: size,
                    width: size,
                    borderRadius: size / 2,
                },
                style,
            ]}/>)}
    </>);
}
const styles = StyleSheet.create({
    dot: {
        backgroundColor: 'white',
        elevation: 3,
        zIndex: 3,
        shadowColor: '#000000',
        shadowOffset: { width: 0, height: 4 },
        shadowOpacity: 0.12,
        shadowRadius: 4,
    }
});
