// @kid-ui/krn/src/components/dialog/dialog.tsx
import React from 'react';
import { Popup } from '../popup';
import { DialogComponent, DialogProps } from './dialog-wrapper';

type DialogApi = typeof DialogComponent & {
    alert: typeof show;
    confirm: typeof show;
};
export const Dialog: DialogApi = DialogComponent as any;
const getProps = (arg: DialogProps) => {
    const {
        useModal,
        showMask,
        maskBgColor,
        maskClosable,
        position,
        top,
        left,
        animate,
        animateDuration,
        isControlAnimationIn,
        popupStyle,
        beforeClose,
        onAnimateEnd,
        onShow,
        onDismiss,
        popupSet = {},
        ...dialogWrapperProps
    } = arg;
    const popupProps = {
        ...popupSet,
        useModal,
        showMask,
        maskBgColor,
        maskClosable,
        position,
        top,
        left,
        animate,
        animateDuration,
        isControlAnimationIn,
        popupStyle,
        beforeClose,
        onAnimateEnd,
        onShow,
        onDismiss,
    };
    return [popupProps, dialogWrapperProps];
};
const show = (
    arg?: DialogProps,
    Component: React.FC<DialogProps> = DialogComponent,
    showLeftButton: boolean = true,
) => {
    const [popupProps, dialogWrapperProps] = getProps(arg!);

    const alert = Popup.show({
        children: (
            <Component
                showLeftButton={showLeftButton}
                {...dialogWrapperProps}
                componentApi={false}
            />
        ),
        ...popupProps,
    });
    return {
        close: () => {
            alert.remove();
        },
        update: (props: DialogProps) => {
            const [updatePopupProps, updateDialogWrapperProps] =
                getProps(props);

            alert.update({
                children: (
                    <Component
                        showLeftButton={showLeftButton}
                        {...updateDialogWrapperProps}
                        componentApi={false}
                    />
                ),
                ...updatePopupProps,
            });
        },
    };
};
Dialog.alert = (props: DialogProps = {}) => {
    return show(props, DialogComponent, false);
};
Dialog.confirm = (props: DialogProps = {}) => {
    return show(props, DialogComponent);
};
export const DialogShow = show;
