// @kds/bridge-lite/lib/commonjs/global-config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setGlobalConfig = exports.globalConfig = exports.defaultSerializer = void 0;

var _mode = require("./mode");

var _version = require("./version");

const defaultSerializer = {
  stringify: params => JSON.stringify(params),
  parse: params => JSON.parse(params)
};
exports.defaultSerializer = defaultSerializer;
let rootTag = null;
let serializer = defaultSerializer;
const globalConfig = {
  get mode() {
    return (0, _mode.getCurrentMode)();
  },

  set mode(mode) {
    (0, _mode.setCurrentMode)(mode);
    console.info(`You're currently running version ${_version.version} in ${mode} mode`);
  },

  get rootTag() {
    return rootTag;
  },

  set rootTag(value) {
    rootTag = value;
  },

  get serializer() {
    return serializer;
  },

  set serializer(newSerializer) {
    serializer = newSerializer;
  },

  get isSetSerializer() {
    return serializer !== defaultSerializer;
  }

};
exports.globalConfig = globalConfig;

const setGlobalConfig = nextGlobalConfig => {
  Object.assign(globalConfig, nextGlobalConfig);
};

exports.setGlobalConfig = setGlobalConfig;
