// @kds/bridge-lite/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  addListener: true,
  removeListener: true,
  invoke: true,
  invokeSync: true,
  version: true,
  isBridgeError: true,
  setGlobalConfig: true
};
Object.defineProperty(exports, "addListener", {
  enumerable: true,
  get: function () {
    return _event.addListener;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "invoke", {
  enumerable: true,
  get: function () {
    return _invoke.invoke;
  }
});
Object.defineProperty(exports, "invokeSync", {
  enumerable: true,
  get: function () {
    return _invoke.invokeSync;
  }
});
Object.defineProperty(exports, "isBridgeError", {
  enumerable: true,
  get: function () {
    return _bridgeError.isBridgeError;
  }
});
Object.defineProperty(exports, "removeListener", {
  enumerable: true,
  get: function () {
    return _event.removeListener;
  }
});
Object.defineProperty(exports, "setGlobalConfig", {
  enumerable: true,
  get: function () {
    return _globalConfig.setGlobalConfig;
  }
});
Object.defineProperty(exports, "version", {
  enumerable: true,
  get: function () {
    return _version.version;
  }
});

var _event = require("./event");

var _invoke = require("./invoke");

var _version = require("./version");

var _bridgeError = require("./bridge-error");

var _globalConfig = require("./global-config");

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
var _default = {
  addListener: _event.addListener,
  removeListener: _event.removeListener,
  invoke: _invoke.invoke,
  invokeSync: _invoke.invokeSync,
  version: _version.version,
  isBridgeError: _bridgeError.isBridgeError,
  setGlobalConfig: _globalConfig.setGlobalConfig
};
exports.default = _default;
