// @locallife/biz-component/lib/module/components/popUp/index.js
import React from 'react';
import { KidIcon, AlertBottomCard } from '@kid-ui/krn';
import { View, StyleSheet } from 'react-native';
import { screenUnit } from '@lux/ts-lbs-krn-utils';
const styles = StyleSheet.create({
  close: {
    position: 'absolute',
    right: 0,
    paddingTop: screenUnit.screenSize(9)
  }
});

/**
 * 文档参考 https://kds-components.corp.kuaishou.com/markdown/kid-ui/components/alert-bottom-card.html
 * 注意使用alertBottomCard父组件需要套一层WrapContext
 * @param title 标题
 * @param renderContent 自定义内容组件
 * @param config 自定义配置 同名配置自动替换
 * @returns alert弹窗
 */
export const popUp = _ref => {
  let {
    title,
    renderContent,
    config = {}
  } = _ref;
  const alert = AlertBottomCard({
    content: '',
    title: title,
    titleStyle: {
      fontWeight: '500',
      fontSize: screenUnit.screenSize(18),
      color: '#222222',
      marginBottom: screenUnit.screenSize(8)
    },
    onConfirm: () => {
      alert.close();
    },
    renderCloseIcon: () => {
      return /*#__PURE__*/React.createElement(View, {
        style: styles.close
      }, /*#__PURE__*/React.createElement(KidIcon, {
        kid: 'common_dialog_colse_normal',
        kidColor: 'color_base_black_8',
        size: screenUnit.screenSize(24)
      }));
    },
    onPressCloseIcon: () => {
      alert.close();
    },
    // 自定义内容区域，在描述文字的下方渲染
    renderContent: () => renderContent,
    renderButton: () => null,
    popupStyle: {
      width: '100%',
      paddingHorizontal: 0,
      borderTopLeftRadius: screenUnit.screenSize(16),
      borderTopRightRadius: screenUnit.screenSize(16)
    },
    position: 'bottom',
    ...config
  });
  return alert;
};
