// @locallife/log/src/LocalLifeFmpLogger.ts
import { NativeModules } from 'react-native';
import LocalLifeInitLogger from './LocalLifeInitLogger';

export default class LocalLifeFmpLogger {
    private static rootTag;
    private static t1;
    private static t2;
    private static t3;
    private static extra;

    // 上报 t1
    // 白屏时间（T1）
    static handleT1() {
        if (this.rootTag) {
            throw new Error('You send fmp log already.');
        }
        let { rootTag, onCreateTimestamp, extra } =
            LocalLifeInitLogger.getAppModel();
        this.rootTag = rootTag;
        this.t1 = onCreateTimestamp;
        this.extra = extra;
        NativeModules?.RootViewPageBridge?.rootViewDidMountWith(
            this.rootTag,
            {},
        );
    }

    // 第一次接口上报时间(T2)
    static handleT2() {
        if (this.t2) {
            return;
        }
        if (this.t1) {
            this.t2 = Date.now() - this.t1;
        }
    }

    // 首屏时间（T3）
    static handleT3() {
        if (this.t3) {
            return;
        }
        if (this.t1) {
            this.t3 = Date.now() - this.t1;
        }
        if (this.t2 && this.t3 && this.t3 > this.t2) {
            NativeModules?.RootViewPageBridge?.reportPageRenderTime(
                this.rootTag,
                this.t2,
                this.t3,
                this.extra,
            );
        }
    }
}
