// @locallife/log/src/LocalLifeInitLogger.ts
import { Weblog } from '@ks/weblogger/lib/log.krnbridge';
import { PVEventOptions } from './types';

export interface _IAppInitProps {
    rootTag: number;
    BundleVersionCode: number;
    bundleId: string;
    componentName: string;

    [prop: string]: any;
}

export default class LocalLifeInitLogger {
    private static rootTag: number;
    private static webLog: Weblog;
    private static appModel: _IAppInitProps;
    private static pvEventOption: PVEventOptions;

    /**
     * deprecated use initWithAppModel(AppModel: _IAppInitProps, webLog: Weblog)
     * @param rootTag
     * @param webLog
     */
    static init(rootTag: number, webLog: Weblog) {
        this.rootTag = rootTag;
        this.webLog = webLog;
    }

    static initWithAppModel(AppModel: _IAppInitProps, webLog: Weblog) {
        this.rootTag = AppModel.rootTag;
        this.appModel = AppModel;
        this.webLog = webLog;
    }

    static setBizInfo(pvEventOption: PVEventOptions) {
        this.pvEventOption = pvEventOption;
    }

    static getPvEventOption(): PVEventOptions {
        if (__DEV__ && !this.pvEventOption) {
            throw new Error(
                'Please call LocalLifeInitLogger function setBizInfo',
            );
        }
        return this.pvEventOption;
    }

    static getAppModel(): _IAppInitProps {
        if (__DEV__ && !this.appModel) {
            console.warn(
                'LocalLifeInitLogger.init 已经过时了，请替换为 LocalLifeInitLogger.initWithAppModel',
            );
        }
        return this.appModel;
    }

    static getRootTag(): number {
        if (__DEV__ && !this.rootTag) {
            throw new Error('Please call LocalLifeInitLogger function init');
        }
        return this.rootTag;
    }

    static getWeblog(): Weblog {
        if (__DEV__ && !this.webLog) {
            throw new Error('Please call LocalLifeInitLogger function init');
        }
        return this.webLog;
    }
}
