// @kid-ui/krn/lib/components/textarea/styles.js
import React from 'react';
import { StyleSheet } from 'react-native';
import { ScaleContext } from '../context';
import useKidColor from '../../theme/useKidColor';
const useStyles = (props, others = {}) => {
    const { rem } = React.useContext(ScaleContext);
    const { count } = props;
    return StyleSheet.create({
        container: {
            paddingVertical: rem(8),
            backgroundColor: useKidColor('cs_common_background_secondary')
        },
        input: {
            marginHorizontal: rem(19),
            marginBottom: count > 0 ? rem(2) : 0,
            fontSize: rem(14),
            color: useKidColor('cs_common_text_title'),
            textAlignVertical: 'top'
        },
        count: {
            height: rem(20),
            marginHorizontal: rem(19),
            flexDirection: 'row',
            justifyContent: 'flex-end',
        },
        countLeft: {
            fontSize: rem(14),
            color: useKidColor('cs_common_tint_red')
        },
        countRight: {
            fontSize: rem(14),
            color: useKidColor('cs_common_text_least')
        }
    });
};
export default useStyles;
