// @kds-devtools/inspector/dist/Bridge.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Inspector_1 = __importDefault(require("./Inspector"));
class BridgeInspector extends Inspector_1.default {
    constructor() {
        super({
            name: 'Bridge',
            type: 'Bridge',
        });
        // @ts-ignore
        if (global.__NATIVE_MODULES_PROXY__) {
            // @ts-ignore
            global.__NATIVE_MODULES_PROXY__.on('invoke', (data) => {
                this.sendMessage({
                    type: 'notification',
                    data: {
                        method: 'Bridge.invoke',
                        params: data,
                    },
                });
            });
            // @ts-ignore
            global.__NATIVE_MODULES_PROXY__.on('callback', (data) => {
                this.sendMessage({
                    type: 'notification',
                    data: {
                        method: 'Bridge.callback',
                        params: data,
                    },
                });
            });
        }
    }
}
exports.default = BridgeInspector;
