// @kds-devtools/inspector/dist/Weblogger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Inspector_1 = __importDefault(require("./Inspector"));
class WebloggerInspector extends Inspector_1.default {
    constructor() {
        super({
            name: 'Weblogger',
            type: 'Weblogger',
        });
        // @ts-ignore
        if (global.__NATIVE_MODULES_PROXY__) {
            // @ts-ignore
            global.__NATIVE_MODULES_PROXY__.on('invoke', (data) => {
                if (data.namespace === 'yoda' && data.params[1] === 'setClientLog') {
                    this.sendMessage({
                        type: 'notification',
                        data: {
                            method: 'Weblogger.invoke',
                            params: data,
                        },
                    });
                }
            });
        }
    }
}
exports.default = WebloggerInspector;
