// @locallife/base-hooks/lib/module/useMemoizedFn/demo/index2.js
/* eslint-disable react-native/no-inline-styles */

/* eslint-disable react/jsx-no-bind */
import React, { useState } from 'react';
import { Button, Text } from 'react-native';
import { View } from 'react-native';
import { useCallback } from 'react';
import useMemoizedFn from '../index';
import { useRef } from 'react'; // 本demo中，可以看到使用 useMemoizedFn 会比 useCallback 减少渲染次数
// 注：点击 加1 按钮，使用 useCallback 的渲染次数会被加1，而 useMemoizedFn 的渲染次数不变
// 按 Command+S，触发页面的刷新，useCallback 和 useMemoizedFn 的渲染次数会同时加1

export const Demo2 = () => {
  const [count, setCount] = useState(0);

  const plusBySetState = () => {
    setCount(c => c + 1);
  };

  const callbackFn = useCallback(() => {
    console.log(`callbackFn：current count is ${count}`);
  }, [count]);
  const memoizedFn = useMemoizedFn(() => {
    console.log(`memoizedFn：current count is ${count}`);
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(Text, null, "count: ", count), /*#__PURE__*/React.createElement(Button, {
    title: "\u52A01",
    onPress: plusBySetState
  })), /*#__PURE__*/React.createElement(View, {
    style: {
      marginTop: 15
    }
  }, /*#__PURE__*/React.createElement(Text, null, "Component with useCallback function: "), /*#__PURE__*/React.createElement(ExpensiveTree, {
    title: "callback",
    showCount: callbackFn
  })), /*#__PURE__*/React.createElement(View, {
    style: {
      marginTop: 10
    }
  }, /*#__PURE__*/React.createElement(Text, null, "Component with useMemoizedFn function: "), /*#__PURE__*/React.createElement(ExpensiveTree, {
    title: "useMemoized",
    showCount: memoizedFn
  })));
};
const ExpensiveTree = /*#__PURE__*/React.memo(_ref => {
  let {
    title,
    showCount
  } = _ref;
  const renderCountRef = useRef(0);
  renderCountRef.current += 1;
  return /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(Text, null, "Render Count: ", renderCountRef.current), /*#__PURE__*/React.createElement(Button, {
    title: title,
    onPress: showCount
  }));
});
