// @krn/live-bridge/src/liveStream/index.ts
/* eslint-disable max-lines */
import {
    ELiveScene,
    ELiveType,
    ELoggerType,
    EMicChatType,
    IMockEffectInfoParams,
    ELogTaskEventParams,
    IAcceptPkOneMoreGameParams,
    IRejectPkOneMoreGameParams,
    IInvitePkOneMoreGameParams,
} from './interface';
import { ELiveSocketMessageType } from './messageType';
import {
    subscribeSocketMessage,
    unsubscribeSocketMessage,
} from './socketMessage';
import { liveRnBridgeExecute } from '../utils/liveRnBridgeExecute';
import type {
    ICdnUrl,
    IGetLiveSceneInfoResult,
    IInitLiveAPPProps,
    ILiveUserInfo,
    ILoggerContentPackage,
    ILoggerContentWrapper,
    ILoggerElementPackage,
    ILoggerUserPackage,
    IProfilePagePrefetchInfo,
    IWalletResponse,
    IEnableDismissTouchOutside,
    IFetchResourcePath,
    IFollowUser,
    IInjectCacheParams,
    ILiveRoomInfoResult,
    ILoggerParam,
    INavigateToH5,
    INotifySendGiftSuccessV2,
    IOpenCustomH5Params,
    IOpenLiveProfileCardParams,
    IOpenNonImmersiveH5Params,
    IProvideCache,
    IProvideTreasureBoxParams,
    IProvideTreasureBoxResult,
    IToastParams,
    IQueryWalletBalanceResult,
    ILogGiftPackage,
    ILogPhotoPackage,
    ILogResultPackage,
    ILogCommonParams,
    ILogTaskEventParams,
    IBatchFetchResourceParams,
} from './interface';
import type { TSocketMessageCallback } from './socketMessage';

/**
 * 设置H5点击空白区域是否可关闭 </br>
 * 支持版本: 9.9.20+ </br>
 * 注意：在无返回按钮的H5页面，禁用了点击空白区域关闭H5时候，要在合适时候开启，避免用户一直不能关闭页面。（最好能做兜底，在禁用点击空白位置关闭时，在H5页面增加关闭按钮）</br>
 * @param enableDismissTouchOutside
 */
export function enableDismissTouchOutside(
    params: IEnableDismissTouchOutside,
): Promise<object> {
    return liveRnBridgeExecute('enableDismissTouchOutside', params);
}

/**
 * 获取资源路径 </br>
 * 支持版本: 9.8.50+ </br>
 * @param params
 */
export function fetchResourcePath(
    params: IFetchResourcePath,
): Promise<{ resourcePath: string }> {
    return liveRnBridgeExecute('fetchResourcePath', params);
}

/**
 * 关注用户 </br>
 * 支持版本: 9.8.50+ </br>
 */
export function followUser(params: IFollowUser): Promise<object> {
    return liveRnBridgeExecute('followUser', params);
}

/**
 * 获取直播间场景信息 </br>
 * 支持版本: 9.8.50+ </br>
 */
export function getLiveSceneInfo(): Promise<IGetLiveSceneInfoResult> {
    return liveRnBridgeExecute('getLiveSceneInfo', {});
}

/**
 * 在rn容器上展示webview </br>
 * 支持版本: 9.8.50+ </br>
 * @param params
 */
export function navigateToH5(params: INavigateToH5): Promise<object> {
    return liveRnBridgeExecute('navigateToH5', params);
}

/**
 * 获取一些设备信息  </br>
 * 支持版本: 9.8.50+ </br>
 */
export function providePhoneInfo(): Promise<{ isLowPhone: boolean }> {
    return liveRnBridgeExecute('providePhoneInfo', {});
}

/**
 * 通知客户端送礼成功 </br>
 * 需要做 刷新钱包信息、发送LiveSentGiftEvent事件 的操作 </br>
 * 支持版本: 9.8.10+ </br>
 * @param walletResponse
 */
export function notifySendGiftSuccessV2(
    params: INotifySendGiftSuccessV2,
): Promise<object> {
    return liveRnBridgeExecute('notifySendGiftSuccessV2', params);
}

/**
 * 直播间埋点 </br>
 * Native端默认会带通用参数 liveStreamPackage、urlPackage、referUrlPackge  </br>
 * 支持版本: 9.6.20+ </br>
 * @param parma
 */
export function logger(params: ILoggerParam): Promise<object> {
    const tempParams = { ...params };
    if (!tempParams.contentPackage) {
        tempParams.contentPackage = {};
    }
    return liveRnBridgeExecute('logger', tempParams);
}

/**
 * 关闭当前页面 </br>
 * 支持版本: 9.6.10+ </br>
 * 半屏或弹框会关闭页面，TK挂件会从容器中删除。依赖于具体容器实现
 */
export function closePage(): Promise<object> {
    return liveRnBridgeExecute('closePage', {});
}

/**
 * 退出直播间 </br>
 * 支持版本: 9.10.30+ </br>
 * 退出直播间
 */
export function quitLive(): Promise<object> {
    return liveRnBridgeExecute('quitLive', {});
}

/**
 * 弹 native toast 消息 </br>
 * 支持版本: 9.6.10+ </br>
 */
export function toast(params: IToastParams): Promise<object> {
    return liveRnBridgeExecute('toast', params);
}

/**
 * 打开自定义H5页面 </br>
 * 支持版本: 8.3.60+ </br>
 */
export function openCustomH5(params: IOpenCustomH5Params): Promise<object> {
    return liveRnBridgeExecute('openCustomH5', { param: params });
}

/**
 * bridge打开非沉浸H5页面 </br>
 * 支持版本: 8.3.60+ </br>
 * 这个页面一定会带个Bar(状态栏) </br>
 * @param params
 */
export function openNonImmersiveH5(
    params: IOpenNonImmersiveH5Params,
): Promise<object> {
    return liveRnBridgeExecute('openNonImmersiveH5', { param: params });
}

/**
 * 获取直播间基础信息 </br>
 * 支持版本: 8.0.40+ </br>
 */
export function getLiveRoomInfo(): Promise<ILiveRoomInfoResult> {
    return liveRnBridgeExecute('getLiveRoomInfo', {});
}

/**
 * 获取宝盒业务缓存 </br>
 * 支持版本: 7.7.30+ </br>
 */
export function provideTreasureBoxParams(
    params: IProvideTreasureBoxParams,
): Promise<IProvideTreasureBoxResult> {
    return liveRnBridgeExecute('provideTreasureBoxParams', {
        param: params,
    });
}

/**
 * 向本地存入缓存 </br>
 * 支持版本: 7.7.20+ </br>
 */
export function injectCache(params: IInjectCacheParams): Promise<object> {
    return liveRnBridgeExecute('injectCache', {
        param: params,
    });
}

/**
 * 读取本地缓存 </br>
 * 支持版本: 7.7.20+ </br>
 * @param params
 */
export function provideCache(params: IProvideCache): Promise<object> {
    return liveRnBridgeExecute('provideCache', {
        param: params,
    });
}

/**
 * 打开直播个人卡片页 </br>
 * 支持版本: 7.4.10+ </br>
 * @param params
 */
export function openLiveProfileCard(
    params: IOpenLiveProfileCardParams,
): Promise<object> {
    return liveRnBridgeExecute('openLiveProfileCard', {
        param: params,
    });
}

/**
 * 客户端播放特效 </br>
 * 支持版本：9.10.40+ </br>
 * @param params
 */
export function mockEffectInfo(params: IMockEffectInfoParams) {
    return liveRnBridgeExecute('mockEffectInfo', params);
}

/**
 * 查询账户余额 </br>
 * 支持版本：9.9.30+ </br>
 */
export function queryWalletBalance(): Promise<IQueryWalletBalanceResult> {
    return liveRnBridgeExecute('queryWalletBalance', {});
}

/**
 * 发起重开一场PK的邀请 </br>
 * 支持版本: 10.4.40+ </br>
 */
export function invitePkOneMoreGame(
    params: IInvitePkOneMoreGameParams,
): Promise<object> {
    return liveRnBridgeExecute('invitePkOneMoreGame', params);
}

/**
 * 接受重开一场PK的邀请 </br>
 * 支持版本: 10.4.40+ </br>
 */
export function acceptPkOneMoreGame(
    params: IAcceptPkOneMoreGameParams,
): Promise<object> {
    return liveRnBridgeExecute('acceptPkOneMoreGame', params);
}

/**
 * 拒绝重开一场PK的邀请 </br>
 * 支持版本: 10.4.40+ </br>
 */
export function rejectPkOneMoreGame(
    params: IRejectPkOneMoreGameParams,
): Promise<object> {
    return liveRnBridgeExecute('rejectPkOneMoreGame', params);
}

/**
 * 批量获取文件路径</br>
 * 支持版本：9.10.30+ </br>
 */
export function batchFetchResourcePath(
    resInfoList: Array<IBatchFetchResourceParams>,
): Promise<object> {
    return liveRnBridgeExecute('batchFetchResourcePath', { resInfoList });
}

export {
    subscribeSocketMessage,
    unsubscribeSocketMessage,
    ELiveType,
    ELiveScene,
    EMicChatType,
    ELoggerType,
    ELiveSocketMessageType,
    ELogTaskEventParams,
};

export type {
    IInitLiveAPPProps,
    ICdnUrl,
    IProfilePagePrefetchInfo,
    ILiveUserInfo,
    IEnableDismissTouchOutside,
    IFetchResourcePath,
    IFollowUser,
    IGetLiveSceneInfoResult,
    INavigateToH5,
    IWalletResponse,
    INotifySendGiftSuccessV2,
    ILoggerElementPackage,
    ILoggerContentWrapper,
    ILoggerContentPackage,
    ILoggerUserPackage,
    ILoggerParam,
    IToastParams,
    IOpenCustomH5Params,
    IOpenNonImmersiveH5Params,
    ILiveRoomInfoResult,
    IProvideTreasureBoxParams,
    IProvideTreasureBoxResult,
    IInjectCacheParams,
    IProvideCache,
    IOpenLiveProfileCardParams,
    TSocketMessageCallback,
    IMockEffectInfoParams,
    IQueryWalletBalanceResult,
    ILogGiftPackage,
    ILogPhotoPackage,
    ILogResultPackage,
    ILogCommonParams,
    ILogTaskEventParams,
};
