// @locallife/biz-component/lib/commonjs/components/commonLabels/CommonLabels.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommonLabels = void 0;
var _reactNative = require("react-native");
var _SingleLineIcon = require("./SingleLineIcon");
var _SingleLineText = require("./SingleLineText");
var _react = _interopRequireWildcard(require("react"));
var _TagType = require("./TagType");
var _SingleLineIconAndText = require("./SingleLineIconAndText");
var _useParseBizHook = require("./hooks/useParseBizHook");
var _SingleLineTextIconText = require("./SingleLineTextIconText");
var _baseComponent = require("@locallife/base-component");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const styles = _reactNative.StyleSheet.create({
  externalContainer: {
    overflow: 'hidden',
    flexDirection: 'row',
    alignItems: 'center'
  },
  placeHolder: {
    flex: 1
  }
});
const CommonLabels = /*#__PURE__*/(0, _react.memo)(props => {
  const {
    bizTags,
    onPress,
    autoCut = true,
    adaptDark = true
  } = props;
  const [showBizTags, setShowBizTags] = (0, _useParseBizHook.useParseBizHook)(bizTags);

  // 超出最大宽度限制，需要删除后面元素，删除结束之前展示透明（不展示），避免看到标签删除的动画
  const [opacity, setOpacity] = (0, _react.useState)(0);

  // 标记是否已经执行了删除操作
  const hasDealDelete = (0, _react.useRef)(false);

  // 展示不下，删除元素
  const onLayout = (0, _react.useCallback)(_ref => {
    var _nativeEvent$layout;
    let {
      nativeEvent
    } = _ref;
    // 此view的宽度 <= 0代表标签展示不下了，需要删除标签
    if ((nativeEvent === null || nativeEvent === void 0 ? void 0 : (_nativeEvent$layout = nativeEvent.layout) === null || _nativeEvent$layout === void 0 ? void 0 : _nativeEvent$layout.width) <= 0) {
      const array = [...showBizTags];
      hasDealDelete.current = false;
      // 展示不下，删除最后一个元素
      while (array.length > 0 && !hasDealDelete.current) {
        var _model$contentModel, _model$contentModel$t, _model$contentModel3, _model$contentModel3$;
        const model = array[array.length - 1];
        if ((model === null || model === void 0 ? void 0 : (_model$contentModel = model.contentModel) === null || _model$contentModel === void 0 ? void 0 : (_model$contentModel$t = _model$contentModel.textInfos) === null || _model$contentModel$t === void 0 ? void 0 : _model$contentModel$t.length) ?? -1 > 0) {
          var _model$contentModel2, _model$contentModel2$;
          // 删除最后一个文案元素
          hasDealDelete.current = true;
          model === null || model === void 0 ? void 0 : (_model$contentModel2 = model.contentModel) === null || _model$contentModel2 === void 0 ? void 0 : (_model$contentModel2$ = _model$contentModel2.textInfos) === null || _model$contentModel2$ === void 0 ? void 0 : _model$contentModel2$.pop();
        } else if ((model === null || model === void 0 ? void 0 : (_model$contentModel3 = model.contentModel) === null || _model$contentModel3 === void 0 ? void 0 : (_model$contentModel3$ = _model$contentModel3.iconInfos) === null || _model$contentModel3$ === void 0 ? void 0 : _model$contentModel3$.length) ?? -1 > 0) {
          var _model$contentModel4, _model$contentModel4$;
          // 删除最后一个icon元素
          hasDealDelete.current = true;
          model === null || model === void 0 ? void 0 : (_model$contentModel4 = model.contentModel) === null || _model$contentModel4 === void 0 ? void 0 : (_model$contentModel4$ = _model$contentModel4.iconInfos) === null || _model$contentModel4$ === void 0 ? void 0 : _model$contentModel4$.pop();
        } else {
          // 这一组没有任何元素了，删除这一组
          array.pop();
        }
      }
      // 更新数据源
      setShowBizTags(array);
    } else {
      setOpacity(1);
    }
  }, [setShowBizTags, showBizTags]);
  return /*#__PURE__*/_react.default.createElement(_baseComponent.ShowWithData, {
    data: showBizTags.length > 0
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.externalContainer, props.containerStyle, {
      width: props.maxWidth,
      opacity: autoCut ? opacity : 1
    }]
  }, showBizTags === null || showBizTags === void 0 ? void 0 : showBizTags.map((info, index) => {
    var _info$contentModel;
    switch (info.styleKey) {
      case _TagType.TagType.LOCAL_LIFE_TAG_ITEM_ICON:
      case _TagType.TagType.LOCAL_LIFE_TAG_ITEM_ICON_COUNT_DOWN:
        // 展示一排图片
        return /*#__PURE__*/_react.default.createElement(_SingleLineIcon.SingleLineIcon, {
          data: info === null || info === void 0 ? void 0 : (_info$contentModel = info.contentModel) === null || _info$contentModel === void 0 ? void 0 : _info$contentModel.iconInfos,
          onPress: onPress,
          bizTags: info,
          adaptScreen: props.adaptScreen,
          key: index,
          adaptDark: adaptDark
        });
      case _TagType.TagType.LOCAL_LIFE_TAG_ITEM_TEXT:
        // 展示一排文案
        return /*#__PURE__*/_react.default.createElement(_SingleLineText.SingleLineText, {
          data: info === null || info === void 0 ? void 0 : info.contentModel,
          onPress: onPress,
          bizTags: info,
          adaptScreen: props.adaptScreen,
          autoRatio: props.autoRatio,
          ratio: props.ratio,
          key: index,
          adaptDark: adaptDark
        });
      case _TagType.TagType.LOCAL_LIFE_TAG_ICON_TEXT:
        // 展示左侧一排图标右侧一排文案
        return /*#__PURE__*/_react.default.createElement(_SingleLineIconAndText.SingleLineIconAndText, {
          data: info.contentModel,
          onPress: onPress,
          bizTags: info,
          adaptScreen: props.adaptScreen,
          key: index,
          adaptDark: adaptDark
        });
      case _TagType.TagType.LOCAL_LIFE_TAG_TEXT_ICON_TEXT:
        return /*#__PURE__*/_react.default.createElement(_SingleLineTextIconText.SingleLineTextIconText, {
          data: info.contentModel,
          onPress: onPress,
          bizTags: info,
          adaptScreen: props.adaptScreen,
          key: index,
          adaptDark: adaptDark
        });
    }
  }), autoCut && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.placeHolder,
    onLayout: onLayout
  })));
});
exports.CommonLabels = CommonLabels;
