// @locallife/biz-component/lib/commonjs/components/commonLabels/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BizTags", {
  enumerable: true,
  get: function () {
    return _Models.BizTags;
  }
});
Object.defineProperty(exports, "CommonLabels", {
  enumerable: true,
  get: function () {
    return _CommonLabels.CommonLabels;
  }
});
Object.defineProperty(exports, "CustomIcon", {
  enumerable: true,
  get: function () {
    return _CustomIcon.CustomIcon;
  }
});
Object.defineProperty(exports, "CustomText", {
  enumerable: true,
  get: function () {
    return _CustomText.CustomText;
  }
});
Object.defineProperty(exports, "SingleLineIcon", {
  enumerable: true,
  get: function () {
    return _SingleLineIcon.SingleLineIcon;
  }
});
Object.defineProperty(exports, "SingleLineIconAndText", {
  enumerable: true,
  get: function () {
    return _SingleLineIconAndText.SingleLineIconAndText;
  }
});
Object.defineProperty(exports, "SingleLineText", {
  enumerable: true,
  get: function () {
    return _SingleLineText.SingleLineText;
  }
});
Object.defineProperty(exports, "SingleLineTextIconText", {
  enumerable: true,
  get: function () {
    return _SingleLineTextIconText.SingleLineTextIconText;
  }
});
Object.defineProperty(exports, "TagContent", {
  enumerable: true,
  get: function () {
    return _Models.TagContent;
  }
});
Object.defineProperty(exports, "TagIconInfo", {
  enumerable: true,
  get: function () {
    return _Models.TagIconInfo;
  }
});
Object.defineProperty(exports, "TagIconStyle", {
  enumerable: true,
  get: function () {
    return _Models.TagIconStyle;
  }
});
Object.defineProperty(exports, "TagTextInfo", {
  enumerable: true,
  get: function () {
    return _Models.TagTextInfo;
  }
});
Object.defineProperty(exports, "TagTextStyle", {
  enumerable: true,
  get: function () {
    return _Models.TagTextStyle;
  }
});
var _CommonLabels = require("./CommonLabels");
var _CustomText = require("./CustomText");
var _CustomIcon = require("./CustomIcon");
var _SingleLineIcon = require("./SingleLineIcon");
var _SingleLineText = require("./SingleLineText");
var _SingleLineIconAndText = require("./SingleLineIconAndText");
var _SingleLineTextIconText = require("./SingleLineTextIconText");
var _Models = require("./Models");
