// @locallife/biz-component/lib/commonjs/components/commonLabels/SingleLineText.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SingleLineText = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _CustomText = require("./CustomText");
var _useCommonStyleHook = require("./hooks/useCommonStyleHook");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
// 一排文案

const styles = _reactNative.StyleSheet.create({
  container: {
    overflow: 'hidden',
    flexDirection: 'row',
    alignItems: 'center'
  }
});
// 一排文案
const SingleLineText = props => {
  var _data$textInfos;
  const {
    data,
    adaptScreen,
    adaptDark = true
  } = props;
  const {
    textInfos,
    contentStyle
  } = data || {};
  const onPressCard = (0, _react.useCallback)(() => {
    if (props.onPress) {
      props.onPress(props.bizTags);
    }
  }, [props]);
  const customStyle = (0, _useCommonStyleHook.useCommonStyleHook)(contentStyle, adaptScreen);
  return textInfos && (textInfos === null || textInfos === void 0 ? void 0 : textInfos.length) > 0 ? /*#__PURE__*/_react.default.createElement(_reactNative.TouchableWithoutFeedback, {
    onPress: onPressCard
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, customStyle]
  }, data === null || data === void 0 ? void 0 : (_data$textInfos = data.textInfos) === null || _data$textInfos === void 0 ? void 0 : _data$textInfos.map((item, index) => {
    return /*#__PURE__*/_react.default.createElement(_CustomText.CustomText, {
      textModel: item,
      key: index,
      adaptScreen: props.adaptScreen,
      autoRatio: props.autoRatio,
      adaptDark: adaptDark,
      ratio: props.ratio
    });
  }))) : /*#__PURE__*/_react.default.createElement(_reactNative.View, null);
};
exports.SingleLineText = SingleLineText;
