// @locallife/log/lib/module/LocalLifeRequestLogger.js
import { NativeModules, Platform } from 'react-native';
import LocalLifeInitLogger from './LocalLifeInitLogger';
import LocalLifeCustomLogger from './LocalLifeCustomLogger';
import AppVersion from './AppVersion';
import { VersionUtil } from './VersionUtil';

function inValidVersion() {
  if (Platform.OS === 'android') {
    return VersionUtil.currentGreaterThanEqualVersion(AppVersion.VERSION_11_4_40);
  } else if (Platform.OS === 'ios') {
    return VersionUtil.currentGreaterThanEqualVersion(AppVersion.VERSION_11_4_40);
  }

  return false;
}

export default class LocalLifeRequestLogger {
  static async logRequest(path, result, errCode, errMessage, params) {
    try {
      let args = params || {};
      let rootTag = LocalLifeInitLogger.getRootTag();
      let {
        bundleId,
        componentName
      } = LocalLifeInitLogger.getAppModel() || {};

      if (!bundleId || !componentName) {
        let bundleInfo = NativeModules.KRNBasic.getBundleInfo(rootTag);
        bundleId = bundleInfo.bundleId;
        componentName = bundleInfo.componentName;
      }

      args = { ...args,
        bundleId,
        componentName
      };

      if (inValidVersion()) {
        NativeModules.Kds.invoke('poi', 'reportRequestLog', JSON.stringify({
          path,
          result,
          errCode: errCode || -1,
          errMessage: errMessage || '',
          params: args || {},
          rootTag
        }));
      } else {
        LocalLifeCustomLogger.logCustomEvent('LOCAL_LIFE_LOG_REQUEST', {
          path,
          result,
          errCode: errCode || -1,
          errMessage: errMessage || '',
          params: args || {}
        });
      }

      console.log('[customEventKey: LOCAL_LIFE_LOG_REQUEST]' + '[path:' + path + '][result:' + result + '][errCode:' + (errCode || -1) + '][errMessage:' + (errMessage ? JSON.stringify(errMessage) : '') + '][params:' + (params ? JSON.stringify(params) : {}) + ']');
    } catch (e) {
      console.error(`LocalLifeRequestLogger error:${e || ''}`);
    }
  }

}
