// @locallife/log/lib/module/LocalLifeInitLogger.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

export default class LocalLifeInitLogger {
  /**
   * deprecated use initWithAppModel(AppModel: _IAppInitProps, webLog: Weblog)
   * @param rootTag
   * @param webLog
   */
  static init(rootTag, webLog) {
    this.rootTag = rootTag;
    this.webLog = webLog;
  }

  static initWithAppModel(AppModel, webLog) {
    this.rootTag = AppModel.rootTag;
    this.appModel = AppModel;
    this.webLog = webLog;
  }

  static setBizInfo(pvEventOption) {
    this.pvEventOption = pvEventOption;
  }

  static getPvEventOption() {
    if (__DEV__ && !this.pvEventOption) {
      throw new Error('Please call LocalLifeInitLogger function setBizInfo');
    }

    return this.pvEventOption;
  }

  static getAppModel() {
    if (__DEV__ && !this.appModel) {
      console.warn('LocalLifeInitLogger.init 已经过时了，请替换为 LocalLifeInitLogger.initWithAppModel');
    }

    return this.appModel;
  }

  static getRootTag() {
    if (__DEV__ && !this.rootTag) {
      throw new Error('Please call LocalLifeInitLogger function init');
    }

    return this.rootTag;
  }

  static getWeblog() {
    if (__DEV__ && !this.webLog) {
      throw new Error('Please call LocalLifeInitLogger function init');
    }

    return this.webLog;
  }

}

_defineProperty(LocalLifeInitLogger, "rootTag", void 0);

_defineProperty(LocalLifeInitLogger, "webLog", void 0);

_defineProperty(LocalLifeInitLogger, "appModel", void 0);

_defineProperty(LocalLifeInitLogger, "pvEventOption", void 0);
