// @locallife/biz-hooks/lib/module/useOnDarkModeChange/demo/index.js
import React from 'react';
import { Text, View, StyleSheet } from 'react-native';
import useOnDarkModeChange from '..';
const styles = StyleSheet.create({
  container: {
    alignItems: 'center'
  },
  row: {
    flexDirection: 'row'
  },
  smallText: {
    fontSize: 13
  },
  bigText: {
    fontSize: 18
  },
  red: {
    color: 'red'
  },
  margin: {
    marginTop: 20
  }
});
export default (() => {
  const {
    changing,
    isDarkMode
  } = useOnDarkModeChange(isDark => {
    console.log(`isDark = ${isDark}`);
  });
  console.log(`changing = ${changing}`);
  let result;

  switch (isDarkMode) {
    case true:
      result = '是';
      break;

    case false:
      result = '否';
      break;

    default:
      result = '未知';
      break;
  }

  return /*#__PURE__*/React.createElement(View, {
    style: styles.container
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.row
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.bigText
  }, "\u5F53\u524D\u662F\u5426\u662F\u6DF1\u8272\u6A21\u5F0F\uFF1A"), /*#__PURE__*/React.createElement(Text, {
    style: [styles.bigText, styles.red]
  }, result)), /*#__PURE__*/React.createElement(Text, {
    style: [styles.smallText, styles.margin]
  }, "\u8BF7\u5230\u8BBE\u7F6E\u4E2D\u5207\u6362\u6DF1\u8272\u6A21\u5F0F"));
});
