// @kid-ui/krn/lib/components/tag/styles.js
import { StyleSheet, } from 'react-native';
const useStyles = (props, others) => {
    const { size } = props || {};
    const isSmall = size === 'small';
    const isBig = size === 'big';
    const paddingVertical = isBig ? 3 : (isSmall ? 1 : 2);
    const borderRadius = isSmall ? 3 : 4;
    const backgroundColor = '#FE3666';
    const fontSize = isBig ? 12 : (isSmall ? 9 : 11);
    const lineHeight = isBig ? 18 : (isSmall ? 14 : 16);
    const color = '#FFF';
    return StyleSheet.create({
        wrapperStyle: {
            paddingVertical,
            borderRadius,
            backgroundColor,
            paddingHorizontal: 4,
            flexDirection: 'row',
            justifyContent: 'center',
            alignItems: 'center',
        },
        textStyle: {
            fontSize,
            lineHeight,
            color,
            fontWeight: '500',
        },
        leftIconStyle: {
            marginRight: 1
        },
        rightIconStyle: {
            marginLeft: 1
        },
        avatarStyle: {
            borderWidth: 1,
            // borderColor: '#FE3666'
        },
        firstAvatarStyle: {
            zIndex: 999,
        },
        notFirstAvatarStyle: {
        // left: -8
        },
        lineStyle: {
            width: 1,
            height: 10,
            marginHorizontal: 4,
            backgroundColor: 'rgba(255, 255, 255, 0.2)'
        },
        descriptionTextStyle: {
            fontSize,
            lineHeight,
            color,
            fontWeight: '500',
        }
    });
};
export default useStyles;
