// @locallife/biz-hooks/lib/module/useOnTextLayout/demo/index.js
import React from 'react';
import { Text } from 'react-native';
import useOnTextLayout from '..';
import { StyleSheet } from 'react-native';
const textStyle = StyleSheet.create({
  text: {
    fontSize: 14,
    color: 'black',
    marginBottom: 20
  },
  originText: {
    fontSize: 20,
    color: 'red'
  },
  hasNextLine: {
    fontSize: 20,
    color: 'orange'
  },
  singleLine: {
    fontSize: 20,
    color: 'blue'
  },
  nextLine: {
    fontSize: 20,
    color: 'green'
  }
});
export default (() => {
  const text = '鲁迅（1881-1936），原名周樟寿，后改名周树人，字豫才，浙江绍兴人。20世纪中国重要作家，新文化运动的领导人、左翼文化运动的支持者。毛泽东评价他是伟大的无产阶级的文学家、思想家、革命家，是中国文化革命的主将，也被人民称为“民族魂”。';
  const [textInfo, onTextLayout] = useOnTextLayout(text);
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Text, {
    style: textStyle.originText
  }, "\u5C55\u793A\u539F\u6587"), /*#__PURE__*/React.createElement(Text, {
    style: textStyle.text // @ts-ignore
    ,
    onTextLayout: onTextLayout
  }, textInfo.text), /*#__PURE__*/React.createElement(Text, {
    style: textStyle.hasNextLine
  }, "\u662F\u5426\u6709\u4E0B\u4E00\u884C"), /*#__PURE__*/React.createElement(Text, {
    style: textStyle.text
  }, textInfo.hasNextLine ? '有下一行' : '没下一行'), /*#__PURE__*/React.createElement(Text, {
    style: textStyle.singleLine
  }, "\u5C55\u793A\u4E00\u884C"), /*#__PURE__*/React.createElement(Text, {
    style: textStyle.text
  }, textInfo.firstLineText), /*#__PURE__*/React.createElement(Text, {
    style: textStyle.nextLine
  }, "\u5C55\u793A\u5269\u4F59\u90E8\u5206"), /*#__PURE__*/React.createElement(Text, {
    style: textStyle.text
  }, textInfo.nextLineText));
});
