// @ks/weblogger/dist/umd/plugins/kdsradar.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
    typeof define === 'function' && define.amd ? define(factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, global.Radar = factory());
})(this, (function () { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */

    /* global Reflect, Promise */
    var _extendStatics = function extendStatics(d, b) {
      _extendStatics = Object.setPrototypeOf || {
        __proto__: []
      } instanceof Array && function (d, b) {
        d.__proto__ = b;
      } || function (d, b) {
        for (var p in b) {
          if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p];
        }
      };

      return _extendStatics(d, b);
    };

    function __extends(d, b) {
      if (typeof b !== "function" && b !== null) throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");

      _extendStatics(d, b);

      function __() {
        this.constructor = d;
      }

      d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var _assign = function __assign() {
      _assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
          s = arguments[i];

          for (var p in s) {
            if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
          }
        }

        return t;
      };

      return _assign.apply(this, arguments);
    };
    function __awaiter(thisArg, _arguments, P, generator) {
      function adopt(value) {
        return value instanceof P ? value : new P(function (resolve) {
          resolve(value);
        });
      }

      return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) {
          try {
            step(generator.next(value));
          } catch (e) {
            reject(e);
          }
        }

        function rejected(value) {
          try {
            step(generator["throw"](value));
          } catch (e) {
            reject(e);
          }
        }

        function step(result) {
          result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
        }

        step((generator = generator.apply(thisArg, _arguments || [])).next());
      });
    }
    function __generator(thisArg, body) {
      var _ = {
        label: 0,
        sent: function sent() {
          if (t[0] & 1) throw t[1];
          return t[1];
        },
        trys: [],
        ops: []
      },
          f,
          y,
          t,
          g;
      return g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
      }, typeof Symbol === "function" && (g[Symbol.iterator] = function () {
        return this;
      }), g;

      function verb(n) {
        return function (v) {
          return step([n, v]);
        };
      }

      function step(op) {
        if (f) throw new TypeError("Generator is already executing.");

        while (_) {
          try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];

            switch (op[0]) {
              case 0:
              case 1:
                t = op;
                break;

              case 4:
                _.label++;
                return {
                  value: op[1],
                  done: false
                };

              case 5:
                _.label++;
                y = op[1];
                op = [0];
                continue;

              case 7:
                op = _.ops.pop();

                _.trys.pop();

                continue;

              default:
                if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                  _ = 0;
                  continue;
                }

                if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                  _.label = op[1];
                  break;
                }

                if (op[0] === 6 && _.label < t[1]) {
                  _.label = t[1];
                  t = op;
                  break;
                }

                if (t && _.label < t[2]) {
                  _.label = t[2];

                  _.ops.push(op);

                  break;
                }

                if (t[2]) _.ops.pop();

                _.trys.pop();

                continue;
            }

            op = body.call(thisArg, _);
          } catch (e) {
            op = [6, e];
            y = 0;
          } finally {
            f = t = 0;
          }
        }

        if (op[0] & 5) throw op[1];
        return {
          value: op[0] ? op[1] : void 0,
          done: true
        };
      }
    }
    function __spreadArray(to, from, pack) {
      if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
          if (!ar) ar = Array.prototype.slice.call(from, 0, i);
          ar[i] = from[i];
        }
      }
      return to.concat(ar || Array.prototype.slice.call(from));
    }

    var BasePlugin = /** @class */ (function () {
        function BasePlugin() {
        }
        BasePlugin.prototype.apply = function (weblog) {
            this.weblog = weblog;
        };
        return BasePlugin;
    }());

    // 统一的 custom_key
    var KEY = 'radar_log';
    var RADAR_KEY;
    (function (RADAR_KEY) {
        // 页面加载监控
        RADAR_KEY["LOAD"] = "load";
        // 静态资源监控
        RADAR_KEY["RES"] = "resource";
        // API 性能监控
        RADAR_KEY["API"] = "api";
        // 异常
        RADAR_KEY["ERROR"] = "error";
        // 脚本自身上报异常
        RADAR_KEY["SDK_ERROR"] = "sdk_error";
        // 用户自定义性能监控
        RADAR_KEY["CUSTOM"] = "custom";
        // 批量日志，用于处理批量情况
        RADAR_KEY["BATCH"] = "batch";
    })(RADAR_KEY || (RADAR_KEY = {}));
    var LEGACY_RADAR_KEY;
    (function (LEGACY_RADAR_KEY) {
        // 页面加载监控
        LEGACY_RADAR_KEY["LOAD"] = "load_statistics";
        // 静态资源监控
        LEGACY_RADAR_KEY["RES"] = "resource_statistics";
        // API 性能监控
        LEGACY_RADAR_KEY["API"] = "api_statistics";
        // 异常
        LEGACY_RADAR_KEY["ERROR"] = "error_statistics";
        // 脚本自身上报异常
        LEGACY_RADAR_KEY["SDK_ERROR"] = "sdk_error";
        // 用户自定义性能监控
        LEGACY_RADAR_KEY["CUSTOM"] = "custom_statistics";
    })(LEGACY_RADAR_KEY || (LEGACY_RADAR_KEY = {}));
    var ERROR_KEY;
    (function (ERROR_KEY) {
        // 脚本异常
        ERROR_KEY["SCRIPT"] = "script";
        // 资源加载异常
        ERROR_KEY["RES"] = "res";
        // 接口请求异常
        ERROR_KEY["API"] = "api";
        // 视频异常
        ERROR_KEY["VIDEO"] = "video";
    })(ERROR_KEY || (ERROR_KEY = {}));
    var DEFAULT_IGNORE_LIST = [
        '//wlog.kuaishou.com/rest/n/log/web/collect',
        '/rest/wd/common/log/collect/misc2',
    ];

    var reProtocol = /^(https?:)?\/\//;
    function normalizeURL(ignoreList) {
        if (ignoreList === void 0) { ignoreList = []; }
        var list = ignoreList.concat(DEFAULT_IGNORE_LIST);
        // 把协议头剃掉，防止出现协议头不一致
        return list.map(function (url) { return url.replace(reProtocol, ''); });
    }
    // 计算雷达自身耗时
    var radarCost = {};
    // 保留拓展能力
    function radarPoint(name) {
        var costName = radarCost[name];
        if (!costName) {
            radarCost[name] = { startTime: new Date().getTime() };
        }
        else if (!costName.endTime) {
            costName.endTime = new Date().getTime();
            costName.duration = costName.endTime - costName.startTime;
            delete costName.startTime;
            delete costName.endTime;
        }
    }

    // const REPORT_THRESSHOLD = 60 * 1000;
    // const PAGELOG_API = 'sendRadarLog';
    var RadarRoot = /** @class */ (function (_super) {
        __extends(RadarRoot, _super);
        function RadarRoot(options) {
            var _this = _super.call(this) || this;
            _this.plugins = [];
            _this.logQueue = [];
            _this.queueConfig = {
                wait: 1500,
                maxBatchLength: 200,
            };
            _this.batchTimer = 0;
            _this.radarSessionId = '';
            _this.isUsingLegacy = null;
            _this.isUsingDetachedReport = null;
            // 默认忽略列表
            _this.ignoreList = ['https://web-trace.ksapisrv.com/ktrace/collect'];
            radarPoint('radarCreated-firstReport');
            radarPoint('radarCreated-onload');
            _this.options = options;
            _this.queueConfig = _assign(_assign({}, _this.queueConfig), (_this.options ? _this.options.queue : {}));
            _this.ignoreList = _this.ignoreList.concat(normalizeURL(_this.options && _this.options.ignoreList));
            return _this;
        }
        RadarRoot.prototype.use = function (plugin, ctx) {
            this.plugins.push(new plugin(ctx, this.options));
        };
        RadarRoot.prototype.apply = function (weblog) {
            this.weblog = weblog;
            // @ts-ignore
            this.ignoreList.push(this.weblog.logger.url);
            //  生成唯一sessionid
            this.radarSessionId = this.nanoId();
            this.created();
        };
        RadarRoot.prototype.created = function () {
            this.plugins.forEach(function (plugin) {
                plugin.created();
            });
        };
        RadarRoot.prototype.destroy = function () {
            this.plugins.forEach(function (plugin) {
                plugin.destroy();
            });
        };
        RadarRoot.prototype.flush = function () {
            if (!this.options) {
                return;
            }
            if (this.logQueue.length <= 0) {
                return;
            }
            var h5_extra_attr = JSON.stringify(_assign(_assign({}, this.weblog.commonPackage.getH5ExtraAttr()), { url: location.href, hash: location.hash }));
            this.weblog.commonPackage.app_package;
            // const commonParams = {
            //     project_id: this.options.projectId,
            //     app_version_name: app_package.app_version_name,
            //     // @ts-ignore
            //     url_package: this.currentUrlPackage && this.currentUrlPackage.toJSON(),
            //     h5_extra_attr,
            // };
            // 桥上报，目前不支持，注释掉
            // this.addRadarStatEvent({
            //     removeStashedLog: [],
            //     sendImmediate: true,
            //     customData: {
            //         data: [this.logQueue],
            //         ...commonParams,
            //     },
            // });
            var commonLogData = {
                project_id: this.options.projectId,
                radar_session_id: this.radarSessionId,
                h5_extra_attr: h5_extra_attr,
            };
            this.weblog.collect('CUSTOM', {
                key: KEY,
                value: _assign(_assign({}, commonLogData), { data: this.logQueue }),
                // @ts-ignore
                currentUrlPackage: this.currentUrlPackage && this.currentUrlPackage.toJSON(),
                // @ts-ignore
                referUrlPackage: this.referUrlPackage && this.referUrlPackage.toJSON(),
            });
            this.logQueue = [];
        };
        RadarRoot.prototype.decorateLog = function (kv) {
            if (!this.currentUrlPackage) {
                this.currentUrlPackage = this.weblog.currentUrlPackage;
            }
            // 由于 url 是日志公参，因此 url 发生变化时直接冲刷日志
            if (this.currentUrlPackage.page !== this.weblog.currentUrlPackage.page) {
                this.flush();
                this.currentUrlPackage = this.weblog.currentUrlPackage;
            }
        };
        RadarRoot.prototype.logCollect = function (kv) {
            var _this = this;
            if (!this.options) {
                return;
            }
            if (this.batchTimer) {
                clearTimeout(this.batchTimer);
            }
            this.decorateLog(kv);
            this.logQueue.push(kv);
            if (this.logQueue.length > this.queueConfig.maxBatchLength) {
                this.flush();
                return;
            }
            this.batchTimer = setTimeout(function () {
                _this.flush();
            }, this.queueConfig.wait);
        };
        RadarRoot.prototype.addRadarStatEvent = function (log) {
            try {
                var yoda = this.weblog.Utils.yoda;
                if (!yoda) {
                    return false;
                }
                yoda.sendRadarLog(log);
                return true;
            }
            catch (e) {
                return false;
            }
        };
        /**
         * 生成nano id
         * 目前radarsessionid生成逻辑与weblog相同，防止weblog修改，雷达自身保存一份
         * @return {string} 16长度的字符串
         */
        RadarRoot.prototype.nanoId = function () {
            var ts = ((Math.random() * 1e9) >>> 0);
            var randomer = [];
            var seed = '0123456789ABCDEF';
            for (var i = 0; i < 7; i++) {
                randomer.push(seed.charAt(Math.random() * 16));
            }
            return ts + randomer.join('');
        };
        return RadarRoot;
    }(BasePlugin));

    var sdkName = 'webLogger';
    function error() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        try {
            var error_1 = sdkName === 'webLogger_krn' ? console === null || console === void 0 ? void 0 : console.log : console === null || console === void 0 ? void 0 : console.error;
            return error_1 && error_1.call.apply(error_1, __spreadArray([console], args, false));
        }
        catch (e) {
            return;
        }
    }

    /**
     * @file kds Radar 插件
     * @author guwanpo@kuaishou.com
     */
    var Radar = /** @class */ (function (_super) {
        __extends(Radar, _super);
        function Radar(options) {
            var _this = _super.call(this, options) || this;
            if (!_this.validOption()) {
                return _this;
            }
            return _this;
        }
        // 采样控制
        Radar.samplingControl = function (options) {
            if (!options) {
                return false;
            }
            var sampling = options.sampling;
            var rand = Math.random();
            return rand < sampling;
        };
        // 用户自定义系列指标
        Radar.prototype.fmp = function (time) {
            return __awaiter(this, void 0, void 0, function () {
                var radarFmp, yoda, kdsPerf, _a, fmp;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            radarFmp = time || Date.now();
                            yoda = this.weblog.Utils.yoda;
                            _a = yoda;
                            if (!_a) return [3 /*break*/, 2];
                            return [4 /*yield*/, yoda.getKDSWebviewLoadPerf()];
                        case 1:
                            _a = (_b.sent());
                            _b.label = 2;
                        case 2:
                            kdsPerf = _a;
                            console.log(kdsPerf, 'kdsPerf');
                            if (!kdsPerf) {
                                error('[Radar]未拿到kds性能数据');
                                return [2 /*return*/];
                            }
                            fmp = radarFmp - kdsPerf.startLoad;
                            this.logCollect({
                                key: RADAR_KEY.CUSTOM,
                                value: {
                                    fmp: fmp,
                                },
                            });
                            this.flush();
                            return [2 /*return*/];
                    }
                });
            });
        };
        Radar.prototype.validOption = function () {
            if (!this.options) {
                // 未传入参数默认不开启
                return false;
            }
            var _a = this.options, sampling = _a.sampling, projectId = _a.projectId;
            if (typeof projectId !== 'string') {
                error('[error 303]', '[Radar] 必须配置 projectId 属性，类型为 string!');
                return false;
            }
            if (sampling > 1 || sampling < 0) {
                error('[error 304]', '[Radar] 采样率 sampling 必须为 0~1 之间的数字');
                return false;
            }
            return true;
        };
        Radar.key = 'radar';
        return Radar;
    }(RadarRoot));

    return Radar;

}));
