// @ks/weblogger/dist/umd/plugins/radar.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, global.Radar = factory());
})(this, (function () { 'use strict';

  function _typeof(obj) {
    "@babel/helpers - typeof";

    if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
      _typeof = function (obj) {
        return typeof obj;
      };
    } else {
      _typeof = function (obj) {
        return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
      };
    }

    return _typeof(obj);
  }

  /******************************************************************************
  Copyright (c) Microsoft Corporation.

  Permission to use, copy, modify, and/or distribute this software for any
  purpose with or without fee is hereby granted.

  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
  REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
  AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
  LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
  OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  PERFORMANCE OF THIS SOFTWARE.
  ***************************************************************************** */

  /* global Reflect, Promise */
  var _extendStatics = function extendStatics(d, b) {
    _extendStatics = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (d, b) {
      d.__proto__ = b;
    } || function (d, b) {
      for (var p in b) {
        if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p];
      }
    };

    return _extendStatics(d, b);
  };

  function __extends(d, b) {
    if (typeof b !== "function" && b !== null) throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");

    _extendStatics(d, b);

    function __() {
      this.constructor = d;
    }

    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
  }

  var _assign = function __assign() {
    _assign = Object.assign || function __assign(t) {
      for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];

        for (var p in s) {
          if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
      }

      return t;
    };

    return _assign.apply(this, arguments);
  };
  function __spreadArray(to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
      if (ar || !(i in from)) {
        if (!ar) ar = Array.prototype.slice.call(from, 0, i);
        ar[i] = from[i];
      }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
  }

  var _e2 = function e() {
    return (_e2 = Object.assign || function (t) {
      for (var e, n = 1, r = arguments.length; n < r; n++) {
        for (var o in e = arguments[n]) {
          Object.prototype.hasOwnProperty.call(e, o) && (t[o] = e[o]);
        }
      }

      return t;
    }).apply(this, arguments);
  };

  function t(t, a, s, c) {
    return new (s = s || Promise)(function (n, e) {
      function r(t) {
        try {
          i(c.next(t));
        } catch (t) {
          e(t);
        }
      }

      function o(t) {
        try {
          i(c["throw"](t));
        } catch (t) {
          e(t);
        }
      }

      function i(t) {
        var e;
        t.done ? n(t.value) : ((e = t.value) instanceof s ? e : new s(function (t) {
          t(e);
        })).then(r, o);
      }

      i((c = c.apply(t, a || [])).next());
    });
  }

  function n(r, o) {
    var i,
        a,
        s,
        c = {
      label: 0,
      sent: function sent() {
        if (1 & s[0]) throw s[1];
        return s[1];
      },
      trys: [],
      ops: []
    },
        t = {
      next: e(0),
      "throw": e(1),
      "return": e(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function () {
      return this;
    }), t;

    function e(n) {
      return function (t) {
        var e = [n, t];
        if (i) throw new TypeError("Generator is already executing.");

        for (; c;) {
          try {
            if (i = 1, a && (s = 2 & e[0] ? a["return"] : e[0] ? a["throw"] || ((s = a["return"]) && s.call(a), 0) : a.next) && !(s = s.call(a, e[1])).done) return s;

            switch (a = 0, (e = s ? [2 & e[0], s.value] : e)[0]) {
              case 0:
              case 1:
                s = e;
                break;

              case 4:
                return c.label++, {
                  value: e[1],
                  done: !1
                };

              case 5:
                c.label++, a = e[1], e = [0];
                continue;

              case 7:
                e = c.ops.pop(), c.trys.pop();
                continue;

              default:
                if (!(s = 0 < (s = c.trys).length && s[s.length - 1]) && (6 === e[0] || 2 === e[0])) {
                  c = 0;
                  continue;
                }

                if (3 === e[0] && (!s || e[1] > s[0] && e[1] < s[3])) {
                  c.label = e[1];
                  break;
                }

                if (6 === e[0] && c.label < s[1]) {
                  c.label = s[1], s = e;
                  break;
                }

                if (s && c.label < s[2]) {
                  c.label = s[2], c.ops.push(e);
                  break;
                }

                s[2] && c.ops.pop(), c.trys.pop();
                continue;
            }

            e = o.call(r, c);
          } catch (t) {
            e = [6, t], a = 0;
          } finally {
            i = s = 0;
          }
        }

        if (5 & e[0]) throw e[1];
        return {
          value: e[0] ? e[1] : void 0,
          done: !0
        };
      };
    }
  }

  function r$1(t, e) {
    var n = "function" == typeof Symbol && t[Symbol.iterator];
    if (!n) return t;
    var r,
        o,
        i = n.call(t),
        a = [];

    try {
      for (; (void 0 === e || 0 < e--) && !(r = i.next()).done;) {
        a.push(r.value);
      }
    } catch (t) {
      o = {
        error: t
      };
    } finally {
      try {
        r && !r.done && (n = i["return"]) && n.call(i);
      } finally {
        if (o) throw o.error;
      }
    }

    return a;
  }

  function o$1(t, e, n) {
    if (n || 2 === arguments.length) for (var r, o = 0, i = e.length; o < i; o++) {
      !r && o in e || ((r = r || Array.prototype.slice.call(e, 0, o))[o] = e[o]);
    }
    return t.concat(r || Array.prototype.slice.call(e));
  }

  var l$3,
      f$4,
      i = "ksurl.cn/csHfgepO",
      a$2 = console,
      c$1 = a$2.log.bind(a$2, i + ", radar:"),
      s$2 = a$2.warn.bind(a$2, i + ", radar:"),
      u$4 = a$2.error.bind(a$2, i + ", radar:"),
      p$3 = function () {
    function t() {
      for (var t, e = [], n = 0; n < arguments.length; n++) {
        e[n] = arguments[n];
      }

      this.middlewares = [], (t = this.middlewares).push.apply(t, o$1([], r$1(e), !1));
    }

    return t.prototype.use = function () {
      for (var t, e = [], n = 0; n < arguments.length; n++) {
        e[n] = arguments[n];
      }

      var r = this.middlewares.pop();
      (t = this.middlewares).push.apply(t, o$1([], r$1(e), !1)), this.middlewares.push(r);
    }, t.prototype.invoke = function (t) {
      return a = this.middlewares, r = t, i = -1, function t(e) {
        if (e <= i) return new Error("`next` 被重复调用");
        var n = a[i = e];
        if (n = e === a.length ? o : n) try {
          return n(r, t.bind(null, e + 1));
        } catch (e) {
          return e;
        }
      }(0);
      var r, o, i, a;
    }, t;
  }(),
      d$3 = ((f$4 = l$3 = l$3 || {}).LOAD = "load", f$4.RES = "resource", f$4.API = "api", f$4.ERROR = "error", f$4.SDK_ERROR = "sdk_error", f$4.CUSTOM = "custom", f$4.BATCH = "batch", f$4.EVENT = "event", f$4.BLOOD = "blood", function () {
    function t(t) {
      void 0 === t && (t = {});
      var e = this;
      this.broadcastSampleFb = function () {
        var t = e.getSample(e.baseOptions.sampling);
        t !== e.sampling && (e.sampling = t, e.sampling && (e.isHit = e.samplingControl(e.sampling)));
      }, void 0 === t && De("100"), t.key || De("101"), !t.core || "object" == _typeof(t.core.sampleData) && "function" == typeof t.core.collect || De("121"), this.baseOptions = t, this.key = t.key, this.version = t.version, this.name = t.name, this.isHit = !1, t.core && this.updateCore(t.core);
    }

    return t.prototype.updateCore = function (t) {
      var e;
      "object" == _typeof(t.sampleData) && "function" == typeof t.collect || De("121"), null != (e = this.core) && e.broadcastSample.off(this.broadcastSampleFb), this.core = t, this.sampling = this.getSample(this.baseOptions.sampling), this.sampling && (this.isHit = this.samplingControl(this.sampling)), null != (t = null == (e = this.core) ? void 0 : e.broadcastSample) && t.on(this.broadcastSampleFb);
    }, t.prototype.getSample = function (t) {
      var e = null != (e = null == (e = this.core) ? void 0 : e.sampleData.core) ? e : 1;
      if (this.key === l$3.LOAD) return e;
      if ("number" != typeof (t = void 0 === t ? 1 : t) || t < 0 || 1 < t) return s$2("102", "".concat(this.key, "，采用Core的采样率").concat(null == (n = this.core) ? void 0 : n.sampleData.core)), e;
      var n = null != (n = this.core) && n.isInYoda ? void 0 === (null == (n = this.core) ? void 0 : n.sampleData[this.key]) ? 1 : this.core.sampleData[this.key] : t;
      return e && e < n ? (s$2("115", "".concat(this.key)), e) : n;
    }, t.prototype.samplingControl = function (t) {
      var e;
      return !(null == (e = this.core) || !e.sampleData.isHit) && (null == (e = this.core) ? void 0 : e.samplingControl(t / this.core.sampleData.core));
    }, t.prototype.baseCollect = function (t) {
      if (this.core) {
        if (this.isHit) {
          var r = t.value || {},
              r = Object.keys(r).filter(function (t) {
            return null !== r[t] && void 0 !== r[t];
          }).reduce(function (t, n) {
            return _e2(_e2({}, t), ((t = {})[n] = r[n], t));
          }, {});

          try {
            this.core.collect(t.key || this.key, _e2(_e2({
              collect_version: this.version,
              collect_name: this.name
            }, t.dimension), {
              sample_rate: this.sampling
            }), r);
          } catch (t) {
            s$2("126");
          }
        }
      } else s$2("110", "".concat(this.key));
    }, t;
  }());

  function h$3() {
    var t,
        e = "undefined" != typeof window ? window.performance : null;
    return e && e.timing ? "function" == typeof e.timing.toJSON ? ((t = e.timing.toJSON()) && !t.redirectStart && (t.redirectStart = t.navigationStart), t) : e.timing : null;
  }

  var v$4 = ["//wlog.kuaishou.com/rest/n/log/web/collect", "/rest/wd/common/log/collect/misc2", "/rest/wd/common/log/collect/radar", "/rest/kd/log/collect"],
      m$2 = window,
      _$3 = "__radarSeedStore__",
      b$3 = "isCoreReady",
      x = "isRuning",
      S$2 = "radarSessionId",
      R = "resource",
      O$1 = "error",
      E = "event",
      k$1 = "Queue",
      T = O$1 + k$1,
      j = R + k$1,
      A$1 = E + k$1,
      M = {
    start_time: {
      custom: function custom(t, e) {
        return Math.round(e.startTime);
      }
    },
    fetch_start: {
      custom: function custom(t, e) {
        return Math.round(e.fetchStart);
      }
    },
    dns: {
      end: "domainLookupEnd",
      start: "domainLookupStart",
      cachedSkip: !0,
      crossOriginSkip: !0
    },
    tcp: {
      end: "connectEnd",
      start: "connectStart",
      cachedSkip: !0,
      crossOriginSkip: !0
    },
    ssl: {
      custom: function custom(t, e) {
        return e.connectEnd && e.secureConnectionStart ? e.connectEnd - e.secureConnectionStart : null;
      }
    },
    ttfb: {
      end: "responseStart",
      start: "requestStart",
      cachedSkip: !0,
      crossOriginSkip: !0
    },
    trans: {
      end: "responseEnd",
      start: "responseStart",
      cachedSkip: !0,
      crossOriginSkip: !0
    },
    fb: {
      end: "responseStart",
      start: "fetchStart",
      cachedSkip: !0,
      crossOriginSkip: !0
    },
    total: {
      end: "responseEnd",
      start: "fetchStart"
    },
    size: {
      custom: function custom(t, e) {
        return e.encodedBodySize;
      },
      cachedSkip: !0,
      crossOriginSkip: !0
    },
    protocol: {
      custom: function custom(t, e) {
        return e.nextHopProtocol;
      }
    },
    file: {
      custom: function custom(t, e) {
        return e.name;
      }
    },
    res_type: {
      custom: function custom(t, e) {
        return e.initiatorType;
      }
    },
    cached: {
      custom: function custom(t, e) {
        return 0 === e.transferSize || 0 === e.duration || null == e.transferSize && (0 === e.domainLookupStart || e.responseStart === e.fetchStart);
      },
      crossOriginSkip: !0
    },
    has_timing_allow: {
      custom: function custom(t, e) {
        return G(e);
      }
    }
  },
      N = /^(https?:)?\/\//;

  function D(t, e) {
    if (!t) return !1;
    var n = t.replace(N, "");
    return e.some(function (t) {
      return 0 <= n.indexOf(t.replace(N, ""));
    });
  }

  function H(t) {
    if ("string" != typeof t) return "";

    try {
      var e = t.lastIndexOf(".");
      return t.substr(e + 1);
    } catch (t) {
      return "";
    }
  }

  function W(e) {
    return "object" == _typeof(e) && Object.keys(e).every(function (t) {
      return "number" == typeof e[t];
    });
  }

  function X(t) {
    return "function" == typeof t;
  }

  function U(t) {
    return t && X(t.then) && X(t["catch"]);
  }

  function G(t) {
    return void 0 === t.nextHopProtocol ? null : "" !== t.nextHopProtocol;
  }

  function J(t) {
    return "resource" === t.entryType ? "fetch" !== (null == t ? void 0 : t.initiatorType) && "xmlhttprequest" !== (null == t ? void 0 : t.initiatorType) || "js" === H(t.name) || "css" === H(t.name) ? "resource" : "api" : "other";
  }

  function K(n, r) {
    var o = {};
    return Object.keys(n).forEach(function (t) {
      var e = Y(t, n, r);
      null != e && (o[t] = e);
    }), o;
  }

  function Y(t, e, n) {
    var r = e[t];
    if (r || De("129，".concat(t)), (!r.crossOriginSkip || G(n)) && (!r.cachedSkip || !Y("cached", e, n))) return "function" == typeof r.custom ? r.custom(t, n) : (e = r.end, t = r.start, !r.notCatchIfUndefined || void 0 !== n[e] && void 0 !== n[t] ? (+n[e] || 0) - (+n[t] || 0) : void 0);
  }

  function B(t, e, n, r) {
    return t.addEventListener(e, n, r);
  }

  function $(t, e, n, r) {
    return t.removeEventListener(e, n, r);
  }

  function Q(t) {
    return t.message || null != t.lineno;
  }

  var Z = ["ktrace-str"];

  function ee(t) {
    return "__radar_".concat(t = void 0 === t ? "" : t).concat(Math.random().toString(36).substring(2, 8));
  }

  function te(o, l) {
    var i = window.XMLHttpRequest;
    return i ? (window.XMLHttpRequest = function () {
      var n,
          s,
          r,
          c = new i(),
          t = (c._radarRequest = {}, c._radarResponse = {}, ee()),
          u = (c._radarRequest.id = t, {});
      return c.addEventListener("readystatechange", function () {
        if (4 === c.readyState && "function" == typeof l) {
          c._radarResponse.status = c.status, c._radarResponse.statusText = c.statusText;

          var t,
              e,
              n,
              r,
              o,
              i = function (t, e) {
            switch (null == e && (e = {}), _typeof(t)) {
              case "object":
                var n,
                    r = [];

                for (n in t) {
                  var o = t[n],
                      i = n.toLowerCase();
                  r.push("".concat(i, ":\t").concat(o));
                }

                return r.join("\n") + "\n";

              case "string":
                for (var a = 0, s = (r = t.split("\n")).length; a < s; a++) {
                  var c = r[a];
                  /([^:]+):\s*(.+)/.test(c) && (c = null != (c = RegExp.$1) ? c.toLowerCase() : void 0, o = RegExp.$2, c && null == e[c] && (e[c] = o));
                }

                return e;
            }
          }(c.getAllResponseHeaders());

          for (t in i) {
            n = i[t], c._radarResponse && c._radarResponse.headers && !c._radarResponse.headers[t] && (e = t.toLowerCase(), c._radarResponse.headers[e] = n);
          }

          if (c.responseType && "text" !== c.responseType) "document" === c.responseType ? (c._radarResponse.xml = c.responseXML, c._radarResponse.data = c.responseXML) : c._radarResponse.data = c.response;else {
            c._radarResponse.text = c.responseText, c._radarResponse.data = c.responseText;

            try {
              c._radarResponse.xml = c.responseXML;
            } catch (t) {}
          }
          "responseURL" in c && (c._radarResponse.finalUrl = c.responseURL);

          try {
            var a = null != (r = u["ktrace-str"]) ? r : "";
            a && window.parseWedTraceId && (c._radarRequest.trace_id = window.parseWedTraceId(null != (o = a.split("|")[1]) ? o : ""));
          } catch (t) {
            s$2("traceId 解析异常");
          }

          l(c._radarRequest, c._radarResponse);
        }
      }), "function" == typeof c.setRequestHeader && (n = c.setRequestHeader, c.setRequestHeader = function () {
        for (var t = [], e = 0; e < arguments.length; e++) {
          t[e] = arguments[e];
        }

        return Z.includes(t[0]) && (u[t[0]] = t[1]), n.apply(this, t);
      }), "function" == typeof c.open && (s = c.open, c.open = function (t, n, r, o, i) {
        var a = [].slice.call(arguments);
        return this._radarRequest = _e2({
          method: t,
          url: n,
          async: r,
          user: o,
          pass: i
        }, this._radarRequest), this._radarResponse.headers = {}, s.apply(this, a);
      }), "function" == typeof c.send && (r = c.send, c.send = function () {
        var t = [].slice.call(arguments);
        return "function" == typeof o && (this._radarRequest._config = t[0], o(this._radarRequest)), r.apply(this, t);
      }), c;
    }, ["DONE", "HEADERS_RECEIVED", "LOADING", "OPENED", "UNSENT"].forEach(function (t) {
      window.XMLHttpRequest[t] = i[t];
    }), window.XMLHttpRequest.prototype = i.prototype, function () {
      i && (window.XMLHttpRequest = i);
    }) : (u$4("118"), !1);
  }

  function ne(n, r) {
    var t = function (t) {
      if ("function" != typeof t.fetch) return null;
      var n,
          e = t.fetch;
      return t.fetch = (n = t.fetch, function () {
        for (var t = [], e = 0; e < arguments.length; e++) {
          t[e] = arguments[e];
        }

        return ie.apply(void 0, o$1([n], r$1(t), !1));
      }), {
        register: function register(e) {
          return re.push(e), function () {
            var t = re.indexOf(e);
            0 <= t && re.splice(t, 1);
          };
        },
        clear: function clear() {
          re = [];
        },
        unRegister: function unRegister() {
          e && (t.fetch = e);
        }
      };
    }("function" == typeof importScripts ? self : window);

    return t && t.register({
      request: function request(t, e) {
        "function" == typeof n && n(t);
      },
      response: function response(t, e) {
        "function" == typeof r && r(e, t);
      },
      responseError: function responseError(t, e) {
        "function" == typeof r && r(e, {
          status: 0,
          statusText: t
        });
      }
    }), null == t ? void 0 : t.unRegister;
  }

  var re = [],
      oe = ["application/json", "text/html", "text/plain"];

  function ie(e) {
    for (var r = this, o = [], i = 1; i < arguments.length; i++) {
      o[i - 1] = arguments[i];
    }

    var a = re.reduce(function (t, e) {
      return [e].concat(t);
    }, []),
        s = "string" == typeof o[0] ? new Request(o[0], o[1]) : o[0],
        c = (s.id = ee(), a.forEach(function (t) {
      t = t.request, s = t(s) || s;
    }), s.clone()),
        u = e.apply(void 0, o$1([], r$1(o), !1)),
        l = null,
        p = null,
        d = !1;
    return u.then(function (t) {
      var e,
          n = null == (e = t.headers) ? void 0 : e.get("content-type");
      return n && oe.some(function (t) {
        return -1 < n.indexOf(t);
      }) ? (d = !0, e = t.clone(), l = e.clone(), e.clone().text()) : (l = t, "[object Object]");
    }).then(function (e) {
      return t(r, void 0, void 0, function () {
        return n(this, function (t) {
          switch (t.label) {
            case 0:
              p = e, t.label = 1;

            case 1:
              return t.trys.push([1, 3,, 4]), [4, c.text()];

            case 2:
              return [2, t.sent()];

            case 3:
              return t.sent(), [2, "Radar: Request.text()失败"];

            case 4:
              return [2];
          }
        });
      });
    }).then(function (t) {
      s._config = t;
      var e = null;
      return d ? (e = l.clone()).data = p : e = l, a.forEach(function (t) {
        (t = t.response) && (e = t(e, s) || e);
      }), l;
    }, function (e) {
      throw a.forEach(function (t) {
        (t = t.responseError) && (e = t(e, s) || e);
      }), e;
    });
  }

  function ae() {
    var n = [];

    function e(t) {
      -1 !== (t = n.indexOf(t)) && n.splice(t, 1);
    }

    function t(t) {
      return n.push(t), e.bind(void 0, t);
    }

    return {
      on: t,
      off: e,
      once: function once(e) {
        var n = t(function (t) {
          e(t), n();
        });
        return n;
      },
      emit: function emit(e) {
        n.forEach(function (t) {
          return t(e);
        });
      }
    };
  }

  function se(t) {
    return (t = new RegExp("(?:^|; ?)".concat(t, "=(1|0)(;|$)")).exec(document.cookie)) && t[1];
  }

  var ue = ["radar_init", "h5_load", "fmp", "main_api", "h5_custom_metric"];

  function le(t) {
    var e = t.dimension,
        n = t.value;
    if (t = e.event_name, ue.indexOf(t) < 0) return !1;
    if (!W(n)) return !1;

    switch (t) {
      case "fmp":
        var r = n.fmp_time;
        return !(!n.fmp || !r);

      case "h5_load":
        return r = n.total, !(!n.load_event_time || !r);

      case "h5_custom_metric":
        return r = n.custom_metric_timestamp, !(!e.custom_metric_name || !r);
    }

    return !0;
  }

  function fe(t) {
    var e = t.dimension,
        n = (t = t.value, e.custom_failed),
        r = e.api,
        o = e.status,
        i = e.res_type,
        a = e.protocol,
        s = (e = e.cached, t.duration),
        c = t.size;
    return "boolean" == typeof (void 0 !== n && n) && "string" == typeof (void 0 === r ? "" : r) && "number" == typeof (void 0 === o ? 0 : o) && "string" == typeof (void 0 === i ? "" : i) && "string" == typeof (void 0 === a ? "" : a) && "boolean" == typeof (void 0 !== e && e) && "number" == typeof (void 0 === s ? 0 : s) && "number" == typeof (void 0 === c ? 0 : c) && W(t);
  }

  function pe(t) {
    var e = (t = t.dimension).error_type;
    if ([t.error_cons_type, e, t.msg].some(function (t) {
      return null == t;
    })) return !1;
    var n,
        e = t.col,
        r = t.line;
    return t.sample_rate, t.event_client_timestamp, t = function (t, e) {
      var n = {};

      for (o in t) {
        Object.prototype.hasOwnProperty.call(t, o) && e.indexOf(o) < 0 && (n[o] = t[o]);
      }

      if (null != t && "function" == typeof Object.getOwnPropertySymbols) for (var r = 0, o = Object.getOwnPropertySymbols(t); r < o.length; r++) {
        e.indexOf(o[r]) < 0 && Object.prototype.propertyIsEnumerable.call(t, o[r]) && (n[o[r]] = t[o[r]]);
      }
      return n;
    }(t, ["col", "line", "sample_rate", "event_client_timestamp"]), !(e && "number" != typeof e || r && "number" != typeof r) && "object" == _typeof(n = t) && Object.keys(n).every(function (t) {
      return void 0 === n[t] || "string" == typeof n[t];
    });
  }

  function de(t) {
    var e = t.dimension,
        n = (t = t.value, e.failed);
    if ("boolean" != typeof n) return !1;

    if (n) {
      var r = e.res_path;
      if ("string" != typeof e.file || "string" != typeof r) return !1;
    }

    return !!n || "string" == typeof e.file && W(t);
  }

  function he(t) {
    var e = t.value;
    return t = t.dimension.name, !(void 0 !== e && !W(e) || !t);
  }

  function ve(t) {
    var e = !0;

    if ((t = t.dimension.data) instanceof Array) {
      if (0 === t.length) return !1;
      t.forEach(function (t) {
        "string" != typeof t.com && (e = !1);
      });
    }

    return e;
  }

  var me = {
    load: le,
    api: fe,
    resource: de,
    error: pe,
    event: he,
    blood: ve
  };

  function ge(t) {
    return t ? Object.keys(l$3).map(function (t) {
      return l$3[t];
    }).indexOf(t.key) < 0 ? (u$4("117"), !1) : me[t.key](t) : (u$4("116"), !1);
  }

  function we(n) {
    function r() {
      n = n.replace(/^\s*/, "");
    }

    function t(t) {
      return r(), n[0] === t && (n = n.slice(1), 1);
    }

    function o() {
      r();
      var t = /^([^\s,;=\\]+)(.*)/.exec(n);
      if (t) return n = t[2], t[1];
    }

    if (n && "string" == typeof n) {
      for (var e, i = []; e = o();) {
        for (var a = {
          name: e
        }, s = {}; t(";");) {
          var c,
              u,
              l = void 0;
          (l = o()) && (c = void 0, t("=") && (c = function () {
            if (r(), '"' !== n[0]) return o();
            n = n.slice(1);

            for (var t = ""; n.length;) {
              var e = /^([^"\\]*)(.*)/.exec(n);

              if (e) {
                if (t += e[1], '"' === e[2][0]) return n = e[2].slice(1), t;
                t += e[2][1], n = e[2].slice(2);
              }
            }
          }(), u = void 0, (u = /([,;].*)/.exec(n)) && (n = u[1])), s[u = l.toLowerCase()] || (s[u] = 1, "dur" === u ? a.duration = parseFloat(c) || 0 : "desc" === u ? a.description = c || "" : a[l] = c));
        }

        if ("sec" === a.unit && a.duration && (a.duration *= 1e3), delete a.unit, i.push(a), !t(",")) break;
      }

      return i.length ? i : void 0;
    }
  }

  function _e() {
    m$2[_$3] && (m$2[_$3][b$3] = !0);
  }

  function be() {
    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function (t) {
      var e = 16 * Math.random() | 0;
      return ("x" === t ? e : 3 & e | 8).toString(16);
    });
  }

  function Re(t, e) {
    var n;
    return null != (n = null == m$2 ? void 0 : m$2[_$3]) && n[t], null != (n = null == m$2 ? void 0 : m$2[_$3]) && n[x] ? m$2[_$3][t] : void 0 === e ? null : e;
  }

  function Oe(t, e) {
    try {
      m$2[_$3][t] = e;
    } catch (t) {
      u$4("1000");
    }
  }

  function Te(t) {
    return JSON.parse(JSON.stringify(t));
  }

  function je(t) {
    try {
      if (window && window.localStorage) {
        var e = window.localStorage.getItem(t);
        if (e) try {
          return JSON.parse(e);
        } catch (t) {
          return e;
        }
      }
    } catch (t) {
      return null;
    }

    return null;
  }

  function Ae(t, e) {
    try {
      if (window && window.localStorage) return window.localStorage.setItem(t, JSON.stringify(e)), !0;
    } catch (t) {
      return !1;
    }

    return !1;
  }

  function Ie(t) {
    try {
      var e;
      if (window && window.localStorage && "undefined" != typeof Storage && window.localStorage instanceof Storage) return Ae(t, (e = 1e8 < (e = Number(je(t)) || 0) + 1 ? 0 : e) + 1), e;
    } catch (t) {}
  }

  function Le(t) {
    for (var e = [], n = 1; n < arguments.length; n++) {
      e[n - 1] = arguments[n];
    }

    1 == +Ce(location.search.slice(1)).debug && s$2.apply(void 0, o$1(["[radar] ".concat(t, ": ")], r$1(e), !1));
  }

  function Ce(t) {
    var e,
        n,
        r = {};
    t = t.split("&");

    try {
      for (var o = function (t) {
        var e = "function" == typeof Symbol && Symbol.iterator,
            n = e && t[e],
            r = 0;
        if (n) return n.call(t);
        if (t && "number" == typeof t.length) return {
          next: function next() {
            return {
              value: (t = t && r >= t.length ? void 0 : t) && t[r++],
              done: !t
            };
          }
        };
        throw new TypeError(e ? "Object is not iterable." : "Symbol.iterator is not defined.");
      }(t), i = o.next(); !i.done; i = o.next()) {
        var a = r$1(i.value.split("="), 2),
            s = a[0],
            c = a[1];
        s in r ? r[s] instanceof Array ? r[s].push(c) : r[s] = [r[s], c] : r[s] = c;
      }
    } catch (t) {
      e = {
        error: t
      };
    } finally {
      try {
        i && !i.done && (n = o["return"]) && n.call(o);
      } finally {
        if (e) throw e.error;
      }
    }

    return r;
  }

  function qe(t, e) {
    null !== window && void 0 !== window && window.__RADAR_PERFORMANCE_TIMING_ || (window.__RADAR_PERFORMANCE_TIMING_ = {}), window.__RADAR_PERFORMANCE_TIMING_[t] = e;
  }

  function Ne(t) {
    if (-1 === (n = (t = void 0 === t ? location.href : t).lastIndexOf("?"))) return {
      page: t
    };
    var e = t.slice(0, n),
        n = t.slice(n + 1);
    return {
      url: t,
      page: e,
      params: Ce(n)
    };
  }

  function De(t) {
    throw new Error("radar error " + t);
  }

  var _v$ = function v$3() {
    return (_v$ = Object.assign || function (t) {
      for (var e, n = 1, r = arguments.length; n < r; n++) {
        for (var o in e = arguments[n]) {
          Object.prototype.hasOwnProperty.call(e, o) && (t[o] = e[o]);
        }
      }

      return t;
    }).apply(this, arguments);
  };

  function _$2(t, e) {
    var n = "function" == typeof Symbol && t[Symbol.iterator];
    if (!n) return t;
    var r,
        o,
        i = n.call(t),
        a = [];

    try {
      for (; (void 0 === e || 0 < e--) && !(r = i.next()).done;) {
        a.push(r.value);
      }
    } catch (t) {
      o = {
        error: t
      };
    } finally {
      try {
        r && !r.done && (n = i["return"]) && n.call(i);
      } finally {
        if (o) throw o.error;
      }
    }

    return a;
  }

  function y$4(t, e, n) {
    if (n || 2 === arguments.length) for (var r, o = 0, i = e.length; o < i; o++) {
      !r && o in e || ((r = r || Array.prototype.slice.call(e, 0, o))[o] = e[o]);
    }
    return t.concat(r || Array.prototype.slice.call(e));
  }

  var b$2 = ["c_dimension1", "c_dimension2", "c_dimension3", "release_tag", "product_version", "sub_app"],
      w = function () {
    function t(t) {
      var e,
          a = this,
          n = (this.queueConfig = {
        wait: 1e3,
        maxBatchLength: 200
      }, this.logQueue = [], this.preStoredLogQueue = [], this.batchTimer = 0, this.eventName = "onpagehide" in window ? "pagehide" : "beforeunload", this.customDimension = {}, this.runtimeIdMap = {}, this.randomSampling = Math.random(), this.commonDimension = {}, this.isInYoda = !1, this.sampleData = {
        isHit: !0,
        core: 1
      }, this.ignoreList = ["https://web-trace.ksapisrv.com/ktrace/collect"], this.broadcastSample = ae(), this.didRate = void 0, this.collect = function (t, e, n) {
        var r = Te({
          key: t,
          dimension: e = void 0 === e ? {} : e,
          value: n = void 0 === n ? {} : n
        });

        if (!a.sampleData || !1 !== a.sampleData.isHit) {
          if (!ge(r)) return u$4("107", "key: ".concat(r.key)), void a.collect(l$3.EVENT, {
            sample_rate: e.sample_rate,
            message: "radar上报日志类型错误",
            name: "radar_log_error",
            extra_info: JSON.stringify(r)
          }, {});
          Object.keys(r.value).forEach(function (t) {
            r.value[t] = r.value[t].toString();
          }), a.decorateLog(r);
          var o = a.getCommonData(),
              i = o.app_version_name;
          if (delete o.app_version_name, (t = new p$3(function (t, e) {
            t.dimension.app_version_name = i, t.dimension.did_rate = a.didRate;
            var n = (null == (n = null == t ? void 0 : t.dimension) ? void 0 : n.event_client_timestamp) || Date.now(),
                r = (null == (r = null == t ? void 0 : t.dimension) ? void 0 : r.radar_session_id) || a.radarSessionId;
            return _v$(_v$(_v$({}, t), o), {
              event_client_timestamp: n,
              event_trigger_source: "H5",
              radar_session_id: r
            });
          })).use(function (t, e) {
            return t.key && "load" === t.key && (t.dimension = _v$(_v$({}, t.dimension), {
              load_device_increase_id: Ie("LOAD_DEVICE_INCREASE_ID"),
              load_session_increase_id: a.genSessionIncreaseId("LOAD_SESSION_INCREASE_ID")
            })), t.key && "load" !== t.key && (t.dimension = _v$(_v$({}, t.dimension), {
              other_device_increase_id: Ie("OTHER_DEVICE_INCREASE_ID"),
              other_session_increase_id: a.genSessionIncreaseId("OTHER_SESSION_INCREASE_ID")
            })), e();
          }), (r = t.invoke(r)) instanceof Error) throw r;

          try {
            window.dispatchEvent(new CustomEvent("get-radar-next-data", {
              detail: {
                kv: Te(r)
              }
            }));
          } catch (t) {}

          if (delete r.dimension.broadcast_info, a.weblog ? a.logQueue.push(r) : (r.base_href = location.href, a.preStoredLogQueue.push(r)), Le("radarLogNext", JSON.parse(JSON.stringify(r))), a.logQueue.length > a.queueConfig.maxBatchLength - 1) return a.flush(), clearTimeout(a.batchTimer), void (a.batchTimer = 0);
          a.batchTimer || (a.batchTimer = window.setTimeout(function () {
            a.flush(), clearTimeout(a.batchTimer), a.batchTimer = 0;
          }, a.queueConfig.wait));
        }
      }, this.flush = function () {
        var t;
        a.logQueue.length <= 0 || (t = null, (t = a.config.httpReportFirst ? t : a.reportByBridge()) && !a.config.httpReportFirst || a.reportByHttp(), a.logQueue = []);
      }, "object" == _typeof(t) && null !== t || De("108"), "string" != typeof t.projectId && De("109"), Re(S$2));
      this.radarSessionId = null != n ? n : be(), this.config = t, this.sampling = "number" == typeof (e = null != (n = this.config.sampling) ? n : 1) && 0 <= e && e <= 1 ? e : (u$4("110"), 1), this.bindWeblogger(t.weblogger), null != (n = this.config) && n.ignoreList && Array.isArray(null == (t = this.config) ? void 0 : t.ignoreList) && (this.ignoreList = this.ignoreList.concat(this.config.ignoreList.map(function (t) {
        return t.replace(/^(https?:)?\/\//, "");
      }))), this.queueConfig = _v$(_v$({}, this.queueConfig), null != (n = this.config) && n.queueConfig ? this.config.queueConfig : {}), null != (t = this.config) && t.customDimensions && this.setDimensions(null == (n = this.config) ? void 0 : n.customDimensions), B(window, this.eventName, this.flush), "undefined" != typeof window && window.yodaCollectErrorDataDestroy && window.yodaCollectErrorDataDestroy();
    }

    return t.prototype.genSessionIncreaseId = function (t) {
      return this.runtimeIdMap[t] || (this.runtimeIdMap[t] = 0), this.runtimeIdMap[t]++;
    }, t.prototype.bindWeblogger = function (t) {
      var e,
          s = this;
      if (null != t && "object" != _typeof(t)) De("106");else {
        if (this.weblog) return void s$2("122");
        this.weblog = t, this.updateUrlPackage(), void 0 !== (null == (e = null == t ? void 0 : t.logger) ? void 0 : e.radarUrl) && this.ignoreList.push(null == (e = null == t ? void 0 : t.logger) ? void 0 : e.radarUrl), void 0 !== (null == (e = null == t ? void 0 : t.logger) ? void 0 : e.url) && this.ignoreList.push(null == (e = null == t ? void 0 : t.logger) ? void 0 : e.url), this.isInYoda = !(null == (e = null == (t = this.weblog) ? void 0 : t.Utils) || !e.yoda), this.getSampleData();
      }
      0 < this.preStoredLogQueue.length && (this.preStoredLogQueue.forEach(function (t) {
        if (void 0 !== (e = s.sampleData["load" === t.key ? "core" : t.key]) && (t.dimension.sample_rate = e), "load" === t.key ? s.sampleData.core : t.dimension.sample_rate * s.sampleData.core >= s.randomSampling) {
          var e = s.getCommonData(),
              n = (t.dimension.app_version_name = e.app_version_name, t.h5_extra_attr = JSON.stringify(_v$(_v$({}, null == (e = s.weblog.commonPackage) ? void 0 : e.getH5ExtraAttr()), JSON.parse(t.h5_extra_attr))), Ne(t.base_href)),
              r = null == (e = s.weblog.logConfig) ? void 0 : e.urlMap,
              e = {};
          if ("function" == typeof r) try {
            var o = r(n);
            "string" == typeof o ? e = {
              page: o
            } : "object" == _typeof(o) && (e = {
              page: o.page,
              params: o.params
            });
          } catch (t) {} else if ("object" == _typeof(r)) {
            var i,
                a = "";

            for (i in r) {
              if (-1 < (n.page || n.url || "").indexOf(i)) {
                a = r[i];
                break;
              }
            }

            a && (e = {
              page: a
            });
          }
          t.url_package = {
            page: e && e.page || n.page
          }, o = s.referUrlPackage && s.referUrlPackage.toJSON(), t.refer_url_package = o, delete t.base_href, s.logQueue.push(t);
        }
      }), this.preStoredLogQueue = [], this.flush());
    }, t.prototype.getSampleData = function () {
      var t,
          e,
          n = this;
      this.isInYoda ? (this.sampleData = {
        isHit: !0,
        core: this.sampling,
        api: 1,
        resource: .1,
        error: 1,
        event: 1
      }, e = window.updateYodaSampleRateWithParams, window.updateYodaSampleRateWithParams = function (t) {
        c$1("104", t), n.updateYodaSampleRate(t), n.broadcastSample.emit(n.sampleData), e && e(t);
      }, window.__yodaCommonDataObject__ && window.__yodaCommonDataObject__.sampleData ? (this.updateYodaSampleRate(window.__yodaCommonDataObject__.sampleData), this.collect(l$3.EVENT, {
        name: "客户端采样率注入成功",
        event_type: "radar_sample_inject_success"
      }, {})) : "function" == typeof (null == (t = null == (t = null == (t = this.weblog) ? void 0 : t.Utils) ? void 0 : t.yoda) ? void 0 : t.getWebviewLoadPerf) && null != (t = this.weblog) && t.Utils.yoda.getWebviewLoadPerf().then(function (t) {
        null != t && t.sampleData && n.updateYodaSampleRate(t.sampleData);
      })) : this.sampleData = {
        isHit: this.samplingControl(this.sampling),
        core: this.sampling
      };
    }, t.prototype.updateYodaSampleRate = function (t) {
      var e = this,
          n = null;

      try {
        var r = t,
            n = "string" == typeof r ? JSON.parse(r) : r;
      } catch (t) {
        return void c$1("error: ", t);
      }

      n && Object.keys(n).forEach(function (t) {
        "load" === t ? e.sampleData.core = n[t] : "did_rate" === t ? e.didRate = n[t] : e.sampleData[t] = n[t];
      });
    }, t.prototype.updateUrlPackage = function () {
      var t;
      null != (t = this.weblog) && t.currentUrlPackage && (this.currentUrlPackage = this.weblog.currentUrlPackage, this.referUrlPackage = this.weblog.referUrlPackage);
    }, t.prototype.setDimensions = function (e) {
      var n = this,
          t = Object.keys(e);
      t.some(function (t) {
        return -1 === b$2.indexOf(t);
      }) && u$4("113", "".concat(t.join("、"))), t.forEach(function (t) {
        null == e[t] ? delete n.customDimension[t] : n.customDimension[t] = e[t];
      });
    }, t.prototype.setCommonDimension = function (t, e, n) {
      this.commonDimension[t] = (n = void 0 === n || n) || null == (n = this.commonDimension[t]) ? e : n;
    }, t.prototype.decorateLog = function (t) {
      this.currentUrlPackage || this.updateUrlPackage(), 0 < Object.keys(this.customDimension).length && Object.assign(t.dimension, this.customDimension), 0 < Object.keys(this.commonDimension).length && Object.assign(t.dimension, this.commonDimension), this.currentUrlPackage && this.currentUrlPackage.page !== (null == (t = null == (t = this.weblog) ? void 0 : t.currentUrlPackage) ? void 0 : t.page) && (this.flush(), this.updateUrlPackage());
    }, t.prototype.samplingControl = function (t) {
      return null !== t && "number" == typeof t && this.randomSampling < t;
    }, t.prototype.getCommonData = function () {
      var t,
          e = null == (e = null == (e = null == (e = this.weblog) ? void 0 : e.commonPackage) ? void 0 : e.app_package) ? void 0 : e.version_name,
          e = {
        h5_extra_attr: JSON.stringify(_v$(_v$({}, null == (t = null == (t = this.weblog) ? void 0 : t.commonPackage) ? void 0 : t.getH5ExtraAttr()), {
          app_version_name: e,
          url: location.href,
          hash: location.hash,
          online: null === navigator || void 0 === navigator ? void 0 : navigator.onLine,
          downlink: null == (t = null === navigator || void 0 === navigator ? void 0 : navigator.connection) ? void 0 : t.downlink,
          rtt: null == (t = null === navigator || void 0 === navigator ? void 0 : navigator.connection) ? void 0 : t.rtt
        })),
        url_package: this.currentUrlPackage && this.currentUrlPackage.toJSON(),
        refer_url_package: this.referUrlPackage && this.referUrlPackage.toJSON(),
        app_version_name: e,
        project_id: null == (t = this.config) ? void 0 : t.projectId
      };
      return "1" === se("radar_report_test") && (e.is_report_test = !0), e;
    }, t.prototype.reportByHttp = function () {
      var t = this.getCommonData(),
          e = (t = {
        project_id: this.config.projectId,
        radar_session_id: this.radarSessionId,
        h5_extra_attr: t.h5_extra_attr
      }, this.logQueue.filter(function (t) {
        return "load" === t.key;
      })),
          n = this.logQueue.filter(function (t) {
        return "load" !== t.key;
      });
      Array.isArray(e) && e.length && this.collectLog(t, e, "radarSDK"), Array.isArray(n) && n.length && this.collectLog(t, n, "radarSDKSupplement");
    }, t.prototype.collectLog = function (t, e, n) {
      var r;
      "function" != typeof (null == (r = this.weblog) ? void 0 : r.collect) && "function" != typeof (null == (r = this.weblog) ? void 0 : r.sendByHttp) ? u$4("114") : (this.weblog.sendByHttp || this.weblog.collect).call(this.weblog, "RADAR", {
        name: "radar_log",
        params: _v$(_v$({}, t), {
          data: y$4([], _$2(e), !1)
        }),
        serviceName: n
      });
    }, t.prototype.reportByBridge = function () {
      var e = null == (n = null == (n = this.weblog) ? void 0 : n.Utils) ? void 0 : n.yoda;
      if (!e) return !1;
      var t,
          n = this.logQueue.filter(function (t) {
        return "load" === t.key;
      }),
          r = this.logQueue.filter(function (t) {
        return "load" !== t.key;
      });

      try {
        return n.forEach(function (t) {
          "function" == typeof (null == (t = e.sendSummarizedLog({
            key: "h5_load",
            data: t
          })) ? void 0 : t["catch"]) && t["catch"](function (t) {
            u$4("123", t);
          });
        }), r.length && "function" == typeof (null == (t = e.sendRadarLog({
          data: r
        })) ? void 0 : t["catch"]) && t["catch"](function (t) {
          u$4("124", t);
        }), !0;
      } catch (t) {
        return !1;
      }
    }, t;
  }(),
      _d$ = function d$2(t, e) {
    return (_d$ = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (t, e) {
      t.__proto__ = e;
    } || function (t, e) {
      for (var n in e) {
        Object.prototype.hasOwnProperty.call(e, n) && (t[n] = e[n]);
      }
    })(t, e);
  },
      _p$ = function p$2() {
    return (_p$ = Object.assign || function (t) {
      for (var e, n = 1, r = arguments.length; n < r; n++) {
        for (var o in e = arguments[n]) {
          Object.prototype.hasOwnProperty.call(e, o) && (t[o] = e[o]);
        }
      }

      return t;
    }).apply(this, arguments);
  };

  function l$2(t, e) {
    var n = {};

    for (o in t) {
      Object.prototype.hasOwnProperty.call(t, o) && e.indexOf(o) < 0 && (n[o] = t[o]);
    }

    if (null != t && "function" == typeof Object.getOwnPropertySymbols) for (var r = 0, o = Object.getOwnPropertySymbols(t); r < o.length; r++) {
      e.indexOf(o[r]) < 0 && Object.prototype.propertyIsEnumerable.call(t, o[r]) && (n[o[r]] = t[o[r]]);
    }
    return n;
  }

  var u$3 = {
    redirect: {
      end: "redirectEnd",
      start: "redirectStart"
    },
    dns: {
      end: "domainLookupEnd",
      start: "domainLookupStart"
    },
    dns_start: {
      end: "domainLookupEnd",
      start: "navigationStart"
    },
    tcp: {
      end: "connectEnd",
      start: "connectStart"
    },
    tcp_start: {
      end: "connectEnd",
      start: "navigationStart"
    },
    redirect_count: {
      custom: function custom(t, e) {
        return e.redirectCount;
      }
    },
    ssl: {
      custom: function custom(t, e) {
        return e.connectEnd && e.secureConnectionStart ? e.connectEnd - e.secureConnectionStart : null;
      }
    },
    ssl_start: {
      custom: function custom(t, e) {
        return e.secureConnectionStart && e.navigationStart ? e.secureConnectionStart - e.navigationStart : null;
      }
    },
    ttfb: {
      end: "responseStart",
      start: "requestStart"
    },
    request_start: {
      end: "requestStart",
      start: "navigationStart"
    },
    trans: {
      end: "responseEnd",
      start: "responseStart"
    },
    dom_parse: {
      end: "domInteractive",
      start: "responseEnd"
    },
    dom_parse_start: {
      end: "domInteractive",
      start: "navigationStart"
    },
    blank: {
      end: "domLoading",
      start: "navigationStart"
    },
    fb: {
      end: "responseStart",
      start: "navigationStart"
    },
    dom_ready: {
      end: "domContentLoadedEventEnd",
      start: "navigationStart"
    },
    total: {
      end: "loadEventStart",
      start: "navigationStart"
    },
    resource: {
      end: "loadEventStart",
      start: "domInteractive"
    },
    fp: {
      custom: function custom() {
        if ("function" != typeof performance.getEntries) return null;
        var t = performance.getEntries().filter(function (t) {
          return "first-paint" === t.name;
        })[0];
        return t ? Math.ceil(t.startTime + t.duration) : null;
      }
    },
    js_cost: {
      custom: function custom() {
        if ("function" != typeof performance.getEntries) return null;
        var t = h$3(),
            e = performance.getEntries(),
            n = performance.getEntries().filter(function (t) {
          return "navigation" === t.initiatorType;
        })[0],
            e = e.filter(function (t) {
          return "script" === t.initiatorType;
        })[0],
            n = n ? n.domComplete : t.domComplete - t.navigationStart;
        return e && n ? n - e.fetchStart : null;
      }
    },
    tti: {
      end: "domInteractive",
      start: "navigationStart"
    },
    protocol: {
      custom: function custom(t, e) {
        return e.nextHopProtocol;
      }
    },
    dom_num: {
      custom: function custom() {
        return "undefined" != typeof document && void 0 !== document.all ? document.all.length : null;
      }
    }
  },
      f$3 = {
    main_api: {
      end: "apiEnd",
      start: "apiStart",
      notCatchIfUndefined: !0
    },
    main_api_offset: {
      end: "apiStart",
      start: "responseEnd",
      notCatchIfUndefined: !0
    },
    fmp: {
      end: "fmp_time",
      start: "navigationStart",
      notCatchIfUndefined: !0
    }
  },
      _$1 = {
    navigation_start_time: "navigationStart",
    fetch_start_time: "fetchStart",
    dns_start_time: "domainLookupStart",
    dns_end_time: "domainLookupEnd",
    connect_start_time: "connectStart",
    ssl_start_time: "secureConnectionStart",
    request_start_time: "requestStart",
    response_start_time: "responseStart",
    response_end_time: "responseEnd",
    dom_interactive_time: "domInteractive",
    dom_loading_time: "domLoading",
    dom_ready_time: "domContentLoadedEventEnd",
    load_event_time: "loadEventStart",
    connect_end_time: "connectEnd",
    load_event_end: "loadEventEnd",
    dom_complete: "domComplete",
    dom_content_loaded_event_start: "domContentLoadedEventStart",
    redirect_start: "redirectStart",
    redirect_end: "redirectEnd",
    radar_init_time: "radar_init_time",
    fmp_time: "fmp_time"
  };

  function h$2(n, t) {
    if (!Array.isArray(t)) return -1;
    var r = -1;
    return t.forEach(function (t, e) {
      new RegExp(t).test(n) && (r = e);
    }), r;
  }

  var v$2 = function (e) {
    function t(t) {
      var g = this;
      return (g = e.call(this, {
        core: null == t ? void 0 : t.core,
        key: l$3.LOAD,
        version: "1.2.5",
        name: "RadarNavigationCollect"
      }) || this).hasFMP = !1, g.syncMetricNum = 0, g.customKeys = [], g.custom_metric = {}, g.optionsMainAPI = [], g.mainAPIList = [], g.radarMainApiList = [], g.mainJSList = [], g.radarMainJSList = [], g.radarTiming = {
        radar_init_time: void 0,
        fmp_time: void 0,
        fmp: void 0,
        apiStart: void 0,
        apiEnd: void 0
      }, g.h5Load = function () {
        var t = g.getPerformanceData();
        g.collect(_p$({
          event_name: "h5_load"
        }, t));
      }, g.dealAPI = function (t) {
        return g.cacheAndCalculateMainAPI(t), g.addMainApiResult(t), g.mainAPIList.filter(function (t) {
          return !!t.url;
        }).length !== g.mainAPIList.length || void 0 === g.radarTiming.fmp;
      }, g.addMainJSResult = function (t) {
        if (!g.mainJSList || 0 === g.mainJSList.length) return !1;
        var e,
            n,
            r,
            o,
            i,
            a,
            s,
            c,
            u,
            l,
            p,
            d,
            f,
            m,
            h,
            v = h$2((t.file || "").replace(/\?.+$/, ""), g.mainJSList);
        return -1 < v && !g.radarMainJSList[v].file && (e = t.file, n = t.cached, r = t.size, o = t.failed, i = t.res_type, a = t.protocol, s = t.fb, c = t.dns, u = t.tcp, l = t.ssl, p = t.ttfb, d = t.trans, f = t.total, m = t.fetch_start, h = t.duration, t = t.start_time, g.radarMainJSList[v] = {
          file: e,
          cached: n,
          size: r,
          failed: o,
          res_type: i,
          protocol: a,
          fb: s,
          dns: c,
          tcp: u,
          ssl: l,
          ttfb: p,
          trans: d,
          total: f,
          fetch_start: m,
          duration: h,
          start_time: t,
          reg: g.mainJSList[v]
        }), g.radarMainJSList.filter(function (t) {
          return !!t.file;
        }).length !== g.mainJSList.length;
      }, t.core || De("212"), "string" == typeof t.mainApi ? (g.optionsMainAPI.push(t.mainApi), g.mainAPIList.push({}), g.radarMainApiList.push({})) : Array.isArray(t.mainApi) && (3 < t.mainApi.length && s$2("207"), g.optionsMainAPI = t.mainApi.slice(0, 3), g.optionsMainAPI.forEach(function () {
        g.mainAPIList.push({}), g.radarMainApiList.push({});
      })), "string" == typeof t.mainJs ? (g.mainJSList.push(t.mainJs), g.radarMainJSList.push({})) : Array.isArray(t.mainJs) && (3 < t.mainJs.length && s$2("211"), g.mainJSList = t.mainJs.slice(0, 3), g.mainJSList.forEach(function () {
        return g.radarMainJSList.push({});
      })), performance.timing ? g.init() : (null != (t = g.core) && t.collect("event", {
        name: "radar_error",
        message: "performance.timing 不存在无法收集数据",
        category: l$3.LOAD
      }), u$4("208")), g;
    }

    var n = t,
        r = e;
    if ("function" != typeof r && null !== r) throw new TypeError("Class extends value " + String(r) + " is not a constructor or null");

    function o() {
      this.constructor = n;
    }

    return _d$(n, r), n.prototype = null === r ? Object.create(r) : (o.prototype = r.prototype, new o()), t.prototype.init = function () {
      this.radarTiming.radar_init_time = Date.now();
      var t,
          e = this.getPerformanceData();
      this.collect(_p$({
        event_name: "radar_init"
      }, e)), null != (t = null == (e = this.core) ? void 0 : e.setCommonDimension) && t.call(e, "happen_stage", "during_fmp"), 0 < performance.timing.loadEventStart ? this.h5Load() : B(window, "load", this.h5Load);
    }, t.prototype.customStage = function (t, e) {
      var n, r;
      void 0 === e && (e = {
        sendTimeline: !0
      }), this.isHit ? (void 0 !== t && "string" == typeof t && "" !== t || De("206"), this.isValidTimestamp(e.timestamp, "custom_stage " + t) && (n = e.timestamp || Date.now(), r = this.customKeys.length - this.syncMetricNum, this.customKeys.includes(t) ? u$4("204", t) : e.sendTimeline && 2 < this.syncMetricNum || !e.sendTimeline && 5 < r ? u$4("205", t) : n && (this.custom_metric[t + "_time"] = n, this.customKeys.push(t), !e || !0 !== e.sendTimeline && null != e.sendTimeline || (r = this.getPerformanceData(), this.collect(_p$(_p$({}, r), {
        stack: new Error().stack,
        event_name: "h5_custom_metric",
        custom_metric_name: t,
        custom_metric_timestamp: n
      })), this.syncMetricNum++)))) : u$4("200");
    }, t.prototype.isValidTimestamp = function (t, e) {
      var n;
      if (void 0 === e && (e = "fmp"), void 0 !== t && "number" != typeof t) De("201");else if (void 0 !== t && t < performance.timing.navigationStart) return s$2("210", "".concat(e)), null != (n = this.core) && n.collect("event", {
        name: "radar_error",
        message: "【210】".concat(e),
        extra_info: {
          timestamp: t,
          navigationStart: performance.timing.navigationStart
        }
      }), !1;
      return !0;
    }, t.prototype.fmp = function (t) {
      var e, n;
      this.isHit ? this.isValidTimestamp(t) && (t = t || Date.now(), this.hasFMP ? s$2("202") : t && performance.timing && (null != (n = null == (e = this.core) ? void 0 : e.setCommonDimension) && n.call(e, "happen_stage", "after_fmp"), this.hasFMP = !0, this.radarTiming.fmp_time = t, performance.timing.radarFmp = t, n = this.getPerformanceData(), performance.timing.fmp = this.radarTiming.fmp = n.fmp, this.collect(_p$({
        event_name: "fmp",
        stack: new Error().stack
      }, n)))) : s$2("200");
    }, t.prototype.getPerformanceData = function () {
      var e = _p$(_p$({}, Te(performance.timing)), this.radarTiming),
          n = K(_p$(_p$({}, u$3), f$3), e);

      return Object.keys(_$1).forEach(function (t) {
        0 <= e[_$1[t]] && (n[t] = e[_$1[t]]);
      }), ["radar_main_js_start", "lcp", "cls", "fid", "fcp"].forEach(function (t) {
        var e = window.__RADAR_PERFORMANCE_TIMING_ && window.__RADAR_PERFORMANCE_TIMING_[t];
        !e && 0 !== e || (n[t] = e);
      }), n;
    }, t.prototype.cacheAndCalculateMainAPI = function (t) {
      var e;
      void 0 === this.radarTiming.fmp && (t.before_hook_time && (e = this.radarTiming.apiStart, this.radarTiming.apiStart = e ? Math.min(e, t.before_hook_time) : t.before_hook_time), t.after_hook_time && (e = this.radarTiming.apiEnd, this.radarTiming.apiEnd = e ? Math.max(e, t.after_hook_time) : t.after_hook_time));
    }, t.prototype.addMainApiResult = function (t) {
      var e,
          n,
          r,
          o,
          i,
          a,
          s,
          c,
          u,
          l,
          p,
          d,
          f,
          m,
          h,
          v,
          g,
          _,
          y = h$2((t.api || "").replace(/\?.+$/, ""), this.optionsMainAPI);

      -1 < y && (null == (_ = this.mainAPIList[y]) || !_.url) && (_ = t.api, e = t.status, n = t.size, r = t.cached, o = t.method, c = t.perfTime, i = t.protocol, a = t.custom_failed, s = t.res_type, g = l$2(t, ["api", "status", "size", "cached", "method", "perfTime", "protocol", "custom_failed", "res_type"]), this.mainAPIList[y] = _p$({
        url: _,
        status: e,
        size_radar: n,
        cached: r,
        method: o,
        perfTime: c,
        protocol: i,
        res_type: s,
        custom_failed: a
      }, g), c = g.fb, u = g.dns, l = g.tcp, p = g.ssl, d = g.ttfb, f = g.trans, m = g.total, h = g.fetch_start, v = g.duration, g = g.start_time, this.radarMainApiList[y] = {
        api: _,
        status: e,
        size: n,
        cached: r,
        method: o,
        protocol: i,
        fb: c,
        dns: u,
        tcp: l,
        ssl: p,
        ttfb: d,
        trans: f,
        total: m,
        fetch_start: h,
        duration: v,
        custom_failed: a,
        res_type: s,
        reg: this.optionsMainAPI[y],
        start_time: g
      }, this.radarTiming.fmp_time && t.after_hook_time && t.after_hook_time < this.radarTiming.fmp_time && (_ = this.getPerformanceData(), this.collect(_p$({
        event_name: "main_api"
      }, _))));
    }, t.prototype.destroy = function () {
      $(window, "load", this.h5Load), this.customStage = this.fmp = function () {
        s$2("213");
      };
    }, t.prototype.collect = function (t) {
      var e = t.protocol,
          n = t.event_name,
          r = (t.custom_metric, t.custom_metric_name),
          o = t.stack;
      t.main_api_list, t = l$2(t, ["protocol", "event_name", "custom_metric", "custom_metric_name", "stack", "main_api_list"]), this.baseCollect({
        value: t,
        dimension: {
          protocol: e,
          event_name: n,
          custom_metric_name: r,
          custom_metric: JSON.stringify(this.custom_metric),
          main_api_list: this.mainAPIList,
          radar_main_api_list: this.radarMainApiList,
          radar_main_js_list: this.radarMainJSList,
          is_official: 1,
          stack: o
        }
      }), null != (t = this.core) && t.flush();
    }, t;
  }(d$3),
      _g$ = function g$3(t, e) {
    return (_g$ = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (t, e) {
      t.__proto__ = e;
    } || function (t, e) {
      for (var n in e) {
        Object.prototype.hasOwnProperty.call(e, n) && (t[n] = e[n]);
      }
    })(t, e);
  },
      _y$ = function y$3() {
    return (_y$ = Object.assign || function (t) {
      for (var e, n = 1, r = arguments.length; n < r; n++) {
        for (var o in e = arguments[n]) {
          Object.prototype.hasOwnProperty.call(e, o) && (t[o] = e[o]);
        }
      }

      return t;
    }).apply(this, arguments);
  };

  function b$1(t, e) {
    var n = {};

    for (o in t) {
      Object.prototype.hasOwnProperty.call(t, o) && e.indexOf(o) < 0 && (n[o] = t[o]);
    }

    if (null != t && "function" == typeof Object.getOwnPropertySymbols) for (var r = 0, o = Object.getOwnPropertySymbols(t); r < o.length; r++) {
      e.indexOf(o[r]) < 0 && Object.prototype.propertyIsEnumerable.call(t, o[r]) && (n[o[r]] = t[o[r]]);
    }
    return n;
  }

  var k = M,
      L = {
    navigation_start: "startTime",
    redirect_start: "redirectStart",
    redirect_end: "redirectEnd",
    dns_start: "domainLookupStart",
    dns_end: "domainLookupEnd",
    connect_start: "connectStart",
    connect_end: "connectEnd",
    ssl_start: "secureConnectionStart",
    request_start: "requestStart",
    response_start: "responseStart",
    response_end: "responseEnd"
  };

  function O(t, e) {
    if (t) return "function" == typeof t.get ? t.get(e) : t[e.toLowerCase()];
  }

  function I(t) {
    var e = "";
    if ("string" == typeof t) e = t;else try {
      e = JSON.stringify(t);
    } catch (n) {
      e = String(t);
    }

    for (var n = 0, r = 0; r < e.length; r++) {
      var o = e.charCodeAt(r);
      n += o < 128 ? 1 : o < 2048 ? 2 : o < 65536 ? 3 : o < 1 << 21 ? 4 : o < 1 << 26 ? 5 : o < 1 << 31 ? 6 : Number.NaN;
    }

    return n;
  }

  function P(t) {
    var e = (n = document.location).protocol,
        n = n.host;
    return 0 === t.indexOf("//") ? e + t : 0 === t.indexOf("/") ? "".concat(e, "//").concat(n).concat(t) : t;
  }

  function S$1(t) {
    return t ? "function" == typeof t.get ? t.get("Content-Type") : t["content-type"] : null;
  }

  var A = function (e) {
    function t(t) {
      var l = e.call(this, {
        core: null == (t = void 0 === t ? {} : t) ? void 0 : t.core,
        sampling: null == t ? void 0 : t.sampling,
        key: l$3.API,
        version: "1.2.5",
        name: "RadarAPICollect"
      }) || this;
      return l.ignoreList = v$4, l.queue = [], l.logList = [], l.perfList = [], l.broadcastApi = ae(), l.initiatorTypeList = [], l.beforeHook = function (t) {
        var e;
        0 === (null == (e = t.url) ? void 0 : e.indexOf("data:")) || D(t.url, l.ignoreList) || D(t.url, (null == (e = l.core) ? void 0 : e.ignoreList) || []) || "js" === H(t.url) || "css" === H(t.url) || (e = {
          api: P(t.url),
          method: t.method,
          request: t,
          startTime: Date.now()
        }, l.queue.push(e));
      }, l.afterHook = function (e, t) {
        var n = l.queue.find(function (t) {
          return t.request === e;
        });

        if (n) {
          delete n.request;
          var r = n.startTime,
              o = (n = b$1(n, ["startTime"]), Date.now()),
              i = o - r,
              a = t.status;
          if ("function" == typeof l.customizeRadarStatus) try {
            a = l.customizeRadarStatus(t);
          } catch (t) {
            console.error(t);
          }

          var s = l.customHook({
            request: e,
            response: t,
            duration: i
          }),
              c = +O(t.headers, "Content-Length") || 0,
              u = we(O(t.headers, "Server-Timing")),
              i = _y$(_y$(_y$({
            duration: i,
            logTime: Date.now(),
            status: a,
            size: +c,
            content_type: S$1(t.headers),
            custom_failed: !(200 <= a && a < 300)
          }, n), s), {
            before_hook_time: r,
            after_hook_time: o,
            trace_id: e.trace_id,
            server_timing: u
          });

          2 === (null == (a = null == (c = O(t.headers, "Intercept-Result")) ? void 0 : c.split(";")) ? void 0 : a.length) && null != (n = l.core) && n.setCommonDimension && (l.core.setCommonDimension("intercept_result", a[0], !1), l.core.setCommonDimension("intercept_result_code", a[1], !1), i.api_intercept_result = a[0], i.api_intercept_result_code = a[1]), t.finalUrl && (i.api = t.finalUrl), "string" == typeof t.text && (i.responseData = t.text), 0 !== _y$(_y$({}, t), {
            status: i.status
          }).status || s.intercept_report ? l.mergeAPIPerf(void 0, i) || l.logList.push(i) : (i.size && 0 != i.size || !i.responseData || (i.size = I(i.responseData)), delete i.responseData, delete i.server_timing, l.collect(i)), l.queue = l.queue.filter(function (t) {
            return t.request && t.request !== e;
          });
        }
      }, void 0 !== (l.option = t) && null != t && t.core || De("310"), void 0 !== t.ignoreList && (Array.isArray(t.ignoreList) || De("309"), t.ignoreList.forEach(function (t) {
        "string" != typeof t && De("309");
      }), l.ignoreList = l.ignoreList.concat(t.ignoreList)), l.APIHook = t.APIHook, l.customizeRadarStatus = t.customizeRadarStatus, l.init(), l;
    }

    var n = t,
        r = e;
    if ("function" != typeof r && null !== r) throw new TypeError("Class extends value " + String(r) + " is not a constructor or null");

    function o() {
      this.constructor = n;
    }

    return _g$(n, r), n.prototype = null === r ? Object.create(r) : (o.prototype = r.prototype, new o()), t.prototype.init = function () {
      var e,
          n = this;

      try {
        this.observer = new PerformanceObserver(function (t) {
          return n.observeAPI(t);
        }), this.observer.observe({
          entryTypes: ["resource"]
        }), this.initiatorTypeList = [], !0 !== this.option.ignoreXHR && (this.initiatorTypeList.push("xmlhttprequest"), this.unhookXML = te(function (t) {
          return n.beforeHook(t);
        }, function (t, e) {
          return n.afterHook(t, e);
        })), !0 !== this.option.ignoreFetch && (this.initiatorTypeList.push("fetch"), this.unhookFetch = ne(function (t) {
          return n.beforeHook(t);
        }, function (t, e) {
          return n.afterHook(t, e);
        }));
      } catch (t) {
        u$4("308"), null != (e = this.core) && e.collect("event", {
          name: "radar_error",
          message: "PerformanceObserver 不存在无法收集数据",
          category: l$3.API
        });
      }
    }, t.prototype.customHook = function (t) {
      var e = {};
      if (void 0 === this.APIHook) return e;
      if ("function" != typeof this.APIHook) return s$2("300"), e;
      var n = {};

      try {
        n = this.APIHook(t);
      } catch (t) {
        return s$2("301", t), {};
      }

      if (!n) return s$2("302"), e;
      t = n.response_msg;
      var r = n.response_code,
          o = n.status,
          i = n.custom_failed,
          a = n.intercept_report,
          n = n.broadcast_info;
      if (!0 === Boolean(a) && (e.intercept_report = !0), void 0 !== t) try {
        var s = JSON.stringify(t);
        100 < s.length && s$2("307"), e.response_msg = s.slice(0, 100);
      } catch (t) {
        e.response_msg = "response_msg返回结果异常", s$2("303");
      }
      return void 0 !== r && (a = Number(r), isNaN(a) ? s$2("304") : e.response_code = a), void 0 !== o && (t = Number(o), isNaN(t) ? s$2("305") : e.status = t), void 0 !== i && (!0 === i || !1 === i ? e.custom_failed = i : s$2("306")), e.broadcast_info = n, e;
    }, t.prototype.observeAPI = function (t) {
      var o = this;
      t.getEntriesByType("resource").forEach(function (e) {
        var t, n, r;
        if (0 === (null == (r = e.name) ? void 0 : r.indexOf("data:")) || D(e.name, o.ignoreList) || D(e.name, (null == (r = o.core) ? void 0 : r.ignoreList) || [])) return !1;
        "api" === J(e) && ((r = K(k, e)).perfTime = Date.now(), null != (t = null == (t = e.serverTiming) ? void 0 : t.map(function (t) {
          return {
            name: t.name,
            duration: t.duration,
            description: t.description || void 0
          };
        })) && t.length || (t = void 0), n = {
          encoded_body_size: e.encodedBodySize,
          decoded_body_size: e.decodedBodySize
        }, Object.keys(L).forEach(function (t) {
          n[t] = Number(e[L[t]] + performance.timing.navigationStart);
        }), r = _y$(_y$(_y$(_y$({}, r), {
          api: r.file
        }), n), {
          server_timing: t
        }), o.mergeAPIPerf(r) || o.perfList.push(r));
      });
    }, t.prototype.destroy = function () {
      this.unhookXML && this.unhookXML(), this.unhookFetch && this.unhookFetch(), this.observer && this.observer.disconnect();
    }, t.prototype.mergeAPIPerf = function (t, e) {
      if (t ? e = this.takeItem(t, this.logList) : e && (t = this.takeItem(e, this.perfList)), !t || !e) return !1;
      if (!0 === e.intercept_report) return !0;
      var n,
          r = Object.assign({}, e, t, {
        size: t.size || e.size
      });
      return delete r.file, r.size && 0 != r.size || !r.responseData || (r.size = I(r.responseData)), delete r.responseData, (e = e.server_timing || t.server_timing) && (n = null, e.map(function (t) {
        "total-timing" === t.name && (n = null != (t = t.duration) ? t : null);
      }), r.total_server_timing = n, r.server_data_list = JSON.stringify(e)), delete r.server_timing, this.collect(r), !0;
    }, t.prototype.takeItem = function (t, e) {
      var n,
          r = (t.api || "").replace(/\?.+$/, "");
      if (-1 !== (t = e.findIndex(function (t) {
        return t = (t.api || "").replace(/\?.+$/, ""), r === t;
      }))) return n = e[t], e.splice(t, 1), n;
    }, t.prototype.collect = function (o) {
      var i = this,
          t = o.res_type,
          e = o.protocol,
          n = o.api,
          r = o.custom_failed,
          a = o.status,
          s = o.cached,
          c = o.method,
          u = o.response_code,
          l = o.response_msg,
          p = (o.has_fmp, o.server_data_list),
          d = o.content_type,
          f = o.broadcast_info,
          m = o.before_hook_time,
          h = o.after_hook_time,
          v = o.api_intercept_result,
          g = o.api_intercept_result_code,
          _ = {
        dimension: {
          before_hook_time: m,
          after_hook_time: h,
          res_type: t,
          protocol: e,
          api: n,
          custom_failed: r,
          status: a,
          cached: s,
          method: c,
          response_code: u,
          response_msg: l,
          content_type: d,
          broadcast_info: f,
          is_official: 1,
          server_data_list: p,
          trace_id: o.trace_id,
          api_intercept_result: v,
          api_intercept_result_code: g,
          has_timing_allow: o.has_timing_allow
        },
        value: b$1(o, ["res_type", "protocol", "api", "custom_failed", "status", "cached", "method", "response_code", "response_msg", "has_fmp", "server_data_list", "content_type", "broadcast_info", "before_hook_time", "after_hook_time", "api_intercept_result", "api_intercept_result_code", "trace_id", "has_timing_allow"])
      };
      this.broadcastApi.emit(Te(o)), 0 === a && X(this.option.isIgnoreInvalidStatusCode) ? new Promise(function (e, t) {
        var n, r;
        i.option.isIgnoreInvalidStatusCode && (U(n = i.option.isIgnoreInvalidStatusCode(o)) ? (r = setTimeout(function () {
          return e(!1);
        }, 2e3), n.then(function (t) {
          clearTimeout(r), e(!0 === t);
        }, t)) : e(!0 === n));
      }).then(function (t) {
        t ? s$2("异步拦截日志") : i.baseCollect(_);
      }, function (t) {
        i.baseCollect(_), u$4(t);
      }) : this.baseCollect(_);
    }, t;
  }(d$3),
      _g$2 = function g$2(t, e) {
    return (_g$2 = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (t, e) {
      t.__proto__ = e;
    } || function (t, e) {
      for (var n in e) {
        Object.prototype.hasOwnProperty.call(e, n) && (t[n] = e[n]);
      }
    })(t, e);
  },
      _2 = function _() {
    return (_2 = Object.assign || function (t) {
      for (var e, n = 1, r = arguments.length; n < r; n++) {
        for (var o in e = arguments[n]) {
          Object.prototype.hasOwnProperty.call(e, o) && (t[o] = e[o]);
        }
      }

      return t;
    }).apply(this, arguments);
  },
      y$2 = Object.assign({}, M, {
    duration: {
      end: "responseEnd",
      start: "fetchStart"
    },
    cached_type: {
      custom: function custom(t, e) {
        var n = "hit_browser";
        if (0 === e.transferSize || 0 === e.duration) return n;
        if (null == e.transferSize && (0 === e.domainLookupStart || e.responseStart === e.fetchStart)) return n;

        if (e.serverTiming && e.serverTiming.some(function (t) {
          return "cdn" === t.name;
        })) {
          if (e.serverTiming.some(function (t) {
            return "miss" === t.name;
          })) return "miss_cdn";
          if (e.serverTiming.some(function (t) {
            return "hit" === t.name;
          })) return "hit_cdn";
        }

        return "miss_browser";
      },
      crossOriginSkip: !0
    },
    cdn_duration: {
      custom: function custom(t, e) {
        return e.serverTiming && e.serverTiming.some(function (t) {
          return "cdn" === t.name;
        }) ? null == (e = null == (e = e.serverTiming) ? void 0 : e.find(function (t) {
          return "total-timing" === t.name;
        })) ? void 0 : e.duration : null;
      },
      crossOriginSkip: !0,
      cachedSkip: !0
    },
    cdn_ip: {
      custom: function custom(t, e) {
        if (e.serverTiming && e.serverTiming.some(function (t) {
          return "cdn" === t.name;
        })) return null == (e = e.serverTiming.find(function (t) {
          return "miss" === t.name || "hit" === t.name;
        })) ? void 0 : e.description;
      },
      crossOriginSkip: !0,
      cachedSkip: !0
    }
  }),
      v$1 = {
    fetch_start_time: "fetchStart",
    redirect_start: "redirectStart",
    redirect_end: "redirectEnd",
    dns_start: "domainLookupStart",
    dns_end: "domainLookupEnd",
    connect_start: "connectStart",
    connect_end: "connectEnd",
    ssl_start: "secureConnectionStart",
    request_start: "requestStart",
    response_start: "responseStart",
    response_end: "responseEnd"
  };

  function b(t) {
    return Array.isArray(t) ? t.map(function (t) {
      var e = void 0 === (e = t.tagName) ? "" : e,
          n = t.id;
      return t = t.getAttribute ? t.getAttribute("class") : t.className || "", e = e.toLowerCase(), n && (e += "#".concat(n)), t && (e += t.split(/\s+/g).map(function (t) {
        return ".".concat(t);
      }).join("")), e;
    }).filter(function (t) {
      return t;
    }).join(",") : "";
  }

  var r,
      o,
      l,
      u,
      S = function (n) {
    function t(t) {
      var e,
          o = this;
      return (o = n.call(this, {
        core: null == t ? void 0 : t.core,
        sampling: null != (e = null == t ? void 0 : t.sampling) ? e : .1,
        version: "1.2.5",
        key: l$3.RES,
        name: "RadarResourceCollect"
      }) || this).ignoreList = v$4, o.broadcastJs = ae(), o.onResError = function (t, e) {
        var n;
        Q(t) || (n = (t.target.src || t.target.href || "").toString(), n = {
          failed: !0,
          file: n = "function" != typeof t.target.getAttribute || "" !== t.target.getAttribute("src") && "" !== t.target.getAttribute("href") ? n : "",
          res_path: b(t && t.path || t && t.composedPath && t.composedPath())
        }, o.collect(n, e), o.broadcastJs.emit(n));
      }, o.observeResource = function (t) {
        t.getEntriesByType("resource").forEach(function (e) {
          var t, n, r;
          if (0 === e.name.indexOf("data:") || D(e.name, o.ignoreList) || D(e.name, (null == (t = o.core) ? void 0 : t.ignoreList) || [])) return !1;
          "resource" === J(e) && (t = K(y$2, e), n = {
            encoded_body_size: e.encodedBodySize,
            decoded_body_size: e.decodedBodySize
          }, Object.keys(v$1).forEach(function (t) {
            n[t] = Number(e[v$1[t]] + performance.timing.navigationStart);
          }), r = _2(_2({
            failed: !1
          }, t), n), o.collect(r), setTimeout(function () {
            o.broadcastJs.emit(r);
          }));
        });
      }, o.resourceHook = t.resourceHook, void 0 !== t.ignoreList && (Array.isArray(t.ignoreList) || De("802"), t.ignoreList.forEach(function (t) {
        "string" != typeof t && De("802");
      }), o.ignoreList = o.ignoreList.concat(t.ignoreList)), o.init(), o;
    }

    var e = t,
        r = n;
    if ("function" != typeof r && null !== r) throw new TypeError("Class extends value " + String(r) + " is not a constructor or null");

    function o() {
      this.constructor = e;
    }

    return _g$2(e, r), e.prototype = null === r ? Object.create(r) : (o.prototype = r.prototype, new o()), t.prototype.init = function () {
      var e,
          n = this;

      try {
        this.observeResource(performance), this.observer = new PerformanceObserver(function (t) {
          return n.observeResource(t);
        }), this.observer.observe({
          entryTypes: ["resource"]
        });
      } catch (t) {
        u$4("801"), null != (e = this.core) && e.collect("event", {
          name: "radar_error",
          message: "PerformanceObserver 不存在无法收集数据",
          category: l$3.RES
        });
      }

      B(window, "error", function (t) {
        return n.onResError(t);
      }, !0), this.parseSeedLog();
    }, t.prototype.parseSeedLog = function () {
      var n = this,
          t = Re(j, []);
      t.length && (t.map(function (t) {
        var e = t.data;
        t = t.dimension || {}, n.onResError(e, t);
      }), Oe(j, []));
    }, t.prototype.collect = function (t, e) {
      var n = t.cached,
          r = t.failed,
          o = t.file,
          i = t.protocol,
          a = t.res_type,
          s = t.res_path,
          c = t.cached_type,
          u = t.cdn_ip,
          l = t.has_timing_allow;

      t = function (t, e) {
        var n = {};

        for (o in t) {
          Object.prototype.hasOwnProperty.call(t, o) && e.indexOf(o) < 0 && (n[o] = t[o]);
        }

        if (null != t && "function" == typeof Object.getOwnPropertySymbols) for (var r = 0, o = Object.getOwnPropertySymbols(t); r < o.length; r++) {
          e.indexOf(o[r]) < 0 && Object.prototype.propertyIsEnumerable.call(t, o[r]) && (n[o[r]] = t[o[r]]);
        }
        return n;
      }(t, ["cached", "failed", "file", "protocol", "res_type", "res_path", "cached_type", "cdn_ip", "has_timing_allow"]);

      try {
        var p = this.resourceHook && this.resourceHook({
          file: o,
          cached: n,
          protocol: i,
          res_type: a,
          failed: r
        });
        if (null != p && p.intercept_report) return;
      } catch (t) {
        u$4("803", t);
      }

      this.baseCollect({
        value: t,
        dimension: _2({
          cached: n,
          failed: r,
          file: o,
          protocol: i,
          res_type: a,
          res_path: s,
          sample_rate: this.sampling,
          is_official: 1,
          cached_type: c,
          cdn_ip: u,
          has_timing_allow: l
        }, e)
      });
    }, t.prototype.destroy = function () {
      this.isHit && ($(window, "error", this.onResError, !0), this.observer && this.observer.disconnect());
    }, t;
  }(d$3),
      _a$ = function a$1(t, e) {
    return (_a$ = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (t, e) {
      t.__proto__ = e;
    } || function (t, e) {
      for (var n in e) {
        Object.prototype.hasOwnProperty.call(e, n) && (t[n] = e[n]);
      }
    })(t, e);
  },
      c = function c() {
    return "".concat(Date.now(), "-").concat(Math.floor(8999999999999 * Math.random()) + 1e12);
  },
      s$1 = function s$1(t) {
    return {
      name: t,
      value: 1 < arguments.length && void 0 !== arguments[1] ? arguments[1] : -1,
      delta: 0,
      entries: [],
      id: c(),
      isFinal: !1
    };
  },
      u$2 = function u$2(t, e) {
    try {
      var n;
      if (PerformanceObserver.supportedEntryTypes.includes(t)) return (n = new PerformanceObserver(function (t) {
        return t.getEntries().map(e);
      })).observe({
        type: t,
        buffered: !0
      }), n;
    } catch (t) {}
  },
      l$1 = !1,
      p$1 = !1,
      f$2 = function f$2(t) {
    l$1 = !t.persisted;
  },
      m$1 = function m$1() {
    addEventListener("pagehide", f$2), addEventListener("beforeunload", function () {});
  },
      d$1 = function d$1(e) {
    var t = 1 < arguments.length && void 0 !== arguments[1] && arguments[1];
    p$1 || (m$1(), p$1 = !0), addEventListener("visibilitychange", function (t) {
      t = t.timeStamp, "hidden" === document.visibilityState && e({
        timeStamp: t,
        isUnloading: l$1
      });
    }, {
      capture: !0,
      once: t
    });
  },
      v = function v(t, e, n, r) {
    var o;
    return function () {
      n && e.isFinal && n.disconnect(), 0 <= e.value && (r || e.isFinal || "hidden" === document.visibilityState) && (e.delta = e.value - (o || 0), (e.delta || e.isFinal || void 0 === o) && (t(e), o = e.value));
    };
  },
      h$1 = function h$1() {
    return void 0 === r && (r = "hidden" === document.visibilityState ? 0 : 1 / 0, d$1(function (t) {
      return t = t.timeStamp, r = t;
    }, !0)), {
      get timeStamp() {
        return r;
      }

    };
  },
      y$1 = function y$1() {
    return o = o || new Promise(function (e) {
      return ["scroll", "keydown", "pointerdown"].map(function (t) {
        addEventListener(t, e, {
          once: !0,
          passive: !0,
          capture: !0
        });
      });
    });
  },
      F = [{
    name: "fid",
    collectFn: function collectFn(t) {
      function e(t) {
        t.startTime < r.timeStamp && (n.value = t.processingStart - t.startTime, n.entries.push(t), n.isFinal = !0, i());
      }

      var n = s$1("FID"),
          r = h$1(),
          o = u$2("first-input", e),
          i = v(t, n, o);
      o ? d$1(function () {
        o.takeRecords().map(e), o.disconnect();
      }, !0) : window.perfMetrics && window.perfMetrics.onFirstInputDelay && window.perfMetrics.onFirstInputDelay(function (t, e) {
        e.timeStamp < r.timeStamp && (n.value = t, n.isFinal = !0, n.entries = [{
          entryType: "first-input",
          name: e.type,
          target: e.target,
          cancelable: e.cancelable,
          startTime: e.timeStamp,
          processingStart: e.timeStamp + t
        }], i());
      });
    }
  }, {
    name: "lcp",
    collectFn: function collectFn(t) {
      function e(t) {
        var e = t.startTime;
        e < i.timeStamp ? (o.value = e, o.entries.push(t)) : o.isFinal = !0, n();
      }

      var n,
          r = 1 < arguments.length && void 0 !== arguments[1] && arguments[1],
          o = s$1("LCP"),
          i = h$1(),
          a = u$2("largest-contentful-paint", e);
      a && (n = v(t, o, a, r), r = function r() {
        o.isFinal || (a.takeRecords().map(e), o.isFinal = !0, n());
      }, y$1().then(r), d$1(r, !0));
    }
  }, {
    name: "cls",
    collectFn: function collectFn(t) {
      function e(t) {
        t.hadRecentInput || (o.value += t.value, o.entries.push(t), n());
      }

      var n,
          r = 1 < arguments.length && void 0 !== arguments[1] && arguments[1],
          o = s$1("CLS", 0),
          i = u$2("layout-shift", e);
      i && (n = v(t, o, i, r), d$1(function (t) {
        t = t.isUnloading, i.takeRecords().map(e), t && (o.isFinal = !0), n();
      }));
    }
  }, {
    name: "fcp",
    collectFn: function collectFn(t) {
      var e,
          n = s$1("FCP"),
          r = h$1(),
          o = u$2("paint", function (t) {
        "first-contentful-paint" === t.name && t.startTime < r.timeStamp && (n.value = t.startTime, n.isFinal = !0, n.entries.push(t), e());
      });
      o && (e = v(t, n, o));
    }
  }],
      g$1 = function (e) {
    function t(t) {
      var r = e.call(this, {
        core: t.core,
        sampling: t.sampling,
        version: "1.2.5",
        key: l$3.EVENT,
        name: "RadarChromeMetricsCollect"
      }) || this;
      return r.options = {
        lcp: !1,
        cls: !1,
        fid: !1,
        fcp: !1
      }, r.init = function () {
        var n = r.options;
        F.forEach(function (t) {
          var e = t.name;
          n && !n[e] || t.collectFn(function (t) {
            var e = t.name;
            t = t.value, qe(e.toLocaleLowerCase(), Number(t)), r.baseCollect({
              dimension: {
                name: e.toLocaleLowerCase(),
                is_official: 1
              },
              value: {
                duration: Number(t)
              }
            });
          });
        });
      }, Object.assign(r.options, t.options), F.forEach(function (t) {
        var e = se("radar_" + t.name);
        e && (r.options[t.name] = "1" === e);
      }), r.init(), r;
    }

    var n = t,
        r = e;
    if ("function" != typeof r && null !== r) throw new TypeError("Class extends value " + String(r) + " is not a constructor or null");

    function o() {
      this.constructor = n;
    }

    return _a$(n, r), n.prototype = null === r ? Object.create(r) : (o.prototype = r.prototype, new o()), t;
  }(d$3),
      _f$ = function f$1(t, e) {
    return (_f$ = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (t, e) {
      t.__proto__ = e;
    } || function (t, e) {
      for (var n in e) {
        Object.prototype.hasOwnProperty.call(e, n) && (t[n] = e[n]);
      }
    })(t, e);
  },
      _s = function s() {
    return (_s = Object.assign || function (t) {
      for (var e, n = 1, r = arguments.length; n < r; n++) {
        for (var o in e = arguments[n]) {
          Object.prototype.hasOwnProperty.call(e, o) && (t[o] = e[o]);
        }
      }

      return t;
    }).apply(this, arguments);
  },
      u$1 = function (n) {
    function t(t) {
      var e = n.call(this, {
        core: t.core,
        sampling: t.sampling,
        key: l$3.EVENT,
        version: "1.2.5",
        name: "RadarEventCollect"
      }) || this;
      return e.action = {
        start: function start(t) {
          e._checkActionDimension(t), e.event({
            name: t.name,
            extra_info: t.extra_info,
            event_type: "radar_action_start"
          });
        },
        end: function end(t) {
          e._checkActionDimension(t), e.event({
            name: t.name,
            extra_info: t.extra_info,
            result_type: t.result_type,
            event_type: "radar_action_end"
          });
        }
      }, e.parseSeedLog(), e;
    }

    var e = t,
        r = n;
    if ("function" != typeof r && null !== r) throw new TypeError("Class extends value " + String(r) + " is not a constructor or null");

    function o() {
      this.constructor = e;
    }

    return _f$(e, r), e.prototype = null === r ? Object.create(r) : (o.prototype = r.prototype, new o()), t.prototype.parseSeedLog = function () {
      var n = this,
          t = Re(A$1, []);
      t.length && (t.map(function (t) {
        var e = t.dimension;
        t = t.value, n.event(e, t);
      }), Oe(A$1, []));
    }, t.prototype.event = function (t, e) {
      var n;
      this.isHit ? (t || De("402"), void 0 !== t.name && "string" == typeof t.name || De("402"), 127 < t.name.length && (u$4("403"), t.name = t.name.slice(0, 127)), t.is_official = 1, e && void 0 !== e.duration && ("number" != typeof e.duration || e.duration < 0) && De("404"), e && void 0 !== e.event_count && ("number" != typeof e.event_count || e.event_count < 0) && De("405，".concat(e.event_count)), n = {
        dimension: t,
        value: e
      }, t && t.extra_info && "object" == _typeof(t.extra_info) && (n = {
        dimension: _s(_s({}, t), {
          extra_info: JSON.stringify(t.extra_info)
        }),
        value: e
      }), this.baseCollect(n)) : s$2("401");
    }, t.prototype.destroy = function () {
      this.event = function () {
        s$2("407");
      }, this.action = {
        start: function start() {
          s$2("407");
        },
        end: function end() {
          s$2("407");
        }
      };
    }, t.prototype._checkActionDimension = function (t) {
      if (!t) throw Error("408");
      void 0 === t.extra_info || t.extra_info instanceof Object || De("406");
    }, t;
  }(d$3),
      _a = function a(t, e) {
    return (_a = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (t, e) {
      t.__proto__ = e;
    } || function (t, e) {
      for (var n in e) {
        Object.prototype.hasOwnProperty.call(e, n) && (t[n] = e[n]);
      }
    })(t, e);
  },
      _p = function p() {
    return (_p = Object.assign || function (t) {
      for (var e, n = 1, r = arguments.length; n < r; n++) {
        for (var o in e = arguments[n]) {
          Object.prototype.hasOwnProperty.call(e, o) && (t[o] = e[o]);
        }
      }

      return t;
    }).apply(this, arguments);
  };

  function f(t, e) {
    var n = {};

    for (o in t) {
      Object.prototype.hasOwnProperty.call(t, o) && e.indexOf(o) < 0 && (n[o] = t[o]);
    }

    if (null != t && "function" == typeof Object.getOwnPropertySymbols) for (var r = 0, o = Object.getOwnPropertySymbols(t); r < o.length; r++) {
      e.indexOf(o[r]) < 0 && Object.prototype.propertyIsEnumerable.call(t, o[r]) && (n[o[r]] = t[o[r]]);
    }
    return n;
  }

  (u = l = l || {}).SCRIPT = "script", u.RES = "res", u.API = "api", u.VIDEO = "video";
  var y = [];

  function m(t, e, n) {
    var r, o;

    if (e && (e instanceof Error || "object" == _typeof(e) && "string" == typeof e.message)) {
      if ("BridgeError" === (r = e.name || "ErrorLikeObject") && "function" == typeof e.toJSON) try {
        var i = e.toJSON();
        delete i.name, delete i.message, o = "".concat(e.message, " (").concat(JSON.stringify(i), ")");
      } catch (t) {
        o = e.message || "";
      } else o = e.message || "";
      i = e.stack || void 0;
    } else o = null == e && "Script error." === (null == n ? void 0 : n.message) ? (r = "CrossOriginError", "".concat(t, " - message: ").concat((null == n ? void 0 : n.message) || "<UNKNOWN>")) : null === e && null != n && n.message && "Uncaught null" !== n.message ? (r = "Error", "".concat(t, " - message: ").concat(n.message)) : (r = null == e ? "".concat(e) : _typeof(e), n = function (t) {
      var e;

      try {
        e = t.toString(), t && "object" == _typeof(t) && (e = JSON.stringify(t));
      } catch (t) {}

      return "string" != typeof e || "" === e ? e = null === t ? "<NULL>" : void 0 === t ? "<UNDEFINED>" : "" === e ? "<EMPTY>" : "<UNKNOWN>" : 2e3 < e.length && (e = e.substring(0, 2e3) + "<<< ……"), e;
    }(e), "".concat(t, " - value: ").concat(n));

    return {
      error_type: l.SCRIPT,
      error_cons_type: r,
      msg: o,
      stack: i
    };
  }

  function g(t) {
    for (var e = 0; e < y.length; e++) {
      if (d(y[e], t)) return e;
    }

    return -1;
  }

  function d(t, e) {
    return t.msg === e.msg && t.stack === e.stack && t.file === e.file && t.line === e.line && t.col === e.col;
  }

  var h = function (e) {
    function t(t) {
      var r = e.call(this, {
        core: t.core,
        sampling: t.sampling,
        key: l$3.ERROR,
        version: "1.2.5",
        name: "RadarErrorCollect"
      }) || this;
      return r.onError = function (t, e) {
        var n;
        !(n = t).message && null == n.lineno || (t = _p(_p({}, m("Global Uncaught Exception", t.error, t)), {
          file: t.filename,
          line: t.lineno,
          col: t.colno
        }), r.isSameErrorAndReport(t, e));
      }, r.onPromiseError = function (t, e) {
        t = m("Unhandled Promise Rejection", t.reason), r.isSameErrorAndReport(t, e);
      }, r.core = t.core, r.init(), r;
    }

    var n = t,
        r = e;
    if ("function" != typeof r && null !== r) throw new TypeError("Class extends value " + String(r) + " is not a constructor or null");

    function o() {
      this.constructor = n;
    }

    return _a(n, r), n.prototype = null === r ? Object.create(r) : (o.prototype = r.prototype, new o()), t.prototype.init = function () {
      y.splice(0), B(window, "error", this.onError, !0), B(window, "unhandledrejection", this.onPromiseError), this.parseSeedLog();
    }, t.prototype.destroy = function () {
      $(window, "error", this.onError, !0), $(window, "unhandledrejection", this.onPromiseError), this.captureException = function () {
        s$2("600");
      };
    }, t.prototype.parseSeedLog = function () {
      var n = this,
          t = Re(T, []);
      t.length && (t.map(function (t) {
        var e = t.data;
        t = t.dimension || {}, "unhandledrejection" === (null == e ? void 0 : e.type) ? n.onPromiseError(e, t) : n.onError(e, t);
      }), Oe(T, []));
    }, t.prototype.isSameErrorAndReport = function (t, e) {
      0 <= g(t) || (y.push(t), setTimeout(function () {
        y.splice(g(t), 1);
      }, 2e3), this.collect(t, e));
    }, t.prototype.collect = function (t, e) {
      var n, r, o, i, a, s, c, u;
      r = (n = t).error_cons_type, o = n.file, i = n.error_type, a = n.msg, s = n.col, c = n.line, u = n.stack, t = {
        key: "error",
        value: f(n, ["error_cons_type", "file", "error_type", "msg", "col", "line", "stack"]),
        dimension: {
          error_cons_type: r,
          file: o,
          error_type: i,
          msg: a,
          col: s,
          line: c,
          stack: u
        }
      }, e = _p(_p({}, t.dimension), e || {}), this.baseCollect({
        dimension: e,
        value: t.value
      });
    }, t.prototype.captureException = function (t) {
      this.isSameErrorAndReport(m("Captured Exception", t));
    }, t;
  }(d$3),
      Radar$1 = function () {
    function t(t) {
      var e,
          n = this;
      this.core = new w({
        weblogger: t.weblog,
        projectId: t.projectId,
        sampling: null != (e = t.sampling) ? e : 1,
        httpReportFirst: t.httpReportFirst,
        customDimensions: t.customDimensions
      }), this.navigationCollect = new v$2({
        core: this.core,
        mainApi: t.mainApi,
        mainJs: t.mainJs
      }), t.unuseAPIPlugin || (this.apiCollect = new A({
        core: this.core,
        sampling: t.apiSampling,
        APIHook: t.APIHook,
        customizeRadarStatus: t.customizeRadarStatus,
        ignoreList: t.ignoreList,
        ignoreXHR: t.ignoreXHR,
        ignoreFetch: t.ignoreFetch,
        isIgnoreInvalidStatusCode: t.isIgnoreInvalidStatusCode
      }), this.apiCollect.broadcastApi.on(function (t) {
        return n.navigationCollect.dealAPI(t);
      })), t.unuseResourcePlugin || (this.resourceCollect = new S({
        sampling: t.resourceSampling,
        core: this.core,
        ignoreList: t.ignoreList,
        resourceHook: t.resourceHook
      }), this.resourceCollect.broadcastJs.on(function (t) {
        return n.navigationCollect.addMainJSResult(t);
      })), this.chromeMetricsCollect = new g$1({
        core: this.core,
        options: {
          lcp: t.lcp,
          cls: t.cls,
          fid: t.fid,
          fcp: t.fcp
        }
      }), this.eventCollect = new u$1({
        core: this.core
      }), this.action = this.eventCollect.action, t.unuseErrorPlugin || (this.errorCollect = new h({
        core: this.core
      })), _e();
    }

    return t.prototype.apply = function (t) {
      this.bindWeblogger(t);
    }, t.prototype.bindWeblogger = function (t) {
      this.weblogger = t, this.core.bindWeblogger(this.weblogger);
    }, t.prototype.event = function (t, e) {
      this.eventCollect.event(t, e);
    }, t.prototype.customStage = function (t, e) {
      this.navigationCollect.customStage(t, e);
    }, t.prototype.setDimensions = function (t) {
      this.core.setDimensions(t);
    }, t.prototype.fmp = function (t) {
      "number" == typeof t ? this.navigationCollect.fmp(t) : "object" == _typeof(t) && t.timestamp ? this.navigationCollect.fmp(t.timestamp) : this.navigationCollect.fmp();
    }, t.prototype.captureException = function (t) {
      var e;
      this.errorCollect ? null != (e = this.errorCollect) && e.captureException(t) : console.warn("902");
    }, t.prototype.destroy = function () {
      var t;
      null != (t = this.navigationCollect) && t.destroy(), null != (t = this.apiCollect) && t.destroy(), null != (t = this.resourceCollect) && t.destroy(), null != (t = this.eventCollect) && t.destroy(), null != (t = this.errorCollect) && t.destroy();
    }, t.key = "radar", t.version = "1.2.5", t;
  }();

  "undefined" != typeof window && "function" != typeof window.Radar && (window.Radar = Radar$1);

  var BasePlugin = /** @class */ (function () {
      function BasePlugin() {
      }
      BasePlugin.prototype.apply = function (weblog) {
          this.weblog = weblog;
      };
      return BasePlugin;
  }());

  function warn() {
      var args = [];
      for (var _i = 0; _i < arguments.length; _i++) {
          args[_i] = arguments[_i];
      }
      try {
          var warn_1 = console === null || console === void 0 ? void 0 : console.warn;
          return warn_1 && warn_1.call.apply(warn_1, __spreadArray([console], args, false));
      }
      catch (e) {
          return;
      }
  }
  var sdkName = 'webLogger';
  function error() {
      var args = [];
      for (var _i = 0; _i < arguments.length; _i++) {
          args[_i] = arguments[_i];
      }
      try {
          var error_1 = sdkName === 'webLogger_krn' ? console === null || console === void 0 ? void 0 : console.log : console === null || console === void 0 ? void 0 : console.error;
          return error_1 && error_1.call.apply(error_1, __spreadArray([console], args, false));
      }
      catch (e) {
          return;
      }
  }

  /**
   * @file Radar SDK
   * @author guwanpo
   */
  var CUSTOM_COLLECT_KEY = [];
  var Radar = /** @class */ (function (_super) {
      __extends(Radar, _super);
      function Radar(weblog, options) {
          var _this = _super.call(this) || this;
          // 有options证明是用配置的方式初始化的，否则是new的方式初始化的，将第一个参数给this.options
          _this.options = options ? options : weblog;
          // 验证配置内容
          if (!_this.validOption()) {
              return _this;
          }
          // 配置方式初始化radar时，weblogger会把weblogger对象传进来，而且不会调用apply方法，这里我们需要手动调用一下
          // https://git.corp.kuaishou.com/ks-frontend/modules/webLogger/-/blob/master/src/entries/browser-full.ts#L49
          if (weblog === null || weblog === void 0 ? void 0 : weblog.collect) {
              _this.apply(weblog);
          }
          return _this;
      }
      Radar.prototype.apply = function (weblog) {
          var _a;
          this.weblog = weblog;
          this.radarInstance = new Radar$1(_assign(_assign({}, this.options), { weblog: weblog }));
          this.action = (_a = this.radarInstance) === null || _a === void 0 ? void 0 : _a.action;
      };
      Radar.prototype.destroy = function () {
          var _a;
          (_a = this.radarInstance) === null || _a === void 0 ? void 0 : _a.destroy();
      };
      Radar.prototype.collect = function (key, duration) {
          var _a;
          if (key.indexOf('c_') !== 0 && CUSTOM_COLLECT_KEY.indexOf(key) < 0) {
              error('[error 305]', "[Radar] \u4F20\u5165\u7684\u81EA\u5B9A\u4E49\u6307\u6807 ".concat(key, " \u5FC5\u987B\u4E3A c_ \u5F00\u5934"));
              return;
          }
          warn('[radar]:Radar+之后将不再支持collect方法，请用event方法替代。');
          (_a = this.radarInstance) === null || _a === void 0 ? void 0 : _a.event({ name: key }, { duration: duration });
      };
      Radar.prototype.fmp = function (time) {
          var _a;
          (_a = this.radarInstance) === null || _a === void 0 ? void 0 : _a.fmp(time);
      };
      Radar.prototype.event = function (dimension, value) {
          var _a;
          (_a = this.radarInstance) === null || _a === void 0 ? void 0 : _a.event(dimension, value);
      };
      Radar.prototype.customStage = function (name, options) {
          var _a;
          (_a = this.radarInstance) === null || _a === void 0 ? void 0 : _a.customStage(name, options);
      };
      // 用于无头浏览器环境手动模拟onload事件内容进行perf数据收集
      Radar.prototype.headlessTestReport = function () {
          warn('[radar]:Radar+已不支持不再支持headlessTestReport方法');
      };
      Radar.prototype.playClicked = function (time) {
          warn('[radar]:Radar+已不支持不再支持playClicked方法');
      };
      Radar.prototype.playIDR = function (time) {
          warn('[radar]:Radar+已不支持不再支持playIDR方法');
      };
      Radar.prototype.playBlock = function (time) {
          warn('[radar]:Radar+已不支持不再支持playBlock方法');
      };
      Radar.prototype.playError = function (time) {
          warn('[radar]:Radar+已不支持不再支持playError方法');
      };
      Radar.prototype.setDimensions = function (param) {
          var _a;
          (_a = this.radarInstance) === null || _a === void 0 ? void 0 : _a.setDimensions(param);
      };
      Radar.prototype.validOption = function () {
          var _this = this;
          if (!this.options) {
              // 未传入参数默认不开启
              return false;
          }
          var _a = this.options, _b = _a.sampling, sampling = _b === void 0 ? 1 : _b, projectId = _a.projectId;
          if (typeof projectId !== 'string') {
              // error('[error 303]', '[Radar] 必须配置 projectId 属性，类型为 string!');
              return false;
          }
          if (sampling > 1 || sampling < 0) {
              // error('[error 304]', '[Radar] 采样率 sampling 必须为 0~1 之间的数字');
              return false;
          }
          [
              'apiSampling', 'errorSampling', 'resourceSampling', 'eventSampling',
          ].forEach(function (optionName) {
              var sampling = _this.options[optionName];
              if (sampling === undefined) {
                  return;
              }
              if (typeof sampling === 'number' && sampling === 0) {
                  error('[error 304]', "[Radar] \u91C7\u6837\u7387 ".concat(optionName, " \u4E0D\u80FD\u88AB\u8BBE\u7F6E\u4E3A 0\uFF0C\u5C06\u6539\u7528\u9ED8\u8BA4\u503C"));
              }
          });
          return true;
      };
      /**
       * 用于主动记录某个错误，并在雷达后台中跟踪它们。
       *
       * ## 场景 一
       *
       * 对于使用 try...catch... 捕获的错误，若想要记录并跟踪它们，则可以使用该方法：
       *
       * ```javascript
       * try {
       *     doSomething();
       * } catch (error) {
       *     weblog.plugins.radar.captureException(error);
       * }
       * ```
       *
       * ## 场景 二
       *
       * 用于记录 Vue 组件中的未捕获异常：
       *
       * ```javascript
       * Vue.config.errorHandler = function (error) {
       *     weblog.plugins.radar.captureException(error);
       * }
       * ```
       *
       * @param error - 错误信息，最好是传入一个 Error 对象，因为这可以让我们记录下该错误发生的地点以及其它有用的信息。
       */
      Radar.prototype.captureException = function (error) {
          var _a;
          (_a = this.radarInstance) === null || _a === void 0 ? void 0 : _a.captureException(error);
      };
      Radar.key = 'radar';
      return Radar;
  }(BasePlugin));

  return Radar;

}));
