// @locallife/biz-component/lib/commonjs/utils/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.REFUNDED_RULE = exports.PRODUCT_DETAIL_PAGE_HORIZONTAL_PADDING = exports.PRODUCT_DETAIL_PAGE_HORIZONTAL_MARGIN = exports.PRODUCT_DETAIL_COMPONENT_MARGIN_GAP = exports.PRODUCT_DETAIL_COMPONENT_HORIZONTAL_PADDING = exports.BUY_LIMIT_RULE = void 0;
const PRODUCT_DETAIL_COMPONENT_MARGIN_GAP = 12;
exports.PRODUCT_DETAIL_COMPONENT_MARGIN_GAP = PRODUCT_DETAIL_COMPONENT_MARGIN_GAP;
const PRODUCT_DETAIL_PAGE_HORIZONTAL_MARGIN = 11;
exports.PRODUCT_DETAIL_PAGE_HORIZONTAL_MARGIN = PRODUCT_DETAIL_PAGE_HORIZONTAL_MARGIN;
const PRODUCT_DETAIL_PAGE_HORIZONTAL_PADDING = 12;
exports.PRODUCT_DETAIL_PAGE_HORIZONTAL_PADDING = PRODUCT_DETAIL_PAGE_HORIZONTAL_PADDING;
const PRODUCT_DETAIL_COMPONENT_HORIZONTAL_PADDING = 14;

// 退款规则
exports.PRODUCT_DETAIL_COMPONENT_HORIZONTAL_PADDING = PRODUCT_DETAIL_COMPONENT_HORIZONTAL_PADDING;
let REFUNDED_RULE = /*#__PURE__*/function (REFUNDED_RULE) {
  REFUNDED_RULE["NO_WRITE_OFF_REFUND_ANY_TIME"] = "\u672A\u4F7F\u7528\u968F\u65F6\u9000\u3001\u8FC7\u671F\u81EA\u52A8\u9000";
  REFUNDED_RULE["CAN_NOT_REFUND"] = "\u4E0D\u53EF\u9000";
  REFUNDED_RULE["WITH_CONDITION_REFUND"] = "\u6709\u6761\u4EF6\u9000\uFF0C\u5177\u4F53\u89C4\u5219\u8BF7\u54A8\u8BE2\u5BA2\u670D";
  REFUNDED_RULE["EFFECTIVE_NOT_REFUND_EXPIRE_AUTO_REFUND"] = "\u6709\u6548\u671F\u5185\u4E0D\u652F\u6301\u53D1\u8D77\u9000\u6B3E\uFF0C\u8FC7\u671F\u672A\u4F7F\u7528\uFF0C\u5E73\u53F0\u5C06\u81EA\u52A8\u9000\u6B3E";
  return REFUNDED_RULE;
}({}); // 限购规则文案映射关系
exports.REFUNDED_RULE = REFUNDED_RULE;
const BUY_LIMIT_RULE = {
  UNLIMITED_PURCHASE: [{
    key: 'onLimit',
    text: '不限购'
  }],
  SINGLE_ORDER_AMOUNT: [{
    key: 'singleOrderLimitNum',
    text: '每笔订单限购{num}份'
  }],
  ORDER: [{
    key: 'orderLeastNum',
    text: '每笔订单最少购买{num}份，'
  }, {
    key: 'orderMostNum',
    text: '最多购买{num}份'
  }],
  MOBILE_PHONE: [{
    key: 'phoneInDay',
    text: '同一手机号在{num}天内，'
  }, {
    key: 'phoneMostNum',
    text: '最多购买{num}份'
  }],
  CARD_ID: [{
    key: 'identityCardInDay',
    text: '同一身份证号在{num}天内，'
  }, {
    key: 'identityCardMostNum',
    text: '最多购买{num}份'
  }],
  ACCOUNT: [{
    key: 'accountInDay',
    text: '同一账号{num}天内，'
  }, {
    key: 'accountMostNum',
    text: '最多购买{num}份'
  }],
  MOBILE_PHONE_NO_TIME: [{
    key: 'phoneNoTimeLimitNum',
    text: '每个手机号最多购买{num}份'
  }],
  MOBILE_PHONE_PLAY_TIME: [{
    key: 'phonePlayTimeLimitNum',
    text: '每个手机号游玩日期最多购买{num}份'
  }],
  CARD_ID_PLAY_TIME: [{
    key: 'idCardPlayTimeLimitNum',
    text: '每个身份证游玩日期最多购买{num}份'
  }],
  ACCOUNT_NO_TIME: [{
    key: 'accountNoTimeLimitNum',
    text: '每个账号最多购买{num}份'
  }]
};
exports.BUY_LIMIT_RULE = BUY_LIMIT_RULE;
