// @kid-ui/krn/src/components/slide-operation/types.ts
import { ReactNode } from 'react';
import { TextStyle } from 'react-native';

export interface ActionType {
    bgColor?: string;
    text?: string;
    textStyle?: TextStyle;
    render?: () => ReactNode;
    onClick?: () => void;
}

export interface ActionItemType {
    item: ActionType;
    index: number;
    actionWidth: number;
}

export interface CustomStyleProps {
    containerHeight?: number;
    actionTextStyle?: TextStyle;
    actionActiveOpacity?: number;
}

export interface SlideOperationProps {
    rootTag?: number;
    customStyle?: CustomStyleProps;
    enableLeftSlideBack?: boolean;
    leftActions?: Array<ActionType>;
    leftActionWidth?: number;
    rightActions?: Array<ActionType>;
    rightActionWidth?: number;
    actionWidth?: number;
    children?: ReactNode;
    canSlideBackSpace?: number;
    index?: number;
    onCallBack?: (inx: number) => void;
    close?: boolean;
    closeAfterClickedBtn?: boolean;
    numberOfLines?: number;
    ellipsizeMode?: "head" | "clip" | "middle" | "tail" | undefined
}
