// @krn/cli/lib/utils/doctor-diagnose/symptoms/metro-react-native-babel-preset-version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetroReactNativeBabelPresetVersion = void 0;
const base_1 = require("./base");
const fs = require("fs-extra");
const path = require("path");
class MetroReactNativeBabelPresetVersion extends base_1.Symptom {
    constructor() {
        super(...arguments);
        this.desc = 'metro-react-native-babel-preset 依赖版本';
        this.advice = '请将 metro-react-native-babel-preset 版本升级至 0.59.0，否则可能出现打包失败或开发环境红屏报错，安装完成后如果是开发环境，请使用 yarn start -r 重启应用';
        this.level = 'error';
    }
    async check() {
        var _a;
        const packageJson = fs.readJSONSync(path.resolve('package.json'));
        return /0\.59\.0$/.test(((_a = packageJson.devDependencies) === null || _a === void 0 ? void 0 : _a['metro-react-native-babel-preset']) ||
            '');
    }
    async checkWithBuildErrorMessage(msg) {
        return !msg.includes('Unexpected token punc «:», expected punc «,» in file node_modules/react-native/Libraries/Blob/URL.js');
    }
    async fix() {
        const packageJson = fs.readJSONSync(path.resolve('package.json'));
        packageJson.devDependencies['metro-react-native-babel-preset'] =
            '0.59.0';
        fs.writeJsonSync('./package.json', packageJson, {
            spaces: '    ',
        });
        return true;
    }
}
exports.MetroReactNativeBabelPresetVersion = MetroReactNativeBabelPresetVersion;
const metroReactNativeBabelPresetVersion = new MetroReactNativeBabelPresetVersion();
exports.default = metroReactNativeBabelPresetVersion;
