// @locallife/biz-component/src/components/commonLabels/Models/index.ts
export interface CDNUrl {
    url: string;
    cdn: string;
    [key: string]: string;
}

export interface BizTags {
    uniqKey: string;
    name: string;
    jumpUrl?: string;
    styleKey: string;
    content: string; // 解析之前是string，解析之后是TagContent
    /** *非协议内容，content解析之后的内容** */
    contentModel?: TagContent;
}

export interface TagContent {
    textInfos?: Array<TagTextInfo>;
    iconInfos?: Array<TagIconInfo>;
    contentStyle?: StyleInfo;
}

export interface StyleInfo {
    //border 相关
    borderRadius: number;
    borderColor: string;
    borderWidth: number;
    borderTopLeftRadius: number;
    borderTopRightRadius: number;
    borderBottomLeftRadius: number;
    borderBottomRightRadius: number;
    //margin 相关
    marginLeft?: number;
    marginRight?: number;
    //bg 相关
    bgColor?: string;
    //padding 相关
    paddingLeft?: number;
    paddingRight?: number;
    paddingBottom?: number;
    paddingTop?: number;
    height?: number;
}

export interface TagTextInfo {
    text: string;
    textStyle: TagTextStyle;
}

export interface TagTextStyle extends StyleInfo {
    textColor?: string;
    textColorDark?: string;
    fontSize?: number;
    fontWeight?:
        | 'normal'
        | 'bold'
        | '100'
        | '200'
        | '300'
        | '400'
        | '500'
        | '600'
        | '700'
        | '800'
        | '900';
    fontFamily?: string;
    width?: number;
    height?: number;
    top?: number;
    bottom?: number;
}

export interface TagIconInfo {
    iconUrl: CDNUrl;
    iconUrls: Array<CDNUrl>;
    darkIconUrls: Array<CDNUrl>;
    iconStyle: TagIconStyle;
}

export interface TagIconStyle extends StyleInfo {
    iconWidth: string;
    iconHeight: string;
}
