// @locallife/biz-component/lib/commonjs/components/commonText/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CommonText;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _tsLbsKrnUtils = require("@lux/ts-lbs-krn-utils");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const {
  screenSize
} = _tsLbsKrnUtils.screenUnit;
const styles = _reactNative.StyleSheet.create({
  container: {
    position: 'relative',
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center'
  },
  title: {
    fontSize: screenSize(14),
    lineHeight: screenSize(18),
    fontFamily: 'PingFangSC-Semibold',
    fontWeight: '600',
    color: '#222222',
    marginLeft: screenSize(20)
  },
  ulPoint: {
    position: 'absolute',
    top: screenSize(9),
    backgroundColor: '#666666',
    borderRadius: screenSize(3),
    width: screenSize(3),
    height: screenSize(3)
  },
  desc: {
    fontSize: screenSize(14),
    lineHeight: screenSize(18),
    fontWeight: '400',
    paddingLeft: screenSize(8),
    color: '#666666'
  }
});
function CommonText(props) {
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, {
      marginVertical: props.text ? screenSize(4) : 0
    }]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.ulPoint, props.customizationUlPoint && props.customizationUlPoint]
  }), props.text ? /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.desc, props.style]
  }, props.text) : props.children);
}
