// @locallife/biz-component/lib/module/components/commonLabels/CustomText.js
// 一个文案
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { transAdaptScreenSize } from './utils/transAdaptScreenSize';
import { useCommonStyleHook } from './hooks/useCommonStyleHook';
import { useAdaptThemeValueHook } from './hooks/useAdaptThemeValueHook';
const styles = StyleSheet.create({
  labelsContainer: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  text: {
    fontSize: 11,
    fontWeight: '400',
    fontFamily: 'PingFang SC',
    color: '#666666'
  }
});
// 一个文案
export const CustomText = /*#__PURE__*/React.memo(props => {
  const {
    textModel,
    adaptScreen,
    autoRatio,
    ratio,
    adaptDark = true
  } = props;
  const {
    textStyle
  } = textModel;
  const style = useCommonStyleHook(textStyle, adaptScreen);
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.labelsContainer, style]
  }, /*#__PURE__*/React.createElement(Text, {
    style: [styles.text, {
      color: useAdaptThemeValueHook({
        light: textModel.textStyle.textColor ?? styles.text.color,
        dark: textModel.textStyle.textColorDark ?? styles.text.color
      }, adaptDark),
      width: transAdaptScreenSize(textStyle === null || textStyle === void 0 ? void 0 : textStyle.width, adaptScreen),
      height: transAdaptScreenSize(textStyle === null || textStyle === void 0 ? void 0 : textStyle.height, adaptScreen),
      top: transAdaptScreenSize(textStyle === null || textStyle === void 0 ? void 0 : textStyle.top, adaptScreen),
      bottom: transAdaptScreenSize(textStyle === null || textStyle === void 0 ? void 0 : textStyle.bottom, adaptScreen)
    }, {
      fontSize: transAdaptScreenSize((textStyle === null || textStyle === void 0 ? void 0 : textStyle.fontSize) ?? styles.text.fontSize, adaptScreen, autoRatio, ratio),
      fontFamily: textModel.textStyle.fontFamily ?? styles.text.fontFamily,
      fontWeight: textModel.textStyle.fontWeight ?? styles.text.fontWeight
    }]
    // ellipsizeMode="tail"
    ,
    numberOfLines: 1
  }, textModel.text ?? ''));
});
