// @locallife/biz-component/lib/commonjs/components/popUp/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.popUp = void 0;
var _react = _interopRequireDefault(require("react"));
var _krn = require("@kid-ui/krn");
var _reactNative = require("react-native");
var _tsLbsKrnUtils = require("@lux/ts-lbs-krn-utils");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const styles = _reactNative.StyleSheet.create({
  close: {
    position: 'absolute',
    right: 0,
    paddingTop: _tsLbsKrnUtils.screenUnit.screenSize(9)
  }
});

/**
 * 文档参考 https://kds-components.corp.kuaishou.com/markdown/kid-ui/components/alert-bottom-card.html
 * 注意使用alertBottomCard父组件需要套一层WrapContext
 * @param title 标题
 * @param renderContent 自定义内容组件
 * @param config 自定义配置 同名配置自动替换
 * @returns alert弹窗
 */
const popUp = _ref => {
  let {
    title,
    renderContent,
    config = {}
  } = _ref;
  const alert = (0, _krn.AlertBottomCard)({
    content: '',
    title: title,
    titleStyle: {
      fontWeight: '500',
      fontSize: _tsLbsKrnUtils.screenUnit.screenSize(18),
      color: '#222222',
      marginBottom: _tsLbsKrnUtils.screenUnit.screenSize(8)
    },
    onConfirm: () => {
      alert.close();
    },
    renderCloseIcon: () => {
      return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: styles.close
      }, /*#__PURE__*/_react.default.createElement(_krn.KidIcon, {
        kid: 'common_dialog_colse_normal',
        kidColor: 'color_base_black_8',
        size: _tsLbsKrnUtils.screenUnit.screenSize(24)
      }));
    },
    onPressCloseIcon: () => {
      alert.close();
    },
    // 自定义内容区域，在描述文字的下方渲染
    renderContent: () => renderContent,
    renderButton: () => null,
    popupStyle: {
      width: '100%',
      paddingHorizontal: 0,
      borderTopLeftRadius: _tsLbsKrnUtils.screenUnit.screenSize(16),
      borderTopRightRadius: _tsLbsKrnUtils.screenUnit.screenSize(16)
    },
    position: 'bottom',
    ...config
  });
  return alert;
};
exports.popUp = popUp;
