// @ks/weblogger/dist/umd/log.hybrid.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@yoda/bridge')) :
    typeof define === 'function' && define.amd ? define(['exports', '@yoda/bridge'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.Weblog = {}, global.KSYoda));
})(this, (function (exports, bridge) { 'use strict';

    /**
     * 兼容 SSR、react native、worker 等非浏览器场景
     */
    var commonGlobal = typeof globalThis !== 'undefined' ? globalThis : typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {};
    (function () {
        if (typeof commonGlobal.window === 'undefined') {
            commonGlobal.window = {
                addEventListener: function () { },
                removeEventListener: function () { },
                setTimeout: function () { },
                setInterval: function () { },
                _WEBLOGGER_MOCKED_WINDOW_: true,
                global: commonGlobal
            };
        }
        if (typeof commonGlobal.document === 'undefined') {
            commonGlobal.document = commonGlobal.window.document || {
                cookie: '',
                addEventListener: function () { },
                removeEventListener: function () { },
                querySelector: function () { },
                querySelectorAll: function () { },
            };
            if (typeof commonGlobal.window.document === 'undefined') {
                commonGlobal.window.document = commonGlobal.document;
            }
        }
        if (typeof commonGlobal.navigator === 'undefined') {
            commonGlobal.navigator = commonGlobal.window.navigator || {
                userAgent: ''
            };
            if (typeof commonGlobal.window.navigator === 'undefined') {
                commonGlobal.window.navigator = commonGlobal.navigator;
            }
        }
        if (typeof commonGlobal.screen === 'undefined') {
            commonGlobal.screen = commonGlobal.window.screen || {};
            if (typeof commonGlobal.window.screen === 'undefined') {
                commonGlobal.window.screen = commonGlobal.screen;
            }
        }
        if (typeof commonGlobal.history === 'undefined') {
            commonGlobal.history = commonGlobal.window.history || {};
            if (typeof commonGlobal.window.history === 'undefined') {
                commonGlobal.window.history = commonGlobal.history;
            }
        }
        if (typeof commonGlobal.location === 'undefined') {
            commonGlobal.location = commonGlobal.window.location || {
                hostname: '',
                search: '',
                href: '',
                origin: ''
            };
            if (typeof commonGlobal.window.location === 'undefined') {
                commonGlobal.window.location = commonGlobal.location;
            }
        }
        try {
            if (typeof commonGlobal.localStorage === 'undefined') {
                commonGlobal.localStorage = commonGlobal.window.localStorage || {
                    getItem: function () { },
                    setItem: function () { }
                };
                if (typeof commonGlobal.window.localStorage === 'undefined') {
                    commonGlobal.window.localStorage = commonGlobal.localStorage;
                }
            }
        }
        catch (err) { }
    })();

    /**
     * Object.defineProperty 容错暂时处理
     */
    if (!Object.defineProperty) {
        // @ts-ignore
        Object.defineProperty = function (target, key, descriptor) {
            // @ts-ignore
            target[key] = Object.prototype.toString.call(descriptor) === '[object Object]'
                && descriptor.hasOwnProperty('value')
                ? descriptor.value
                : descriptor;
        };
    }
    /**
     * Object.assign polyfill
     */
    if (!Object.assign) {
        // Must be writable: true, enumerable: false, configurable: true
        Object.defineProperty(Object, 'assign', {
            value: function (target) {
                if (target === null) { // TypeError if undefined or null
                    throw new TypeError('Cannot convert undefined or null to object');
                }
                var to = Object(target);
                for (var index = 1; index < arguments.length; index++) {
                    var nextSource = arguments[index];
                    if (nextSource !== null) { // Skip over if undefined or null
                        for (var nextKey in nextSource) {
                            // Avoid bugs when hasOwnProperty is shadowed
                            if (Object.prototype.hasOwnProperty.call(nextSource, nextKey)) {
                                to[nextKey] = nextSource[nextKey];
                            }
                        }
                    }
                }
                return to;
            },
            writable: true,
            configurable: true,
        });
    }
    /**
     * bind polyfill
     */
    if (!Function.prototype.bind) {
        Function.prototype.bind = function (oThis) {
            if (typeof this !== 'function') {
                // closest thing possible to the ECMAScript 5
                // internal IsCallable function
                throw new TypeError('Function.prototype.bind - what is trying to be bound is not callable');
            }
            var aArgs = Array.prototype.slice.call(arguments, 1);
            var fToBind = this;
            var FNOP = /** @class */ (function () {
                function FNOP() {
                }
                return FNOP;
            }());
            var fBound = function () {
                var currentThis = this instanceof FNOP
                    ? this
                    : oThis;
                // this instanceof FNOP === true时,说明返回的fBound被当做new的构造函数调用
                return fToBind.apply(currentThis, 
                // 获取调用时(fBound)的传参.bind 返回的函数入参往往是这么传递的
                aArgs.concat(Array.prototype.slice.call(arguments)));
            };
            // 维护原型关系
            if (this.prototype) {
                // Function.prototype doesn't have a prototype property
                FNOP.prototype = this.prototype;
            }
            // 下行的代码使fBound.prototype是FNOP的实例,因此
            // 返回的fBound若作为new的构造函数,new生成的新对象作为this传入fBound,新对象的__proto__就是FNOP的实例
            fBound.prototype = new FNOP();
            return fBound;
        };
    }
    /**
     * indexOf polyfill
     */
    // Production steps of ECMA-262, Edition 5, 15.4.4.14
    // Reference: http://es5.github.io/#x15.4.4.14
    if (!Array.prototype.indexOf) {
        Array.prototype.indexOf = function (searchElement, fromIndex) {
            var k;
            // 1. Let O be the result of calling ToObject passing
            //    the this value as the argument.
            if (this == null) {
                throw new TypeError('"this" is null or not defined');
            }
            var O = Object(this);
            // 2. Let lenValue be the result of calling the Get
            //    internal method of O with the argument "length".
            // 3. Let len be ToUint32(lenValue).
            var len = O.length >>> 0;
            // 4. If len is 0, return -1.
            if (len === 0) {
                return -1;
            }
            // 5. If argument fromIndex was passed let n be
            //    ToInteger(fromIndex); else let n be 0.
            var n = +fromIndex || 0;
            if (Math.abs(n) === Infinity) {
                n = 0;
            }
            // 6. If n >= len, return -1.
            if (n >= len) {
                return -1;
            }
            // 7. If n >= 0, then Let k be n.
            // 8. Else, n<0, Let k be len - abs(n).
            //    If k is less than 0, then let k be 0.
            k = Math.max(n >= 0 ? n : len - Math.abs(n), 0);
            // 9. Repeat, while k < len
            while (k < len) {
                // a. Let Pk be ToString(k).
                //   This is implicit for LHS operands of the in operator
                // b. Let kPresent be the result of calling the
                //    HasProperty internal method of O with argument Pk.
                //   This step can be combined with c
                // c. If kPresent is true, then
                //    i.  Let elementK be the result of calling the Get
                //        internal method of O with the argument ToString(k).
                //   ii.  Let same be the result of applying the
                //        Strict Equality Comparison Algorithm to
                //        searchElement and elementK.
                //  iii.  If same is true, return k.
                if (k in O && O[k] === searchElement) {
                    return k;
                }
                k++;
            }
            return -1;
        };
    }
    if (!Object.keys) {
        Object.keys = (function () {
            var hasOwnProperty = Object.prototype.hasOwnProperty;
            var hasDontEnumBug = !({ toString: null }).propertyIsEnumerable('toString');
            var dontEnums = [
                'toString',
                'toLocaleString',
                'valueOf',
                'hasOwnProperty',
                'isPrototypeOf',
                'propertyIsEnumerable',
                'constructor',
            ];
            return function (obj) {
                if (typeof obj !== 'object' && typeof obj !== 'function' || obj === null) {
                    throw new TypeError('Object.keys called on non-object');
                }
                var result = [];
                for (var prop in obj) {
                    if (hasOwnProperty.call(obj, prop))
                        result.push(prop);
                }
                if (hasDontEnumBug) {
                    for (var _i = 0, dontEnums_1 = dontEnums; _i < dontEnums_1.length; _i++) {
                        var key = dontEnums_1[_i];
                        if (hasOwnProperty.call(obj, key)) {
                            result.push(key);
                        }
                    }
                }
                return result;
            };
        })();
    }
    // Production steps of ECMA-262, Edition 5, 15.4.4.18
    // Reference: http://es5.github.io/#x15.4.4.18
    if (!Array.prototype.forEach) {
        Array.prototype.forEach = function (callback, thisArg) {
            var T;
            var k;
            if (this == null) {
                throw new TypeError(' this is null or not defined');
            }
            // 1. Let O be the result of calling toObject() passing the
            // |this| value as the argument.
            var O = Object(this);
            // 2. Let lenValue be the result of calling the Get() internal
            // method of O with the argument "length".
            // 3. Let len be toUint32(lenValue).
            var len = O.length >>> 0;
            // 4. If isCallable(callback) is false, throw a TypeError exception.
            // See: http://es5.github.com/#x9.11
            if (typeof callback !== 'function') {
                throw new TypeError(callback + ' is not a function');
            }
            // 5. If thisArg was supplied, let T be thisArg; else let
            // T be undefined.
            if (arguments.length > 1) {
                T = thisArg;
            }
            // 6. Let k be 0
            k = 0;
            // 7. Repeat, while k < len
            while (k < len) {
                var kValue = void 0;
                // a. Let Pk be ToString(k).
                //    This is implicit for LHS operands of the in operator
                // b. Let kPresent be the result of calling the HasProperty
                //    internal method of O with argument Pk.
                //    This step can be combined with c
                // c. If kPresent is true, then
                if (k in O) {
                    // i. Let kValue be the result of calling the Get internal
                    // method of O with argument Pk.
                    kValue = O[k];
                    // ii. Call the Call internal method of callback with T as
                    // the this value and argument list containing kValue, k, and O.
                    callback.call(T, kValue, k, O);
                }
                // d. Increase k by 1.
                k++;
            }
            // 8. return undefined
        };
    }

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */

    /* global Reflect, Promise */
    var _extendStatics = function extendStatics(d, b) {
      _extendStatics = Object.setPrototypeOf || {
        __proto__: []
      } instanceof Array && function (d, b) {
        d.__proto__ = b;
      } || function (d, b) {
        for (var p in b) {
          if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p];
        }
      };

      return _extendStatics(d, b);
    };

    function __extends(d, b) {
      if (typeof b !== "function" && b !== null) throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");

      _extendStatics(d, b);

      function __() {
        this.constructor = d;
      }

      d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var _assign = function __assign() {
      _assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
          s = arguments[i];

          for (var p in s) {
            if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
          }
        }

        return t;
      };

      return _assign.apply(this, arguments);
    };
    function __rest(s, e) {
      var t = {};

      for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
      }

      if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
        if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
      }
      return t;
    }
    function __awaiter(thisArg, _arguments, P, generator) {
      function adopt(value) {
        return value instanceof P ? value : new P(function (resolve) {
          resolve(value);
        });
      }

      return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) {
          try {
            step(generator.next(value));
          } catch (e) {
            reject(e);
          }
        }

        function rejected(value) {
          try {
            step(generator["throw"](value));
          } catch (e) {
            reject(e);
          }
        }

        function step(result) {
          result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
        }

        step((generator = generator.apply(thisArg, _arguments || [])).next());
      });
    }
    function __generator(thisArg, body) {
      var _ = {
        label: 0,
        sent: function sent() {
          if (t[0] & 1) throw t[1];
          return t[1];
        },
        trys: [],
        ops: []
      },
          f,
          y,
          t,
          g;
      return g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
      }, typeof Symbol === "function" && (g[Symbol.iterator] = function () {
        return this;
      }), g;

      function verb(n) {
        return function (v) {
          return step([n, v]);
        };
      }

      function step(op) {
        if (f) throw new TypeError("Generator is already executing.");

        while (_) {
          try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];

            switch (op[0]) {
              case 0:
              case 1:
                t = op;
                break;

              case 4:
                _.label++;
                return {
                  value: op[1],
                  done: false
                };

              case 5:
                _.label++;
                y = op[1];
                op = [0];
                continue;

              case 7:
                op = _.ops.pop();

                _.trys.pop();

                continue;

              default:
                if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                  _ = 0;
                  continue;
                }

                if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                  _.label = op[1];
                  break;
                }

                if (op[0] === 6 && _.label < t[1]) {
                  _.label = t[1];
                  t = op;
                  break;
                }

                if (t && _.label < t[2]) {
                  _.label = t[2];

                  _.ops.push(op);

                  break;
                }

                if (t[2]) _.ops.pop();

                _.trys.pop();

                continue;
            }

            op = body.call(thisArg, _);
          } catch (e) {
            op = [6, e];
            y = 0;
          } finally {
            f = t = 0;
          }
        }

        if (op[0] & 5) throw op[1];
        return {
          value: op[0] ? op[1] : void 0,
          done: true
        };
      }
    }
    function __spreadArray(to, from, pack) {
      if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
          if (!ar) ar = Array.prototype.slice.call(from, 0, i);
          ar[i] = from[i];
        }
      }
      return to.concat(ar || Array.prototype.slice.call(from));
    }

    var GlobalConfig = {
        sessionId: '',
        // 小程序专用
        appDevicePackageReady: false,
        identityPackageReady: false, // identity数据是否设置完成，调用setPublicParam后 设置完成。
    };
    var updateConfig = function (key, value) {
        GlobalConfig[key] = value;
    };

    var UrlPackage$1 = /** @class */ (function () {
        function UrlPackage(url, pageType) {
            // 一级页面。参考埋点需求填写即可
            this.page = '';
            // 实例的全局唯一标识，使用UUID即可
            this.identity = '';
            switch (pageType) {
                case 'web':
                    this.page_type = 2;
                    break;
                case 'native':
                    this.page_type = 1;
                    break;
                case 'mina':
                    this.page_type = 3;
                    break;
                default:
                    this.page_type = 0; // unknown
            }
            this.update(url.page, url.params);
        }
        // 更新urlPackage
        UrlPackage.prototype.update = function (page, params) {
            if (page === void 0) { page = ''; }
            if (page && page !== this.page) {
                this.page = page;
                this.identity = this.generatePageId(page);
            }
            if (params) {
                this.params = Object.assign(this.params || {}, params);
            }
        };
        UrlPackage.prototype.toJSON = function () {
            return {
                page: this.page,
                identity: this.identity,
                page_type: this.page_type,
                params: JSON.stringify(this.params),
                category: this.category,
            };
        };
        /**
         * 生成页面唯一id
         * 未自定义页面唯一ID生成方法，请自行添加后重新尝试
         * @param page
         */
        UrlPackage.prototype.generatePageId = function (page) {
            return page + new Date().getTime();
        };
        return UrlPackage;
    }());

    /**
     * 快手系应用UA标示
     */
    // TODO 海外 APP List 要求不区分大小写判断
    var ksOverseaAppList = ['UVideo', 'Vstatus', 'Kwaigo', 'MvMaster'];
    // TODO这么加没个完了，后面想办法搞成自行传入吧
    var ksAppList = [
        'Kwai',
        'Kwai_Lite',
        'Kwai_Pro',
        'livemate',
        'ksthanos',
        'ksNebula',
        'ksnebula',
        'kwaiying',
        'pearl',
        'kinder',
        'm2u',
        'LOLita',
        'XFunCore',
        'ACVideoCore',
        'ZIKZAK', // ZIKZAK
    ].concat(ksOverseaAppList);

    /**
     * 创建一个cookie字符串，可被赋值给document.cookie.
     * @param {string} name cookie名称.
     * @param {string} value cookie值.
     * @param {boolean} encode 是否对值进行编码.
     * @param {Object} options (Optional) 配置项.
     * @return {string} 返回 cookie 字符串.
     */
    function createCookieString(name, value, encode, options) {
        if (options === void 0) { options = {}; }
        var text = "".concat(encodeURIComponent(name), "=").concat((encode ? encodeURIComponent(value) : value));
        var expires = options.expires;
        var path = options.path || '/';
        var domain = options.domain || '';
        if (expires instanceof Date) {
            text += "; expires=".concat(expires.toUTCString());
        }
        if (typeof expires === 'number') {
            text += "; max-age==".concat(expires);
        }
        if (path !== '') {
            text += "; path=".concat(path);
        }
        if (domain !== '') {
            text += "; domain=".concat(domain);
        }
        if (options.secure === true) {
            text += '; secure';
        }
        return text;
    }
    /**
     * 将cookie序列化成对象.
     * @param {string} text cookie值.
     * @param {boolean=} shouldDecode cookie是否需要被解码. 默认为true.
     * @param {object=} options 如何读取cookie的配置.
     * @return {object} 可访问cookie的序列化结果
     */
    function parseCookieString(text, shouldDecode, options) {
        var _a;
        var cookies = {};
        if (text.length > 0) {
            var decodeValue = !shouldDecode
                ? function (s) {
                    return s;
                }
                : decodeURIComponent;
            var cookieParts = text.split(/;\s/g);
            var cookieName = null;
            var cookieValue = null;
            var cookieNameValue = null;
            for (var i = 0, len = cookieParts.length; i < len; i++) {
                cookieNameValue = (_a = cookieParts[i]) === null || _a === void 0 ? void 0 : _a.match(/([^=]+)=/i);
                if (cookieNameValue && cookieNameValue !== null) {
                    try {
                        cookieName = decodeURIComponent(cookieNameValue[1]);
                        cookieValue = decodeValue(cookieParts[i].substring(cookieNameValue[1].length + 1));
                    }
                    catch (ex) {
                        // 忽略
                    }
                }
                else {
                    // cookie 中没有 "=", 当成布尔值
                    cookieName = decodeURIComponent(cookieParts[i]);
                    cookieValue = '';
                }
                if (cookieName !== null) {
                    // 默认后面的cookie覆盖前面的
                    cookies[cookieName] = cookieValue;
                }
            }
        }
        return cookies;
    }
    var cookiesMap = {};
    var prevCookie;
    /**
     * 获取cookie值.
     * @param {string} name cookie名称.
     * @param {object=} options 配置项
     *                    raw: 是否进行编码，true则是未编码
     * @return {*} 不存在返回null.
     */
    function getCookie(name, options, force) {
        if (options === void 0) { options = {}; }
        if (force === void 0) { force = false; }
        try {
            if (!force && name in cookiesMap) {
                return cookiesMap[name];
            }
            if (!(window === null || window === void 0 ? void 0 : window.document))
                return;
            // 处理 cookie 动态更新情形
            var curCookie = document.cookie || '';
            if (curCookie === prevCookie)
                return cookiesMap[name];
            prevCookie = curCookie;
            cookiesMap = parseCookieString(curCookie, !options.raw, options);
            return cookiesMap[name];
        }
        catch (err) { }
    }
    /**
     * 给定的键值设置cookie
     * @param {string} name 名称.
     * @param {*} value cookie值.
     * @param {object=} options (Optional) 配置项:
     *      path 路径
     *      domain 域名
     *      expires 过期时间，一个Date对象
     *      secure 是否安全cookie
     *      raw (true/false). Setting raw to true indicates that the cookie should not be URI encoded before being set.
     */
    function setCookie(name, value, options) {
        if (options === void 0) { options = {}; }
        try {
            document.cookie = createCookieString(name, value, !options.raw, options);
            cookiesMap[name] = value;
        }
        catch (err) { }
    }
    var cookieUtils = {
        getCookie: getCookie,
        setCookie: setCookie,
        parseCookieString: parseCookieString
    };

    function log() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        try {
            var log_1 = console === null || console === void 0 ? void 0 : console.log;
            return log_1 && log_1.call.apply(log_1, __spreadArray([console], args, false));
        }
        catch (e) {
            return;
        }
    }
    function warn() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        try {
            var warn_1 = console === null || console === void 0 ? void 0 : console.warn;
            return warn_1 && warn_1.call.apply(warn_1, __spreadArray([console], args, false));
        }
        catch (e) {
            return;
        }
    }
    var sdkName = 'webLogger';
    function error() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        try {
            var error_1 = sdkName === 'webLogger_krn' ? console === null || console === void 0 ? void 0 : console.log : console === null || console === void 0 ? void 0 : console.error;
            return error_1 && error_1.call.apply(error_1, __spreadArray([console], args, false));
        }
        catch (e) {
            return;
        }
    }

    /**
     * 获取 localStorage 缓存
     */
    function getLocalStorageItem(key) {
        try {
            if (window && window.localStorage) {
                var value = window.localStorage.getItem(key);
                if (value) {
                    try {
                        return JSON.parse(value);
                    }
                    catch (err) {
                        return value;
                    }
                }
            }
        }
        catch (err) {
            return null;
        }
        return null;
    }
    /**
     * 设置 localStorage 缓存
     */
    function setLocalStorageItem(key, value) {
        try {
            if (window && window.localStorage) {
                window.localStorage.setItem(key, JSON.stringify(value));
                return true;
            }
        }
        catch (err) {
            return false;
        }
        return false;
    }
    /**
     * 从目标arr中删除item
     * @param {*} arr  目标数组
     * @param {*} item  item
     */
    function deleteArray(arr, item) {
        var index = arr.indexOf(item);
        if (index > -1) {
            arr.splice(index, 1);
        }
    }
    /**
     * 事件监听兼容处理
     * @param {*} evtTarget  事件监听目标
     * @param {*} evtName   事件名
     * @param {*} callback  回调函数
     * @param {*} options  相关参数设置
     */
    function addMonitor(evtTarget, evtName, callback, options) {
        if ('attachEvent' in evtTarget) {
            return evtTarget.attachEvent('on' + evtName, callback);
        }
        return evtTarget.addEventListener(evtName, callback, options);
    }
    /**
     * 事件监听解绑兼容处理
     * @param {*} evtTarget  事件监听目标
     * @param {*} evtName   事件名
     * @param {*} callback  事件回调
     */
    function removeMonitor(evtTarget, evtName, callback, options) {
        if ('attachEvent' in evtTarget) {
            return evtTarget.detachEvent('on' + evtName, callback);
        }
        return evtTarget.removeEventListener(evtName, callback, options);
    }
    function leftPad(str, len, ch) {
        if (ch === void 0) { ch = ' '; }
        len = len - str.length;
        if (len <= 0) {
            return str;
        }
        var pad = '';
        while (len) {
            if (len & 1) {
                pad += ch;
            }
            len >>= 1;
            ch += ch;
        }
        return pad + str;
    }
    /**
     * 是否是快手系APP
     */
    var isInApp = function (ua) {
        if (ua === void 0) { ua = window.navigator.userAgent.toLowerCase(); }
        try {
            for (var i = 0; i < ksAppList.length; i++) {
                if (ua.indexOf(ksAppList[i].toLowerCase()) > -1) {
                    return true;
                }
            }
        }
        catch (err) {
            return false;
        }
    };
    function parseQueryString(queryString) {
        var result = {};
        var queries = queryString.split('&');
        for (var _i = 0, queries_1 = queries; _i < queries_1.length; _i++) {
            var query = queries_1[_i];
            var _a = query.split('='), key = _a[0], value = _a[1];
            if (!(key in result)) {
                result[key] = value;
                continue;
            }
            if (result[key] instanceof Array) {
                result[key].push(value);
                continue;
            }
            result[key] = [result[key], value];
        }
        return result;
    }
    /**
     * 检查依赖版本号是否不小于推荐版本
     */
    function checkVersion(recommendVersion, version) {
        if (!version)
            return false;
        if (version.indexOf('alpha') !== -1)
            return false;
        if (version.indexOf('beta') !== -1)
            return false;
        try {
            var matches = recommendVersion.match(/\d+\.\d+\.\d+/);
            recommendVersion = (matches === null || matches === void 0 ? void 0 : matches[0]) || '';
            if (!recommendVersion)
                return true;
            matches = version.match(/\d+\.\d+\.\d+/);
            version = (matches === null || matches === void 0 ? void 0 : matches[0]) || '';
            if (!version)
                return false;
            var _a = recommendVersion.split('.').map(function (i) { return Number(i); }), rMajor = _a[0], rMinor = _a[1], rPatch = _a[2];
            var _b = version.split('.').map(function (i) { return Number(i); }), major = _b[0], minor = _b[1], patch = _b[2];
            if (major < rMajor)
                return false;
            if (major > rMajor)
                return true;
            if (minor < rMinor)
                return false;
            if (minor > rMinor)
                return true;
            if (patch < rPatch)
                return false;
            return true;
        }
        catch (err) {
            return false;
        }
    }
    function checkSupportCloseCoPage() {
        var version = getCookie('appver', undefined, true);
        return version && checkVersion('10.8.30', version);
    }
    function getUrlAndQueryFromUrl(url) {
        if (url === void 0) { url = location.href; }
        var queryIndex = url.lastIndexOf('?');
        if (queryIndex === -1) {
            return {
                page: url
            };
        }
        var page = url.slice(0, queryIndex);
        var queryString = url.slice(queryIndex + 1);
        return {
            page: page,
            params: parseQueryString(queryString),
        };
    }
    var getMappingPage = function (urlPackage, urlMap) {
        var url = urlPackage.url, page = urlPackage.page, params = urlPackage.params, pageId = urlPackage.pageId;
        if (!page && typeof url === 'string') {
            var parsed = getUrlAndQueryFromUrl(url);
            page = parsed.page;
            params = parsed.params;
        }
        if (typeof urlMap === 'function') {
            try {
                var res = urlMap({ url: url, page: page, params: params });
                if (typeof res === 'string') {
                    page = res;
                }
                else if (typeof res === 'object') {
                    if (res.page) {
                        page = res.page;
                    }
                    if (res.params) {
                        params = res.params;
                    }
                    if (res.pageId) {
                        pageId = res.pageId;
                    }
                }
            }
            catch (err) { }
        }
        else if (typeof urlMap === 'object') {
            var mappingPage = '';
            for (var key in urlMap) {
                if ((url || page || '').indexOf(key) > -1) {
                    mappingPage = urlMap[key];
                    break;
                }
            }
            if (mappingPage) {
                page = mappingPage;
            }
        }
        return {
            page: page,
            params: params,
            pageId: pageId
        };
    };
    /**
     * 非空纯对象判定
     * @param obj
     * @returns
     */
    var isObj = function (obj) { return obj && typeof obj === 'object' && !Array.isArray(obj); };
    /**
     * 对象打平赋值
     * @param target
     * @param options
     * @returns
     */
    var flattenAssign = function (target, options) {
        if (!isObj(target) || !isObj(options))
            return;
        var merge = function (obj1, obj2, key) {
            if (isObj(obj1[key]) && isObj(obj2[key])) {
                Object.assign(obj1[key], obj2[key]);
            }
            else if (!isObj(obj1[key]) && !isObj(obj2[key])) {
                obj1[key] = obj2[key];
            }
        };
        for (var key in options) {
            if (target.hasOwnProperty(key)) {
                merge(target, options, key);
                continue;
            }
            for (var deepKey in target) {
                if (isObj(target[deepKey]) && target[deepKey].hasOwnProperty(key)) {
                    merge(target[deepKey], options, key);
                }
            }
        }
    };
    function pageUrlValidate(page) {
        // 对于上报到v2，page为url的埋点，给予警告
        if (page && /^(https?:)?\/\//.test(page)) {
            error('[error 108]', "\u8BF7\u6CE8\u610F\u5F53\u524D\u57CB\u70B9\u9875\u9762\u4FE1\u606F\u4E3A ".concat(page, "\uFF0C\u4E0D\u7B26\u5408\u89C4\u8303\uFF0C\u4E0A\u62A5\u5931\u8D25\uFF01"));
            return false;
        }
        return true;
    }
    var getResolution = (function () {
        var resolution = '';
        return function () {
            if (resolution) {
                return resolution;
            }
            try {
                var ratio = window.devicePixelRatio || 1;
                var width = Math.floor(screen.width * ratio); // 有的devicePixelRatio为小数，在此转为整数
                var height = Math.floor(screen.height * ratio);
                resolution = "".concat(width, "x").concat(height);
                return resolution;
            }
            catch (err) {
                return '';
            }
        };
    })();
    // 从vue-router里面拿的……
    function supportsPushState() {
        var ua = window.navigator.userAgent;
        if ((ua.indexOf('Android 2.') !== -1 || ua.indexOf('Android 4.0') !== -1)
            && ua.indexOf('Mobile Safari') !== -1
            && ua.indexOf('Chrome') === -1
            && ua.indexOf('Windows Phone') === -1) {
            return false;
        }
        return window.history && 'pushState' in window.history;
    }
    function isMobileAndIpad() {
        var _a, _b;
        var ua = ((_a = window === null || window === void 0 ? void 0 : window.navigator) === null || _a === void 0 ? void 0 : _a.userAgent) || '';
        var isMobile = /mobile|tablet|ip(ad|hone|od)|android|(windows phone)/i.test(ua);
        // 兼容最新版本 ipad 版本 ua 判断
        var isiPad = (window === null || window === void 0 ? void 0 : window.navigator.platform) === 'MacIntel' && ((_b = window === null || window === void 0 ? void 0 : window.navigator) === null || _b === void 0 ? void 0 : _b.maxTouchPoints) > 1;
        return isMobile || isiPad;
    }
    /**
     * 是否支持 worker gzip 压缩
     * @returns
     */
    var enableAsyncGzip = function () { return !!((window === null || window === void 0 ? void 0 : window.Worker // 可以新建 worker
    )
        && (window === null || window === void 0 ? void 0 : window.Uint8Array) // 支持二进制存储
        && window.URL // 可以创建文件路径
        && window.Promise // 可以使用 promise
        && !isMobileAndIpad() // 移动端禁止使用压缩
    ); };
    /**
     * 返回正整数数值
     * @param value
     */
    var getPositiveInteger = function (value) {
        try {
            return Math.abs(Math.floor(value));
        }
        catch (e) {
            return value;
        }
    };
    function stringifyParams(params) {
        return typeof params === 'string' ? params : typeof params === 'object' ? JSON.stringify(params) : '';
    }

    // 为了减少点代码体积，自己搞一个简易的EventEmitter
    var EventEmitter = /** @class */ (function () {
        function EventEmitter() {
            this.events = {};
        }
        EventEmitter.prototype.on = function (eventName, callback) {
            var _a;
            if (callback && typeof callback === 'function') {
                this.events[eventName] = this.events[eventName] || [];
                (_a = this.events[eventName]) === null || _a === void 0 ? void 0 : _a.push(callback);
            }
        };
        EventEmitter.prototype.off = function (eventName, callback) {
            if (!this.events[eventName]) {
                return;
            }
            if (callback && typeof callback === 'function') {
                var array = this.events[eventName];
                if (array) {
                    deleteArray(array, callback);
                }
            }
            if (!callback) {
                this.events[eventName] = [];
            }
        };
        EventEmitter.prototype.emit = function (eventName) {
            var _a;
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            if (!this.events[eventName]) {
                return;
            }
            (_a = this.events[eventName]) === null || _a === void 0 ? void 0 : _a.forEach(function (callback) {
                callback.apply(void 0, args);
            });
        };
        return EventEmitter;
    }());

    var initConfig = (function (options) {
        var logConfig = _assign({ env: 'production', proto: 'v3', timeout: 30000, wait: 1000, maxBatchLength: 50, sampleRate: 1, 
            // 尝试从全局中获取 yoda
            yoda: typeof window !== 'undefined' && window.yoda, forbidV2HttpUrlPage: true }, (options || {}));
        return logConfig;
    });

    var pluginIndex = 0;
    var Weblog$1 = /** @class */ (function (_super) {
        __extends(Weblog, _super);
        /**
         * @param {?object} logConfig Logger初始化配置
         * @param {object=} options 额外配置参数
         *                    - base 公参的覆盖
         */
        function Weblog(logConfig, base) {
            var _this = _super.call(this) || this;
            _this.version = '3.10.23';
            // 插件
            _this.plugins = {};
            _this.flush = function () {
                _this.logger.flush();
            };
            // 初始化初始化参数配置和埋点基础信息配置
            _this.baseOption = _assign({}, base);
            _this.logConfig = initConfig(logConfig);
            // 初始化页面信息
            _this.initUrlPackage();
            // 记录首个实例
            if (!Weblog.__instance) {
                Weblog.__instance = _this;
            }
            return _this;
        }
        Object.defineProperty(Weblog.prototype, "sessionId", {
            get: function () {
                return GlobalConfig.sessionId;
            },
            enumerable: false,
            configurable: true
        });
        Weblog.prototype.initUrlPackage = function () {
            this.updateCurrentUrlPackage();
        };
        /**
         * 更新当前 url 信息
         * @param url
         * @param type
         */
        Weblog.prototype.updateCurrentUrlPackage = function (url, type) {
            if (url === void 0) { url = {}; }
            if (type === void 0) { type = 'web'; }
            // 该接口不改变 refer 信息
            if (typeof url === 'object' && url.page) {
                if (this.currentUrlPackage && url.page === this.currentUrlPackage.page) {
                    return this.currentUrlPackage.update(url.page, url.params);
                }
            }
            this.currentUrlPackage = new UrlPackage$1(url, type);
        };
        /**
         * 更新 refer url 信息
         * @param url
         * @param type
         */
        Weblog.prototype.updateReferUrlPackage = function (url, type) {
            if (url === void 0) { url = this.currentUrlPackage; }
            if (type === void 0) { type = 'web'; }
            if (url instanceof UrlPackage$1) {
                this.referUrlPackage = url;
            }
            else {
                this.referUrlPackage = new UrlPackage$1(url, type);
            }
        };
        /**
         * 重设 common_package 信息
         * @param base
         */
        Weblog.prototype.updateBase = function (base) {
            this.updateCommonPackage(base);
        };
        /**
         * 更新现有 common_package 信息
         */
        Weblog.prototype.updateCommonPackage = function (options) {
            if (typeof options !== 'object')
                return;
            this.commonPackage.update(options);
        };
        /**
         * 初始化插件
         * @param plugins
         */
        Weblog.prototype.addPlugins = function () {
            var _this = this;
            if (!this.logConfig.plugins || !this.logConfig.plugins.length)
                return;
            // 类似 webpack plugins 的实例数组
            this.logConfig.plugins.forEach(function (plugin) {
                if (typeof plugin === 'object' && typeof plugin.apply === 'function') {
                    _this.addPluginInstance(plugin);
                }
            });
        };
        /**
         * 加载插件实例
         */
        Weblog.prototype.addPluginInstance = function (pluginInstance) {
            if (pluginInstance) {
                var key = pluginInstance.key || pluginInstance.constructor && pluginInstance.constructor.key || "plugin_auto_key_".concat(pluginIndex++);
                if (typeof pluginInstance.apply === 'function'
                    && (!pluginInstance.weblog || pluginInstance.weblog !== this)) {
                    pluginInstance.apply(this);
                }
                this.plugins[key] = pluginInstance;
            }
        };
        /**
         * 加载插件构造函数
         * @param plugin
         * @param pluginOptions
         */
        Weblog.prototype.plug = function (plugin, pluginOptions) {
            if (this.plugins[plugin.key]) {
                return error('[code 301]', "".concat(plugin.key, " \u63D2\u4EF6\u91CD\u590D\u52A0\u8F7D\uFF01"));
            }
            this.addPluginInstance(new plugin(this, pluginOptions));
        };
        Weblog.prototype.unplug = function (name) {
            var plugin = this.plugins[name];
            if (plugin) {
                plugin.destroy();
                delete this.plugins[name];
            }
        };
        Weblog.prototype.unplugAll = function () {
            for (var key in this.plugins) {
                if (this.plugins[key]) {
                    this.unplug(key);
                }
            }
        };
        /**
         * 扩展类实现
         * @param eventType
         * @param eventOptions
         * @returns
         */
        Weblog.prototype.generateLog = function (eventType, eventOptions) {
            return {};
        };
        Weblog.prototype.send = function (type, options, immediately) {
            var log = this.generateLog(type.toUpperCase(), options);
            // @ts-ignore
            if (typeof this.beforeSend === 'function') {
                // @ts-ignore
                this.beforeSend(type, options, log);
            }
            return this.logger.send(log, !!immediately, options.callback);
        };
        /**
         * 收集上报
         * @param type
         * @param options
         * @returns
         */
        Weblog.prototype.collect = function (type, options) {
            this.send(type, options);
        };
        Weblog.prototype.sendImmediately = function (type, options) {
            this.send(type, options, true);
        };
        Weblog.prototype.sendPackage = function (data, callback) {
            this.logger.sendPackage(data, callback);
        };
        Weblog.prototype.destroy = function () {
            this.unplugAll();
        };
        return Weblog;
    }(EventEmitter));

    function uuid() {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    }
    /**
     * 生成nano id
     * @return {string} 16长度的字符串
     */
    function nanoId() {
        var ts = ((Math.random() * 1e9) >>> 0);
        var randomer = [];
        var seed = '0123456789ABCDEF';
        for (var i = 0; i < 7; i++) {
            randomer.push(seed.charAt(Math.random() * 16));
        }
        return ts + randomer.join('');
    }
    /**
     * 提取顶级域名
     * @param {?string=} hostname
     * @return {string} 返回顶级域名
     */
    function resolveTopLevelDomain(hostname) {
        if (hostname === void 0) { hostname = typeof location !== 'undefined' ? location.hostname : ''; }
        if (!hostname) {
            return '';
        }
        var ret = hostname.split('.');
        var len = ret.length;
        return (len <= 2 ? '' : ".".concat(ret[len - 2], ".").concat(ret[len - 1]));
    }
    var didKey = 'did';
    /**
     * 生成did
     * @return {string}
     */
    function generateDID() {
        try {
            // 兼容一下原下发的did逻辑，如果能从did获取，那么就走did
            // 如果不能，那就通过log sdk自行维护的_did来搞
            var did = getCookie(didKey) || getCookie('_did');
            if (!did) {
                did = "web_".concat(nanoId());
                var expires = new Date();
                expires.setFullYear(expires.getFullYear() + 1);
                setCookie('_did', did, {
                    expires: expires,
                    domain: resolveTopLevelDomain(),
                    path: '/',
                });
            }
            return did;
        }
        catch (err) {
            return '';
        }
    }
    var sessionKey = 'WEBLOGGER_SESSIONID';
    // session_id 用来串联在一次打开 App 期间跳转不同页面的用户行为
    var getSessionId = function () {
        var sessionId = '';
        var defaultKey = 'session_id';
        try {
            if (isInApp()) {
                sessionId = getCookie(defaultKey) || getCookie('sid'); // 端内从 cookie 中获取客户端传来的 session_id
                if (sessionId)
                    return sessionId;
            }
            else if (typeof sessionStorage !== 'undefined') {
                sessionId = sessionStorage.getItem(sessionKey) || '';
                if (!sessionId) {
                    sessionId = uuid();
                    sessionStorage.setItem(sessionKey, sessionId);
                }
            }
        }
        catch (err) {
            sessionId = uuid();
        }
        return sessionId || uuid();
    };

    var UrlPackage = /** @class */ (function (_super) {
        __extends(UrlPackage, _super);
        function UrlPackage(url, pageType, urlMap) {
            var _this = _super.call(this, {}, pageType) || this;
            if (typeof url === 'string') {
                // TODO: 避免兼容问题，暂不处理
                // try {
                //     url = window.decodeURIComponent(url);
                // } catch (err) {}
                var obj = getUrlAndQueryFromUrl(url);
                _this.page = obj.page;
                _this.params = obj.params;
            }
            else {
                _this.page = (url || {}).page || '';
                _this.params = (url || {}).params || undefined;
            }
            _this.identity = _this.generatePageId();
            if (urlMap && typeof url === 'string') {
                _this.init(url, urlMap);
            }
            return _this;
        }
        UrlPackage.prototype.init = function (url, urlMap) {
            if (urlMap === void 0) { urlMap = {}; }
            var _a = getMappingPage({ url: url, page: this.page, params: this.params }, urlMap), page = _a.page, params = _a.params;
            this.update(page, params);
        };
        UrlPackage.prototype.attachUrl = function () {
            if (!this.params) {
                this.params = {};
            }
            if (location === null || location === void 0 ? void 0 : location.href) {
                // 电商需求，页面信息携带原始 url 和 pathname
                this.params.origin_url = this.params.origin_url || (location === null || location === void 0 ? void 0 : location.href);
                var _a = getUrlAndQueryFromUrl(location.href).page, page = _a === void 0 ? '' : _a;
                this.params.origin_pathname = page;
            }
        };
        UrlPackage.prototype.getRealUrlPackage = function () {
            var _a = getUrlAndQueryFromUrl(location.href), page = _a.page, params = _a.params;
            return {
                page: page,
                params: JSON.stringify(_assign({ page_code: this.page, url: location.href, query: params }, params)),
                page_type: this.page_type,
                identity: this.identity
            };
        };
        UrlPackage.prototype.generatePageId = function () {
            return uuid();
        };
        return UrlPackage;
    }(UrlPackage$1));

    var dataTrackHost = 'https://data-track.corp.kuaishou.com';
    var overseaDataTrackHost = 'https://data-track-sgp.corp.kuaishou.com';
    var testHost = dataTrackHost + '/';
    var PROD = 'production';
    var TEST = 'test';
    var DEV = 'development';
    var OVERSEA = 'oversea';
    var LOGGER = 'logger';
    var LOGGER_SGP = 'logger-oversea';
    var v3Api = 'rest/wd/common/log/collect/';
    var Envs = [PROD, TEST, DEV, LOGGER, OVERSEA];
    var overseaHost = 'https://logsdk.kwai-pro.com/';
    var Apis = {
        // v2 接口一定要带上 _json=1 参数，不然服务端会解析失败
        v2: "rest/kd/log/collect?_json=1&biz=",
        v3: "".concat(v3Api, "misc2"),
        // 雷达都上报到 v3
        radar: "".concat(v3Api, "radar")
    };
    var Hosts = {
        v2: {
            // 正式环境随机域名分流
            production: ['https://wlog.ksapisrv.com/', 'https://wlog.gifshow.com/'][Math.round(Math.random())],
            development: testHost,
            test: testHost,
            oversea: overseaHost
        },
        v3: {
            production: 'https://log-sdk.ksapisrv.com/',
            development: testHost,
            test: testHost,
            oversea: overseaHost
        }
    };
    var buildUrl = function (env, radar, proto) {
        if (env === void 0) { env = PROD; }
        if (radar === void 0) { radar = false; }
        if (proto === void 0) { proto = 'v3'; }
        if (Envs.indexOf(env) === -1) {
            return env + Apis[proto];
        }
        if (radar && Hosts.v3[env]) {
            return Hosts.v3[env] + Apis.radar;
        }
        if (Hosts[proto][env] && Apis[proto]) {
            return Hosts[proto][env] + Apis[proto];
        }
        return Hosts.v3.production + Apis.v3;
    };
    var formatUrlQuery = function (url, querys) {
        if (!querys)
            return url;
        var reg = /\?(.+?)$/;
        try {
            return reg.test(url) ? url.replace(/\?(.+?)$/, "?".concat(querys, "&$1")) : url + "?".concat(querys);
        }
        catch (err) {
            return url;
        }
    };

    var BatchLog = /** @class */ (function () {
        function BatchLog(config, commonPackage) {
            var _this = this;
            // 立即发送队列
            this.asyncQueue = [];
            // 存储待发送的消息
            this.throttleQueue = [];
            // 存储失败或者超时的数据
            this.errorQueue = [];
            this.sendingQueue = {};
            this.url = '';
            this.isV2 = false;
            this.isDebug = false;
            this.radarUrl = '';
            this.drained = false; // 是否未排干日志队列
            this.batchCount = 50;
            this.isSetSamplingResult = false;
            // 有值时阻塞日志发送
            this.sendingYield = null;
            /**
             * 清空当前日志队列
             * @param callback
             */
            this.flush = function (callback) {
                _this.sendLogs(_this.throttleQueue.concat(_this.asyncQueue), callback);
                _this.throttleQueue = [];
                _this.asyncQueue = [];
            };
            /**
             * 页面关闭前排干所有未发送的和发送失败的日志
             */
            this.drain = function () {
                _this.drained = true;
                _this.flush();
                _this.flushErrorLogs();
                // 处理 beforunload 中断，实际页面未退出的情况
                setTimeout(function () {
                    _this.drained = false;
                }, 1000);
            };
            this.config = config;
            if (this.config.maxBatchLength && this.config.maxBatchLength > 1) {
                this.batchCount = Math.min(50, this.config.maxBatchLength);
            }
            this.commonPackage = commonPackage;
            this.isDebug = this.config.logger || this.config.env === 'logger';
            this.isV2 = this.config.proto === 'v2';
            this.updateUrls();
        }
        BatchLog.prototype.sendData = function (data, callback) {
            if (typeof this.config.sender === 'function') {
                return this.config.sender(data, callback);
            }
            return this.baseSendData(data, callback);
        };
        Object.defineProperty(BatchLog.prototype, "responseSamplingStorageKey", {
            get: function () {
                var _a = this.commonPackage.app_package, product_name = _a.product_name, product = _a.product;
                var device_id = this.commonPackage.identity_package.device_id;
                // tslint:disable-next-line:max-line-length
                return "RESPONSE_SAMPLING_STORAGE_KEY_".concat(product_name || product, "_").concat(device_id);
            },
            enumerable: false,
            configurable: true
        });
        /**
         * 根据配置生成上报的 url 地址
         */
        BatchLog.prototype.updateUrls = function () {
            var env = this.config.env;
            // 线上地址 -> 抓包查找工具后端，用以测试后端
            // 兼容方式：若小程序里有扫码工具，扫码后直接进行切换
            if (env && /^(https?:)?\/\//.test(env)) {
                this.url = env;
            }
            else {
                this.url = buildUrl(env);
            }
            this.formatUrl();
        };
        BatchLog.prototype.formatUrl = function () {
            if (!this.radarUrl) {
                this.radarUrl = this.url.replace(this.url.indexOf(Apis.v2) !== -1 ? Apis.v2 : Apis.v3, Apis.radar);
                // v2 域名不能上报雷达事件
                if (this.radarUrl.indexOf(Hosts.v2.production || '') !== -1) {
                    this.radarUrl.replace(Hosts.v2.production || '', Hosts.v3.production);
                }
            }
            var _a = this.commonPackage.app_package, product_name = _a.product_name, product = _a.product;
            var querys = "v=3.10.23&kpn=".concat(product_name || product);
            this.url = formatUrlQuery(this.url, querys);
            this.radarUrl = formatUrlQuery(this.radarUrl, querys);
        };
        /**
         * 之所以使用 JSON 而不是 commonPackage 对象，是因为对象可能会变化，而 JSON 则是当时的一个快照，一旦生成不会变化
         */
        BatchLog.prototype.getCommonPackageJSON = function () {
            return this.commonPackage.toJSON();
        };
        /**
         * 进入发送流程
         * @param data
         * @param immediate
         * @param callback 用户自定义 callback
         * @returns
         */
        BatchLog.prototype.send = function (data, immediate, callback) {
            if (immediate === void 0) { immediate = false; }
            // 有回调的立即上报
            if (callback || this.drained) {
                return this.sendLogs([data], callback);
            }
            if (immediate) {
                this.sendAsync(data, callback);
                return;
            }
            this.sendThrottle(data);
        };
        /**
         * 短时异步上报
         */
        BatchLog.prototype.sendAsync = function (data, callback) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            this.asyncQueue.push(data);
                            if (this.asyncQueue.length >= this.batchCount) {
                                this.flush();
                                return [2 /*return*/];
                            }
                            if (!this.sendingYield) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.sendingYield];
                        case 1:
                            _a.sent();
                            this.sendingYield = null;
                            _a.label = 2;
                        case 2:
                            _a.trys.push([2, 4, , 5]);
                            return [4 /*yield*/, Promise.resolve()];
                        case 3:
                            _a.sent();
                            return [3 /*break*/, 5];
                        case 4:
                            _a.sent();
                            return [3 /*break*/, 5];
                        case 5:
                            if (!this.asyncQueue.length)
                                return [2 /*return*/];
                            this.flush();
                            return [2 /*return*/];
                    }
                });
            });
        };
        /**
         * 节流上报
         */
        BatchLog.prototype.sendThrottle = function (data) {
            var _this = this;
            this.throttleQueue.push(data);
            // 积攒超过最大
            if (this.throttleQueue.length >= this.batchCount) {
                this.flush();
                return;
            }
            // debounce逻辑
            clearTimeout(this.batchWaitTimer);
            this.batchWaitTimer = setTimeout(function () {
                _this.flush();
            }, this.config.wait);
        };
        /**
         * 批量发送日志
         * @param logs
         * @param callback 用户自定义 callback
         */
        BatchLog.prototype.sendLogs = function (logs, callback) {
            var _this = this;
            if (!logs || !logs.length) {
                return typeof callback === 'function' && callback();
            }
            var data = this.buildLogPackage(logs, this.url);
            // 网络发送错误回调
            var callbackFn = function (error) {
                // 如果网络异常，日志发送失败，则将错误日志缓存，在网络恢复时清空一次错误队列
                if (error) {
                    _this.errHandler(data);
                }
                else {
                    _this.flushErrorLogs();
                }
                typeof callback === 'function' && callback(error);
            };
            this.sendPackage(data, callbackFn);
        };
        /**
         * 将日志包通过接口发送
         * @param data
         * @param callback
         */
        BatchLog.prototype.sendPackage = function (data, callback) {
            var timeout = this.config.timeout;
            try {
                // 如果业务方配置了 sender
                this.sendData(_assign(_assign({}, data), { isDebug: this.isDebug, isDrained: this.drained, timeout: timeout }), callback);
            }
            catch (err) {
                typeof callback === 'function' && callback(err);
            }
        };
        /**
         * 每个 logPackage 包含了所有上报所需的信息，上报地址和上报数据
         * @param logs
         * @param url
         */
        BatchLog.prototype.buildLogPackage = function (logs, url) {
            if (this.isV2) {
                return this.buildV2Package(logs, url);
            }
            return this.buildV3Package(logs, url);
        };
        BatchLog.prototype.buildV2Package = function (logs, url) {
            return {
                url: url,
                data: {
                    log: {
                        event: logs
                    }
                }
            };
        };
        BatchLog.prototype.buildV3Package = function (logs, url, commonPackageOptions) {
            var common = this.getCommonPackageJSON();
            commonPackageOptions && Object.assign(common, commonPackageOptions);
            return {
                url: url,
                data: {
                    common: common,
                    logs: logs
                }
            };
        };
        /**
         * 日志上报出错处理，一般为网络异常
         * @param logPackage
         */
        BatchLog.prototype.errHandler = function (logPackage) {
            var _a;
            if (this.isV2) {
                this.errorQueue.unshift(logPackage);
                return;
            }
            var data = logPackage.data;
            if (!data.logs.length)
                return;
            // http_seq_id 和 client_timestamp 不是必须的，合并时删除
            delete data.common.h5_extra_attr.http_seq_id;
            delete data.common.h5_extra_attr.client_timestamp;
            // 尝试合并相同 url 和 common 的请求
            var merged = false;
            for (var i = 0; i < this.errorQueue.length; i++) {
                var currentLog = this.errorQueue[i];
                if (currentLog.url === logPackage.url &&
                    currentLog.data.logs.length + data.logs.length <= 100 &&
                    JSON.stringify(currentLog.data.common) === JSON.stringify(data.common)) {
                    merged = true;
                    (_a = currentLog.data.logs).push.apply(_a, data.logs);
                    break;
                }
            }
            if (!merged) {
                // 先进后出
                // 最大错误数量限制为 5
                if (this.errorQueue.length >= 5) {
                    this.errorQueue.pop();
                }
                this.errorQueue.unshift(logPackage);
            }
        };
        // 当网络恢复时尝试一次上报
        BatchLog.prototype.flushErrorLogs = function () {
            var _this = this;
            this.errorQueue.forEach(function (logPackage) {
                _this.sendPackage(logPackage);
            });
            this.errorQueue = [];
        };
        BatchLog.prototype.destory = function () {
            this.batchWaitTimer && clearTimeout(this.batchWaitTimer);
            this.compensateTimer && clearTimeout(this.compensateTimer);
        };
        /**
         * 雷达专用上报（雷达上报接口和其他事件上报接口做了区分，用于分流统计）
         * @param radar
         */
        BatchLog.prototype.sendRadar = function (radar, service_name) {
            var options = service_name ? { service_name: service_name } : undefined;
            var radarPackage = this.buildV3Package([radar], this.radarUrl || this.url, options);
            this.sendPackage(radarPackage);
        };
        return BatchLog;
    }());

    var xhrErrorLog = function (err, url, data) { return error('[error 400]', '埋点上报接口请求报错', '\nurl:', url, '\ndata', data, '\nerror:', err || 'server decode log failed'); };
    var logXhrResponseCheck = function (status, response, url, data) {
        var err;
        if (status < 200 || (status >= 300 && status !== 304)) ;
        else {
            try {
                var _a = JSON.parse(response), exception = _a.exception, result = _a.result, error_msg = _a.error_msg;
                if (exception) {
                    err = exception;
                }
                if (result !== 1) {
                    err = error_msg || "result is ".concat(result);
                }
            }
            catch (e) {
                err = e.message;
            }
        }
        err && xhrErrorLog(err, url, data);
        return err;
    };

    function send$1(_a, callback) {
        var url = _a.url, data = _a.data, timeout = _a.timeout;
        return new Promise(function (resolve) {
            var xhr = new XMLHttpRequest();
            // async 同步情况会报错，而且影响性能，取消使用同步
            xhr.open('POST', url);
            xhr.setRequestHeader('Content-Type', 'text/plain;charset=UTF-8');
            if (typeof data === 'object' && !(data instanceof Uint8Array)) {
                data = JSON.stringify(data);
            }
            if (timeout) {
                xhr.timeout = timeout;
            }
            xhr.onload = function () {
                var err = logXhrResponseCheck(xhr.status, xhr.response, url, data);
                callback && callback(err);
                resolve({
                    error: err,
                    response: err ? undefined : xhr.response,
                });
            };
            xhr.ontimeout = xhr.onerror = function (err) {
                if (err)
                    xhrErrorLog(err, url, data);
                // err 不能直接传入 postMessage，会导致 postMessage 直接报错，误认为 worker 内部直接逻辑报错，以后发送埋点直接无法压缩
                callback && callback(err ? 'networkTimeout' : '');
                resolve({
                    error: 'networkTimeout',
                });
            };
            xhr.send(data);
            return xhr;
        });
    }
    var get = function (url) {
        var errorHandler = function (err) { return error('[error 401]', "GET \u8BF7\u6C42\u51FA\u9519 url: ".concat(url), err); };
        try {
            var xhr = new XMLHttpRequest();
            xhr.open('GET', url);
            xhr.setRequestHeader('Content-Type', 'text/plain;charset=UTF-8');
            xhr.send();
            xhr.onerror = errorHandler;
        }
        catch (err) {
            errorHandler(err);
        }
    };

    var navigator$1 = window === null || window === void 0 ? void 0 : window.navigator;
    var navigatorEnable = navigator$1 && navigator$1.userAgent.indexOf('Chrome') !== -1;
    var functionEnable = typeof navigator$1.sendBeacon === 'function';
    /**
     *
     * @param url 发送地址
     * @param data 发送内容
     * @param forceSendBeacon 是否强制使用 sendBeacon 方式发送埋点。
     */
    function send(_a) {
        var url = _a.url, data = _a.data, forceSendBeacon = _a.forceSendBeacon;
        if (!navigatorEnable && !forceSendBeacon || !functionEnable)
            return false;
        // xhook 如果改写了 FormData
        if (typeof data === 'object' && data.fd) {
            data = data.fd;
        }
        try {
            return navigator$1.sendBeacon(url, data);
        }
        catch (err) {
            error('[error 403]', 'navigator.sendBeacon 报错', err);
            return false;
        }
    }

    function sendData(_a, callback) {
        var url = _a.url, data = _a.data, timeout = _a.timeout; _a.isDebug; var isDrained = _a.isDrained, forceSendBeacon = _a.forceSendBeacon;
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        data = JSON.stringify(data);
                        // 优先走 sendBeacon，如果成功把数据添加到 beacon 通道，则认为发送成功
                        if (isDrained && send({ url: url, data: data, forceSendBeacon: forceSendBeacon })) {
                            callback && callback();
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, send$1({ url: url, data: data, timeout: timeout }, callback)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/, result.error ? undefined : result.response];
                }
            });
        });
    }

    var getReportUrl = function (viewHost, reportHost, sessionId) {
        if (viewHost === void 0) { viewHost = dataTrackHost; }
        var loggerUrl = "".concat(viewHost, "/#/logger/index?sessionId=").concat(sessionId);
        log("%c\u57CB\u70B9\u6293\u5305\u6821\u9A8C: %c".concat(loggerUrl), 'color:#1abf89;font-size:1.2em;line-height:2.4em;', 'font-size:1.2em;');
        return "".concat(reportHost, "/").concat(sessionId, "/");
    };

    /**
     * 异步模块加载
     */
    var ModuleStates;
    (function (ModuleStates) {
        ModuleStates["loading"] = "loading";
        ModuleStates["loaded"] = "loaded";
    })(ModuleStates || (ModuleStates = {}));
    var LoadedModules = Object.create(null);
    /**
     * 加载 js 脚本
     * @param script
     * @param async
     */
    var loadScript = function (script, async, callback) {
        if (async === void 0) { async = true; }
        if (LoadedModules[script]) {
            return callback && callback("".concat(script, " load repeat"));
        }
        var mod = LoadedModules[script] = {
            state: ModuleStates.loading,
        };
        var scriptEl = document.createElement('script');
        scriptEl.crossOrigin = 'anonymous';
        scriptEl.src = script;
        if (async) {
            scriptEl.async = true;
        }
        var referEl = document.getElementsByTagName('script')[0];
        if (referEl && referEl.parentNode) {
            referEl.parentNode.insertBefore(scriptEl, referEl);
        }
        else {
            document.head.appendChild(scriptEl);
        }
        scriptEl.onload = function () {
            mod.state = ModuleStates.loaded;
            callback && callback();
        };
        scriptEl.onerror = function (e) {
            mod.state = ModuleStates.loaded;
            callback && callback("".concat(script, " loaded failed, ").concat(e));
        };
    };

    var SWICTH = 'weblogger_switch';
    /**
     * 端内环境根据cookie值切换发送地址
     */
    var getCookieReportUrl = function () {
        try {
            var webloggerSwitch = getCookie(SWICTH) ||
                typeof sessionStorage !== 'undefined' && sessionStorage.getItem(SWICTH);
            if (!webloggerSwitch)
                return;
            var _a = JSON.parse(webloggerSwitch), loggerSessionId = _a.loggerSessionId, reportHost = _a.reportHost, loggerHost = _a.loggerHost, handshakeApi = _a.handshakeApi;
            // 若cookie内存在loggerSessionId，则会自动切换发送路径到测试地址，否则返回空，代表不切换地址
            if (!loggerSessionId) {
                return '';
            }
            get(handshakeApi || "".concat(reportHost, "/").concat(loggerSessionId));
            return getReportUrl(loggerHost, reportHost, loggerSessionId);
        }
        catch (err) {
            return '';
        }
    };
    /**
     * env 为 logger 时上报的 url
     */
    var getLoggerReportUrl = function (did, isOversea) {
        if (did === void 0) { did = generateDID(); }
        if (isOversea === void 0) { isOversea = false; }
        var url = isOversea ? overseaDataTrackHost : dataTrackHost;
        get("".concat(url, "/rest/").concat(did));
        return getReportUrl(url, "".concat(url, "/rest"), did);
    };
    /**
     * 端外H5 or pc页面根据当前的url参数切换发送地址
     */
    var getWebPageReportUrl = function (url) {
        if (url === void 0) { url = location.href; }
        // const { storage, io } = Global.adapter;
        var queryIndex = url.lastIndexOf('?');
        if (queryIndex === -1) {
            return '';
        }
        var queryString = url.slice(queryIndex + 1);
        var webloggerSwitch = parseQueryString(queryString).webloggerSwitch;
        if (!webloggerSwitch) {
            return '';
        }
        try {
            var data = decodeURIComponent(webloggerSwitch);
            var _a = JSON.parse(data), loggerSessionId = _a.loggerSessionId, reportHost = _a.reportHost, loggerHost = _a.loggerHost, handshakeApi = _a.handshakeApi;
            if (!loggerSessionId) {
                return '';
            }
            // 存到 sessionStorage 中使同域名页面都可以调试
            // storage.setSession(SWICTH, data);
            // io.get({url: handshakeApi || `${reportHost}/${loggerSessionId}`});
            if (typeof sessionStorage !== 'undefined') {
                sessionStorage.setItem(SWICTH, data);
            }
            get(handshakeApi || "".concat(reportHost, "/").concat(loggerSessionId));
            return getReportUrl(loggerHost, reportHost, loggerSessionId);
        }
        catch (err) {
            return '';
        }
    };
    var getTemporaryLoggerUrl = function (proto) {
        var url = getCookieReportUrl() || getWebPageReportUrl();
        if (url) {
            return buildUrl(url, false, proto);
        }
    };
    var Logger = /** @class */ (function (_super) {
        __extends(Logger, _super);
        // 采样存储 key，作为实例化的一个键值，避免多实例互相干扰
        function Logger(config, commonPackage) {
            var _this = _super.call(this, config, commonPackage) || this;
            _this.baseSendData = _this.sendLog;
            _this.getResponseSamplingStorage();
            return _this;
        }
        // 获取 storage 判断是否有过期时间
        Logger.prototype.getResponseSamplingStorage = function () {
            return __awaiter(this, void 0, void 0, function () {
                var storageEffectiveTime, e_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 2, , 3]);
                            // 直接异步获取 localstorage 信息。保证初始化 weblogger 性能。
                            // 因为 sendImmediately 也是微任务发送，所以获取本地存储信息，是可以保证在发送 pv 之前,获取到采样信息
                            return [4 /*yield*/, Promise.resolve()];
                        case 1:
                            // 直接异步获取 localstorage 信息。保证初始化 weblogger 性能。
                            // 因为 sendImmediately 也是微任务发送，所以获取本地存储信息，是可以保证在发送 pv 之前,获取到采样信息
                            _a.sent();
                            storageEffectiveTime = Number(localStorage.getItem(this.responseSamplingStorageKey));
                            if (storageEffectiveTime && (Date.now() < storageEffectiveTime)) {
                                this.sendEffectiveTime = storageEffectiveTime;
                            }
                            else if (storageEffectiveTime) {
                                // 如果采样已经过期，清除 storage
                                localStorage.removeItem(this.responseSamplingStorageKey);
                            }
                            return [3 /*break*/, 3];
                        case 2:
                            e_1 = _a.sent();
                            console.error(e_1);
                            return [3 /*break*/, 3];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        };
        /**
         * 返回 true 代表不发送埋点
         * @param log
         * @param immediate
         * @param callback
         */
        Logger.prototype.isResponseSampling = function () {
            return !!(this.sendEffectiveTime && (Date.now() < this.sendEffectiveTime));
        };
        // 埋点返回结果处理
        Logger.prototype.setResponseSamplingConfig = function (responsePromise) {
            var _this = this;
            if (this.isSetSamplingResult) {
                return;
            }
            this.isSetSamplingResult = true;
            responsePromise === null || responsePromise === void 0 ? void 0 : responsePromise.then(function (res) {
                var response;
                try {
                    response = res && JSON.parse(res);
                }
                catch (e) {
                    response = {
                        result: 1,
                        hostName: '',
                    };
                }
                var expireSeconds = response === null || response === void 0 ? void 0 : response.expireSeconds;
                if (expireSeconds) {
                    var effectiveTime = Date.now() + expireSeconds * 1000;
                    _this.sendEffectiveTime = effectiveTime;
                    localStorage.setItem(_this.responseSamplingStorageKey, String(effectiveTime));
                }
            });
        };
        /**
         * 根据配置生成上报的 url 地址
         */
        Logger.prototype.updateUrls = function () {
            var _a = this.config, _b = _a.env, env = _b === void 0 ? PROD : _b, logger = _a.logger, proto = _a.proto, isBridge = _a.isBridge, disableCompress = _a.disableCompress;
            var url = getTemporaryLoggerUrl(proto);
            if (url) {
                this.url = url;
                this.isDebug = true;
            }
            else if (logger || env === LOGGER || env === LOGGER_SGP) {
                var loggerEnv = getLoggerReportUrl(this.commonPackage.identity_package.device_id, env === LOGGER_SGP);
                this.url = buildUrl(loggerEnv, false, proto);
                this.isDebug = true;
            }
            else if (env && /^(https?:)?\/\//.test(env)) {
                this.url = env;
            }
            else {
                this.isDebug = env !== PROD;
                this.url = buildUrl(env, false, proto);
                this.radarUrl = buildUrl(env, true, proto);
            }
            this.enableAsyncGzip = !disableCompress // 用户没有配置禁止压缩
                && proto === 'v3' // 必须是 v3 上报, v2 虽然压缩上报接口虽然没报错，但是服务端不做解析。
                && !isBridge // 非桥接上报
                && enableAsyncGzip() // 平台兼容性
                && env === PROD // 必须在 prod 环境
                && !url // 非埋点抓包查找
                && !logger; // 非埋点抓包查找
            this.enableAsyncGzip && loadScript('https://static.yximgs.com/udata/pkg/ks-track-platform-new/weblogger/3.10.23/async/gzipper.min.js');
            this.formatUrl();
        };
        Logger.prototype.sendLog = function (data, callback) {
            var forceSendBeacon = this.config.forceSendBeacon;
            if (!this.isV2) {
                //  根据埋点接口判断是否需要采样
                if (this.isResponseSampling()) {
                    return;
                }
            }
            if (this.drained || !this.enableAsyncGzip || !this.sendGzip(data, callback)) {
                this.setResponseSamplingConfig(sendData(_assign(_assign({}, data), { forceSendBeacon: forceSendBeacon }), callback));
            }
        };
        /**
         *
         * @param data
         * @param callback weblogger 系统级错误回调，会清除缓存的发送错误的日志。
         */
        Logger.prototype.sendGzip = function (data, callback) {
            var _this = this;
            var _a;
            if (!((_a = Logger.Gzipper) === null || _a === void 0 ? void 0 : _a.sendData))
                return false;
            var workerCallback = function (e) {
                if (e) {
                    _this.setResponseSamplingConfig(sendData(data, callback));
                }
            };
            // 如果发送压缩代码成功，会返回一个 promise
            var promise = Logger.Gzipper.sendData(data, workerCallback);
            this.setResponseSamplingConfig(promise);
            return !!promise;
        };
        /**
         * 埋点发送前获取 common_package 信息
         * @param data
         * @returns
         */
        Logger.prototype.getCommonPackageJSON = function () {
            return this.commonPackage.toJSON();
        };
        Logger.prototype.send = function (data, immediate, callback) {
            if (immediate === void 0) { immediate = false; }
            // v2 是每个 log 都有 common_package
            if (this.isV2) {
                // v2 log 的自增 id 是存在 common 信息中
                this.commonPackage.setAdditionalSeqIdPackage(data.getEventType());
                data.common_package = this.getCommonPackageJSON();
            }
            _super.prototype.send.call(this, data, immediate, callback);
        };
        return Logger;
    }(BatchLog));

    /**
     * v2、v3 PB 相同部分
     */
    var Base;
    (function (Base) {
        (function (ElementStatus) {
            ElementStatus[ElementStatus["UNKNOWN_STATUS"] = 0] = "UNKNOWN_STATUS";
            ElementStatus[ElementStatus["CHECKED"] = 1] = "CHECKED";
            ElementStatus[ElementStatus["UNCHECKED"] = 2] = "UNCHECKED";
        })(Base.ElementStatus || (Base.ElementStatus = {}));
        (function (PageShowAction) {
            PageShowAction[PageShowAction["UNKNOWN_ACTION"] = 0] = "UNKNOWN_ACTION";
            PageShowAction[PageShowAction["ENTER"] = 1] = "ENTER";
            PageShowAction[PageShowAction["LEAVE"] = 2] = "LEAVE";
            PageShowAction[PageShowAction["RESUME"] = 3] = "RESUME";
        })(Base.PageShowAction || (Base.PageShowAction = {}));
        (function (ActionStatus) {
            ActionStatus[ActionStatus["UNKNOWN_STATUS"] = 0] = "UNKNOWN_STATUS";
            ActionStatus[ActionStatus["SUCCESS"] = 1] = "SUCCESS";
            ActionStatus[ActionStatus["FAIL"] = 2] = "FAIL";
        })(Base.ActionStatus || (Base.ActionStatus = {}));
        (function (ActionType) {
            ActionType[ActionType["UNKNOWN_ACTION_TYPE"] = 0] = "UNKNOWN_ACTION_TYPE";
            ActionType[ActionType["CLICK"] = 1] = "CLICK";
            ActionType[ActionType["LEFT_PULL"] = 2] = "LEFT_PULL";
            ActionType[ActionType["RIGHT_PULL"] = 3] = "RIGHT_PULL";
            ActionType[ActionType["UP_PULL"] = 4] = "UP_PULL";
            ActionType[ActionType["DOWN_PULL"] = 5] = "DOWN_PULL";
        })(Base.ActionType || (Base.ActionType = {}));
        (function (SubAction) {
            SubAction[SubAction["UNKNOWN_SUB_ACTION"] = 0] = "UNKNOWN_SUB_ACTION";
            SubAction[SubAction["PAGE_ENTER"] = 1] = "PAGE_ENTER";
            SubAction[SubAction["PAGE_LEAVE"] = 2] = "PAGE_LEAVE";
            SubAction[SubAction["PAGE_RESUME"] = 3] = "PAGE_RESUME";
            SubAction[SubAction["PAGE_PAUSE"] = 4] = "PAGE_PAUSE";
        })(Base.SubAction || (Base.SubAction = {}));
        (function (Direction) {
            Direction[Direction["UNKNOWN2"] = 0] = "UNKNOWN2";
            Direction[Direction["UP"] = 1] = "UP";
            Direction[Direction["DOWN"] = 2] = "DOWN";
            Direction[Direction["LEFT"] = 3] = "LEFT";
            Direction[Direction["RIGHT"] = 4] = "RIGHT";
        })(Base.Direction || (Base.Direction = {}));
        (function (TaskStatus) {
            TaskStatus[TaskStatus["UNKNOWN_STATUS"] = 0] = "UNKNOWN_STATUS";
            TaskStatus[TaskStatus["START"] = 1] = "START";
            TaskStatus[TaskStatus["RETRY"] = 2] = "RETRY";
            TaskStatus[TaskStatus["PAUSE"] = 3] = "PAUSE";
            TaskStatus[TaskStatus["RESUME"] = 4] = "RESUME";
            TaskStatus[TaskStatus["PENDING"] = 5] = "PENDING";
            TaskStatus[TaskStatus["PROCESSING"] = 6] = "PROCESSING";
            TaskStatus[TaskStatus["SUCCESS"] = 7] = "SUCCESS";
            TaskStatus[TaskStatus["FAIL"] = 8] = "FAIL";
            TaskStatus[TaskStatus["CANCEL"] = 9] = "CANCEL";
            TaskStatus[TaskStatus["FINISH"] = 10] = "FINISH";
        })(Base.TaskStatus || (Base.TaskStatus = {}));
        (function (OperationType) {
            OperationType[OperationType["UNKNOWN_OPERATION"] = 0] = "UNKNOWN_OPERATION";
            OperationType[OperationType["CLICK"] = 1] = "CLICK";
            OperationType[OperationType["DOUBLE_CLICK"] = 2] = "DOUBLE_CLICK";
            OperationType[OperationType["TRIPLE_CLICK"] = 3] = "TRIPLE_CLICK";
            OperationType[OperationType["LONG_PRESS"] = 4] = "LONG_PRESS";
            OperationType[OperationType["PULL"] = 5] = "PULL";
            OperationType[OperationType["DRAG"] = 6] = "DRAG";
            OperationType[OperationType["SCALE"] = 7] = "SCALE";
            OperationType[OperationType["PULL_DOWN"] = 8] = "PULL_DOWN";
            OperationType[OperationType["PULL_UP"] = 9] = "PULL_UP";
            OperationType[OperationType["RIGHT_CLICK"] = 10] = "RIGHT_CLICK";
            OperationType[OperationType["AUTO"] = 11] = "AUTO";
        })(Base.OperationType || (Base.OperationType = {}));
    })(Base || (Base = {}));

    var ClientEvent$1;
    (function (ClientEvent) {
        (function (ShowType) {
            ShowType[ShowType["UNKNOWN_TYPE"] = 0] = "UNKNOWN_TYPE";
            ShowType[ShowType["PAGE_AUTO"] = 10] = "PAGE_AUTO";
            ShowType[ShowType["PAGE_CUSTOM"] = 11] = "PAGE_CUSTOM";
            ShowType[ShowType["ELEMENT"] = 12] = "ELEMENT";
        })(ClientEvent.ShowType || (ClientEvent.ShowType = {}));
        (function (TaskEventType) {
            TaskEventType[TaskEventType["UNKNOWN_TYPE"] = 0] = "UNKNOWN_TYPE";
            TaskEventType[TaskEventType["USER_OPERATION"] = 1] = "USER_OPERATION";
            TaskEventType[TaskEventType["STAY_LENGTH_STAT_EVENT"] = 2] = "STAY_LENGTH_STAT_EVENT";
            TaskEventType[TaskEventType["BACKGROUND_TASK_EVENT"] = 3] = "BACKGROUND_TASK_EVENT";
        })(ClientEvent.TaskEventType || (ClientEvent.TaskEventType = {}));
    })(ClientEvent$1 || (ClientEvent$1 = {}));

    var ClientEvent;
    (function (ClientEvent) {
        (function (ShowType) {
            ShowType[ShowType["UNKNOWN2"] = 0] = "UNKNOWN2";
            ShowType[ShowType["PAGE"] = 1] = "PAGE";
        })(ClientEvent.ShowType || (ClientEvent.ShowType = {}));
    })(ClientEvent || (ClientEvent = {}));

    // TODO: 这两个变量有副作用，待优化
    var firstLoad = true;
    var enterTime = new Date().valueOf();
    var ShowEvent = (function (options, isV2) {
        var _a, _b;
        if (isV2 === void 0) { isV2 = false; }
        var type = options.type, currentUrlPackage = options.currentUrlPackage, referUrlPackage = options.referUrlPackage, _c = options.name, name = _c === void 0 ? '' : _c, action = options.action, beginTime = options.beginTime, params = options.params, contentPackage = options.contentPackage, status = options.status, actionType = options.actionType, // 这值目前似乎没传进来
        auto = options.auto;
        var commonInfo = (_a = {
                status: status ? (Base.ActionStatus[status] || Base.ActionStatus.UNKNOWN_STATUS) : Base.ActionStatus.SUCCESS
            },
            _a[isV2 ? 'show_type' : 'action_type'] = actionType ? (Base.ActionType[actionType] || Base.ActionType.UNKNOWN_ACTION_TYPE) : Base.ActionType.CLICK,
            _a.url_package = currentUrlPackage,
            _a.refer_url_package = referUrlPackage,
            _a[isV2 ? 'content_wrapper' : 'content_package'] = contentPackage,
            _a);
        if (type === 'PV') {
            var sub_action = Base.SubAction.PAGE_ENTER;
            var actionPB = Base.PageShowAction.ENTER;
            var first_load = firstLoad;
            var stay_length = 0;
            firstLoad = false;
            switch (action) {
                case 'leave':
                    actionPB = Base.PageShowAction.LEAVE;
                    sub_action = Base.SubAction.PAGE_LEAVE;
                    stay_length = getPositiveInteger(new Date().valueOf() - (beginTime || enterTime));
                    break;
                case 'enter':
                    actionPB = Base.PageShowAction.ENTER;
                    sub_action = Base.SubAction.PAGE_ENTER;
                    enterTime = new Date().valueOf();
                    break;
                case 'visible':
                    actionPB = Base.PageShowAction.RESUME;
                    sub_action = Base.SubAction.PAGE_RESUME;
                    enterTime = new Date().valueOf();
                    break;
                case 'hidden':
                    actionPB = Base.PageShowAction.LEAVE;
                    sub_action = Base.SubAction.PAGE_PAUSE;
                    stay_length = getPositiveInteger(new Date().valueOf() - (beginTime || enterTime));
                    break;
            }
            var showType = auto ? ClientEvent$1.ShowType.PAGE_AUTO : ClientEvent$1.ShowType.PAGE_CUSTOM;
            if (isV2) {
                first_load = undefined;
                showType = ClientEvent.ShowType.PAGE;
            }
            return {
                show_event: _assign({ action: actionPB, sub_action: sub_action, type: showType, first_load: first_load, time_cost: 0, stay_length: stay_length }, commonInfo)
            };
        }
        return {
            show_event: _assign({ action: isV2 ? 0 : Base.PageShowAction.ENTER, type: isV2 ? 0 : ClientEvent$1.ShowType.ELEMENT, sub_action: isV2 ? 0 : Base.SubAction.PAGE_ENTER, element_package: (_b = {},
                    _b[isV2 ? 'action2' : 'action'] = name,
                    _b.params = JSON.stringify(params),
                    _b) }, commonInfo)
        };
    });

    var TaskEvent = (function (options, isV2) {
        var _a, _b;
        if (isV2 === void 0) { isV2 = false; }
        var sessionId = options.sessionId, currentUrlPackage = options.currentUrlPackage, referUrlPackage = options.referUrlPackage, name = options.name, params = options.params, type = options.type, contentPackage = options.contentPackage, status = options.status, taskType = options.taskType, operationDirection = options.operationDirection;
        var commonInfo = (_a = {
                url_package: currentUrlPackage,
                refer_url_package: referUrlPackage,
                element_package: (_b = {},
                    _b[isV2 ? 'action2' : 'action'] = name,
                    _b.params = JSON.stringify(params),
                    _b)
            },
            _a[isV2 ? 'content_wrapper' : 'content_package'] = contentPackage,
            _a);
        if (isV2) {
            // v2 click_event 是一个单独的事件类型，v3 则属于 task_event
            var isV2Click = taskType === 'USER_OPERATION' || type === 'CLICK' && !taskType;
            var operationType = type && Base.OperationType[type] || Base.OperationType.CLICK;
            if (isV2Click) {
                return {
                    click_event: _assign({ type: operationType, direction: operationDirection && Base.Direction[operationDirection] || Base.Direction.UNKNOWN2 }, commonInfo)
                };
            }
            return {
                task_event: _assign({ action2: name, status: status && Base.TaskStatus[status] || Base.TaskStatus.UNKNOWN_STATUS }, commonInfo)
            };
        }
        return {
            task_event: _assign({ type: taskType && ClientEvent$1.TaskEventType[taskType] || ClientEvent$1.TaskEventType.USER_OPERATION, status: status && Base.TaskStatus[status] || Base.TaskStatus.UNKNOWN_STATUS, operation_type: Base.OperationType[type] || Base.OperationType.CLICK, operation_direction: operationDirection && Base.Direction[operationDirection] || Base.Direction.UNKNOWN2, session_id: sessionId }, commonInfo)
        };
    });

    /**
     * 自定义事件上报
     */
    var CustomStatEvent = (function (options) {
        var name = options.name, params = options.params, currentUrlPackage = options.currentUrlPackage, referUrlPackage = options.referUrlPackage, extra = __rest(options, ["name", "params", "currentUrlPackage", "referUrlPackage"]);
        return {
            key: name,
            value: JSON.stringify(_assign(_assign({ url_package: currentUrlPackage, refer_url_package: referUrlPackage }, params), extra))
        };
    });

    var VideoStatEvent = (function (params) {
        var currentUrlPackage = params.currentUrlPackage, referUrlPackage = params.referUrlPackage, options = params.options;
        return {
            video_stat_event: _assign(_assign({}, options), { url_package: currentUrlPackage, refer_url_package: referUrlPackage }),
        };
    });

    // 常规业务事件自增 id key
    var LOCAL_INCREAMENT_ID_KEY = 'WEBLOGGER_INCREAMENT_ID_KEY';
    // 自定义事件自增 id key
    var LOCAL_CUSTOM_INCREAMENT_ID_KEY = 'WEBLOGGER_CUSTOM_INCREAMENT_ID_KEY';
    // Bridge 常规事件自增 id 本地缓存 key
    var H5_SEQ_ID_KEY = 'WEBLOGGER_H5_SEQ_ID';
    // Bridge 自定义事件自增 id 本地缓存 key
    var H5_CUSTOM_SEQ_ID_KEY = 'WEBLOGGER_H5_CUSTOM_SEQ_ID';
    // v2 自增 id 本地缓存 key
    var V2_INCREMENT_ID_KEY = 'WEBLOGGER_V2_SEQ_ID';
    var CHANNEL_INCREMENT_ID_KEY = 'WEBLOGGER_CHANNEL_SEQ_ID';
    // 自增 id 最大值
    var maxId = 1e8;
    // 如果没有 localStorage 则使用运行时自增 id，用于非浏览器场景，比如 React Native
    var runtimeIdMap = {};
    var genId = function (key) {
        try {
            if (window && window.localStorage && typeof Storage !== 'undefined' && window.localStorage instanceof Storage) {
                var id = Number(getLocalStorageItem(key)) || 0;
                if (id + 1 > maxId) {
                    id = 0;
                }
                setLocalStorageItem(key, id + 1);
                return id;
            }
        }
        catch (err) { }
        if (!runtimeIdMap[key]) {
            runtimeIdMap[key] = 0;
        }
        return runtimeIdMap[key]++;
    };
    /**
     * 常规事件自增 id
     */
    var getNomalIncrementId = function () {
        return genId(LOCAL_INCREAMENT_ID_KEY);
    };
    /**
     * 自定义事件自增 id
     */
    var getCustomIncrementId = function () {
        return genId(LOCAL_CUSTOM_INCREAMENT_ID_KEY);
    };
    /**
     * 桥接常规事件自增 id
     */
    var getBridgeNormalIncrementId = function () {
        return genId(H5_SEQ_ID_KEY);
    };
    /**
     * 桥接自定义事件自增 id
     */
    var getBridgeCustomIncrementId = function () {
        return genId(H5_CUSTOM_SEQ_ID_KEY);
    };
    /**
     * v2 channel 自增 id
     */
    var getChannelIncrementId = function (channel) {
        if (channel === void 0) { channel = ''; }
        return genId("".concat(CHANNEL_INCREMENT_ID_KEY, "_").concat(channel));
    };
    /**
     * v2 事件自增 id
     */
    var getV2EventIncrementId = function (event) {
        return genId("".concat(V2_INCREMENT_ID_KEY, "_").concat(event));
    };

    var GMT = (function () {
        var timeZoneGMTOffset = (new Date()).getTimezoneOffset() / 60;
        return timeZoneGMTOffset <= 0
            ? "GMT+".concat(leftPad(-timeZoneGMTOffset + '', 2, '0'), ":00")
            : "GMT-".concat(leftPad(timeZoneGMTOffset + '', 2, '0'), ":00");
    })();
    // 为防止每个传出都带common_package，所以用base保留大多数信息
    // 部分信息放到每个消息体内维护，比如client_timestamp
    var ReportEvent = /** @class */ (function () {
        function ReportEvent(options, isV2) {
            if (isV2 === void 0) { isV2 = false; }
            // uint64 客户端时间, Event或Stat发生时的时间, 而不是上传时的时间
            this.client_timestamp = getPositiveInteger(new Date().valueOf());
            // uint64 客户端自增Id, 每新增一条日志+1, 必须是连续的, 用于校检数据完整性
            this.client_increment_id = 0;
            // string 会话id，设备冷启动或者切到后台超过5分钟再切回前台时，重新生成一个id
            // 端内页面不可缺省
            // Tip: 小程序的global字段不一致
            this.session_id = GlobalConfig.sessionId;
            // 埋点唯一主键， PB 中已不推荐使用
            this.event_id = '';
            // v2 没有 time_zone 字段
            if (!isV2) {
                this.time_zone = GMT;
            }
            Object.assign(this, options);
            this.genIncrementId();
        }
        // 设置自增 id，自定义事件和常规事件使用不同的 localStorage key
        ReportEvent.prototype.genIncrementId = function () {
            this.client_increment_id = this.isCustomStatEvent() ? getCustomIncrementId() : getNomalIncrementId();
        };
        // 是否是自定义事件
        ReportEvent.prototype.isCustomStatEvent = function () {
            return !!(this.stat_package && 'custom_stat_event' in this.stat_package);
        };
        // 获取埋点事件类型
        ReportEvent.prototype.getEventType = function () {
            if (this.event_package) {
                var _a = this.event_package, task_event = _a.task_event, show_event = _a.show_event, click_event = _a.click_event, custom_event = _a.custom_event;
                if (show_event)
                    return 'showEvent';
                if (click_event)
                    return 'clickEvent';
                if (task_event)
                    return 'taskEvent';
                if (custom_event)
                    return 'customEvent';
            }
            return 'customEvent';
        };
        return ReportEvent;
    }());
    // 微信小程序事件处理
    var StatEvent = function (eventOptions) {
        eventOptions.currentUrlPackage; eventOptions.referUrlPackage; eventOptions.taskType; var event = __rest(eventOptions, ["currentUrlPackage", "referUrlPackage", "taskType"]);
        return event;
    };
    function logFactory$1(type, data, isV2) {
        var _a;
        if (isV2 === void 0) { isV2 = false; }
        var event_id = data.eventId, currentUrlPackage = data.currentUrlPackage, referUrlPackage = data.referUrlPackage, contentPackage = data.contentPackage, name = data.name, params = data.params;
        if (isV2 && type !== 'RADAR' && type !== 'CUSTOM') {
            // v2 需要特殊处理，先保留 ts-ignore
            if (currentUrlPackage) {
                // @ts-ignore
                currentUrlPackage.page2 = currentUrlPackage.page;
                // @ts-ignore
                delete currentUrlPackage.page;
            }
            if (referUrlPackage) {
                // @ts-ignore
                referUrlPackage.page2 = referUrlPackage.page;
                // @ts-ignore
                delete referUrlPackage.page;
            }
        }
        switch (type) {
            case 'PV':
            case 'SHOW':
                return new ReportEvent({
                    event_package: ShowEvent({
                        type: type,
                        status: data.status,
                        currentUrlPackage: currentUrlPackage,
                        referUrlPackage: referUrlPackage,
                        action: data.type,
                        beginTime: data.beginTime,
                        actionType: data.actionType,
                        name: name,
                        params: params,
                        contentPackage: contentPackage,
                        operationDirection: data.operationDirection,
                        auto: data.auto,
                    }, isV2),
                    event_id: event_id,
                }, isV2);
            case 'CUSTOM':
            case 'CUSTOM_STAT_EVENT':
                var customData = CustomStatEvent({
                    name: name,
                    params: params,
                    currentUrlPackage: currentUrlPackage,
                    referUrlPackage: referUrlPackage,
                });
                return new ReportEvent(isV2 ?
                    {
                        event_package: {
                            custom_event: customData,
                        },
                        event_id: event_id,
                    }
                    :
                        {
                            stat_package: {
                                custom_stat_event: customData,
                            },
                            event_id: event_id,
                        }, isV2);
            case 'RADAR':
                return new ReportEvent({
                    stat_package: {
                        custom_stat_event: CustomStatEvent({
                            name: name,
                            params: params,
                            currentUrlPackage: currentUrlPackage,
                            referUrlPackage: referUrlPackage,
                        }),
                    },
                    event_id: event_id,
                }, isV2);
            // 小程序自定义的事件名称
            case 'HEART_BEAT_EVENT':
            case 'LAUNCH_EVENT':
            case 'APP_USAGE_STAT_EVENT':
            case 'EXCEPTION_EVENT':
            case 'DEVICE_STAT_EVENT':
                return new ReportEvent({
                    stat_package: (_a = {},
                        _a[type.toLowerCase()] = StatEvent(data),
                        _a),
                }, isV2);
            case 'VIDEO':
                if (isV2) {
                    return new ReportEvent({
                        stat_package: VideoStatEvent({
                            currentUrlPackage: currentUrlPackage,
                            referUrlPackage: referUrlPackage,
                            options: params.params,
                        }),
                        event_id: event_id,
                    }, isV2);
                }
                return new ReportEvent({
                    event_package: TaskEvent({
                        type: type,
                        status: data.status,
                        taskType: data.taskType,
                        sessionId: GlobalConfig.sessionId,
                        currentUrlPackage: currentUrlPackage,
                        referUrlPackage: referUrlPackage,
                        name: name,
                        params: params.params,
                        contentPackage: contentPackage,
                    }, isV2),
                    event_id: event_id,
                }, isV2);
            default:
                return new ReportEvent({
                    event_package: TaskEvent({
                        type: type,
                        status: data.status,
                        taskType: data.taskType,
                        sessionId: GlobalConfig.sessionId,
                        currentUrlPackage: currentUrlPackage,
                        referUrlPackage: referUrlPackage,
                        name: name,
                        params: params,
                        contentPackage: contentPackage,
                    }, isV2),
                    event_id: event_id,
                }, isV2);
        }
    }

    var _a;
    var uaInfo;
    var lastUA = ((_a = window === null || window === void 0 ? void 0 : window.navigator) === null || _a === void 0 ? void 0 : _a.userAgent) || '';
    var getUAInfo = function (ua) {
        var _a, _b, _c, _d;
        if (ua === void 0) { ua = (_a = window === null || window === void 0 ? void 0 : window.navigator) === null || _a === void 0 ? void 0 : _a.userAgent; }
        if (uaInfo && ua === lastUA) {
            return uaInfo;
        }
        var UNKNOWN = 'unknow';
        var androidReg = /android/i;
        var MacOS = 'Mac OS';
        var Windows = 'Windows';
        var Android = 'Android';
        var iPhone = 'iPhone';
        var iPad = 'iPhone';
        var WinPhone = 'Windows Phone';
        var Linux = 'Linux';
        uaInfo = {
            os: {
                name: UNKNOWN,
                version: UNKNOWN
            },
            model: UNKNOWN
        };
        try {
            var uaMatches = ua.match(/\((.*?)\)/);
            if (!uaMatches) {
                var isAndroid = androidReg.test(ua);
                if (isAndroid) {
                    uaInfo.os.name = Android;
                }
                return uaInfo;
            }
            var uaMain = uaMatches[1] + ')';
            var uaArr = uaMain.split(';').map(function (str) { return str.trim(); });
            var osName = void 0, osVersion = void 0, model = void 0;
            if (ua.indexOf(WinPhone) > -1) {
                model = osName = WinPhone;
                var matches = uaMain.match(/Windows\sPhone\s(.*?)[;\)\s]/);
                if (matches) {
                    osVersion = matches[1];
                }
                model = (_b = uaArr[uaArr.length - 1]) === null || _b === void 0 ? void 0 : _b.replace(')', '');
            }
            else if (androidReg.test(ua)) {
                // 安卓 ua 信息解析
                osName = Android;
                var modelMatches = uaMain.match(/android.*?;(.*?)build\//i);
                if (modelMatches) {
                    model = (_c = modelMatches[1]) === null || _c === void 0 ? void 0 : _c.split(';').pop();
                    model = model && model.trim();
                }
                for (var i = 0; i < uaArr.length; i++) {
                    if (androidReg.test(uaArr[i])) {
                        osVersion = (_d = uaArr[i]) === null || _d === void 0 ? void 0 : _d.replace(androidReg, '');
                        if (!model) {
                            model = uaArr[i + 1];
                        }
                        break;
                    }
                }
            }
            else if (ua.indexOf(iPhone) > -1 || ua.indexOf(iPad) > -1) {
                // iOS 系统信息解析
                model = ua.indexOf(iPhone) > -1 ? iPhone : iPad;
                osName = 'iOS';
                var matches = uaMain.match(/OS\s(.*?)\slike/);
                if (matches) {
                    osVersion = matches[1];
                }
            }
            else if (ua.indexOf(MacOS) > -1) {
                // Mac OS 系统信息解析
                model = osName = MacOS;
                var matches = uaMain.match(/OS\sX\s(.*?)[;\)\s]/);
                if (matches) {
                    osVersion = matches[1];
                }
            }
            else if (ua.indexOf(Windows) > -1) {
                model = osName = Windows;
                var matches = uaMain.match(/Windows\s(.*?)[;\)]/);
                if (matches) {
                    osVersion = matches[1];
                }
            }
            else if (ua.indexOf('Nokia') > -1) {
                osName = 'Symbian';
                var vMatches = ua.match(/Symbian.*?\/(.*?);/);
                if (vMatches) {
                    osVersion = vMatches[1];
                }
                var mMatches = ua.match(/Nokia(.*?)\//);
                if (mMatches) {
                    model = mMatches[1];
                }
            }
            else if (uaMain.indexOf(Linux) > -1) {
                osName = model = Linux;
            }
            uaInfo = {
                os: {
                    name: osName || UNKNOWN,
                    version: osVersion && osVersion.replace(/_/g, '.').trim() || UNKNOWN
                },
                model: model || UNKNOWN
            };
        }
        catch (err) {
            console.error('[109]', 'UA 信息解析错误', err.message);
        }
        return uaInfo;
    };
    /**
     * http://git.corp.kuaishou.com/ks-frontend/modules/web-util/blob/master/src/device.js
     */
    var kpnFromUA = function () {
        if (typeof window === 'undefined')
            return '';
        var ua = window.navigator.userAgent;
        // 注意顺序很重要
        var rules = [
            [/ Kwai\//, 'KUAISHOU'],
            [/ ksthanos\//, 'THANOS'],
            [/ ksNebula\//i, 'NEBULA'],
            // iOS 审核移除敏感信息
            // [/ wxwork\//, 'workWechat'],
            // [/ MicroMessenger\//, 'wechat'],
            // [/ QQ\//, 'qq'],
            // [/__weibo__([^_]+)/, 'weibo'],
            // [/ livemate\//, 'livemate'],
            // [/ baiduboxapp\//, 'baidu'],
            // [/ MQQBrowser\//, 'qqBrowser'],
            // [/\(MSIE ([^;]+);/, 'ie'],
        ];
        for (var i = 0; i < rules.length; i++) {
            var tmp = rules[i];
            var rule = tmp[0];
            var name_1 = tmp[1];
            var rRes = rule.exec(ua);
            if (rRes) {
                return name_1;
            }
        }
        return '';
    };
    /**
     * 获取默认的 KPN,TODO: 副作用函数
     */
    var defaultKpn;
    function getDefaultKpn() {
        if (defaultKpn === undefined) {
            defaultKpn = isInApp() && getCookie('kpn') || kpnFromUA() || '';
        }
        return defaultKpn;
    }
    /**
     * 判断是否是 kds 环境
     */
    var isKDSNative = (function () {
        var res;
        return function () {
            return res;
        };
    })();
    var uaUtils = {
        supportsPushState: supportsPushState,
        getUAInfo: getUAInfo,
        getDefaultKpn: getDefaultKpn
    };

    /**
     * hybrid 以及 browser 包使用此文件，将 invoke 定义和 bridge 包隔离，避免 browser 以及 hybrid 包必须引用 yoda 包 ts 定义问题。
     */
    var invoke;
    var isInYoda = function () {
        var _a;
        // @ts-ignore
        return typeof ((_a = window === null || window === void 0 ? void 0 : window.__yodaBridge__) === null || _a === void 0 ? void 0 : _a.invoke) === 'function';
    };
    /**
     * 是否支持桥接 canIUse
     */
    var isSupportBridge = function (name, namespace) {
        if (namespace === void 0) { namespace = 'tool'; }
        return __awaiter(void 0, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (typeof invoke === 'undefined')
                            return [2 /*return*/, false];
                        return [4 /*yield*/, invoke('tool.canIUse', {
                                namespace: namespace,
                                name: name
                            })];
                    case 1:
                        res = _a.sent();
                        return [2 /*return*/, (res === null || res === void 0 ? void 0 : res.canUse) || false];
                }
            });
        });
    };
    /**
     * 是否支持桥接上报埋点接口
     */
    var isSupportBridgeLog = function () {
        return isSupportBridge('setClientLog');
    };
    /**
     * 传入 invoke 方法
     */
    var postInvoke = function (invokeOrYoda) {
        if (typeof invokeOrYoda === 'function') {
            invoke = invokeOrYoda;
        }
        else {
            invoke = invokeOrYoda.invoke;
        }
    };
    /**
     * 发送业务事件
     */
    var setClientLog = function (type, data) { return __awaiter(void 0, void 0, void 0, function () {
        var err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, invoke('tool.setClientLog', { type: type, data: data })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    err_1 = _a.sent();
                    error('[error 206]', "yoda.tool.setClientLog() \u62A5\u9519: ".concat(err_1.message));
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    /**
     * 发送雷达事件
     */
    var sendRadarLog = function (params) { return __awaiter(void 0, void 0, void 0, function () {
        var err_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, invoke('tool.sendRadarLog', params)];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    err_2 = _a.sent();
                    error('[error 207]', "yoda.tool.sendRadarLog() \u62A5\u9519: ".concat(err_2.message));
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    /**
     * 发送web日志，支持日志回捞
     */
    var sendWebLog = function (params) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            try {
                return [2 /*return*/, invoke('tool.sendWebLog', params)];
            }
            catch (err) {
                error('[error 214]', "yoda.tool.sendWebLog() \u62A5\u9519: ".concat(err.message));
            }
            return [2 /*return*/];
        });
    }); };
    /**
     * 发送汇总数据
     */
    var sendSummarizedLog = function (params) { return __awaiter(void 0, void 0, void 0, function () {
        var err_3;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, invoke('tool.sendSummarizedLog', params)];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    err_3 = _a.sent();
                    error('[error 208]', "yoda.tool.sendSummarizedLog() \u62A5\u9519: ".concat(err_3.message));
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    /**
     * 获取应用信息
     * @returns
     */
    var getAppInfo = function () { return __awaiter(void 0, void 0, void 0, function () {
        var err_4;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, invoke('system.getAppInfo')];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    err_4 = _a.sent();
                    error('[error 209]', "yoda.system.getAppInfo() \u62A5\u9519: ".concat(err_4.message));
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    /**
     * 获取 webview 信息
     */
    var getWebviewLoadPerf = function (params, options) {
        if (params === void 0) { params = {}; }
        return __awaiter(void 0, void 0, void 0, function () {
            var err_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, invoke('webview.getPageLoadData', params, options)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        err_5 = _a.sent();
                        error('[error 209]', "yoda.webview.getPageLoadData() \u62A5\u9519: ".concat(err_5.message));
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 获取 ksd相关webview 信息
     */
    var getKDSWebviewLoadPerf = function (params) {
        if (params === void 0) { params = {}; }
        return __awaiter(void 0, void 0, void 0, function () {
            var err_7;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, invoke('system.getPageLoadData', params)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        err_7 = _a.sent();
                        error('[error 209]', "yoda.system.getPageLoadData() \u62A5\u9519: ".concat(err_7.message));
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 获取设备信息
     * @returns
     */
    var getDeviceInfo = function () { return __awaiter(void 0, void 0, void 0, function () {
        var err_8;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, invoke('system.getDeviceInfo')];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    err_8 = _a.sent();
                    error('[error 209]', "yoda.system.getDeviceInfo() \u62A5\u9519: ".concat(err_8.message));
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    var yodaUtils = {
        setClientLog: setClientLog,
        sendRadarLog: sendRadarLog,
        sendWebLog: sendWebLog,
        sendSummarizedLog: sendSummarizedLog,
        getWebviewLoadPerf: getWebviewLoadPerf,
        getKDSWebviewLoadPerf: getKDSWebviewLoadPerf,
        isSupportBridgeLog: isSupportBridgeLog,
        isSupportBridge: isSupportBridge,
        invoke: invoke
    };

    /**
     * 事件
     */
    // 网络类型
    var networkType = {
        unknown: 0,
        none: 1,
        wifi: 2,
        '4g': 3,
        '3g': 4,
        '2g': 5,
        '5g': 7,
        'slow-2g': 5,
    };
    // 任务事件类型
    var TaskOperations;
    (function (TaskOperations) {
        TaskOperations["CLICK"] = "CLICK";
        TaskOperations["DOUBLE_CLICK"] = "DOUBLE_CLICK";
        TaskOperations["TRIPLE_CLICK"] = "TRIPLE_CLICK";
        TaskOperations["LONG_PRESS"] = "LONG_PRESS";
        TaskOperations["PULL"] = "PULL";
        TaskOperations["DRAG"] = "DRAG";
        TaskOperations["SCALE"] = "SCALE";
        TaskOperations["PULL_DOWN"] = "PULL_DOWN";
        TaskOperations["PULL_UP"] = "PULL_UP";
        TaskOperations["AUTO"] = "AUTO";
    })(TaskOperations || (TaskOperations = {}));
    // 事件类型
    var EventTypes;
    (function (EventTypes) {
        EventTypes["PV"] = "PV";
        EventTypes["SHOW"] = "SHOW";
        EventTypes["VIDEO"] = "VIDEO";
        // Costume
        EventTypes["CUSTOM"] = "CUSTOM";
        EventTypes["CUSTOM_EVENT"] = "CUSTOM_EVENT";
        EventTypes["RADAR"] = "RADAR";
    })(EventTypes || (EventTypes = {}));

    var CommonPackage$1 = /** @class */ (function () {
        function CommonPackage(options) {
            // 默认值为 undefined 减小上报时字段
            // 标识信息
            this.identity_package = {
                device_id: undefined,
                global_id: undefined,
                user_id: undefined,
                union_id: undefined,
                open_id: undefined,
                iu_id: undefined
            };
            // 应用信息
            this.app_package = {
                product: undefined,
                language: undefined,
                platform: undefined,
                container: 'H5',
                package_name: undefined,
                product_name: undefined,
                version_name: undefined,
                channel: undefined,
                version_code: undefined,
            };
            this.experiment = undefined;
            this.service_name = undefined;
            this.safety_id = undefined;
            this.sub_biz = undefined;
            // 设备信息
            this.device_package = {
                os_version: undefined,
                model: undefined,
                ua: undefined,
            };
            this.need_encrypt = false;
            this.network_package = {
                type: networkType.unknown,
            };
            // H5 附加参数
            this.h5_extra_attr = {
                // SDK 版本信息
                sdk_name: 'webLogger',
                sdk_version: '3.10.23',
                sdk_bundle: 'log.hybrid.js',
            };
            this.global_attr = {
                // 染色参数
                entry_tag: [],
            };
            this.location_package = {
                country: undefined,
                province: undefined,
                city: undefined,
                county: undefined,
                street: undefined,
                latitude: undefined,
                longitude: undefined, // 经度
            };
            this.update(options);
            if (this.app_package.version_name) {
                var lastDotIndex = this.app_package.version_name.lastIndexOf('.');
                this.app_package.version_code = +this.app_package.version_name.slice(lastDotIndex + 1) || 0;
            }
            if (!this.app_package.version_name) {
                this.app_package.version_name = undefined;
            }
            if (!this.app_package.version_code) {
                this.app_package.version_code = undefined;
            }
        }
        CommonPackage.prototype.getH5ExtraAttr = function (extraInfo) {
            return Object.assign({}, this.h5_extra_attr, extraInfo);
        };
        /**
         * 更新 commonPackage 信息
         * @param options
         * @returns
         */
        CommonPackage.prototype.update = function (options) {
            if (typeof options !== 'object')
                return;
            flattenAssign(this, options);
            // network type 特殊处理
            var network = options.network_type;
            if (network && networkType[network]) {
                this.network_package.type = networkType[network];
            }
        };
        /**
         * 更新 global_attr 信息
         * @param options
         */
        CommonPackage.prototype.updateGlobalAttr = function (options) {
            Object.assign(this.global_attr || {}, options);
        };
        /**
         * 将 commonPackage 转换为 json 输出
         * @returns
         */
        CommonPackage.prototype.toJSON = function () {
            // 处理 v2 空字符串接口会报错问题
            if (!this.identity_package.user_id) {
                this.identity_package.user_id = undefined;
            }
            var json = _assign(_assign({}, this), { toJSON: function () {
                    return _assign(_assign({}, json), { h5_extra_attr: JSON.stringify(json.h5_extra_attr), global_attr: JSON.stringify(json.global_attr) });
                } });
            json.global_attr = _assign({}, this.global_attr);
            if (this.global_attr.entry_tag && this.global_attr.entry_tag.length) {
                json.global_attr.entry_tag = this.global_attr.entry_tag.slice();
            }
            else {
                // 节省上报字段
                delete json.global_attr.entry_tag;
            }
            if (!Object.keys(this.global_attr).length) {
                // 节省上报字段
                delete json.global_attr;
            }
            var location = this.location_package;
            if (location) {
                Object.keys(location).forEach(function (key) {
                    if (typeof location[key] === 'undefined') {
                        delete location[key];
                    }
                });
            }
            return json;
        };
        return CommonPackage;
    }());

    var devices = [
        ['a7', '640x1136', ['iPhone 5', 'iPhone 5s']],
        ['a7', '1536x2048', ['iPad Air', 'iPad Mini 2', 'iPad Mini 3']],
        ['a8', '640x1136', ['iPod touch (6th gen)']],
        ['a8', '750x1334', ['iPhone 6']],
        ['a8', '1242x2208', ['iPhone 6 Plus']],
        ['a8', '1536x2048', ['iPad Air 2', 'iPad Mini 4']],
        ['a9', '640x1136', ['iPhone SE']],
        ['a9', '750x1334', ['iPhone 6s']],
        ['a9', '1242x2208', ['iPhone 6s Plus']],
        ['a9x', '1536x2048', ['iPad Pro (1st gen 9.7-inch)']],
        ['a9x', '2048x2732', ['iPad Pro (1st gen 12.9-inch)']],
        ['a10', '750x1334', ['iPhone 7']],
        ['a10', '1242x2208', ['iPhone 7 Plus']],
        ['a10x', '1668x2224', ['iPad Pro (2th gen 10.5-inch)']],
        ['a10x', '2048x2732', ['iPad Pro (2th gen 12.9-inch)']],
        ['a11', '750x1334', ['iPhone 8']],
        ['a11', '1242x2208', ['iPhone 8 Plus']],
        ['a11', '1125x2436', ['iPhone X']],
        ['a12', '828x1792', ['iPhone Xr']],
        ['a12', '1125x2436', ['iPhone Xs']],
        ['a12', '1242x2688', ['iPhone Xs Max']],
        ['a12x', '1668x2388', ['iPad Pro (3rd gen 11-inch)']],
        ['a12x', '2048x2732', ['iPad Pro (3rd gen 12.9-inch)']],
    ];
    var pickDeviceByResolution = function () {
        var resolution = getResolution();
        var perhaps = [];
        for (var _i = 0, devices_1 = devices; _i < devices_1.length; _i++) {
            var device = devices_1[_i];
            if (resolution === device[1]) {
                perhaps = perhaps.concat(device[2]);
            }
        }
        return perhaps.length && perhaps;
    };
    var getModel = (function () {
        return function () {
            try {
                var uaInfo = getUAInfo();
                var inApp = isInApp();
                var model_1 = uaInfo.model || uaInfo.os.name || 'unknown';
                if (uaInfo.os.name !== 'iOS' || inApp) {
                    return model_1;
                }
                var device = pickDeviceByResolution();
                if (device) {
                    return device.join(' or ');
                }
                return model_1;
            }
            catch (err) {
                return 'unknown';
            }
        };
    })();

    var platformPB = 0;
    function platform() {
        try {
            var osName = getUAInfo().os.name;
            if (!osName) {
                return platformPB = 0;
            }
            if (osName === 'Android') {
                return platformPB = isInApp() ? 6 : 8;
            }
            if (osName === 'iOS') {
                return platformPB = isInApp() ? 7 : 9;
            }
            return platformPB = 10;
        }
        catch (err) {
            return 0;
        }
    }
    function network() {
        var networkType = 'unknown';
        var navigator = window === null || window === void 0 ? void 0 : window.navigator;
        var connection = (navigator === null || navigator === void 0 ? void 0 : navigator.connection) || (navigator === null || navigator === void 0 ? void 0 : navigator.mozConnection) || (navigator === null || navigator === void 0 ? void 0 : navigator.webkitConnection);
        if (connection) {
            networkType = connection.type || connection.effectiveType;
        }
        return networkType;
    }
    var CommonPackage = /** @class */ (function (_super) {
        __extends(CommonPackage, _super);
        function CommonPackage(options) {
            var _this = _super.call(this, _assign({ platform: platform(), container: platformPB === 10 ? 'WEB' : 'H5', version_name: isInApp() && getCookie('appver', undefined, true) || '', network_type: network(), device_id: generateDID(), user_id: getCookie('userId'), global_id: isInApp() && getCookie('egid') || '', app_package: {
                    language: navigator.language,
                }, device_package: {
                    os_version: getUAInfo().os.version,
                    model: getModel(),
                    ua: navigator.userAgent,
                } }, options)) || this;
            Object.assign(_this.h5_extra_attr, {
                host_product: getDefaultKpn(),
                resolution: getResolution(),
                screen_width: getPositiveInteger(screen.width),
                screen_height: getPositiveInteger(screen.height),
                device_pixel_ratio: window.devicePixelRatio || 1,
                domain: window.location.origin,
                fromIframe: (window === null || window === void 0 ? void 0 : window.self) !== (window === null || window === void 0 ? void 0 : window.top),
            }, options.h5_extra_attr);
            delete _this.location_package;
            return _this;
        }
        CommonPackage.prototype.getVersionName = function () {
            return this.app_package.version_name;
        };
        /**
        * 更新 commonPackage 信息
        * @param options
        * @returns
        */
        CommonPackage.prototype.update = function (options) {
            _super.prototype.update.call(this, options);
            // 主站版本号较为严格，直接通过版本名读取
            var version_name = this.app_package.version_name;
            if (typeof version_name === 'string') {
                var lastDotIndex = version_name.lastIndexOf('.');
                this.app_package.version_code = +version_name.slice(lastDotIndex + 1) || 0;
                this.h5_extra_attr.app_version_name = version_name;
            }
        };
        // v2 丢失统计
        CommonPackage.prototype.setAdditionalSeqIdPackage = function (eventType) {
            this.additional_seq_id_package = {
                channel: 3,
                channel_seq_id: getChannelIncrementId('NORMAL'),
                custom_type: eventType,
                custom_seq_id: getV2EventIncrementId(eventType)
            };
        };
        // 用于统计桥接丢失率
        CommonPackage.prototype.increaseH5SeqId = function (isCustom) {
            this.h5_extra_attr.client_timestamp = getPositiveInteger(new Date().valueOf());
            this.h5_extra_attr.seq_id = isCustom ? getBridgeCustomIncrementId() : getBridgeNormalIncrementId();
        };
        return CommonPackage;
    }(CommonPackage$1));

    var oldApiTransformer = function (type, name, params, contentPackage) {
        if (name === void 0) { name = {}; }
        var realType;
        if (typeof type !== 'string') {
            realType = type.type;
        }
        realType = type.toUpperCase();
        var options = {};
        if (typeof name === 'string') {
            console.warn('[warn 105]', '接口参数为旧式，后续版本将不再兼容，推荐使用新的参数形式');
            options.contentPackage = contentPackage;
            if (type === 'CUSTOM') {
                options.key = name;
                options.value = params;
            }
            else {
                options.params = params;
                if (type === 'PV') {
                    options.page = name;
                }
                else {
                    options.action = name;
                }
            }
        }
        else {
            options = name;
            if (['CUSTOM', 'CUSTOM_EVENT'].includes(realType)) {
                var customParams = name;
                if (customParams.action && !customParams.key) {
                    options.key = customParams.action;
                }
                if (customParams.params && !customParams.value) {
                    options.value = customParams.params;
                }
            }
        }
        return [
            realType,
            options
        ];
    };

    var WeblogNamespace;
    var WEBLOGGER_SILENCE_KEY = 'WEBLOGGER_SILENCED';
    // 在浏览器环境下统一一个命名空间，用来统计
    if (typeof window !== 'undefined') {
        if (typeof window._WEBLOGGER !== 'object') {
            window._WEBLOGGER = {
                _silenced: false,
                constructors: [],
                instances: [],
                // 全局静默
                silence: function (state) {
                    try {
                        if (state) {
                            sessionStorage.setItem(WEBLOGGER_SILENCE_KEY, '1');
                        }
                        else {
                            sessionStorage.removeItem(WEBLOGGER_SILENCE_KEY);
                        }
                        window._WEBLOGGER.instances.forEach(function (i) {
                            i === null || i === void 0 ? void 0 : i.silence(state);
                        });
                    }
                    catch (err) { }
                }
            };
        }
        WeblogNamespace = window._WEBLOGGER;
    }
    var initSilence = function () {
        if (WeblogNamespace === null || WeblogNamespace === void 0 ? void 0 : WeblogNamespace.silence) {
            try {
                WeblogNamespace._silenced = !!sessionStorage.getItem(WEBLOGGER_SILENCE_KEY);
            }
            catch (err) { }
        }
    };
    /**
     * 记录一个 Weblog 类
     */
    var addClass = function (C) {
        // bootstrap 分级加载命名空间，后续废弃
        if (typeof window._GLOBAL_KS_WEBLOGGER_ !== 'undefined') {
            window._GLOBAL_KS_WEBLOGGER_.Factory = C;
        }
        else {
            window.Weblog = C;
        }
        if ((WeblogNamespace === null || WeblogNamespace === void 0 ? void 0 : WeblogNamespace.constructors) && C && WeblogNamespace.constructors.indexOf(C) === -1) {
            // 加个 id 区分
            C._classId = WeblogNamespace.constructors.length;
            WeblogNamespace.constructors.push(C);
        }
    };
    /**
     * 记录一个 Weblog 实例
     */
    var addInstance = function (i) {
        if ((WeblogNamespace === null || WeblogNamespace === void 0 ? void 0 : WeblogNamespace.instances) && i && WeblogNamespace.instances.indexOf(i) === -1) {
            WeblogNamespace.instances.push(i);
            if (WeblogNamespace._silenced) {
                i === null || i === void 0 ? void 0 : i.silence(true);
            }
        }
    };
    /**
     * 移除一个 Weblog 实例
     */
    var removeInstance = function (i) {
        if (WeblogNamespace === null || WeblogNamespace === void 0 ? void 0 : WeblogNamespace.instances) {
            var index = WeblogNamespace.instances.indexOf(i);
            if (index !== -1) {
                WeblogNamespace.instances.splice(index, 1);
            }
        }
    };

    var getDefaultUserId = function () { return getCookie('userId') || getCookie('userName') || undefined; };
    var isWillClosed = false;
    var randomValue = Math.random();
    var Weblog = /** @class */ (function (_super) {
        __extends(Weblog, _super);
        function Weblog(logConfig, base) {
            if (logConfig === void 0) { logConfig = {}; }
            var _this = _super.call(this, logConfig, base) || this;
            _this.beforeUnload = function (event) {
                if (isWillClosed)
                    return;
                _this.logger.drain();
                for (var key in _this.plugins) {
                    var plugin = _this.plugins[key];
                    if (typeof plugin.beforeUnload === 'function') {
                        plugin.beforeUnload(event);
                    }
                }
                isWillClosed = true;
                setTimeout(function () {
                    isWillClosed = false;
                }, 2000);
            };
            _this.isBridge = logConfig.bridgeMode && logConfig.yoda && isInYoda();
            _this.logConfig.isBridge = _this.isBridge;
            console.warn('[webLogger] 您正在使用的是', logConfig.bridgeMode ?
                "\u3010\u6865\u63A5\u4E0A\u62A5\u3011\u7684 js \u5305\u3002".concat(_this.isBridge ? '' : '当前是端外打开，会自动改为 http 上报！上报链路为 ' + (logConfig.proto === 'v2' ? 'v2' : 'v3'))
                :
                    "\u3010http \u4E0A\u62A5\u3011\u7684 js \u5305\u3002\u4E0A\u62A5\u94FE\u8DEF\u4E3A ".concat(logConfig.proto === 'v2' ? 'v2' : 'v3'));
            updateConfig('sessionId', base && base.session_id || getCookie('sid') || getSessionId());
            _this.isV2 = _this.logConfig.proto === 'v2';
            _this.getSampled(_this.logConfig);
            // 参数默认值处理
            _this.presetBaseOption(_this.baseOption);
            _this.commonPackage = new CommonPackage(_this.baseOption);
            _this.logger = new Logger(_this.logConfig, _this.commonPackage);
            _this.initYoda();
            _this.addPlugins();
            addMonitor(window, 'pagehide', _this.beforeUnload);
            addMonitor(window, 'beforeunload', _this.beforeUnload);
            addInstance(_this);
            return _this;
        }
        Weblog.prototype.getSampled = function (logConfig) {
            var _a = logConfig || {}, sampleRateFn = _a.sampleRateFn, sampleRate = _a.sampleRate;
            if (typeof sampleRateFn === 'function') {
                this.sampled = !!sampleRateFn();
            }
            else if (typeof sampleRate !== 'undefined') {
                this.sampled = randomValue < Number(sampleRate);
            }
        };
        /**
         * 参数预处理
         * @param base
         */
        Weblog.prototype.presetBaseOption = function (base) {
            if (!this.isBridge) {
                if (this.isV2) {
                    if (typeof base.product !== 'number')
                        error("[error 101]", "\u8BF7\u8BBE\u7F6E\u6709\u6548\u7684 product \u503C\uFF0C\u503C\u7C7B\u578B\u4E3A number!");
                    delete base.product_name;
                }
                else {
                    if (!base.product_name || typeof base.product_name !== 'string')
                        error("[error 100]", "\u8BF7\u8BBE\u7F6E\u6709\u6548\u7684 product_name \u503C\uFF0C\u503C\u7C7B\u578B\u4E3A string!");
                    delete base.product;
                }
                if (!base.user_id) {
                    base.user_id = getDefaultUserId();
                    if (!base.user_id)
                        warn("[warn 102]", "user_id \u4E3A\u7A7A\uFF0C\u8BF7\u786E\u5B9A\u662F\u5426\u4F20\u5165\uFF0C\u5982\u679C\u662F\u5F02\u6B65\u901A\u8FC7 updateCommonPackage \u63A5\u53E3\u4F20\u5165\uFF0C\u53EF\u4EE5\u5FFD\u7565\u8BE5\u63D0\u793A!");
                }
                if (!base.device_id) {
                    base.device_id = generateDID();
                    if (!base.device_id)
                        error("[error 103]", "device_id \u4E3A\u7A7A\uFF0C\u8BF7\u8BBE\u7F6E\u6709\u6548\u7684\u8BBE\u5907 id!");
                }
                if (!base.global_id && isInApp()) {
                    base.global_id = getCookie('egid') || '';
                }
            }
        };
        Object.defineProperty(Weblog.prototype, "yoda", {
            get: function () {
                return this.logConfig.yoda || (window === null || window === void 0 ? void 0 : window.KSYoda) || (window === null || window === void 0 ? void 0 : window.yoda);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Weblog.prototype, "Utils", {
            // TIP:
            get: function () {
                return {
                    yoda: isInYoda() && this.yoda ? yodaUtils : null,
                    cookie: cookieUtils,
                    ua: uaUtils,
                    io: {
                        sendData: sendData,
                    },
                };
            },
            enumerable: false,
            configurable: true
        });
        Weblog.prototype.initYoda = function () {
            var _this = this;
            if (!isInYoda() || !this.yoda)
                return;
            postInvoke(this.yoda);
            // 更新设备信息
            getDeviceInfo().then(function (info) {
                if (info === null || info === void 0 ? void 0 : info.mod) {
                    _this.commonPackage.device_package.model = info.mod;
                }
            });
            // 更新应用信息
            getAppInfo().then(function (info) {
                if (typeof info === 'object') {
                    var did = info.did, userId = info.userId, appver = info.appver;
                    did && (_this.commonPackage.identity_package.device_id = did);
                    userId && !_this.commonPackage.identity_package.user_id && (_this.commonPackage.identity_package.user_id = userId);
                    appver && (_this.commonPackage.app_package.version_name = appver);
                }
            });
        };
        Weblog.prototype.initUrlPackage = function () {
            _super.prototype.initUrlPackage.call(this);
            var referer = this.logConfig.referer;
            var referUrl;
            var referType = 'web';
            if (referer) {
                referUrl = referer.value;
                referType = referer.type && referer.type || referType;
            }
            else if (typeof document !== 'undefined' && document.referrer) {
                referUrl = document.referrer;
            }
            if (referUrl) {
                this.updateReferUrlPackage(referUrl, referType);
            }
        };
        Weblog.prototype.updateCurrentUrlPackage = function (url, type) {
            if (url === void 0) { url = (location === null || location === void 0 ? void 0 : location.href) || ''; }
            if (type === void 0) { type = 'web'; }
            if (typeof url === 'object' && !url.force && this.currentUrlPackage) {
                var _a = url.page, page = _a === void 0 ? this.currentUrlPackage.page : _a, _b = url.params, params = _b === void 0 ? {} : _b;
                if (page === this.currentUrlPackage.page) {
                    return this.currentUrlPackage.update(page, params);
                }
            }
            this.currentUrlPackage = new UrlPackage(url, type, this.logConfig.urlMap);
            if (this.logConfig.attachUrl)
                this.currentUrlPackage.attachUrl();
        };
        /**
         * 更新 refer url 信息
         * @param url
         * @param type
         */
        Weblog.prototype.updateReferUrlPackage = function (url, type) {
            if (url === void 0) { url = this.currentUrlPackage; }
            if (type === void 0) { type = 'web'; }
            if (url instanceof UrlPackage) {
                this.referUrlPackage = url;
            }
            else {
                this.referUrlPackage = new UrlPackage(url, type, this.logConfig.urlMap);
            }
        };
        Object.defineProperty(Weblog.prototype, "isSendSampled", {
            get: function () {
                return (typeof this.sampled === 'undefined' ? true : this.sampled) || this.logger.isDebug;
            },
            enumerable: false,
            configurable: true
        });
        Weblog.prototype.generateLog = function (eventType, eventOptions, factory) {
            if (factory === void 0) { factory = logFactory$1; }
            var _a = eventOptions, action = _a.action, params = _a.params, status = _a.status, type = _a.type, eventId = _a.eventId, contentPackage = _a.contentPackage, urlPage = _a.urlPage;
            var currentUrlPackage;
            if (contentPackage) {
                contentPackage = typeof contentPackage === 'string' ? contentPackage : JSON.stringify(contentPackage);
            }
            var isPVEvent = eventType === 'PV';
            // 若是 pv 先更新currentUrlPackage
            if (isPVEvent) {
                var _b = eventOptions, page = _b.page, type_1 = _b.type;
                if (!type_1)
                    eventOptions.type = type_1 = 'enter';
                // enter 则重新生成新的 currentUrlPackage，否则局部更新
                if (type_1 === 'enter' && page) {
                    this.updateCurrentUrlPackage({ page: page, params: params });
                }
                else {
                    this.currentUrlPackage.update(page, params);
                }
            }
            else if (urlPage && urlPage.page) { // 如果参数中带有 urlPage，用来解决页面飘移问题，只影响该埋点
                currentUrlPackage = {
                    page: urlPage.page,
                    identity: urlPage.identity || uuid(),
                    params: JSON.stringify(urlPage.params),
                    page_type: this.currentUrlPackage.page_type
                };
            }
            var commonInfo = {
                currentUrlPackage: currentUrlPackage || this.currentUrlPackage.toJSON(),
                referUrlPackage: this.referUrlPackage ? this.referUrlPackage.toJSON() : undefined,
                contentPackage: contentPackage
            };
            // 除了雷达之外的事件都希望能按照规范上报埋点，使 page 不为url
            if (eventType !== 'RADAR' && (currentUrlPackage === null || currentUrlPackage === void 0 ? void 0 : currentUrlPackage.page.indexOf('http')) === 0) {
                console.warn('[warn 104]', "\u57CB\u70B9\u9875\u9762\u4FE1\u606F ".concat(currentUrlPackage === null || currentUrlPackage === void 0 ? void 0 : currentUrlPackage.page, " \u4E0D\u7B26\u5408\u89C4\u8303\uFF0C\u63A8\u8350\u4F7F\u7528\u5927\u5199\u82F1\u6587 + \u4E0B\u5212\u7EBF\u683C\u5F0F"));
            }
            if (isPVEvent) {
                var type_2 = eventOptions.type;
                // PV 类型
                return factory(eventType, _assign(_assign(_assign({}, eventOptions), { type: type_2 }), commonInfo), this.isV2);
            }
            if (eventType === 'SHOW') {
                // SHOW 类型
                return factory(eventType, _assign(_assign(_assign({}, eventOptions), { name: action }), commonInfo), this.isV2);
            }
            if (eventType === 'VIDEO') {
                return factory(eventType, _assign({ params: eventOptions, name: 'VIDEO' }, commonInfo), this.isV2);
            }
            // Radar 日志特殊处理
            if (eventType === 'RADAR') {
                return factory(eventType, eventOptions, this.isV2);
            }
            if (['CUSTOM', 'RADAR', 'CUSTOM_EVENT'].indexOf(eventType) !== -1) {
                // CUSTOM 类型
                var _c = eventOptions, key = _c.key, value = _c.value;
                return factory('CUSTOM', _assign(_assign(_assign({}, eventOptions), { params: value, name: key }), commonInfo), this.isV2);
            }
            // TASK 类型
            return factory(eventType, _assign(_assign(_assign({}, eventOptions), { params: params, status: status, taskType: type, eventId: eventId, name: action }), commonInfo), this.isV2);
        };
        Weblog.prototype.collect = function (type, name, params, contentPackage) {
            var _a = oldApiTransformer(type, name, params, contentPackage), logType = _a[0], logParams = _a[1];
            return this.send(logType, logParams, false);
        };
        Weblog.prototype.sendImmediately = function (type, name, params, contentPackage) {
            var _a = oldApiTransformer(type, name, params, contentPackage), logType = _a[0], logParams = _a[1];
            return this.send(logType, logParams, true);
        };
        Weblog.prototype.send = function (type, params, immediately) {
            if (this._silenced)
                return;
            var log = this.generateLog(type, params);
            this.emit('event', { type: type, action: params === null || params === void 0 ? void 0 : params.action, log: log });
            // this.beforeSend(type, params, log);
            // 雷达日志单独处理
            if (type === 'RADAR') {
                return this.logger.sendRadar(log, params === null || params === void 0 ? void 0 : params.serviceName);
            }
            if (this.isV2 && this.logConfig.forbidV2HttpUrlPage && !pageUrlValidate(this.currentUrlPackage.page))
                return;
            // 如果该用户不被采样，则所有日志都不会上报
            if (!this.isSendSampled) {
                return;
            }
            var callback = typeof params === 'object' && params.callback || undefined;
            this.logger.send(log, !!immediately, callback);
        };
        Weblog.prototype.destroy = function () {
            _super.prototype.destroy.call(this);
            removeMonitor(window, 'pagehide', this.beforeUnload);
            removeMonitor(window, 'beforeunload', this.beforeUnload);
            removeInstance(this);
        };
        /**
         * 是否不发送日志，true 暂停请求，false 则会继续请求
         */
        Weblog.prototype.silence = function (state) {
            this._silenced = state;
        };
        Weblog.Logger = Logger;
        return Weblog;
    }(Weblog$1));

    function logFactory(type, data) {
        var _a = data, action = _a.action, params = _a.params, eventId = _a.eventId, contentPackage = _a.contentPackage, currentUrlPackage = _a.currentUrlPackage, _b = _a.status, status = _b === void 0 ? 'SUCCESS' : _b; _a.name; var feedLogCtx = _a.feedLogCtx;
        if (type === 'PV') {
            var _c = data, actionType = _c.type, beginTime = _c.beginTime;
            return _assign(_assign({}, currentUrlPackage), { actionType: actionType, contentPackage: contentPackage, status: status, beginTime: beginTime, eventId: eventId, pageType: isKDSNative() ? 'NATIVE' : 'H5' });
        }
        var urlPage;
        if (currentUrlPackage) {
            var page = currentUrlPackage.page, params_1 = currentUrlPackage.params, identity = currentUrlPackage.identity, coPage = currentUrlPackage.coPage;
            urlPage = {
                page: page,
                params: params_1,
                identity: identity,
                coPage: coPage,
            };
        }
        if (['RADAR', 'CUSTOM', 'CUSTOM_EVENT'].includes(type)) {
            var _d = data, key = _d.key, value = _d.value, biz = _d.biz;
            return {
                key: key,
                value: JSON.stringify(_assign(_assign({}, value), { url_package: currentUrlPackage })),
                biz: biz,
                eventId: eventId,
                urlPage: urlPage
            };
        }
        var elementLogParams = {
            action: action,
            params: JSON.stringify(params),
            contentPackage: contentPackage,
            feedLogCtx: feedLogCtx,
            eventId: eventId,
            urlPage: urlPage
        };
        if (type !== 'SHOW') {
            var taskType = data.type;
            Object.assign(elementLogParams, {
                type: taskType || 'USER_OPERATION',
                status: status || 'UNKNOWN_STATUS',
                operationType: type,
                operationDirection: 'UNKNOWN2',
            });
        }
        return elementLogParams;
    }

    /**
     * 判断当前yoda的状态
     * INIT: 初始状态
     * CHECKING：检查是否可用中，此时上报的埋点都会推入等待队列
     * READY：可用，此时上报的埋点会以桥接方式上报
     * DISABLED：不可用，此时上报的埋点会以 http 方式上报
     */
    var YodaStatus;
    (function (YodaStatus) {
        YodaStatus[YodaStatus["INIT"] = -1] = "INIT";
        YodaStatus[YodaStatus["CHECKING"] = 0] = "CHECKING";
        YodaStatus[YodaStatus["READY"] = 1] = "READY";
        YodaStatus[YodaStatus["DISABLED"] = 2] = "DISABLED";
    })(YodaStatus || (YodaStatus = {}));
    var getMockYoda = function () {
        try {
            if (isInYoda()) {
                return {
                    invoke: bridge.invoke,
                    version: bridge.version
                };
            }
        }
        catch (err) { }
        console.warn('Not in yoda webview!');
        return;
    };
    var BridgeWeblogger = /** @class */ (function (_super) {
        __extends(BridgeWeblogger, _super);
        function BridgeWeblogger(logConfig, base) {
            var _this = _super.call(this, _assign(_assign({}, logConfig), { yoda: getMockYoda(), bridgeMode: !(logConfig === null || logConfig === void 0 ? void 0 : logConfig.disableBridge) }), base) || this;
            _this.yodaStatus = YodaStatus.INIT;
            // 记录页面是否已发送过 enter pv;
            _this.entered = '';
            if ((logConfig === null || logConfig === void 0 ? void 0 : logConfig.disableBridge) || !isInYoda()) {
                _this.yodaStatus = YodaStatus.DISABLED;
            }
            else {
                _this.checkSupportBridge();
            }
            // @ts-ignore
            if (!_this.waitYodaQueue) {
                _this.waitYodaQueue = [];
            }
            return _this;
        }
        BridgeWeblogger.prototype.checkSupportBridge = function () {
            return __awaiter(this, void 0, void 0, function () {
                var isSupport;
                var _this = this;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, isSupportBridgeLog()];
                        case 1:
                            isSupport = _a.sent();
                            this.yodaStatus = isSupport ? YodaStatus.READY : YodaStatus.DISABLED;
                            this.waitYodaQueue.forEach(function (item) {
                                var type = item.type, options = item.options, immediately = item.immediately;
                                if (isSupport) {
                                    _this.sendByBridge(type, options, immediately);
                                }
                                else {
                                    _this.sendByHttp(type, options, immediately);
                                }
                            });
                            this.waitYodaQueue = [];
                            return [2 /*return*/];
                    }
                });
            });
        };
        /**
         * 桥接上报
         */
        BridgeWeblogger.prototype.sendByBridge = function (eventType, eventOptions, immediately) {
            var _a, _b;
            var _c = this.logConfig, configDisablePV = _c.disablePV, biz = _c.biz;
            var _d = eventOptions, contentPackage = _d.contentPackage, eventDisablePV = _d.disablePV, _e = _d.bridgeFields, bridgeFields = _e === void 0 ? {} : _e, urlPage = _d.urlPage;
            var disablePV = configDisablePV || eventDisablePV;
            if (contentPackage) {
                contentPackage = typeof contentPackage === 'string' ? contentPackage : JSON.stringify(contentPackage);
            }
            var setClientLogType = 'addTaskEvent';
            var isPVEvent = eventType === 'PV';
            if (isPVEvent) {
                if (disablePV)
                    return;
                // bridge PV 仅支持 enter 和 visible
                var supportPVTypes = ['enter', 'visible'];
                var _f = eventOptions, page = _f.page, type = _f.type, params = _f.params, coPage_1 = _f.coPage, category = _f.category;
                if (!type)
                    eventOptions.type = type = 'enter';
                if (type === 'leave' && this.currentUrlPackage.coPage && checkSupportCloseCoPage()) {
                    setClientLog('closeCoPage', _assign(_assign({}, eventOptions), this.currentUrlPackage));
                    return;
                }
                if (supportPVTypes.indexOf(type) === -1) {
                    return;
                }
                // enter 则强行重新生成新的 currentUrlPackage，否则局部更新
                if (type === 'enter') {
                    var realPage = page || this.currentUrlPackage.page;
                    this.updateCurrentUrlPackage({ page: realPage, params: !page ? this.currentUrlPackage.params : params, force: true });
                    if (coPage_1) {
                        this.currentUrlPackage.coPage = coPage_1;
                    }
                    if (category) {
                        this.currentUrlPackage.category = category;
                    }
                    this.entered = realPage;
                }
                else if (page === this.entered) {
                    this.currentUrlPackage.update(page, params);
                }
                // 对于上报到v2，page为url的埋点，给予提示
                if (!pageUrlValidate(this.currentUrlPackage.page))
                    return;
                setClientLogType = 'setCurrentPage';
            }
            var coPage = !!this.currentUrlPackage.coPage;
            var urlPackage = (urlPage === null || urlPage === void 0 ? void 0 : urlPage.page) && !isPVEvent ? {
                page: urlPage.page,
                identity: urlPage.identity,
                params: stringifyParams(urlPage.params),
                page_type: this.currentUrlPackage.page_type
            } : _assign(_assign({}, this.currentUrlPackage.toJSON()), { coPage: coPage });
            var log = logFactory(eventType, _assign(_assign({ biz: biz }, eventOptions), { contentPackage: contentPackage, currentUrlPackage: disablePV ? undefined : urlPackage })); // TODO: 待优化 logFactory 参数
            var isCustomType = ['CUSTOM', 'RADAR', 'CUSTOM_EVENT'].indexOf(eventType) !== -1;
            // 保持 h5ExtraAttr 来源统一
            this.commonPackage.increaseH5SeqId(isCustomType);
            // 设置 log 公共属性
            var _g = this.baseOption || {}, service_name = _g.service_name, sub_biz = _g.sub_biz, need_encrypt = _g.need_encrypt;
            // TODO: 桥接上报不需要 common_package 信息，可以优化
            log.h5ExtraAttr = JSON.stringify(this.commonPackage.getH5ExtraAttr({ bridge_info: ((_b = (_a = this.logConfig) === null || _a === void 0 ? void 0 : _a.yoda) === null || _b === void 0 ? void 0 : _b.version) || bridge.version || true, coPage: coPage }));
            log.realtime = !!immediately;
            log.serviceName = service_name || '';
            log.subBiz = sub_biz || '';
            log.needEncrypt = need_encrypt || false;
            log.container = isKDSNative() ? 'REACT_NATIVE' : 'H5';
            if (eventType === 'SHOW') {
                setClientLogType = 'addElementShowEvent';
            }
            else if (isCustomType) {
                // 兼容业务一些迁移行为，自动把 custom_stat_event 转为 custom_event
                setClientLogType = (eventType === 'CUSTOM_EVENT' || this.logConfig.customStatToCustom) ? 'addCustomEvent' : 'addCustomStatEvent';
            }
            setClientLog(setClientLogType, _assign(_assign({}, bridgeFields), log));
        };
        BridgeWeblogger.prototype.updateCurrentUrlPackage = function (url, type) {
            if (url === void 0) { url = (location === null || location === void 0 ? void 0 : location.href) || ''; }
            if (type === void 0) { type = 'web'; }
            if (typeof url === 'object' && !url.force && this.currentUrlPackage) {
                var _a = url.page, page = _a === void 0 ? this.currentUrlPackage.page : _a, _b = url.params, params = _b === void 0 ? {} : _b;
                if (!page || page === this.currentUrlPackage.page) {
                    // 兼容降级为 http 上报情况
                    this.currentUrlPackage.update(page, params);
                    var isPageEntered = this.entered === page;
                    if (this.yodaStatus !== YodaStatus.DISABLED && isPageEntered) {
                        // 客户端内直接使用 updateCurrentUrlPackage 更改页面参数是无法成功的
                        // 客户端只认 pv 上报时候带上去的参数，因此此处发送一个visible类型的pv，改变客户端维护的页面参数
                        this.send('PV', {
                            type: 'visible',
                            page: page,
                            params: params,
                        }, true);
                    }
                    return;
                }
            }
            this.currentUrlPackage = new UrlPackage(url, type, this.logConfig.urlMap);
            if (this.logConfig.attachUrl)
                this.currentUrlPackage.attachUrl();
        };
        /**
         * Http 上报
         * @param type
         * @param name
         * @param params
         * @param contentPackage
         * @param immediately
         */
        BridgeWeblogger.prototype.sendByHttp = function (type, options, immediately) {
            return _super.prototype.send.call(this, type, options, immediately);
        };
        /**
         * 复写父类 send 方法
         * @param type
         * @param name
         * @param params
         * @param contentPackage
         * @param immediately
         */
        BridgeWeblogger.prototype.send = function (type, options, immediately) {
            var _a = this.logConfig, enableV3CustomEvent = _a.enableV3CustomEvent, openHttpSender = _a.openHttpSender;
            // 目前只有enableV3CustomEvent打开时强制自定义事件上报到v3，注意此时应限制用户设置proto为v2
            var CustomToV3 = type === 'CUSTOM' && enableV3CustomEvent;
            /**
             * 强制 http 上报或yoda不可用，降级使用 http 上报
             *
             * 其中的 disabled 逻辑与 isInYoda 逻辑判断并不冲突。
             * 增加 isInYoda 判断是为了兼容 autopv 发送，autopv 发送时父BaseLogger类会调用子(也就是本类）的send 方法，
             * 这个时候 yoda 状态，waitYodaQueue 等参数还未初始化)
             *
             * 逻辑说明：
             * 如果是端外情况（不在 yoda 环境），可不用判断，直接使用 http 发送即可。
             * 如果是端内情况（拥有yoda 环境）这个 if 条件不会进入，会走下面将埋点存入 waitYodaQueue 列表逻辑，判断 yoda 使用状态后，再决定如何发送。
              */
            if (CustomToV3 || this.yodaStatus === YodaStatus.DISABLED || !isInYoda()) {
                return this.sendByHttp(type, options, immediately);
            }
            // 双报模式
            if (openHttpSender) {
                this.sendByHttp(type, options, immediately);
            }
            if (this.yodaStatus === YodaStatus.INIT) {
                this.yodaStatus = YodaStatus.CHECKING;
            }
            // yodaStatus 为 undefined 是 autopv 插件最开始发送，hybrid 内部还没有来得及初始化的情况
            if (this.yodaStatus === YodaStatus.CHECKING || typeof this.yodaStatus === 'undefined') {
                log('[weblogger]: waiting yoda ready');
                if (!this.waitYodaQueue) {
                    this.waitYodaQueue = [];
                }
                this.waitYodaQueue.push({
                    type: type,
                    options: options,
                    immediately: immediately,
                });
                return;
            }
            return this.sendByBridge(type, options, immediately);
        };
        return BridgeWeblogger;
    }(Weblog));

    var BasePlugin = /** @class */ (function () {
        function BasePlugin() {
        }
        BasePlugin.prototype.apply = function (weblog) {
            this.weblog = weblog;
        };
        return BasePlugin;
    }());

    /**
     * HTTP 和 BRIDGE 混合版
     * 端内优先桥接上报，端外走 http 上报
     */
    // 注册到全局对象
    addClass(BridgeWeblogger);
    initSilence();

    exports.BasePlugin = BasePlugin;
    exports.Weblog = BridgeWeblogger;
    exports["default"] = BridgeWeblogger;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
