// @kid-ui/krn/src/components/textarea/styles.ts
import React from 'react';
import { StyleSheet, ViewStyle, TextStyle } from 'react-native';
import { TextareaProps } from './types';
import { ScaleContext } from '../context';
import useThemeValue from '../../theme/useThemeValue';
import useKidColor from '../../theme/useKidColor';

interface TextAreaStyle {
  container: ViewStyle;
  input: ViewStyle;
  count: ViewStyle;
  countLeft: TextStyle,
  countRight: TextStyle
}

const useStyles = (props: TextareaProps, others: any = {}) => {
  const { rem } = React.useContext(ScaleContext);
  const { count } = props;
  return StyleSheet.create<TextAreaStyle>({
    container: {
      paddingVertical: rem(8),
      backgroundColor: useKidColor('cs_common_background_secondary')
    },
    input: {
      marginHorizontal: rem(19),
      marginBottom: count! > 0 ? rem(2) : 0,
      fontSize: rem(14),
      color: useKidColor('cs_common_text_title'),
      textAlignVertical: 'top'
    },
    count: {
      height: rem(20),
      marginHorizontal: rem(19),
      flexDirection: 'row',
      justifyContent: 'flex-end',
    },
    countLeft: {
      fontSize: rem(14),
      color: useKidColor('cs_common_tint_red')
    },
    countRight: {
      fontSize: rem(14),
      color: useKidColor('cs_common_text_least')
    }
  })
}

export default useStyles;
