// @locallife/base-hooks/lib/commonjs/useMemoizedFn/demo/index2.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Demo2 = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _index = _interopRequireDefault(require("../index"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* eslint-disable react-native/no-inline-styles */

/* eslint-disable react/jsx-no-bind */
// 本demo中，可以看到使用 useMemoizedFn 会比 useCallback 减少渲染次数
// 注：点击 加1 按钮，使用 useCallback 的渲染次数会被加1，而 useMemoizedFn 的渲染次数不变
// 按 Command+S，触发页面的刷新，useCallback 和 useMemoizedFn 的渲染次数会同时加1
const Demo2 = () => {
  const [count, setCount] = (0, _react.useState)(0);

  const plusBySetState = () => {
    setCount(c => c + 1);
  };

  const callbackFn = (0, _react.useCallback)(() => {
    console.log(`callbackFn：current count is ${count}`);
  }, [count]);
  const memoizedFn = (0, _index.default)(() => {
    console.log(`memoizedFn：current count is ${count}`);
  });
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactNative.View, null, /*#__PURE__*/_react.default.createElement(_reactNative.Text, null, "count: ", count), /*#__PURE__*/_react.default.createElement(_reactNative.Button, {
    title: "\u52A01",
    onPress: plusBySetState
  })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      marginTop: 15
    }
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, null, "Component with useCallback function: "), /*#__PURE__*/_react.default.createElement(ExpensiveTree, {
    title: "callback",
    showCount: callbackFn
  })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      marginTop: 10
    }
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, null, "Component with useMemoizedFn function: "), /*#__PURE__*/_react.default.createElement(ExpensiveTree, {
    title: "useMemoized",
    showCount: memoizedFn
  })));
};

exports.Demo2 = Demo2;

const ExpensiveTree = /*#__PURE__*/_react.default.memo(_ref => {
  let {
    title,
    showCount
  } = _ref;
  const renderCountRef = (0, _react.useRef)(0);
  renderCountRef.current += 1;
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, null, /*#__PURE__*/_react.default.createElement(_reactNative.Text, null, "Render Count: ", renderCountRef.current), /*#__PURE__*/_react.default.createElement(_reactNative.Button, {
    title: title,
    onPress: showCount
  }));
});
