// @kid-ui/krn/lib/components/input/styles.js
import { StyleSheet, Platform } from 'react-native';
import { getThemeValue } from '../../utils';
const android = Platform.OS === 'android';
const LABEL_FONTSIZE = 16;
const EXTRA_FONTSIZE = 16;
const CONTAINER_HEIGHT_NORMAL = 60;
const CONTAINER_HEIGHT_SMALL = 56;
const MARGIN_VERTICAL_NORMAL = 18;
const MARGIN_VERTICAL_SMALL = 16;
export { LABEL_FONTSIZE, CONTAINER_HEIGHT_NORMAL, CONTAINER_HEIGHT_SMALL, EXTRA_FONTSIZE, MARGIN_VERTICAL_NORMAL, MARGIN_VERTICAL_SMALL };
const useStyles = (props, others) => {
    const { required, extra, showLine, showArrow, showStatistics, clearable, multiline, size = 'normal' } = props;
    const margin_vertical = size !== 'small' ? MARGIN_VERTICAL_NORMAL : MARGIN_VERTICAL_SMALL;
    const { theme } = others;
    return StyleSheet.create({
        container: {
            flexDirection: 'row',
            minHeight: size !== 'small' ? CONTAINER_HEIGHT_NORMAL : CONTAINER_HEIGHT_SMALL,
            alignItems: multiline ? 'flex-start' : 'center',
            backgroundColor: getThemeValue({ light: '#FFF', dark: '#222226' }, theme)
        },
        prefix: {
            flexDirection: 'row',
            alignItems: 'center',
            marginTop: multiline ? margin_vertical : 0,
        },
        text: {
            marginLeft: 19,
            fontSize: 16,
            textAlignVertical: 'center',
            color: getThemeValue({ light: '#222', dark: '#B5B5B6' }, theme)
        },
        badge: {
            marginLeft: 3,
        },
        content: {
            flex: 1,
            flexDirection: 'row',
            alignItems: 'center',
            marginVertical: multiline ? margin_vertical : 0,
        },
        line: {
            width: 1,
            height: 24,
            backgroundColor: getThemeValue({ light: '#EAEAEA', dark: '#B5B5B6' }, theme),
            marginHorizontal: 16,
        },
        inputContainer: {
            flexDirection: 'column',
            flex: 1,
        },
        input: {
            backgroundColor: 'transparent',
            paddingTop: 0,
            paddingBottom: 0,
            fontSize: 16,
            textAlignVertical: 'center',
            color: getThemeValue({ light: '#222', dark: '#E6E6E6' }, theme),
            marginLeft: showLine ? 0 : 32,
            marginRight: (showArrow || clearable || extra || showStatistics) ? 0 : 19
        },
        inputDisabled: {
        // backgroundColor: theme.fill_disabled,
        // color: theme.color_text_disabled,
        },
        inputErrorColor: {
        // color: '#f50',
        },
        clear: {
            marginRight: (extra || showStatistics || showArrow) ? 14 : 19,
        },
        arrow: {
            marginRight: 19,
        },
        extra: {
            marginRight: 19,
        },
        statistics: {
            marginRight: 19,
            flexDirection: 'row',
            fontSize: 16,
            color: '#9C9C9C'
        },
        error: {
            color: '#FE3666',
            fontSize: 12,
            lineHeight: 16,
            position: 'absolute',
            bottom: android ? -16 : -20,
            left: (android ? 3 : 0) + (showLine ? 0 : 32)
        },
        bottomLine: {
            height: 1,
            position: 'absolute',
            bottom: 0,
            right: 12,
            backgroundColor: '#EAEAEA'
        }
    });
};
export default useStyles;
