// @krn/live-bridge/dist/liveStream/interface.js
/**
 * 直播间类型
 */
export var ELiveType;
(function (ELiveType) {
    ELiveType[ELiveType["VIDEO"] = 1] = "VIDEO";
    ELiveType[ELiveType["VOICE"] = 2] = "VOICE";
    ELiveType[ELiveType["CHAT"] = 3] = "CHAT";
    ELiveType[ELiveType["GAME"] = 4] = "GAME";
})(ELiveType || (ELiveType = {}));
/**
 * 直播玩法:
 * PK、视频连线、连麦属于
 */
export var ELiveScene;
(function (ELiveScene) {
    ELiveScene[ELiveScene["DEFAULT"] = 0] = "DEFAULT";
    ELiveScene[ELiveScene["PK"] = 1] = "PK";
    ELiveScene[ELiveScene["VIDEO_CHAT"] = 2] = "VIDEO_CHAT";
    ELiveScene[ELiveScene["MIC_CHAT"] = 3] = "MIC_CHAT";
})(ELiveScene || (ELiveScene = {}));
/**
 * 连麦类型
 */
export var EMicChatType;
(function (EMicChatType) {
    EMicChatType[EMicChatType["VOICE"] = 1] = "VOICE";
    EMicChatType[EMicChatType["VIDEO"] = 2] = "VIDEO";
})(EMicChatType || (EMicChatType = {}));
export var ELoggerType;
(function (ELoggerType) {
    ELoggerType["CLICK"] = "click";
    ELoggerType["SHOW"] = "show";
    ELoggerType["TASK"] = "task";
})(ELoggerType || (ELoggerType = {}));
