// @kid-ui/krn/lib/components/nav-bar/styles.js
import { StyleSheet } from "react-native";
import { getThemeValue, rem } from "../../utils";
const useStyles = (props, others) => {
    const { height, backgroundColor } = props || {};
    const { theme } = others;
    return StyleSheet.create({
        containerStyle: {
            width: '100%',
            height: height || rem(44),
            backgroundColor: backgroundColor || getThemeValue({ light: '#FFF', dark: '#222226' }, theme),
            flexDirection: 'row',
            alignItems: 'center',
        },
        leftContainerStyle: {
            height: '100%',
            left: rem(11),
            paddingVertical: rem(2),
            position: 'absolute',
            flexDirection: 'row',
            alignItems: 'center',
        },
        leftTextStyle: {
            fontSize: 17,
            color: getThemeValue({ light: '#222', dark: '#E6E6E6' }, theme),
            paddingHorizontal: rem(8),
            marginRight: rem(2),
        },
        iconContainerStyle: {
            width: rem(40),
            height: rem(40),
            justifyContent: 'center',
            alignItems: 'center',
        },
        btnContainerStyle: {
            height: '100%',
            flexDirection: 'row',
            alignItems: 'center',
        },
        titleTextStyle: {
            flex: 1,
            textAlign: 'center',
            textAlignVertical: 'center',
            fontSize: 17,
            fontWeight: '500',
            lineHeight: rem(25),
            color: getThemeValue({ light: '#222', dark: '#E6E6E6' }, theme),
        },
        middleContainerStyle: {
            flex: 1,
            justifyContent: 'center',
            alignItems: 'center',
            height: '100%'
        },
        rightContainerStyle: {
            height: '100%',
            right: rem(11),
            paddingVertical: rem(2),
            position: 'absolute',
            flexDirection: 'row',
            alignItems: 'center',
        },
        rightTextStyle: {
            fontSize: 17,
            color: getThemeValue({ light: '#222', dark: '#E6E6E6' }, theme),
            paddingHorizontal: rem(8),
            marginRight: rem(2),
        },
        capsuleStyle: {
            height: 'auto',
            borderWidth: rem(1),
            borderColor: getThemeValue({ light: '#EAEAEA', dark: '#EAEAEA' }, theme),
            borderRadius: rem(16),
            justifyContent: 'center',
            alignItems: 'center',
            paddingHorizontal: rem(11),
            paddingVertical: rem(6)
        },
        capsuleLineStyle: {
            width: 1,
            height: 12,
            backgroundColor: getThemeValue({ light: '#EAEAEA', dark: '#EAEAEA' }, theme),
            marginHorizontal: rem(6)
        }
    });
};
export default useStyles;
