// @locallife/biz-component/lib/commonjs/components/foldContent/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FoldContent = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _reactNativeLinearGradient = _interopRequireDefault(require("react-native-linear-gradient"));
var _krn = require("@kid-ui/krn");
var _tsLbsKrnUtils = require("@lux/ts-lbs-krn-utils");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const {
  screenSize
} = _tsLbsKrnUtils.screenUnit;
const styles = _reactNative.StyleSheet.create({
  container: {
    position: 'relative',
    overflow: 'hidden'
  },
  moreContainer: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    position: 'absolute',
    left: 0,
    bottom: 0,
    height: screenSize(98)
  },
  moreIconContainer: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: screenSize(40),
    backgroundColor: '#fff'
  },
  moreText: {
    marginTop: screenSize(8),
    fontSize: screenSize(14),
    lineHeight: screenSize(17),
    color: '#666',
    textAlign: 'center',
    marginRight: screenSize(3)
  },
  moreIcon: {
    marginTop: screenSize(6)
  },
  linearGradient: {
    width: '100%',
    height: screenSize(60)
  }
});
/**
 * 折叠内容
 */
const FoldContent = props => {
  const {
    customStyle = {}
  } = props;
  const [isMoreShow, setMoreVisible] = (0, _react.useState)(false);
  const heightAnim = (0, _react.useRef)(new _reactNative.Animated.Value(0)).current;
  const [maxHeight, setMaxHeight] = (0, _react.useState)(0);
  const isClickMore = (0, _react.useRef)(false);
  const onLayout = event => {
    const height = Math.floor(event.nativeEvent.layout.height || 0);
    if (height > props.foldHeight && !isClickMore.current) {
      setMoreVisible(true);
      setMaxHeight(height);
      heightAnim === null || heightAnim === void 0 ? void 0 : heightAnim.setValue(props.foldHeight);
    }
  };
  const onClickMore = () => {
    _reactNative.Animated.timing(heightAnim, {
      toValue: maxHeight,
      duration: 300,
      easing: _reactNative.Easing.linear,
      useNativeDriver: false
    }).start();
    setMoreVisible(false);
    isClickMore.current = true;
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
    style: [styles.container, maxHeight && {
      height: heightAnim
    }, (customStyle === null || customStyle === void 0 ? void 0 : customStyle.outerStyle) || {}]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    onLayout: onLayout
  }, props.children), isMoreShow && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.moreContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNativeLinearGradient.default, {
    start: {
      x: 0,
      y: 0
    },
    end: {
      x: 0,
      y: 1
    },
    style: styles.linearGradient,
    colors: ['rgba(255,255,255,0.00)', 'rgba(255,255,255,1)']
  }), /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    activeOpacity: 1,
    onPress: onClickMore,
    style: [styles.moreIconContainer, (customStyle === null || customStyle === void 0 ? void 0 : customStyle.moreClickStyle) || {}]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.moreText
  }, "\u5C55\u5F00\u66F4\u591A"), /*#__PURE__*/_react.default.createElement(_krn.KidIcon, {
    kid: 'local_life_locallife_taocanxiangqingzhankai_cdn',
    kidColor: 'color_base_black_6',
    size: screenSize(12),
    style: styles.moreIcon
  }))));
};
exports.FoldContent = FoldContent;
