// @kid-ui/krn/lib/components/popup/popup.js
import { WrapperComponent } from './alert-wrapper';
import React from 'react';
import RootSiblings from 'react-native-root-siblings';
export const Popup = WrapperComponent;
const show = (props) => {
    let newProps = props;
    const sibling = PopupManager.show(<Popup {...props} componentApi={false} onAnimateEnd={() => {
            props.onAnimateEnd && props.onAnimateEnd();
            sibling.destroy();
        }}/>);
    return {
        update: (props) => {
            newProps = props;
            sibling.update(<Popup {...props} visible={true} componentApi={false} onAnimateEnd={() => {
                    props.onAnimateEnd && props.onAnimateEnd();
                    sibling.destroy();
                }}/>);
        },
        remove: () => {
            sibling.update(<Popup {...newProps} visible={false} componentApi={false} onAnimateEnd={() => {
                    props.onAnimateEnd && props.onAnimateEnd();
                    sibling.destroy();
                }}/>);
        },
    };
};
Popup.show = show;
export const PopupManager = {
    show(children) {
        const sibling = new RootSiblings(children);
        return {
            update: (children) => {
                sibling.update(children);
            },
            destroy: () => {
                sibling.destroy();
            },
        };
    },
};
