// @locallife/biz-hooks/lib/commonjs/useOnTextLayout/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useOnTextLayout;

var _react = require("react");

/**
 * 监听文本布局
 *
 * @param textStr 文本字符串
 * @returns [textInfo, onTextLayout] [文本信息, 用来监听文本布局的变化]
 */
function useOnTextLayout(textStr) {
  const initTextInfo = (0, _react.useMemo)(() => {
    return {
      hasNextLine: false,
      firstLineBrand: false,
      firstLineText: '',
      text: textStr,
      nextLineText: ''
    };
  }, [textStr]);
  const [textInfo, setTextInfo] = (0, _react.useState)(initTextInfo);
  const hasLayout = (0, _react.useRef)(false);
  const onTextLayout = (0, _react.useCallback)(e => {
    let newTitleInfo;

    if (hasLayout.current === false) {
      if (e.nativeEvent.lines.length > 1) {
        let {
          text
        } = e.nativeEvent.lines[0];
        let remainText = '';

        for (let i = 1; i < e.nativeEvent.lines.length; i++) {
          remainText += e.nativeEvent.lines[i].text;
        }

        newTitleInfo = {
          hasNextLine: true,
          firstLineBrand: false,
          text: initTextInfo.text,
          nextLineText: remainText,
          firstLineText: text
        };
      } else {
        newTitleInfo = { ...initTextInfo,
          firstLineBrand: true
        };
      }

      hasLayout.current = true;
      setTextInfo(newTitleInfo);
    }
  }, [initTextInfo]);
  return [textInfo, onTextLayout];
}
