// @locallife/biz-request/lib/commonjs/util/RequestRubas.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RequestRubas = exports.REQUEST_SUCCESS = exports.REQUEST_START = exports.REQUEST_PATH = exports.REQUEST_FAILED = void 0;

var _reactNative = require("react-native");

const REQUEST_PATH = 'LOCALLIFE_KRN_REQUEST_PATH';
exports.REQUEST_PATH = REQUEST_PATH;
const REQUEST_START = 'LOCALLIFE_KRN_REQUEST_START';
exports.REQUEST_START = REQUEST_START;
const REQUEST_SUCCESS = 'LOCALLIFE_KRN_REQUEST_SUCCESS';
exports.REQUEST_SUCCESS = REQUEST_SUCCESS;
const REQUEST_FAILED = 'LOCALLIFE_KRN_REQUEST_FAILED';
exports.REQUEST_FAILED = REQUEST_FAILED;

class RequestRubas {
  static publish(event, payload) {
    _reactNative.NativeModules.Kds.invoke('tool', 'publishRubas', JSON.stringify({
      event: event,
      payload: payload
    }));
  }

  static setDimension(dimensionKey, value) {
    _reactNative.NativeModules.Kds.invoke('tool', 'setRubasDimension', JSON.stringify({
      key: dimensionKey,
      value: value
    }));
  }

}

exports.RequestRubas = RequestRubas;
