// @locallife/biz-component/lib/commonjs/components/poiInfo/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PoiInfo = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _krn = require("@kid-ui/krn");
var _styles = _interopRequireDefault(require("./styles"));
var _tsLbsKrnUtils = require("@lux/ts-lbs-krn-utils");
var _decimal = require("../../utils/decimal");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const {
  screenSize
} = _tsLbsKrnUtils.screenUnit;
const PoiInfo = _ref => {
  let {
    productData,
    poiData,
    isShowPhoneIcon,
    isShowMinDistance,
    toPOIList,
    toPOIDetail,
    toGuide,
    isShowPopupVisible
  } = _ref;
  // 点击电话弹窗
  const handlePhoneIcon = () => {
    isShowPopupVisible && isShowPopupVisible();
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.default.container
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.default.header
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: _styles.default.title
  }, "\u9002\u7528\u95E8\u5E97", (productData === null || productData === void 0 ? void 0 : productData.poiSize) > 1 && /*#__PURE__*/_react.default.createElement(_reactNative.Text, null, "\uFF08", productData === null || productData === void 0 ? void 0 : productData.poiSize, " \u5BB6\uFF09")), (productData === null || productData === void 0 ? void 0 : productData.poiSize) > 1 && /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    activeOpacity: 1,
    onPress: toPOIList
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.default.rowFlexAlignCenter
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: _styles.default.moreStoreLink
  }, "\u67E5\u770B\u5168\u90E8"), /*#__PURE__*/_react.default.createElement(_krn.KidIcon, {
    kid: 'common_base_next_24',
    kidColor: 'color_base_black_6',
    size: screenSize(14)
  })))), /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    activeOpacity: 1,
    onPress: toPOIDetail
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [_styles.default.content, _styles.default.rowFlexAlignCenter]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    style: _styles.default.poiImage,
    source: {
      uri: poiData.image
    }
  }), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.default.poiDescWrapper
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.default.poiNameWrap
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    numberOfLines: 1,
    style: [_styles.default.poiName, isShowMinDistance && _styles.default.poiNameWithMinDistance]
  }, poiData.poiName), isShowMinDistance && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.default.minDistance
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: _styles.default.minDistanceText
  }, "\u79BB\u6211\u6700\u8FD1"))), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.default.poiCategoryContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.default.subCategory
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: _styles.default.tag
  }, poiData.subCate)), !!poiData.avgPrice && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [_styles.default.subCategory, _styles.default.subCategoryDistance]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: _styles.default.tag
  }, "\uFFE5", Math.floor(+poiData.avgPrice / 100), "/\u4EBA")), !!(0, _tsLbsKrnUtils.formatHotPeople)(poiData.visitCount) && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [_styles.default.subCategory, _styles.default.subCategoryDistance]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: _styles.default.tag
  }, (0, _tsLbsKrnUtils.formatHotPeople)(poiData.visitCount), "\u4EBA\u60F3\u53BB")))), isShowPhoneIcon && /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    activeOpacity: 1,
    onPress: handlePhoneIcon
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.default.poiPhoneWrapper
  }, /*#__PURE__*/_react.default.createElement(_krn.KidIcon, {
    kid: 'common_baseflat_telephone_24',
    kidColor: 'color_base_black_8',
    size: screenSize(24)
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: _styles.default.poiPhoneText
  }, "\u7535\u8BDD"))))), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.default.dividingLine
  }), /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    activeOpacity: 1,
    onPress: toGuide
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [_styles.default.positionWrapper, _styles.default.rowFlexAlignCenter]
  }, !!poiData.distance && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_krn.KidIcon, {
    kid: 'common_list_location_grey',
    kidColor: 'color_base_black_11',
    size: screenSize(14)
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: _styles.default.distance
  }, (0, _decimal.getDistance)(poiData.distance)), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: _styles.default.divide
  }, "|")), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: _styles.default.address,
    ellipsizeMode: "tail",
    numberOfLines: 1
  }, poiData.address), /*#__PURE__*/_react.default.createElement(_krn.KidIcon, {
    kid: 'common_base_next_24',
    kidColor: 'color_base_black_6',
    size: screenSize(14)
  }))));
};
exports.PoiInfo = PoiInfo;
