// @locallife/log/src/radarlog/index.ts
import RadarKrn from '@ks-radar/radar/krn';
import { RadarExtraInfo } from './radartypes';
import { LocalLifeInitLogger } from '..';

class LocalLifeRadarLogger {
    private radarKrnLog: RadarKrn;
    private rootPropsRadar: object;

    constructor() {
        const appModel = LocalLifeInitLogger.getAppModel();
        const weblog = LocalLifeInitLogger.getWeblog();
        this.radarKrnLog = new RadarKrn({
            weblog: weblog,
            sessionId: `${appModel.sessionId}`,
            bundleId: appModel.bundleId,
            bundleVersionCode: `${appModel.BundleVersionCode}`,
            bundleVersion: `${appModel.bundleVersion}`,
            componentName: appModel.componentName,
        });
        this.rootPropsRadar = appModel;
    }

    radarEvent(name: string, params?: { [key: string]: any }) {
        try {
            this.radarKrnLog.event({
                name, // 必填
                params,
                // eslint-disable-next-line camelcase
                extra_info: {
                    ...params.extraInfo,
                    rootProps: this.rootPropsRadar,
                },
            });
        } catch (error) {
            console.error('LocalLifeRadarLogger.radarEvent Error: ', error);
        }
    }

    logRadar(name: string, extraInfo: RadarExtraInfo) {
        this.radarEvent(name, { extraInfo: extraInfo });
    }
}

export const localLifeRadarLogger = new LocalLifeRadarLogger();
