// @locallife/biz-hooks/lib/commonjs/useOnTextLayout/demo/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const textStyle = _reactNative.StyleSheet.create({
  text: {
    fontSize: 14,
    color: 'black',
    marginBottom: 20
  },
  originText: {
    fontSize: 20,
    color: 'red'
  },
  hasNextLine: {
    fontSize: 20,
    color: 'orange'
  },
  singleLine: {
    fontSize: 20,
    color: 'blue'
  },
  nextLine: {
    fontSize: 20,
    color: 'green'
  }
});

var _default = () => {
  const text = '鲁迅（1881-1936），原名周樟寿，后改名周树人，字豫才，浙江绍兴人。20世纪中国重要作家，新文化运动的领导人、左翼文化运动的支持者。毛泽东评价他是伟大的无产阶级的文学家、思想家、革命家，是中国文化革命的主将，也被人民称为“民族魂”。';
  const [textInfo, onTextLayout] = (0, _.default)(text);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: textStyle.originText
  }, "\u5C55\u793A\u539F\u6587"), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: textStyle.text // @ts-ignore
    ,
    onTextLayout: onTextLayout
  }, textInfo.text), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: textStyle.hasNextLine
  }, "\u662F\u5426\u6709\u4E0B\u4E00\u884C"), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: textStyle.text
  }, textInfo.hasNextLine ? '有下一行' : '没下一行'), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: textStyle.singleLine
  }, "\u5C55\u793A\u4E00\u884C"), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: textStyle.text
  }, textInfo.firstLineText), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: textStyle.nextLine
  }, "\u5C55\u793A\u5269\u4F59\u90E8\u5206"), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: textStyle.text
  }, textInfo.nextLineText));
};

exports.default = _default;
