// @krn/live-bridge/src/liveStreamStreamer/interface.ts
import { ILiveUserInfo } from '../liveStream/interface';

export interface ILiveLineInvitee {
    // UserInfo所需字段与列表item中保持一致
    userInfo: ILiveUserInfo;
    liveStreamId: string;
    displayLocation: string;
    displayAge: string;
    // 付费信息对象
    payInfoV2: object;
}
export interface IStartSpecificLineMatchParam {
    liveLineInvitee: ILiveLineInvitee;
    // int
    sourceType: number;
    // 投资付费透传信息
    extraInfo: string;
}

export interface ISuspendLineInvitationParam {
    lineChatId: string;
    // 挂起的原因，0:指超时默认的稍后。 1:用户点击稍后
    reason: 0 | 1;
}

export interface INotifyAddFansPriceChanged {
    addFansPrice: number;
    liveStreamId: string;
}

export interface ILineInvitationParam {
    lineChatId: string;
}

export interface IAcceptPKInvitationParams {
    /**
     * PKId
     */
    pkId: string;
}

export interface IRejectPKInvitationParams {
    /**
     * PKId
     */
    pkId: string;
}

export interface IStartPKInvitation {
    /**
     * 通用PK邀请信息
     */
    pkInviteInfo: IPKInviteInfo;

    /**
     * 业务自定义透传信息
     */
    extraInfo: string;
}

export interface IPKInviteInfo {
    /**
     * 被邀请用户UserInfo
     */
    userInfo: ILiveUserInfo;
    /**
     * 发起邀请用户直播流Id
     */
    selfLiveStreamId: string;
    /**
     * PK邀请类型：
     * 1 : 好友列表Pk
     * 2 : 推荐Pk
     * 3 : 小时榜Pk(Pk面板)
     * 4 : 小时榜Pk(小时榜面板)
     * 5 : cpsPk卖货
     */
    pkInviteType: number;
    /**
     * PK邀请来源
     * 0 : 未知
     * 1 : Pk面板
     * 2 : 强提醒
     * 3 : 引导小窗
     */
    pkInviteSource: number;
}

export interface IPKMagicFaceInfo {
    /**
     * 魔表Id
     */
    magicFaceId: string;
}

export interface IPKCurrentMagicFaceResult<T> {
    /**
     * 返回码
     */
    result: number;
    /**
     * 业务错误信息，可选
     */
    // eslint-disable-next-line camelcase
    error_msg?: string;
    /**
     * 可选，默认为 null。如果有值，其结构定义见 IPKMagicFaceInfo
     */
    data?: T;
}

/**
 * 发起多人PK参数
 */
export interface IStartMultiPK {
    /**
     * 多人PK基本配置
     */
    multiPKInfo: IMultiPKBaseInfo;
    /**
     * 具体业务额外透传参数
     */
    bizExtraInfo?: string;
}

/**
 * 多人PK基本配置信息
 */
export interface IMultiPKBaseInfo {
    /**
     * 多人PK玩法类型
     */
    playType: number;
    /**
     * 多人PK组队信息
     */
    teams: IMultiPKTeamInfo[];
}

/**
 * 多人PK单个队伍信息
 */
export interface IMultiPKTeamInfo {
    /**
     * 队伍成员userId数组
     */
    players: string[];
    /**
     * 本队id
     */
    teamId: number;
}
