// @kid-ui/krn/src/components/city-selector/style.ts
import { getKidColor, rem } from '../../utils';
import { UIUtils, TITLE_BAR_HEIGHT, windowHeight, isIOS } from './utils';
import { Platform, StyleSheet } from 'react-native';

export const paddingTop =
    Platform.OS === 'ios'
        ? UIUtils.getStatusBarHeight() + TITLE_BAR_HEIGHT
        : TITLE_BAR_HEIGHT;

// 剩余空间大于550
export const canShowLetter: boolean = windowHeight - paddingTop >= 550;

// 设计确定的最大高度为0.62
export const noLetterMaxHeight: number = Math.floor((windowHeight - paddingTop) * 0.62);

export const maxHeight = Math.max(noLetterMaxHeight, 550);

export default (isShowLetter: boolean = false) => {
    return StyleSheet.create({
        header: {
            fontSize: rem(16),
            backgroundColor: '#ffffff',
        },
        title: {
            fontSize: rem(24),
        },
        closeArea: {
            height: rem(paddingTop),
            width: '100%',
            backgroundColor: 'transparent',
            zIndex: rem(99)
        },
        mask: {
            flex: 1,
            overflow: 'hidden',
            backgroundColor: 'rgba(0, 0, 0, 0.3)',
        },
        banner: {
            flexDirection: 'row',
            justifyContent: 'space-between',
            alignItems: 'center',
        },
        cityInfo: {
            borderBottomRightRadius: rem(12),
            borderBottomLeftRadius: rem(12),
            flex: 1,
            maxHeight: isShowLetter ? rem(maxHeight) : rem(noLetterMaxHeight)
        },
        SectionListContainer: {
            paddingLeft: rem(18),
            paddingVertical: rem(20),
            paddingRight: isShowLetter ? rem(40) : rem(18)
        },
        letterSelector: {
            flexBasis: rem(40),
            borderBottomRightRadius: rem(12),
            justifyContent: 'center',
            position: 'absolute',
            right: 0,
            top: 0,
            bottom: 0,
        },
        noData: {
            flex: 1,
            justifyContent: 'center',
            alignItems: 'center',
            padding: rem(20),
        },
        top: {
            width: '100%',
        },
        bottom: {
            width: '100%',
        }
    });
};
