// @kid-ui/krn/src/components/city-selector/utils.ts
import { NativeModules, StatusBar, Platform, Dimensions } from 'react-native';
import { ElementType } from 'react';
import { rem } from '../../utils';
type IProps = {
    data: any;
    children?: ElementType | any;
};


export const screenWidth = Dimensions.get('screen').width;
export const screenHeight = Dimensions.get('screen').height;
export const windowWidth = Dimensions.get('window').width;
export const windowHeight = Dimensions.get('window').height;

export const isAndroid = Platform.OS === 'android';
export const isIOS = Platform.OS === 'ios';

export const TITLE_BAR_HEIGHT = 44;

export const fontWeight = {
    Thin: Platform.OS === 'ios' ? '100' : 'normal',
    UltraLight: Platform.OS === 'ios' ? '200' : 'normal',
    Light: Platform.OS === 'ios' ? '300' : 'normal',
    Regular: Platform.OS === 'ios' ? '400' : 'normal',
    Medium: Platform.OS === 'ios' ? '500' : 'bold',
    Semibold: Platform.OS === 'ios' ? '600' : 'bold',
    Bold: Platform.OS === 'ios' ? '700' : 'bold',
    Heavy: Platform.OS === 'ios' ? '800' : 'bold',
    Black: Platform.OS === 'ios' ? '900' : 'bold',
};

export class UIUtils {
    static isDark(theme: 'light' | 'dark') {
        return theme === 'dark';
    }

    static isIphoneX() {
        return isIOS && windowHeight >= 812;
    }

    static getStatusBarHeight(): number {
        let height: number;
        if (isIOS) {
            height = this.isIphoneX() ? 44 : 20;
        } else if (NativeModules.KSAppearance?.canImmersive?.()) {
            height = StatusBar.currentHeight || 0;
        } else {
            height = 0;
        }
        return height;
    }

    static getBottomBarHeight(): number {
        let height: number;
        if (Platform.OS === 'ios') {
            height = UIUtils.isIphoneX() ? rem(34) : 0;
        } else {
            height = 0;
        }
        return height;
    }
}

export function ShowWithData(props: IProps) {
    if (!props.data) {
        return null;
    }
    return props.children;
}
