// @locallife/biz-component/src/components/commonText/index.tsx
import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { screenUnit } from '@lux/ts-lbs-krn-utils';
const { screenSize } = screenUnit;
interface ICommonText {
    text?: string;
    children?: any;
    style?: any;
    customizationUlPoint?: any; // 定制圆点颜色
}
const styles = StyleSheet.create({
    container: {
        position: 'relative',
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center',
    },
    title: {
        fontSize: screenSize(14),
        lineHeight: screenSize(18),
        fontFamily: 'PingFangSC-Semibold',
        fontWeight: '600',
        color: '#222222',
        marginLeft: screenSize(20),
    },
    ulPoint: {
        position: 'absolute',
        top: screenSize(9),
        backgroundColor: '#666666',
        borderRadius: screenSize(3),
        width: screenSize(3),
        height: screenSize(3),
    },
    desc: {
        fontSize: screenSize(14),
        lineHeight: screenSize(18),
        fontWeight: '400',
        paddingLeft: screenSize(8),
        color: '#666666',
    },
});
export default function CommonText(props: ICommonText) {
    return (
        <View
            style={[
                styles.container,
                { marginVertical: props.text ? screenSize(4) : 0 },
            ]}
        >
            <View
                style={[
                    styles.ulPoint,
                    props.customizationUlPoint && props.customizationUlPoint,
                ]}
            />
            {props.text ? (
                <Text style={[styles.desc, props.style]}>{props.text}</Text>
            ) : (
                props.children
            )}
        </View>
    );
}
