// @locallife/biz-hooks/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "useABTest", {
  enumerable: true,
  get: function () {
    return _useABTest.default;
  }
});
Object.defineProperty(exports, "useActiveCallBack", {
  enumerable: true,
  get: function () {
    return _baseHooks.useActiveCallBack;
  }
});
Object.defineProperty(exports, "useAppOnScroll", {
  enumerable: true,
  get: function () {
    return _useAppOnScroll.default;
  }
});
Object.defineProperty(exports, "useAsync", {
  enumerable: true,
  get: function () {
    return _baseHooks.useAsync;
  }
});
Object.defineProperty(exports, "useBoolean", {
  enumerable: true,
  get: function () {
    return _baseHooks.useBoolean;
  }
});
Object.defineProperty(exports, "useCountdown", {
  enumerable: true,
  get: function () {
    return _useCountdown.default;
  }
});
Object.defineProperty(exports, "useFeedMount", {
  enumerable: true,
  get: function () {
    return _useFeedMount.default;
  }
});
Object.defineProperty(exports, "useIsTestChannel", {
  enumerable: true,
  get: function () {
    return _baseHooks.useIsTestChannel;
  }
});
Object.defineProperty(exports, "useLatest", {
  enumerable: true,
  get: function () {
    return _baseHooks.useLatest;
  }
});
Object.defineProperty(exports, "useLoading", {
  enumerable: true,
  get: function () {
    return _useLoading.default;
  }
});
Object.defineProperty(exports, "useMemoizedFn", {
  enumerable: true,
  get: function () {
    return _baseHooks.useMemoizedFn;
  }
});
Object.defineProperty(exports, "useMount", {
  enumerable: true,
  get: function () {
    return _baseHooks.useMount;
  }
});
Object.defineProperty(exports, "useMountUntilEffect", {
  enumerable: true,
  get: function () {
    return _baseHooks.useMountUntilEffect;
  }
});
Object.defineProperty(exports, "useOnDarkModeChange", {
  enumerable: true,
  get: function () {
    return _useOnDarkModeChange.default;
  }
});
Object.defineProperty(exports, "useOnTextLayout", {
  enumerable: true,
  get: function () {
    return _useOnTextLayout.default;
  }
});
Object.defineProperty(exports, "useSafeMemo", {
  enumerable: true,
  get: function () {
    return _useSafeMemo.default;
  }
});
Object.defineProperty(exports, "useToggle", {
  enumerable: true,
  get: function () {
    return _baseHooks.useToggle;
  }
});
Object.defineProperty(exports, "useUnmount", {
  enumerable: true,
  get: function () {
    return _baseHooks.useUnmount;
  }
});

var _useAppOnScroll = _interopRequireDefault(require("./useAppOnScroll"));

var _useCountdown = _interopRequireDefault(require("./useCountdown"));

var _useFeedMount = _interopRequireDefault(require("./useFeedMount"));

var _useOnDarkModeChange = _interopRequireDefault(require("./useOnDarkModeChange"));

var _useOnTextLayout = _interopRequireDefault(require("./useOnTextLayout"));

var _useABTest = _interopRequireDefault(require("./useABTest"));

var _useLoading = _interopRequireDefault(require("./useLoading"));

var _useSafeMemo = _interopRequireDefault(require("./useSafeMemo"));

var _baseHooks = require("@locallife/base-hooks");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
