// @locallife/biz-component/lib/module/components/purchaseInfo/PricePopup.js
import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { screenUnit } from '@lux/ts-lbs-krn-utils';
import CommonText from '../commonText';
const {
  screenSize
} = screenUnit;
const styles = StyleSheet.create({
  contentContainer: {
    paddingTop: screenSize(20),
    marginBottom: screenSize(24),
    paddingHorizontal: screenSize(19),
    borderTopColor: '#eaeaea',
    borderTopWidth: screenSize(0.5)
  },
  descContainer: {
    marginTop: screenSize(18),
    marginLeft: screenSize(10)
  },
  title: {
    fontSize: screenSize(14),
    lineHeight: screenSize(20),
    fontWeight: '500',
    color: '#222222',
    marginLeft: screenSize(10)
  },
  desc: {
    fontSize: screenSize(14),
    lineHeight: screenSize(20),
    fontWeight: '400',
    color: '#666666',
    textAlign: 'justify',
    marginTop: screenSize(6),
    marginLeft: screenSize(10)
  }
});
export const PricePopup = () => {
  return /*#__PURE__*/React.createElement(View, {
    style: styles.contentContainer
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.descContainer, {
      marginTop: 0
    }]
  }, /*#__PURE__*/React.createElement(CommonText, null, /*#__PURE__*/React.createElement(Text, {
    style: styles.title
  }, "\u56E2\u8D2D\u4EF7\u53CA\u5176\u4ED6\u672A\u5212\u7EBF\u5C55\u793A\u4EF7\u683C")), /*#__PURE__*/React.createElement(Text, {
    style: styles.desc
  }, "\u5546\u5BB6\u5728\u672C\u5E73\u53F0\u5C55\u793A\u7684\u5546\u54C1/\u670D\u52A1\u7684\u552E\u4EF7\uFF0C\u5982\u5B58\u7591\u8BF7\u5148\u54A8\u8BE2\u63D0\u4F9B\u4EF7\u683C\u7684\u5546\u5BB6\u3002\u5B9E\u9645\u6210\u4EA4\u4EF7\u683C\u53EF\u80FD\u4F1A\u56E0\u4E3A\u4F7F\u7528\u4F18\u60E0\u5238\u800C\u53D1\u751F\u53D8\u5316\u3002")), /*#__PURE__*/React.createElement(View, {
    style: styles.descContainer
  }, /*#__PURE__*/React.createElement(CommonText, null, /*#__PURE__*/React.createElement(Text, {
    style: styles.title
  }, "\u5212\u7EBF\u4EF7/\u95E8\u5E02\u4EF7/\u95E8\u5E97\u4EF7")), /*#__PURE__*/React.createElement(Text, {
    style: styles.desc
  }, "\u5546\u5BB6/\u670D\u52A1\u5728\u7279\u5B9A\u65F6\u95F4\u3001\u5730\u57DF\u3001\u5B9E\u4F53\u95E8\u5E97\u6216\u7F51\u7EDC\u5E73\u53F0\u9500\u552E\u7684\u540A\u724C\u4EF7\u3001\u4E13\u67DC\u6807\u4EF7\u3001\u5382\u5546\u5EFA\u8BAE\u96F6\u552E\u4EF7\uFF0C\u7531\u4E8E\u5E02\u573A\u884C\u60C5\u6CE2\u52A8\uFF0C\u8BE5\u4EF7\u683C\u4EC5\u4F9B\u53C2\u8003\u3002")));
};
